---
title: "Structural systematics and phylogeny"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
source("BSPN.R", verbose = F, print.eval = F)

database_dir <- 
    "../data/ncbi_db"

options(timeout = 300)

ncbi_table_list_2 <- 
  get_ncbi_table(db_dir = database_dir) 

saveRDS(ncbi_table_list, sprintf("%s/ncbi_table_list_v2.rds", database_dir))
#saveRDS(ncbi_table_list2, sprintf("%s/ncbi_table_list.rds", database_dir))
ncbi_table_list <- 
    readRDS(sprintf("%s/ncbi_table_list.rds", database_dir))
ncbi_table <- 
    ncbi_table_list$ncbi_table %>%
    filter(!is.na(p))


ncbi_table %>% count(db)
ncbi_table %>% filter(db =="genbank") %>% count(g) %>% arrange(desc(n))


species_summary %>%
    pull(technology)
```



## NCBI setup

```{r}

ncbi_table %>% dplyr::slice(which(ncbi_table$taxid!= ncbi_table$species_taxid)) %>% add_count(species) %>% arrange(desc(n))

ncbi_table %>% dplyr::slice(which(ncbi_table$taxid!= ncbi_table$species_taxid)) %>%
    filter(species=="Salmonella_enterica")

taxonomy <- 
    ncbi_table %>% 
    add_count(species, name = "rows") %>%
    group_by(taxid) %>%
    transmute(
        taxid,
        species,
        genus = g,
        family = f,
        order = o,
        class = c,
        phylum = p,
        phylum_clade = 
            ifelse(
                phylum =="Pseudomonadota", 
                sub("proteobacteria", "",class), 
                phylum
            ),
        rows
    ) %>%
    arrange(desc(rows)) %>%
    distinct(species, .keep_all = TRUE) %>%
    filter(rows>1)

saveRDS(
    taxonomy, 
    "../data/input/taxonomy.rds"
)



ncbi_cc <- 
    ncbi_table %>% 
    filter(db=="refseq") %>%
    add_count(species, name = "spec_count") %>%
    filter(spec_count>=10) %>%
    group_by(p, c, species) %>%
    summarise(spec_count = n(), .groups="keep") %>%
    ungroup %>%
    mutate(genus = sub("([A-Za-z]+).*", "\\1", species)) %>%
    group_by(p, c, genus) %>%
    mutate(genus_count = sum(spec_count)) %>%
    arrange(desc(genus_count), desc(spec_count)) %>%
    mutate(species2 = sub(" ", "_", species)) %>%
    ungroup %>%
    `colnames<-`(
        c(
            "Phylum", "Class", "Species", "Species_count", 
            "Genus", "Genus_count", "Species_"
            )
        ) %>%
    mutate(
        Class = 
            str_remove(Class, "proteobacteria")
    ) %>%
    mutate( # some calss labels to fix
        Class = 
            ifelse(
                Genus %in% 
                    c(
                        "Mycoplasmoides", "Mycoplasmopsis", 
                        "Mesomycoplasma", "Ureaplasma", 
                        "Metamycoplasma"
                        ),
                "Mollicutes",
                Class
            )
    )

ncbi_assembly_summary <- 
    ncbi_cc




ncbi_cc



ncbi_cc %>%
    filter(Phylum=="Mycoplasmatota")

# ncbi_table %>% filter(is.na(p)) %>% View
# 
# ncbi_cc %>% filter(is.na(Phylum))
top_species <- 
    ncbi_cc %>%
    group_by(Phylum, Class) %>%  
    filter(
        Species_count < 1500
    ) %>%
        # !(
        #     Species_ %in% 
        #         c("Escherichia_coli", "Klebsiella_pneumoniae")
        #     )
        # ) %>%
    filter(
        Species_count==max(Species_count)
        ) %>%
    arrange(
        desc(Species_count)
        ) %>%
    filter(Species_count >40)
    

top_species

```


## Single species setup

```{r}

saved_results <- 
    list()

ssn <- 
    list()

length_summary_tables <- 
    list()


species_name <- 
    top_species$Species_[1]
        #(species_list %>% pull(species2))[i]

#set up directories
genus_name <- 
      strsplit(species_name, "_")[[1]][1]
    
clade_name <- 
    filter(ncbi_table_list$ncbi_cc, species_==species_name) %>%
    pull(phylum) %>% gsub(" ", "_", .)

species_variable_names <-
  c(
      "fna_dir",
      "sync_dir", "gff_dir", "identity_dir", "mummer_dir",
      "oriC_dir",
      "delta_dir", "r_vars_dir", "results_dir", 
      "repeats_dir", "invs_dir",
      "disparities_dir",# "c01_dir",
      "subset_dir",
      "breakpoint_dir", "prokka_dir",
      "breakpoint_comparisons_dir", "snp_dir",
      "kmer_dir", "bd_dir", "plot_dir", "roary_dir",
      "sync2_dir", "test_dir", "IS_dir"
  )

species_dirs <- 
  gsub("_dir", "", species_variable_names)

for (var_name in species_variable_names){
  dir_name <- 
    gsub("_dir", "", var_name)
  full_dir_name <- 
    sprintf(
        "../data/processing/%s/%s/%s/%s", 
        dir_name, clade_name, 
        genus_name, species_name
        )
  
  assign(
    var_name, 
    full_dir_name
  )
  if (!dir.exists(full_dir_name)){dir.create(full_dir_name, recursive = T)}
}

species_table <- 
    ncbi_table %>%
    filter(species== sub("_", " ", species_name))

spec_count <- 
    nrow(species_table)

if(spec_count<250){
    subsample_idx <- 
        1:spec_count
} else {
    set.seed(123)
    subsample_idx <- 
        sample(1:spec_count, 200)
}

st_ss <- 
    species_table %>%
    dplyr::slice(subsample_idx)
    
the_rest_idx <- 
    species_table %>%
    filter(!row_number() %in% subsample_idx)
    
species_summary <- #species_table
    get_species_summary(st_ss) 

ssn[[i]] <- 
    tibble(
        spec_ = species_name,
        stn = nrow(species_table),
        ssn = nrow(species_summary)
    )
saveRDS(
    species_summary, 
    sprintf(
        "%s/species_summary.rds",
        r_vars_dir
    )
)

```


### Finding OriC and rearranging
```{r}
ftp_paths <- 
    species_summary$ftp_path

gff_paths <- 
    ftp_paths %>%
            sprintf(
                "%s/%s_genomic.gff.gz",
                ., gsub(".*/", "", .)
            )
# 
# sync_tib <- 
#     pbmclapply(
#         1:nrow(species_summary),
#         function(ii){
#             #for (ii in nrow(species_summary):1){
#             assembly_path <- 
#                 species_summary$ftp_path[ii]
#             
#             #species_summary$accession[ii]
#             sync_summary <- 
#                 sync_dnaA(
#                     assembly_path = 
#                         assembly_path,
#                     sync_dir = 
#                         sync_dir, 
#                     gff_dir = 
#                         gff_dir
#                     ) %>%
#                 mutate(
#                     idx = ii
#                 )
#             return(sync_summary)
#         }, mc.cores = 4
#     ) %>% bind_rows()

sync_tib <-
    readRDS(
        sprintf(
            "%s/sync_tib.rds",
            results_dir
            )
        )


st2 <- 
    sync_tib %>%
    filter(is.na(note))


if (nrow(sync_tib)==nrow(st2)){
    "All genomes rearranged"
} else{
    sprintf(
        "%i/200 sequences had issues",
        sum(!is.na(sync_tib$note))
        )
}

```

## ori verification
```{r}

for (i in 3:4){
species_name <- 
    top_species$Species_[i]
#(species_list %>% pull(species2))[i]
print(species_name )
#set up directories
genus_name <- 
    strsplit(species_name, "_")[[1]][1]

clade_name <- 
    filter(top_species, Species_==species_name) %>%
    pull(Phylum) %>% gsub(" ", "_", .)

species_variable_names <-
    c(
        "fna_dir",
        "sync_dir", "gff_dir", "identity_dir", "mummer_dir",
        "oriC_dir",
        "delta_dir", "r_vars_dir", "results_dir", 
        "repeats_dir", "invs_dir",
        "disparities_dir",# "c01_dir",
        "subset_dir",
        "breakpoint_dir", "prokka_dir",
        "breakpoint_comparisons_dir", "snp_dir",
        "kmer_dir", "bd_dir", "plot_dir", "roary_dir",
        "sync2_dir", "test_dir", "IS_dir"
    )

species_dirs <- 
    gsub("_dir", "", species_variable_names)

for (var_name in species_variable_names){
    dir_name <- 
        gsub("_dir", "", var_name)
    full_dir_name <- 
        sprintf(
            "../data/processing/%s/%s/%s/%s", 
            dir_name, clade_name, 
            genus_name, species_name
        )
    
    assign(
        var_name, 
        full_dir_name
    )
    if (!dir.exists(full_dir_name)){dir.create(full_dir_name, recursive = T)}
}
sync_tib <- 
    readRDS(
        sprintf(
            "%s/sync_tib.rds",
            results_dir
        )
    )
st2 <- 
    sync_tib %>%
    filter(is.na(note))


if (nrow(sync_tib)==nrow(st2)){
    "All genomes rearranged"
} else{
    sprintf(
        "%i/200 sequences had issues",
        sum(!is.na(sync_tib$note))
    )
}


### identities



species_summary <- 
    readRDS(
    sprintf(
        "%s/species_summary.rds",
        r_vars_dir
    ))


oriC_results <- 
    list()
for (j in 1:5){
  #st2[j,]
  
  disparity_file <-
    sprintf(
      "%s/%s.txt",
      disparities_dir,
      st2$accession[j]
    )
  
  
  fasta_file <- 
    st2$out_file[j]
  
  # genome_seq <- 
  #   readDNAStringSet(
  #     fasta_file
  #   )
  
  fasta_len <- 
    st2$len[j]
  
  
  ori_loc <- 
      locate_OriC(
        fasta_file,
        disparity_file,
        fasta_len
      )
      
  oriC_results[[j]] <- 
      ori_loc
          
  
}


saveRDS(
    oriC_results, 
    sprintf(
        "results/%s_Ori_positions.rds",
        species_name
    )
    )

print(i)
}


}
oriC_results[[2]]
readRDS(
    sprintf(
        "results/%s_Ori_positions.rds",
        species_name
    )
)[[2]]$OriC_plot

top_species

ori_rds <- 
    list.files("results", pattern = "Ori_positions.rds", full.names = T)

oriC_plots <- 
    lapply(
        1:length(ori_rds),
        function(i){
            ori_ls_i <- 
                readRDS(ori_rds[i])
            spec_i <- 
                basename(ori_rds[i]) %>%
                sub("_Ori_positions.rds", "", .) %>%
                sub("_", " ", .)
            op_i <- 
                lapply(ori_ls_i[1], function(x){x$OriC_plot})[[1]] +
                labs(x = NULL, y = NULL, title = paste0(spec_i)) +
                theme(
                    axis.title = element_blank(),
                    axis.text = element_blank(),
                    axis.ticks = element_blank(),
                    plot.title = 
                        element_text(hjust = 0.5, face = "bold", size = 22)
                    )

            return(op_i)
            
        }
        )



(ori_rds[1] %>% readRDS())[[4]]

oriC_scores <- 
    lapply(
        1:length(ori_rds),
        function(i){
            ori_ls_i <- 
                readRDS(ori_rds[i])
            spec_i <- 
                basename(ori_rds[i]) %>%
                sub("_Ori_positions.rds", "", .) %>%
                sub("_", " ", .)
            op_i <- 
                lapply(
                    ori_ls_i, 
                    function(x){
                        lapply(x)
                        x$top_regions
                        })[[1]] +
                labs(x = NULL, y = NULL, title = paste0(spec_i)) +
                theme(
                    axis.title = element_blank(),
                    axis.text = element_blank(),
                    axis.ticks = element_blank(),
                    plot.title = 
                        element_text(hjust = 0.5, face = "bold", size = 22)
                    )

            return(op_i)
            
        }
        )



oriC_plots
top_species
p1 <- oriC_plots[[4]]
p2 <- oriC_plots[[5]]
p3 <- oriC_plots[[1]]
p4 <- oriC_plots[[3]]
p5 <- oriC_plots[[2]]

library(patchwork)
library(grid)

#png("results/oriC_grid_plot.png", width = 4000, height = 2500, res = 300)

vps <- c("top_left", "top_mid", "top_right", "bottom_left", "bottom_right")

main_vp <- vpTree(
  viewport(name = "main"),
  vpList(
    viewport(x = 1/6 + 0.03, y = 0.73, width = 0.28, height = 0.44, name = vps[1]),
    viewport(x = 3/6 + 0.00, y = 0.73, width = 0.28, height = 0.44, name = vps[2]),
    viewport(x = 5/6 - 0.03, y = 0.73, width = 0.28, height = 0.44, name = vps[3]),
    viewport(x = 1/3 + 0.025, y = 0.25, width = 0.28, height = 0.44, name = vps[4]),
    viewport(x = 2/3 - 0.025, y = 0.25, width = 0.28, height = 0.44, name = vps[5])
  )
)

# Start plot
grid.newpage()
pushViewport(main_vp)

# Draw each plot
lapply(seq_along(pl), function(i) {
  seekViewport(vps[i])
  grid.draw(ggplotGrob(pl[[i]]))
})
# Shared axis labels — now with adjusted placement
seekViewport("main")
grid.text(
    "Genome position (10 Kb)", y = unit(0.015, "npc"), gp = gpar(fontsize = 20))
grid.text(
    "G–C disparity & \n dnaA box clusters",
    x = unit(0.025, "npc"), rot = 90, gp = gpar(fontsize = 20))

# Finish
dev.off()

# dev.off()
# dev.off()  # Uncomment when saving to file






d
for (j in 1:5){
  #st2[j,]
  
  disparity_file <-
    sprintf(
      "%s/%s.txt",
      disparities_dir,
      st2$accession[j]
    )
  
  
  fasta_file <- 
    st2$out_file[j]
  
  # genome_seq <- 
  #   readDNAStringSet(
  #     fasta_file
  #   )
  
  fasta_len <- 
    st2$len[j]
  
  
  
  locate_OriC <- 
      function(
        fasta_file,
        disparity_file,
        fasta_len
      ){
          
          genome_seq <- 
              readDNAStringSet(
                  fasta_file
                  )
          
      }
  
  
  rearranged_seq <- 
    xscat(
      subseq(genome_seq, start = round(fasta_len/2), end = fasta_len),
      subseq(genome_seq, start = 1, end = round(fasta_len/2)-1)
    )
  
  ### dnaA box check
  genome_seq <- 
    rearranged_seq
  
  dnaA_box_tib <- 
    id_dnaA_boxes(
      genome_seq
    )
  
  
  dbox <- 
      dnaA_box_tib %>%
      transmute(
          loc = midpt, 
          b
      )
  
  #dnaA_box_tib$b %>% boxplot
  
  ### GC check
  writeXStringSet(
    rearranged_seq,
    sprintf(
      "%s/%s.txt",
      oriC_dir, 
      st2$accession[j]
    )
  )
  GC_info <- 
    GC_check(
      sprintf(
        "%s/%s.txt",
        oriC_dir, 
        st2$accession[j]
      ), 
      disparity_file
    )
  
  GC_disp <- 
    fread(
      disparity_file
    ) %>%
    mutate(
      Loci = 
        row_number()
    )
  
  GC_tib <- 
      GC_disp %>%
      transmute(
          GC = `G-C`,
          loc = Loci*1e4
      )
  
  #dbox
  #GC_tib
  dgene <- 
      round(fasta_len/2)
  
  
  
  ### agreement metric
  # ori_score_tib <- 
  #     OriC_score(
  #         GC_disp, dnaA_box_tib, dnaA_gene_pos= fasta_len, 
  #         genome_len = fasta_len
  #         )
  
  ori_scores <- 
      oriC_scores_v2(
          dbox = dbox,
          GC_tib = GC_tib,
          dgene = round(fasta_len / 2),
          genome_len = fasta_len
          ) 
  #
  # ori_score_plot <- 
  #     ori_scores %>%
  #     select(mid, box_z, gc_z, gene_z, ori_score) %>%
  #     pivot_longer(cols = -mid, names_to = "component", values_to = "value") %>%
  #     ggplot(aes(x = mid / 1000, y = value, color = component)) +
  #     geom_line(size = 1) +
  #     labs(x = "Genome position (Kb)", y = "Normalized score",
  #          title = "Component contributions to OriC agreement score") +
  #     theme_classic()
      
  tier_colors <-
      c(
          High = "#B2D8B2",    # faint green
          Moderate = "#FFCC66",# amber
          Low = "#D3D3D3"      # light grey
          )
  top_regions <- 
      ori_scores %>% 
      dplyr::slice(1:3) %>%#
      mutate(
          rank_label = as.character(rank),
          fill_color = tier_colors[confidence_tier]
          )
  top_regions
  plot_OriC_markers
  
  
 
  dnaA_gene_box_GC_disp <- 
      plot_OriC_markers(
          GC_disp,
          dnaA_box_tib,
          fasta_len,
          top_regions
      )
 
  oriC_plot_dir <- 
    sprintf(
      "%s/OriC",
      plot_dir
    )
  if (!dir.exists(oriC_plot_dir)){dir.create(oriC_plot_dir)}
  ggsave(
    sprintf(
      "%s/%s.png",
      oriC_plot_dir,
      st2$accession[j]
    ),
    plot = dnaA_gene_box_GC_disp, 
    width = 8, height = 6, dpi = 300
  )
}
```



### identities
```{r}
saveRDS(
    sync_tib,
    sprintf(
        "%s/sync_tib.rds",
        results_dir
    )
    )


#species_summary %>% filter(accession =="NZ_CP062406.1")

sdc <- 
    sync_dir
#}
list.files(sdc)
## Identities
species_identities <- 
    get_pairwise_identities(
        sync_dir = sdc, 
        identity_dir = identity_dir, 
        species_table = st2, 
        genus_name, species_name
        )


#species_identities$id_dists %>% pull(Mean_distance) %>% boxplot()

species_identities$mash_plot +
    scale_fill_viridis(
                option = "B", 
                limits = 
                    c(min(species_identities$mash_tib$identity),100)
            ) 


#species_identities$mash_plot
saveRDS(
    species_identities, 
    sprintf(
        "%s/species_identities.rds",
        results_dir
    )
    )

sampled_id_table <- 
    species_identities$id_dists %>%
    arrange(desc(Sequence_mean_distance))
    

accessions <- 
    sync_tib %>%
    filter(is.na(note)) %>%
    pull(accession)

most_related <- 
    sampled_id_table %>%
    arrange(Sequence_mean_distance) %>%
    filter(Sequence %in% accessions) %>%
    dplyr::slice(1) %>% 
    pull(Sequence)

top_3 <- 
    sampled_id_table %>%
    arrange(Sequence_mean_distance) %>%
    filter(Sequence %in% accessions) %>%
    dplyr::slice(1:3) %>% 
    pull(Sequence)



```

```{r}


nucmer_commands <- 
    nucmer_ref_v_all(
        most_related, 
        accessions,
        delta_dir =  delta_dir, 
        sync_dir = sdc
        )

minimum_alignment_percentage <- 
    90
minimum_inversion_length <- 
    5e4
filtered_combined_delta <- 
    filter_combined_delta(
        sprintf("%s/ref_v_all", delta_dir), 
        minimum_alignment_percentage, 
        minimum_inversion_length
    ) 


## fix any odd replication 


filtered_combined_delta %>%
    group_by(qid) %>%
    filter(min(qs)!=1)




fcd <- 
    filtered_combined_delta %>%
    filter(!(qid %in% likely_missassembled)) %>%
    filter(qid %in% accessions)

saveRDS(
    fcd, 
    sprintf(
        "%s/fcd.rds",
        results_dir
    )
    )



clustered_genomes_list <- 
    cluster_genomes(fcd)

saveRDS(
    clustered_genomes_list, 
    sprintf(
        "%s/clustered_genomes_list.rds",
        results_dir
    )
    )
    
 
delta_identities <- 
    pbmclapply(
        1:length((list.files(sprintf("%s/ref_v_all", delta_dir)))),
        function(i){
            #for (i in 1:length((list.files(sprintf("%s/ref_v_all", delta_dir))))){
            ff <-     
                list.files(
                        sprintf("%s/ref_v_all", delta_dir), 
                        full.names = T
                        )[i]
            if (file.size(ff)!=0 & grepl("filtered", ff)) {
                delta_tib <- 
                read_delta(
                    list.files(
                        sprintf("%s/ref_v_all", delta_dir), 
                        full.names = T
                        )[i]
                    ) %>%
                rowwise() %>%
                mutate(
                    qs2 = ifelse(qe<qs, qe, qs),
                    qe2 = ifelse(qe<qs, qs, qe),
                ) %>% ungroup
            ref_ranges <- 
                IRanges(
                    start = delta_tib$rs,
                    end = delta_tib$re
                ) %>% reduce
            
            length_dif_prop <- 
                abs(
                    delta_tib$rlen[1] -
                        delta_tib$qlen[1]
                )/
                (max(
                    c(
                        delta_tib$rlen[1], 
                        delta_tib$qlen[1])
                     ))
            
            nucmer_id <- 
                sum(
                    delta_tib$meanlen
                    )/
                (max(
                    c(
                        delta_tib$rlen[1], 
                        delta_tib$qlen[1])
                     ))
            
            out_tib <- 
                tibble(
                    accession = 
                        delta_tib$qid[1],
                    identity = 
                        nucmer_id-length_dif_prop
                )
            } else {
                out_tib <- 
                    tibble(
                        accession = 
                            basename(ff),
                        identity = NA
                    )
            }
                
            
            
            return(out_tib)
            
        }, mc.cores = 4
    ) %>% bind_rows() %>%
    filter(accession %in% accessions) 



all_delta_tibs <-
    pbmclapply(
        1:length((list.files(sprintf("%s/ref_v_all", delta_dir)))),
        function(i){
            delta_tib <- 
                read_delta(
                    list.files(
                        sprintf("%s/ref_v_all", delta_dir), 
                        full.names = T
                        )[i]
                    ) %>%
                rowwise() %>%
                mutate(
                    qs2 = ifelse(qe<qs, qe, qs),
                    qe2 = ifelse(qe<qs, qs, qe),
                ) %>% ungroup
        }
        )

saveRDS(
    all_delta_tibs, 
    sprintf(
        "%s/all_delta_tibs.rds",
        r_vars_dir
        )
)


dt2 <- 
    right_join(
        delta_identities,
        species_identities$id_dists[c(1,2)] %>%
                `colnames<-`(c("accession", "MASH Identity"))
    )

delta_identities$accession

clustered_genomes <- 
   clustered_genomes_list$cluster_summary %>%
    left_join(
        ., 
        sync_tib[,c("accession", "len")] %>% 
            `colnames<-`(c("accessions", "len")),
        by = "accessions"
        ) 




## All structural variants versus identity

qrys <-
    fcd$qid %>% unique
    accessions[!(accessions==most_related)]

best_identity <- 
    inner_join(
        delta_identities %>% `colnames<-`(c("accessions", "identity")),
        clustered_genomes[,-5],
        by = "accessions"
    ) %>%
    mutate(
        identity = 
            ifelse(identity>1,  1, identity),
        identity = 
            ifelse(accessions==most_related,  1.01, identity),
    ) %>%
    arrange(cluster_id)


bis <- 
    best_identity %>% 
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>% 
    filter(!is.na(cluster_id)) %>%
    dplyr::slice(1)

```





Identify and locate structural variants

```{r}



#reference accessions
set.seed(123)
sampled_refs <- 
    c(
        best_identity %>% 
            filter(
                grouped_sequences==max(grouped_sequences)
            ) %>% 
            sample_n(3) %>%
            pull(accessions), ### 3 from the main cluster
        best_identity %>%
            filter(
                grouped_sequences!=max(grouped_sequences)
            ) %>% 
            sample_n(2) %>%
            pull(accessions)
    )
        

best_identity %>% group_by(cluster_id) %>% summarise(accessions = n())


clustered_genomes %>% count(cluster_id)    
#clustered_genomes %>% filter(cluster_id %in% 0:5)
### qry sampled
sampled_qrys <- 
    clustered_genomes[,c(1,2)] %>% 
    filter(
        !(accessions %in% sampled_refs),
        cluster_id %in% 0:5
    ) %>%
    group_by(cluster_id) %>% 
    dplyr::slice(1:2) %>%
    ungroup %>%
    sample_n(5) %>%
    pull(accessions)

##
ref_qry_pairs <- 
    expand_grid(ref = sampled_refs, qry = sampled_qrys)


ref_qry_pairs
#expand_grid(ref = c(1:5), qry = c(1:5))


delta_dir2 <- 
    sprintf(
        "%s/dd2",
        delta_dir
    )

if (!dir.exists(delta_dir2)) {dir.create(delta_dir2)}
nuc_comms <- 
    generate_nucmer_commands(
        ref_qry_pairs, delta_dir = delta_dir2, sync_dir = sync_dir
        )


delta_dir3 <- 
    sprintf(
        "%s/dd3",
        delta_dir
    )

if (!dir.exists(delta_dir3)) {dir.create(delta_dir3)}
nuc_comms <- 
    generate_nucmer_commands(
        ref_qry_pairs, delta_dir = delta_dir3, sync_dir = sync_dir, 
        nuc_remove = F
        )

pbmclapply(
    nuc_comms,
    system
)


bis_1 <- 
    bis[-1,]

if (nrow(bis_1)==0) next
rq_plots <- 
    lapply(
        1:21,#length(list.files(delta_dir2)),
        function(j){
            dfj <- 
                list.files(delta_dir2, full.names = T)[j]
            # if (j==5){
            #     pj <- 
            #         plot_delta(dfj)  +
            #         xlab("Cluster 0")
            # } else {
                pj <- 
                    plot_delta(dfj)  +
                    xlab("")
            #}
            pjo <- 
                pj +
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 22)
                )
            return(pjo)
        }
    )



do.call(
     "grid.arrange",
     c(
         rq_plots,
         ncol=7
         )
     )
    # Close the PNG device to save the image
    



    
```





```{r}
accs_ss <- 
    c(sampled_refs, sampled_qrys)

lapply(
    1:length(accs_ss),
    function(i){
        file.copy(
            sprintf(
                "%s/%s.txt",
                sync_dir, accs_ss[i]
                ),
            sprintf(
                sprintf(
                "%s/%s.txt",
                snp_dir, accs_ss[i]
                ), 
                overwrite =T
            )
            )
    }
)



accs_ss[1]

parsnp  -p 6 \
    -r ../data/processing/snp/Actinomycetota/Mycobacterium/Mycobacterium_tuberculosis/NZ_LR882500.1.txt \
    -c \
    -d ../data/processing/snp/Actinomycetota/Mycobacterium/Mycobacterium_tuberculosis \
    -o ../data/processing/snp/Actinomycetota/Mycobacterium/Mycobacterium_tuberculosis/out2


tree_file = "../data/processing/snp/Actinomycetota/Mycobacterium/Mycobacterium_tuberculosis/out2/parsnp.tree"

# Read the tree

snp_tree <- 
    read.tree(tree_file) 


snp_dist <- cophenetic(snp_tree)
# Plot a simple, rectangular tree
plot.phylo(snp_tree)

list.files(
    "%s/out2",
    snp_dir
    )


xmfa_file <- 
    sprintf(
        "%s/out2/parsnp.xmfa", 
        snp_dir
        )

xmfa_lines <-
    readLines(
        sprintf(
            "%s/out2/parsnp.xmfa", 
            snp_dir
            )
        )


read_xmfa <- 
    function(
        xmfa_file
    ){
        xmfa_lines <-
            readLines(xmfa_file)
        
        xmfa_info_tib <- 
            tibble(
                seq_idx = 
                    xmfa_lines[grepl("^##SequenceIndex", xmfa_lines)] %>%
                    sub("##SequenceIndex ", "", .),
                seq_acc = 
                    xmfa_lines[grepl("^##SequenceHeader", xmfa_lines)] %>%
                    sub("##SequenceHeader >", "", .) 
                )
        
        xmfa_headers <- 
            grep("^>\\d+:\\d+-\\d+ \\+", xmfa_lines, value = TRUE)
        
        header_tib <- 
            str_match(xmfa_headers, "^>([0-9]+):(\\d+)-(\\d+)\\s+\\+") %>%
            as.data.frame() %>% 
            as_tibble %>% 
            `colnames<-`(
                c(
                    "header", "seq_idx", "start0", "end0"
                )
            ) %>%
            mutate(
                start = as.numeric(start0) + 1,
                end   = as.numeric(end0) + 1
            ) %>%
            dplyr::select(-c("start0", "end0"))
        
        joined_tib <- 
            header_tib %>%
            left_join(xmfa_info_tib, by = c("seq_idx")) %>%
            dplyr::select(accession = seq_acc, start, end)
        
        jt_gr <- 
            GRanges(
                seqnames = joined_tib$accession,
                ranges   = IRanges(start = joined_tib$start, end = joined_tib$end)
                ) %>% 
            reduce()
        
        
    }



```


## check replication indicators
```{r}


accs_ss <- 
    c(sampled_refs, sampled_qrys)


box_tibs <- 
    pbmclapply(
        1:length(accs_ss),
        function(i){
            acc_seq <- 
                sprintf(
                    "%s/%s.txt",
                    sync_dir, 
                    accs_ss[i]
                ) %>%
                readDNAStringSet()
            
            dnaA_box_tib <- 
                id_dnaA_boxes(
                    acc_seq
                )
            
            quantile_cutoff <- 
                quantile(dnaA_box_tib$b, 0.975)
            dbs <- 
                dnaA_box_tib %>%
                dplyr::select(midpt, b, midpt, direction) %>%
                filter(
                    b >= quantile_cutoff,
                    midpt >= 5e4,
                    midpt <= (width(acc_seq)-5e4)
                    ) %>%
                mutate(
                    source = accs_ss[i],
                    s_len = width(acc_seq),
                    mp_perc = round(100*midpt/s_len)
                    )
            #plot(dbs$midpt, dbs$b)
            # SV_mps <- 
            #     breakpoint_locations %>%
            #     filter(
            #         source==accs_ss[j]) %>% 
            #     mutate(
            #         mp = round((start+end)/2)
            #         ) %>% 
            #     filter(
            #         width>1e3
            #     ) %>% 
            #     dplyr::select(variant, mp, width)
            # 
            # SV_mps2 <- 
            #   SV_mps %>%
            #   arrange(mp) %>%
            #   mutate(
            #     nearest_distance = sapply(mp, function(x) {
            #       min(abs(dbs$midpt - x))
            #     })
            #   ) %>%
            #     filter(nearest_distance<1e4)
            return(dbs)
            
        }
    )



all_box_tibs <-
    box_tibs %>% 
    bind_rows %>% 
    group_by(source, mp_perc) %>%
    mutate(
        box_density = n(),
    )


plot(dbs$midpt, dbs$b)


ggplot(all_box_tibs, aes(x = mp_perc, y = boxes, color = source)) +
  geom_point(alpha = 0.7) +
  labs(
    title = "Density (b) Across Genomic Midpoints",
    x = "Genomic Midpoint (bp)",
    y = "Density (b)",
    color = "Source"
  ) +
  theme_minimal(base_size = 14)


df2 <- 
    list.files(delta_dir2, full.names = T)

grid.arrange(
    df2[grep(accs_ss[j], df2)][1] %>% plot_delta,
    df2[grep(accs_ss[j], df2)][2] %>% plot_delta,
    df2[grep(accs_ss[j], df2)][3] %>% plot_delta
)


do.call(
     "grid.arrange",
     c(
         mpis,
         ncol=3
         )
     )

#df2[grep("NZ_CP134047.1", df2)][1] %>% read_delta


dbs

# 
# asv %>% 
#     filter(ref %in% "NZ_CP134047.1") %>%
#     mutate(
#         sv_prop = round(100*midpoint/ref_len)
#     ) %>% 
#     group_by(
#         sv_prop
#         ) %>%
#     summarise(
#         sv_count = n(),
#     ) %>% arrange(desc(sv_count))
#     
#     filter(sv_prop==75)

SV_eg <- 
    asv %>% 
    filter(ref== accs_ss[j]) %>%
    #filter(length>1e3) %>%
    dplyr::select( midpoint) %>%
    mutate(type = "SV")

box_eg <- 
    dnaA_box_tib %>%
    dplyr::select(midpt, b) %>%
    filter(b>0.005) %>%
    mutate(type = "dnaA box cluster") %>%
    `colnames<-`(
        c(
            "midpoint", "b", "type"
        )
    ) %>%
    dplyr::select( midpoint, type)

box_eg$midpoint %>% plot


breakpoint_locations %>%
    filter(
        source==accs_ss[j]) %>% 
    mutate(
        mp = round((start+end)/2)
        ) %>% 
    filter(
        width>1e3
    ) %>% 
    dplyr::select(variant, mp)
    
    filter(
        (start >720000 & (end <740000)) |
            (mp >720000) & (mp <740000)
    )

eg_tib <- 
    bind_rows(
        SV_eg, 
        box_eg
    )


ggplot() +
  geom_density(data = breakpoint_locations %>% filter(source == accs_ss[j]), 
               aes(x = start), fill = "#4F84C4", alpha = 0.5, color="#4F84C4") +
  geom_vline(data = eg_tib %>% filter(type == "dnaA box cluster"), 
             aes(xintercept = midpoint), 
             color = "#E69F00", linetype = "dashed", linewidth = 0.8) +
  labs(
    title = "SV Midpoint Density with dnaA Box Cluster Locations",
    x = "Genomic Coordinate (bp)",
    y = "Density",
  ) +

  theme_minimal(base_size = 14) +
  theme(
    plot.title = element_text(face = "bold", size = 16),
    axis.title = element_text(size=13)
  )


#box_eg %>% filter(b>0.005) %>% plot
#abline(v = fasta_len/2)


box_eg %>% plot


for (j in 1:length(accs_ss)){
    #st2[j,]
    
     disparity_file <-
         sprintf(
             "%s/%s.txt",
             disparities_dir,
             accs_ss[j]
             )
        
    fasta_file <- 
        sprintf(
            "%s/%s.txt",
            sync_dir, 
            accs_ss[j]
        )
    
    genome_seq <- 
        readDNAStringSet(
            fasta_file
        )
    
    fasta_len <- 
        width(genome_seq)
    
    rearranged_seq <- 
        xscat(
            subseq(genome_seq, start = round(fasta_len/2), end = fasta_len),
            subseq(genome_seq, start = 1, end = round(fasta_len/2)-1)
        )
    
    ### dnaA box check
    genome_seq <- 
        rearranged_seq
    
    dnaA_box_tib <- 
        id_dnaA_boxes(
            genome_seq
        )
    
    ### GC check
    writeXStringSet(
        rearranged_seq,
        sprintf(
            "%s/%s.txt",
            oriC_dir, 
            st2$accession[j]
        )
    )
    GC_info <- 
        GC_check(
            sprintf(
                "%s/%s.txt",
                oriC_dir, 
                st2$accession[j]
                ), 
            disparity_file
        )
    
    GC_disp <- 
        fread(
            disparity_file
        ) %>%
        mutate(
            Loci = 
                row_number()
        )
    dnaA_gene_box_GC_disp <- 
        ggplot(
            GC_disp, aes(x= Loci, y = `G-C`)
            ) +
            
            geom_point(size = 0.1) +  
            geom_segment(
                data = dnaA_box_tib, 
                aes(
                    x = midpt/1e4, 
                    xend = midpt/1e4, 
                    y = 
                        min(GC_disp$`G-C`)-25e2, 
                    yend = 
                        min(GC_disp$`G-C`) + 
                        (b * ((max(abs(GC_disp$`G-C`))-6e3)/max(b)))-25e2
                             #1e6)  
                  ), 
                  color = "grey", linewidth = 1
                ) +
            geom_point(
                data = dnaA_box_tib, 
                aes(
                    x = midpt / 1e4, 
                    y = min(GC_disp$`G-C`) + 
                        (b * ((max(abs(GC_disp$`G-C`)) - 6e3) / max(b)))-25e2
                    ), 
                color = "blue", size = 2,  alpha = 0.4
                ) +
        geom_line() +
          labs(
              x = "Genome sequence loci (10 Kb)", 
              y = "G-C disparity", 
              #title = ""
              ) +
            annotate(
                "segment",
                x = round(fasta_len / 2e4), xend = round(fasta_len / 2e4),  
                y = 
                    #0,
                    min(GC_disp[["G-C"]]) -25e2,
                yend = 
                    #-5e3,
                    min(GC_disp[["G-C"]]) -55e2,
                color = "black",
                linewidth = 1.5,
                arrow = 
                    arrow(
                        ends = "first", type = "closed", 
                        length = unit(0.5, "cm")
                        )
              ) +
          theme_classic() +
            theme(
                text = element_text(size=20),
                legend.position = "none"
            )
    
    oriC_plot_dir <- 
        sprintf(
            "%s/OriC",
            plot_dir
        )
    if (!dir.exists(oriC_plot_dir)){dir.create(oriC_plot_dir)}
    ggsave(
        sprintf(
            "%s/%s.png",
            oriC_plot_dir,
            st2$accession[j]
        ),
        plot = dnaA_gene_box_GC_disp, 
        width = 8, height = 6, dpi = 300
    )
    }


dnaA_gene_box_GC_disp

dnaA_box_tib %>% 
    # filter(midpt>2400000, midpt <2500000) %>%
    # filter(midpt>2445000, midpt <2450000) %>%
    filter(midpt>3030000, midpt <3050000) %>%
        ggplot() +
    geom_segment(
                aes(
                    x = midpt, 
                    xend = midpt, 
                    y = 
                        0, 
                    yend = 
                        b
                  ), 
                  color = "grey", linewidth = 1
                ) +
    geom_point(
                aes(
                    x = midpt, 
                    y = b
                    ), 
                color = "blue", size = 2,  alpha = 0.4
                ) +
    theme_classic()
    




```






Prokka annotation # 1
```{r}
prokka_accs <- 
    c(sampled_refs, sampled_qrys)

prokka_dir2 <- 
    "../data/prokka2"
pd3 <- 
    sprintf(
        "../data/prokka3/%s",
        species_name
    )

pd4 <- 
    sprintf(
        "../data/prokka4/%s",
        species_name
    )

if (!dir.exists(pd3)){dir.create(pd3, recursive = T)}
if (!dir.exists(pd4)){dir.create(pd4, recursive = T)}
prok_com <- 
    annotate_fna2(
        species_name = 
            species_name,
        sync_dir = 
            sync_dir, 
        prokka_dir = 
            pd3, 
        accessions = 
            prokka_accs
    )



pbmclapply(
    1:length(prok_com$comms),
    function(i){
        system(prok_com$comms[i])
    }, mc.cores = 1
)


list.files(pd3)


lapply(
    1:length(prokka_accs),
    function(i){
        file.copy(
            from = 
                list.files(
                    pd3,
                    pattern = sprintf("%s.gff", prokka_accs[i]), 
                    full.names = T
                    ),
            to=
                sprintf(
                    "%s/%s.gff",
                    pd4, prokka_accs[i]
                )
            
        )
    }
    
)





system(prok_com$comms[i])

system(prok_com$comms[1])


# pb = txtProgressBar(min = 0, max = length(df2), initial = 1) 
# 
# for (i in 1:length(prok_com$comms)){
#     system(prok_com$comms[i])
# }
#close(pb)

list.files(pd3, pattern = ".gff")

gffs <- 
    list.files(prokka_dir2, pattern = ".gff")

lapply(
    1:length(prokka_accs),
    function(i){
        file.copy(
            from = 
                list.files(
                    pd3,
                    pattern = sprintf("%s.gff", prokka_accs[i]), 
                    full.names = T
                    ),
            to=
                sprintf(
                    "%s/%s.gff",
                    pd4, prokka_accs[i]
                )
            
        )
    }
    
)

file.copy(
    list.files(
        prokka_dir2, 
               pattern = sprintf("%s.gff", prokka_accs[i]), 
               full.names = T
               ),
    sprintf(
        "%s/%s.gff",
        pd3, prokka_accs[i]
    )
)
    

writeLines(
    text = 
        prok_com$comms,
    con = 
        prok_com$file
    )

prok_com$comms[1]

prok_com$comms[18]

cat(paste0("sh ", prok_com$file))

prok_com$file %>%readLines()
prok_com$comms[18]

sprintf(
    "%s/%s.txt",
    sync_dir, prokka_accs[1]
) %>% readDNAStringSet(
    
)
list.files(pd3)

prok_com$comms[grepl(prokka_accs[1], prok_com$comms)]
list.files(prokka_dir2, pattern = ".gff")
system("prokka --species 'Salmonella enterica' --outdir ../data/prokka2 --prefix NZ_CP047548.1 ../data/processing/sync/Pseudomonadota/Salmonella/Salmonella_enterica/NZ_CP047548.1.txt --force")

prok_files <- 
    list.files(prokka_dir2, pattern = ".gff")
prokka_accs %in% (prok_files %>% sub( ".gff", "", .))

prokka_list <- 
    pbmclapply(
        1:length(prokka_accs),
        function(i){
            #for (i in 2:length(prokka_accs)){
            gff_i <- 
                read.delim(
                    sprintf(
                        '%s/%s.gff', pd3, prokka_accs[i]
                        ), header=F, comment.char = "#")  %>%
                as_tibble() %>%
                `colnames<-`(
                    c(
                       "accession", "source", "type", "start", 
                       "end", "score", "strand", "phase", "attributes")
                    ) %>%
                filter(!is.na(start)) %>%
                mutate(
                    width = end-start+1,
                    gene = str_extract(attributes, "(?<=gene=)[^;]+"),
                    product = str_extract(attributes, "(?<=product=)[^;]+")
                ) %>%
                dplyr::select(-c(attributes, source))
            return(gff_i)
        }
    )

all_prokka <- 
    bind_rows(prokka_list)


### ISes
eg_file <- 
    sprintf(
        "%s/%s.txt",
        sync_dir, "NZ_AP018036.1"
    )
sprintf(
    "isescan.py --seqfile %s --output %s/%s --nthread 2",
    eg_file, IS_dir, "NZ_AP018036.1"
)


all_prokka$accession %>% unique

# prokka_commands2 <- 
#     annotate_ind(
#         fasta_file, species_name, prokka_dir
#         #species_name, sync_dir, prokka_dir, accessions = c01$accessions
#     )
# 
# 
# 
# writeLines(
#     text = 
#         prokka_commands2$comms,
#     con = 
#         prokka_commands2$file
#     )

unique(all_prokka$accession)


```


## genic vs intergenic regions
```{r}
prokka_list[[1]]

clustered_genomes <- 
    clustered_genomes %>%
    rowwise() %>%
    mutate(
        len = 
            ifelse(
                is.na(len),
                width(
                    readDNAStringSet(
                        sprintf("%s/%s.txt", sync_dir, accessions)
                        )
                    ),
                len
                )
            )
    
gene_regions <- 
    lapply(
        1:length(prokka_list),
        function(i){
            #for (i in 1:length(prokka_list)){
            gff_i <- 
                prokka_list[[i]]
            
            gff_ranges <- 
                IRanges(
                   start = 
                       gff_i$start,
                   end = 
                       gff_i$end
               )
            genes_merged <- 
                reduce(gff_ranges)
            
            gene_coverage <- 
                sum(width(genes_merged))
            
            ref_len <- 
                clustered_genomes %>%
                distinct(accessions, .keep_all = T) %>%
                filter(accessions ==gff_i$accession[1]) %>%
                pull(len)
            
            intergenic_coverage <- 
                ref_len - gene_coverage
            
            genome_coverage_df <- 
                tibble(
                    region = c("Genic", "Intergenic"),
                    bp = c(gene_coverage, intergenic_coverage)
                    ) %>%
                mutate(
                    proportion = bp / sum(bp),
                    acc= 
                        gff_i %>% pull(accession) %>% unique
                    )
            return(genome_coverage_df)
            
        }
    )

gene_regions_tib <- 
    gene_regions %>% bind_rows


saveRDS(
    gene_regions_tib,
    sprintf(
        "%s/gene_regions_tib.rds",
        r_vars_dir
    )
)


```


## chromosome level and local SVs


```{r}

df_tib <- 
    ref_qry_pairs %>%
    mutate(
        expected_basename = str_c(ref, "_v_", qry, "_filtered.delta")
        ) %>% 
    inner_join(
        ., 
        tibble(
            expected_basename = basename(df3),
            fullname = df3
            ), 
        by = "expected_basename"
    ) %>%
    mutate(
        unfiltered_fullname = 
            sprintf(
                "%s/unfiltered/%s.delta",
                dirname(df3[1]),
                str_c(ref, "_v_", qry)
            )
    )

df_inner <- 
    df_tib$fullname

row_i <- 
                df_tib[i,]
            delta_i <- 
                row_i %>%
                pull(fullname)
delta_i %>% read_delta()
## SV calling

all_indels <- 
    pbmclapply(
        1:nrow(df_tib),
        function(i){
            #for (i in 1:nrow(df_tib)){
            row_i <- 
                df_tib[i,]
            delta_i <- 
                row_i %>%
                pull(fullname)
            delta_indels_i <- 
                delta_indels(delta_i) %>%
                mutate(idx = i)
            
            if (nrow(delta_indels_i)==0){
                delta_indels_i <- 
                    tibble(
                        rid = df_tib$ref[i],
                        qid = df_tib$qry[i],
                        idx = i
                    )
                    
            }
            
            #print(i)}
            return(delta_indels_i)
        }
    ) %>% 
    bind_rows %>%
    mutate(
        bin = 
            cut(
                width, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) %>%
    rowwise() %>%
    mutate(rq = paste(rid, qid, sep ="_")) %>%
    group_by(bin, rq) %>%
    mutate(
        SV_count = n()
    )

all_indels %>% ungroup %>% count(idx) %>% pull(n) %>% boxplot
all_indels %>% pull((idx)) %>% unique

saveRDS(
    all_indels,
    sprintf(
        "results/%s_all_indels.rds",
        species_name
    )
)

all_dups <- 
    pbmclapply(
        1:nrow(df_tib),
        function(i){
            row_i <- 
                df_tib[i,]
            delta_i <- 
                row_i %>%
                pull(fullname)
            
            ufd_i <- 
                row_i %>%
                pull(unfiltered_fullname)
            
            delta_dups_i <- 
                delta_duplications(
                    delta_table = delta_i, 
                    unfiltered_delta_table = ufd_i
                    ) 
            
            dd_check <- 
                ncol(delta_dups_i)
            
            delta_dups_i <- 
                delta_dups_i %>%
                mutate(
                    nr = 
                        ifelse(
                            dd_check==2,
                            0,
                            n()
                            ),
                    idx = i
                    )
            
            if (nrow(delta_dups_i)==0){
                delta_dups_i <- 
                    tibble(
                        rid = df_tib$ref[i],
                        qid = df_tib$qry[i],
                        idx = i
                    )
                    
            }
            return(delta_dups_i)
        }, mc.cores = 4
    ) %>% 
    bind_rows %>%
    mutate(
        bin = 
            cut(
                dup_len, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) 
    # group_by(bin, rq) %>%
    # mutate(
    #     SV_count = n()
    # ) 


all_dups %>% 
    dplyr::selecl(c(domains, duplication_type, dup_len, dup_gap))
all_dups %>% ungroup %>% count(idx) %>% pull(n) %>% boxplot

all_dups$dup_len %>%summary

boxplot(
    all_indels %>% ungroup %>% count(idx, sort = T) %>% pull(n),
    all_dups %>% ungroup %>% count(idx) %>% pull(n),
    all_tlocs %>% ungroup %>% count(idx) %>% pull(n)    
    )


boxplot(
    #all_indels %>% ungroup %>% pull(width),
    all_dups %>% ungroup %>% filter(dup_len<5e3) %>% pull(dup_len),
    all_tlocs %>% ungroup %>% filter(width<5e3) %>% pull(width)
    )

boxplot(
    #all_indels %>% ungroup %>% pull(width),
    all_dups %>% ungroup %>% filter(dup_len<5e3) %>% pull(dup_len),
    all_tlocs %>% ungroup %>% filter(width<5e3) %>% pull(width),
    substruct_invs%>% ungroup %>% filter(width<5e3) %>% pull(width)
    )
all_tlocs


all_13s <- 
    bind_rows(
        all_dups %>%
            transmute(
                s= dup_s, e = dup_e, 
                rid = ref_name,qid= qry_name, len = dup_len,
                variant = "dup"
            ),
        all_tlocs %>% ungroup %>%
            transmute(
                s= start, e = end, rid = rid,qid= qid, len = width,
                variant = "tloc"
            ),
        substruct_invs %>% ungroup %>%
            transmute(
                s= rs, e = re, rid = rid,qid= qid, len = meanlen,
                variant = "inv"
            )
    ) %>%
    filter(!is.na(s))%>%
    mutate(
        idx = row_number()
    )


boxplot(
    all_indels %>% ungroup %>% count(idx, sort = T) %>% pull(n),
    all_dups %>% ungroup %>% count(idx) %>% pull(n)    
    )

saveRDS(
    all_dups,
    sprintf(
        "results/%s_all_dups.rds",
        species_name
    )
)


all_dups %>% count(duplication_type)


saveRDS(
    all_dups,
    sprintf(
        "results/%s_all_dups.rds",
        species_name
    )
)

all_dups %>% ungroup %>% count(bin)
all_dups %>% ungroup %>%
    count(replichores)


all_tlocs <- 
    pbmclapply(
        1:nrow(df_tib),
        function(i){
            row_i <- 
                df_tib[i,]
            delta_i <- 
                row_i %>%
                pull(fullname)
            delta_tlocs_i <- 
                delta_translocations(delta_i) %>%
                mutate(idx = i)
            
            
            
            return(delta_tlocs_i)
        }
    ) %>% 
    bind_rows%>% 
    mutate(
        bin = 
            cut(
                width, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) %>%
    rowwise() %>%
    mutate(rq = paste(rid, qid, sep ="_")) %>%
    group_by(bin, rq) %>%
    mutate(
        SV_count = n()
    ) %>%
  mutate(
    replichores = as.character(replichores),
    replichores = sapply(
      strsplit(replichores, "-"),
      function(x) paste(sort(x), collapse = "-")
    ),
    replichores = factor(replichores)           
  ) %>% ungroup()


saveRDS(
    all_tlocs,
    sprintf(
        "results/%s_all_tlocs.rds",
        species_name
    )
)


all_tlocs %>% filter(width <1e4) %>% pull(width) %>% density %>% plot



### inversions
all_SRs <- 
    pbmclapply(
        1:nrow(df_tib),
        function(i){
            #for (i in 1:nrow(df_tib)){
            row_i <- 
                df_tib[i,]
            delta_i <- 
                row_i %>%
                pull(fullname)
            delta_SR_i <- 
                delta_structural_inversions(delta_i) %>%
                mutate(idx = i)
            # delta_SR <- 
            #     delta_SR_i
            delta_subSR_i <- 
                delta_substructural_inversions(delta_i, delta_SR_i)
            
            
            ssr_check <- 
                nrow(delta_subSR_i)
            
            delta_subSR_i <- 
                delta_subSR_i %>%
                mutate(
                    nr = 
                        ifelse(
                            ssr_check==2,
                            0,
                            n()
                            ),
                    idx = i
                    )
            
            #plot_delta(delta_i)
            
            both_types <- 
                list(
                    "structural_invs" = 
                        delta_SR_i,
                    "substructural_invs" =
                        delta_subSR_i
                )
            #print(i)}
            
            return(both_types)
        }
    )


struct_invs <- 
    lapply(
        all_SRs,
        function(x){
            x[[1]]
        }
    ) %>% 
    bind_rows()


substruct_invs %>% filter(width <1e4) %>% pull(width) %>% density %>% plot

substruct_invs$meanlen[which.max(density(substruct_invs$meanlen)$y)]

substruct_invs <- 
    lapply(
        all_SRs,
        function(x){
            x[[2]]
        }
    ) %>% 
    bind_rows() %>%
    mutate(
        width= meanlen,
        bin = 
            cut(
                width, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) %>%
    rowwise() %>%
    mutate(rq = paste(rid, qid, sep ="_")) %>%
    group_by(bin, rq) %>%
    mutate(
        SV_count = n()
    )
substruct_invs$meanlen[which.max(density(substruct_invs$meanlen)$y)]

saveRDS(
    all_tlocs,
    sprintf(
        "results/%s_all_SRs.rds",
        species_name
    )
)

list.files("results")

small_invs <- 
    substruct_invs %>% 
    ungroup %>%
    group_by(rq) %>%
    count(variant)

substruct_invs %>% 
    ungroup %>%
    group_by(bin) %>%
    count(variant)




all_indels

install.packages("pracma")
library(pracma)
x <- density(all_indels$width)
x_peaks <- findpeaks(x$y)
x$x[x_peaks[,2]]

all_SVs <- 
    bind_rows(
        all_indels %>%
            ungroup %>%
            transmute(
                rid,qid, width, 
                start, end,
                refmid = round((start+end)/2),
                #indel_type = variant, 
                bin, 
                variant_specific = variant,
                variant = "Indel",
                species, idx
                ),
        all_dups %>%
            transmute(
                rid = ref_name, 
                qid = qry_name,
                width = dup_len,
                start = dup_s, end = dup_e,
                refmid = round((start+end)/2),
                bin, 
                variant_specific = 
                    ifelse(
                        duplication_type =="tandem",
                        duplication_type,
                        paste(
                            duplication_type, domains, sep = " "
                        )
                    ),
                variant = "Duplication",
                species , idx
            ),
        all_tlocs %>%
            ungroup %>%
            transmute(
                rid, qid, 
                width,
                start, end,
                refmid,
                bin, variant,
                variant_specific =
                    replichores,
                species, idx
            ),
        substruct_invs %>%
            ungroup() %>%
            transmute(
                rid, qid, 
                width = meanlen,
                start = rs, end = re, 
                refmid = round((rs+re)/2),
                bin, 
                variant_specific = 
                    variant,
                variant = "Inversion",
                species, idx
            )
    ) 


saveRDS(
    all_SVs,
    sprintf(
        "results/%s_all_SVs.rds",
        species_name
    )
)



```



```{r}

all_SVs %>% filter(variant=="Duplication") %>% filter(width <2.5e3) %>% pull(width) %>% density %>% plot

ggplot(all_SVs %>% filter(width<1e4), aes(x = width, fill = variant)) +
  geom_density(alpha = 0.5) +
  facet_wrap(~ variant, scales = "free_y") +
  geom_vline(xintercept = 100, linetype = "dashed", color = "red") +
  geom_vline(xintercept = 1300, linetype = "dashed", color = "blue") +
  labs(x = "SV Length (bp)", y = "Density", title = "SV Length Distribution by Type") +
  theme_classic() +
  theme(text = element_text(size = 16))
```





    
```{r}


tib1 <- 
    all_indels %>% filter(width<1e3) %>%
    ungroup

tib2 <- 
    all_prokka %>%
    mutate(
        func_group = 
            case_when(grepl("transposase", product, ignore.case = TRUE) ~ "Transposase",
          grepl("hypothetical", product, ignore.case = TRUE) ~ "Hypothetical protein",
          grepl("putative", product, ignore.case = TRUE) ~ "Putative protein",
          grepl("PPE", product, ignore.case = TRUE) ~ "PPE family",
          grepl("PE[-_ ]?PGRS", product, ignore.case = TRUE) ~ "PE-PGRS family",
          grepl("^PE ", product, ignore.case = TRUE) ~ "PE family",
          grepl("ESAT-6|Esx|type VII|ESX-", product, ignore.case = TRUE) ~ "ESX secretion family",
          grepl("ABC transporter", product, ignore.case = TRUE) ~ "ABC transporter",
          grepl("toxin|antitoxin", product, ignore.case = TRUE) ~ "Toxin-antitoxin system",
          grepl("phage|capsid|tail", product, ignore.case = TRUE) ~ "Phage-associated",
          grepl("dehydrogenase|ligase|reductase|synthase", product, ignore.case = TRUE) ~ "Enzyme",
          grepl("ribosomal protein", product, ignore.case = TRUE) ~ "Ribosomal protein",
          product %like% "kinase" ~ "Signaling",
          product %like% "glucosidase|dipeptidase|transferase|oxygenase" ~ "Enzyme",
          product %like% "transporter" ~ "Transporter",
          gene %in% c("rho", "nusA") ~ "Regulatory",
          gene %in% c("rpfA", "rpfE") ~ "Dormancy/Resuscitation",
          is.na(product) & is.na(gene) ~ NA_character_,
          TRUE ~ "Other"
          )
    )

gene_gr <- 
    GRanges(
        seqnames = all_prokka$accession,
        ranges = IRanges(start = all_prokka$start, end = all_prokka$end),
        gene = all_prokka$gene,
        product = all_prokka$product
        )


merge_gr <- 
    function(
        tib1, tib2
        ){
        ranges1 <- 
            GRanges(
                seqnames = tib1$rid,
                ranges = IRanges(start = tib1$start, end = tib1$end)
            )
        ranges2 <- 
            GRanges(
                seqnames = tib2$accession,
                ranges = IRanges(start = tib2$start, end = tib2$end),
                gene = tib2$gene,
                product = tib2$product,
                prod_group = tib2$func_group
                )
        

        hits <- findOverlaps(ranges1, ranges2)
        overlap_width <- 
            width(pintersect(ranges1[queryHits(hits)], ranges2[subjectHits(hits)]))
  # Combine overlapping pairs
        merged_df <- 
            tib1 %>%
            mutate(row_id = row_number()) %>%
            left_join(
                tibble(
                    row_id = 
                        queryHits(hits),
                      gene = 
                        mcols(ranges2)$gene[subjectHits(hits)],
                      product = 
                        mcols(ranges2)$product[subjectHits(hits)],
                      overlap_len = 
                        width(pintersect(ranges1[queryHits(hits)], ranges2[subjectHits(hits)])),
                    prod_group = 
                        mcols(ranges2)$prod_group[subjectHits(hits)]
                    ),
                    by = "row_id"
                  ) %>%
                  dplyr::select(-row_id)
  
  return(merged_df)
}


all_indels %>% ungroup %>% count(rq)

all_indels %>% filter(width<1e3) %>% pull(width) %>% boxplot


tloc_tib <- 
    all_tlocs
dup_tib <- 
    all_dups %>%
    transmute(
        rid = ref_name, 
        qid = qry_name,
        start = dup_s, end = dup_e, 
        dup_len, dup_gap, domains, duplication_type 
    )



dt2 <- 
    all_dups %>%
    transmute(
        rid = ref_name, 
        qid = qry_name,
        dup_s, dup_e, dup_len, dup_gap, domains, duplication_type
    ) %>% 
    filter(duplication_type =="dispersed") %>%
    mutate(
        idx = row_number()
    )
    #pull(dup_len) %>% density %>% plot
a13 <- 
    GRanges(
  seqnames =  all_13s$rid,
  ranges = IRanges(start = all_13s$s, end = all_13s$e),
  qid = all_13s$qid,
  dup_len = all_13s$len,
  variant = all_13s$variant,
  idx = all_13s$idx
)


dup_gr <- 
    GRanges(
  seqnames = dt2$rid,
  ranges = IRanges(start = dt2$dup_s, end = dt2$dup_e),
  qid = dt2$qid,
  dup_len = dt2$dup_len,
  dup_gap = dt2$dup_gap,
  domains = dt2$domains,
  duplication_type = dt2$duplication_type,
  idx = dt2$idx
)

# 2. Make GRanges for genes from Prokka


# 3. Find overlaps
hits <- findOverlaps(a13, gene_gr)

# 4. Compute reciprocal overlap fractions
ov_w <- width(pintersect(a13[queryHits(hits)], gene_gr[subjectHits(hits)]))
a13_w <- width(a13)[queryHits(hits)]
gene_w <- width(gene_gr)[subjectHits(hits)]

# 5. Keep only overlaps with ≥50% reciprocal overlap
keep_hits <- hits[ov_w / a13_w >= 0.5 | ov_w / gene_w >= 0.5]

# 6. Build tibble of matches
matched <- 
    tibble(
        idx = mcols(a13)$idx[queryHits(keep_hits)],
        gene_start = start(gene_gr)[subjectHits(keep_hits)],
        gene_end   = end(gene_gr)[subjectHits(keep_hits)],
        gene       = mcols(gene_gr)$gene[subjectHits(keep_hits)],
        product    = mcols(gene_gr)$product[subjectHits(keep_hits)]
)

# 7. Left join to dt2 by idx (preserving all duplications)
a13_with_genes <- 
    all_13s %>%
  left_join(matched, by = "idx") %>%
    group_by(
        gene, product
    ) %>%
    ungroup %>% 
    mutate(
        func_group = 
        case_when(
          grepl("transposase", product, ignore.case = TRUE) ~ "Transposase",
          grepl("hypothetical", product, ignore.case = TRUE) ~ "Hypothetical protein",
          grepl("putative", product, ignore.case = TRUE) ~ "Putative protein",
          grepl("PPE", product, ignore.case = TRUE) ~ "PPE family",
          grepl("PE[-_ ]?PGRS", product, ignore.case = TRUE) ~ "PE-PGRS family",
          grepl("^PE ", product, ignore.case = TRUE) ~ "PE family",
          grepl("ESAT-6|Esx|type VII|ESX-", product, ignore.case = TRUE) ~ "ESX secretion family",
          grepl("ABC transporter", product, ignore.case = TRUE) ~ "ABC transporter",
          grepl("toxin|antitoxin", product, ignore.case = TRUE) ~ "Toxin-antitoxin system",
          grepl("phage|capsid|tail", product, ignore.case = TRUE) ~ "Phage-associated",
          grepl("dehydrogenase|ligase|reductase|synthase", product, ignore.case = TRUE) ~ "Enzyme",
          grepl("ribosomal protein", product, ignore.case = TRUE) ~ "Ribosomal protein",
          is.na(product) & is.na(gene) ~ NA_character_,
          TRUE ~ "Other"
          )
  ) %>%
    count(variant, product, sort = T)
    )
    summarise(
        rid = unique(rid),
        qid = unique(qid),
        s = min(s),
        e = min(e),
        len = e-s,
        #dup_gap = round(mean(dup_gap)),
        #domains = unique(domains),
        #dup_type = unique(duplication_type),
        variant = unique(all_13s$variant),
        #gs = unique(gene_start),
        #ge = unique(gene_end),
        #gene = unique(gene),
        # product = unique(product),
        dn = n()
    ) %>%
  mutate(
      gene_product =
          paste0(gene, ": ", product),
    func_group = 
        case_when(
          grepl("transposase", product, ignore.case = TRUE) ~ "Transposase",
          grepl("hypothetical", product, ignore.case = TRUE) ~ "Hypothetical protein",
          grepl("putative", product, ignore.case = TRUE) ~ "Putative protein",
          grepl("PPE", product, ignore.case = TRUE) ~ "PPE family",
          grepl("PE[-_ ]?PGRS", product, ignore.case = TRUE) ~ "PE-PGRS family",
          grepl("^PE ", product, ignore.case = TRUE) ~ "PE family",
          grepl("ESAT-6|Esx|type VII|ESX-", product, ignore.case = TRUE) ~ "ESX secretion family",
          grepl("ABC transporter", product, ignore.case = TRUE) ~ "ABC transporter",
          grepl("toxin|antitoxin", product, ignore.case = TRUE) ~ "Toxin-antitoxin system",
          grepl("phage|capsid|tail", product, ignore.case = TRUE) ~ "Phage-associated",
          grepl("dehydrogenase|ligase|reductase|synthase", product, ignore.case = TRUE) ~ "Enzyme",
          grepl("ribosomal protein", product, ignore.case = TRUE) ~ "Ribosomal protein",
          is.na(product) & is.na(gene) ~ NA_character_,
          TRUE ~ "Other"
          )
  ) 

dups_with_genes %>% ungroup %>%
    count(func_group, sort = T)

dups_with_genes %>% 
    group_by(func_group) %>%
    count(domains, sort = T)

dg2 <- 
    domain_props %>%
      mutate(
          domain_group = 
              case_when(
                domains %in% c("O_O", "R_R", "L_L", "T_T") ~ "Within-domain",
                domains %in% c("O_L", "L_O", "O_R", "R_O") ~ "Origin–replichore",
                domains %in% c("T_L", "L_T", "T_R", "R_T") ~ "Terminus–replichore",
                domains %in% c("O_T", "T_O") ~ "Origin–terminus",
                domains %in% c("L_R", "R_L") ~ "Replichore-crossing",
                TRUE ~ "Other"
              )
        ) %>%
  group_by(func_group) %>%
  filter(sum(n) >= 10) 


dg2 %>%
  group_by(func_group, domain_group) %>%
  summarise(n = n(), .groups = "drop") %>%
  group_by(func_group) %>%
  mutate(prop = n / sum(n)) %>%
  ggplot(aes(x = fct_reorder(func_group, -n), y = prop, fill = domain_group)) +
  geom_col(width = 0.8) +
  scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_brewer(palette = "Set2") +
  labs(
    x = "Functional group",
    y = "% of duplications",
    fill = "Domain group",
    title = "Domain group usage by functional category"
  ) +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))



domain_props %>%
  group_by(func_group, domain_group) %>%
  summarise(n = n(), .groups = "drop") %>%
  group_by(func_group) %>%
  mutate(prop = n / sum(n)) %>%
  ggplot(aes(x = fct_reorder(func_group, -n), y = prop, fill = domain_group)) +
  geom_col(width = 0.8) +
  scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_brewer(palette = "Set2") +
  labs(
    x = "Functional group",
    y = "% of duplications",
    fill = "Domain group",
    title = "Domain group usage by functional category"
  ) +
  theme_minimal(base_size = 14) +



dg2 <- domain_props %>%
  group_by(func_group, domains) %>%
  summarise(n = n(), .groups = "drop") %>%
  group_by(func_group) %>%
  mutate(prop = n / sum(n)) -> domain_props



ggplot(domain_props, aes(x = domains, y = fct_reorder(func_group, -n), fill = prop)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "white", high = "#2166AC") +
  labs(
    x = "Replication domain transition",
    y = "Functional group",
    fill = "Proportion",
    title = "Domain usage by functional group"
  ) +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(domain_props, aes(x = domain_group, y = fct_reorder(func_group, -n), fill = prop)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "white", high = "#2166AC") +
  labs(x = "Replication domain", y = "Functional group", fill = "Proportion") +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


```
    



```{r}
dup_tib


library(patchwork)

p1 <- dup_tib %>%
  count(bin, duplication_type) %>%
  ggplot(aes(bin, n, fill = duplication_type)) +
    geom_col(width=0.7) + theme_classic() +
    labs(y="Count", title="Raw counts")

p2 <- dup_tib %>%
  count(bin, duplication_type) %>%
  group_by(bin) %>%
  mutate(prop = n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(bin, prop, fill=duplication_type)) +
    geom_col(width=0.7) +
    scale_y_continuous(labels=percent_format()) +
    labs(y="Proportion", title="Relative composition") +
    theme_classic() +
    theme(axis.text.x = element_blank())

p1 / p2 + patchwork::plot_layout(heights = c(2,1))
## tlocs and dups
tloc_dup <- 
    all_SVs %>% 
    filter(variant %in% c("Duplication", "Translocation"))


tloc_dup_dist <- 
    tloc_dup %>% 
    filter(
        variant_specific != "tandem"
    ) %>%
    separate(
        col      = rq,                          
        into     = c("rid", "qid"),      
        sep      = "_(?=[^_]+_[^_]+$)",  
        remove   = FALSE,                
        extra    = "merge",              
        fill     = "right"
        ) %>%
    mutate(
        replichores = 
            ifelse(
                variant==
                    Duplication
                    )
    )




```

###
```{r}

```





#####

```{r}
df_disp <- all_dups %>%
  # only dispersed calls
  filter(duplication_type == "dispersed") %>%
  # create a symmetrized domain‐pair
  mutate(
    from_to = strsplit(domains, "_"),
    transition = map_chr(from_to, ~ paste(sort(.x), collapse = "_"))
  ) %>%
  # drop any stray levels and order by median gap
  mutate(
    transition = fct_reorder(transition, dup_gap, .fun = median)
  )

ggplot(df_disp, aes(x = transition, y = dup_len)) +
  geom_boxplot(fill = "#D95F02", alpha = 0.7, outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.3, size = 0.8) +
  scale_y_log10(labels = scales::comma_format()) +
  labs(
    x = "Symmetrized domain transition",
    y = expression("dup_gap (bp, log"[10]*")"),
    title = "Reference‐duplication separation by domain transition (dispersed only)"
  ) +
  theme_classic(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )


```



#####



```{r}

genic_duplications <- 
    

potential_genes <- 
    all_sv_lengths %>%
    group_by(bin) %>%
    mutate(
        bin_counts = 
            n()
            ) %>%
    ungroup %>%
    filter(bin_counts == max(bin_counts)) %>%
    transmute(
        rid, qid, width, refmid, bin, variant, species
    )


all_prokka
    

potential_genes

```



### Visualizations

```{r}
occ <- all_sv_lengths %>%
  mutate(
    v2 = if_else(grepl("inversion", variant), "Inversion", variant),
    # reverse the stacking order so Indel is on top
    v2 = factor(v2, levels = c("Translocation",
                               "Inversion",
                               "Duplication",
                               "Indel")),
    bin = factor(bin,
                 levels = c("[50,500)",
                            "[500,1e+03)",
                            "[1e+03,2e+03)",
                            "[2e+03,1e+04)",
                            "[1e+04,Inf)"))
  ) %>%
  count(bin, v2, name = "Occurrence")

totals <- occ %>%
  group_by(bin) %>%
  summarise(total = sum(Occurrence), .groups = "drop")

# 2) Plot
ggplot() +
  # stacked segments with grey85 outlines BETWEEN them
  geom_col(
    data    = occ,
    aes(x = bin, y = Occurrence, fill = v2),
    width   = 0.7,
    color   = "grey65",
    size    = 0.3
  ) +
  # outer black outline
  geom_col(
    data  = totals,
    aes(x = bin, y = total),
    fill  = NA,
    color = "black",
    size  = 0.6,
    width = 0.7
  ) +
  # total count labels
  geom_text(
    data        = totals,
    aes(x = bin, y = total, label = total),
    vjust       = -0.5,
    size        = 6
  ) +
  # new palette & no legend title
  scale_fill_viridis_d(
    option = "C",
    begin  = 0.2,
    end    = 0.8,
    direction = -1,
    name = NULL
  ) +
  scale_x_discrete(
    labels = c(
      "[50,500)"     = "50–500",
      "[500,1e+03)"  = "500–1,000",
      "[1e+03,2e+03)"= "1,000–2,000",
      "[2e+03,1e+04)"= "2,000–10,000",
      "[1e+04,Inf)"  = ">10,000"
    )
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  labs(
    x = "\n SV length range (bp)",
    y = "SV Occurrance"
  ) +
  theme_classic(base_size = 20) +
  theme(
    legend.position = "top"
  )


ggplot(all_sv_lengths, aes(x = refmid/1e6, y = variant, color = variant)) +
  geom_jitter(height = 0.2, size = 1.5, alpha = 0.6) +
  scale_color_brewer(palette = "Set1", guide = FALSE) +
  scale_x_continuous(name = "Position on reference (Mb)") +
  scale_y_discrete(name = "SV Variant Type") +
  labs(
    title = "SV breakpoint positions along the reference genome"
  ) +
  theme_classic(base_size = 14) +
  theme(
    panel.grid.major.x = element_line(color = "grey90"),
    panel.grid.minor.x = element_blank()
  )


ggplot(all_sv_lengths, aes(x = refmid/1e6, y = variant, fill = variant)) +
  geom_density_ridges(alpha = 0.7, scale = 1) +
  scale_fill_brewer(palette = "Set1", guide = FALSE) +
  scale_x_continuous(name = "Position on reference (Mb)") +
  scale_y_discrete(name = "SV Variant Type") +
  labs(
    title = "SV breakpoint density along the reference genome"
  ) +
  theme_ridges(font_size = 14) +
  theme(
    panel.grid.major.x = element_line(color = "grey90"),
    panel.grid.minor.x = element_blank()
  )



```


#### genic proportions

```{r}

sv_gr <- GRanges(
  seqnames = potential_genes$rid,
  ranges   = IRanges(start = potential_genes$refmid, width = 1)
)
mcols(sv_gr) <- potential_genes  # copies rid, qid, width, refmid, bin, species

# 2) Build GRanges of your Prokka features, with only gene & product as metadata
prokka_gr <- GRanges(
  seqnames = all_prokka$accession,
  ranges   = IRanges(start = all_prokka$start, end = all_prokka$end)
)
mcols(prokka_gr) <- all_prokka %>% select(gene, product)

# 3) Find “midpoint within feature” overlaps
hits <- findOverlaps(sv_gr, prokka_gr, type = "within")

# 4) Pull out metadata and then left-join on the hit indices
sv_df <- as_tibble(mcols(sv_gr)) %>%
  mutate(.sv_row = row_number())

hit_df <- tibble(
  .sv_row  = queryHits(hits),
  gene     = mcols(prokka_gr)$gene[subjectHits(hits)],
  product  = mcols(prokka_gr)$product[subjectHits(hits)]
)

annotated <- 
    sv_df %>%
  left_join(hit_df, by = ".sv_row") %>%
  select(-.sv_row)  %>%
    mutate(
    gene_product =
      paste0(gene, ": ", product)
    
  ) %>%
  mutate(
    func_group = 
        case_when(
            is.na(product) ~ NA_character_,
            str_detect(
                gene_product, regex("transposase",   ignore_case = TRUE)
                ) ~ "Transposase",
            str_detect(
                gene_product, regex("hypothetical",   ignore_case = TRUE)
                ) ~ "Hypothetical protein",
            str_detect(
                gene_product, regex("putative",       ignore_case = TRUE)
                ) ~ "Putative protein",
            str_detect(
                gene_product, regex("PPE family",     ignore_case = TRUE)
                ) ~ "PPE family",
            str_detect(
                gene_product, regex("tRNA",           ignore_case = TRUE)
                ) ~ "tRNA",
            str_detect(
                gene_product, regex("DNA gyrase",     ignore_case = TRUE)
                ) ~ "DNA gyrase",
            TRUE ~ "Other"
    )
  ) %>%
  # 2) if you want to keep only the top 10 groups and lump the rest:
  mutate(
    func_group = fct_lump(func_group, n = 10, other_level = "Other")
  )


annotated %>% filter(func_group =="Other") %>% 
    count(product,sort = T) %>% as.data.frame()

fg_var <- annotated %>%
  group_by(func_group, variant) %>%
  summarise(n = n(), .groups = "drop")

# 2) reorder func_group by total hits (so the biggest groups sit at top after flipping)
fg_var <- fg_var %>%
  group_by(func_group) %>%
  mutate(total = sum(n)) %>%
  ungroup() %>%
  mutate(func_group = fct_reorder(func_group, total))

# 3) plot
ggplot(fg_var, aes(x = func_group, y = n, fill = variant)) +
  geom_col(position = position_dodge(width = 0.8), width = 0.7) +
  coord_flip() +
  scale_fill_brewer(palette = "Set2", name = "SV Type") +
  labs(
    x = NULL,
    y = "Count of SVs",
    title = "Breakpoints by Functional Group and SV Variant"
  ) +
  theme_classic(base_size = 16) +
  theme(
    axis.text.y = element_text(size = 10),
    legend.position = "bottom"
  )



```















## plotting SV stats
```{r}



occ <- 
    all_sv_lengths %>%
  mutate(
      v2 = 
          if_else(grepl("inversion", variant), "Inversion", variant) %>%
          factor(.,
                levels = c("Indel",
                           "Duplication",
                           "Inversion",
                           "Translocation"))
      ) %>%
  count(bin, v2, name = "Occurrence") %>%
  mutate(
      bin = 
          factor(
              bin, 
              levels = 
                  c(
                      "[50,500)", "[500,1e+03)", 
                      "[1e+03,2e+03)", "[2e+03,1e+04)", "[1e+04,Inf)"
                      )
              )
      )%>%
  mutate(bin_ord = as.integer(bin))
  
    
totals <- 
    occ %>%
    group_by(bin) %>%
    summarise(total = sum(Occurrence), .groups="drop")


ggplot(occ, aes(x = bin, y = Occurrence, fill = v2)) +
  geom_col(width = 0.7) +                   # stacking by default
  geom_text(
    data = totals,
    aes(x = bin, y = total, label = total),
    vjust = -0.5, 
    size = 7,
    inherit.aes = FALSE
  ) +                                        # total count labels
  scale_fill_brewer("Variant Type", palette = "Set2") +
  scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
  labs(
    title = "",
    x     = "SV length range (bp)",
    y     = "Count of SVs"
  ) +
  theme_classic() +
  theme(
    #axis.text.x    = element_text(angle = 45, hjust = 1),
      text = element_text(size = 22),
    legend.position = "top"
  ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = "> 10,000"
    )
  ) 


bin_totals <- totals %>%
  mutate(prop = total / sum(total))

ggplot(bin_totals, aes(x = bin, y = prop)) +
  geom_col(fill = "#4C72B0", width = 0.7) +
  geom_text(aes(label = scales::percent(prop, accuracy=1)), 
            vjust = -0.5, size = 3) +
  scale_y_continuous(labels = scales::percent_format(), 
                     expand = expansion(mult = c(0, .05))) +
  labs(
    title = "Proportion of All SVs by Size Bin",
    x     = "SV Size Bin (bp)",
    y     = "Proportion of total SVs"
  ) +
  theme_classic() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  )



ggplot(occ, aes(x = bin_ord, y = Occurrence, group = v2, color = v2)) +
  geom_line(size = 1) +
  geom_point(size = 3) +
  scale_x_continuous(
    breaks = 1:5,
    labels = levels(occ$bin)
  ) +
  scale_color_brewer("Variant Type", palette = "Set2") +
  labs(
    title = "SV Count by Size Bin for Each Variant Type",
    x     = "SV Size Bin (bp)",
    y     = "Count of SVs"
  ) +
  theme_light() +
  theme(
    axis.text.x    = element_text(angle = 45, hjust = 1),
    legend.position = "bottom"
  )

ggplot(occ, aes(x = bin, y = Occurrence, fill = v2)) +
  geom_col(position = "fill", width = 0.7) +
  scale_y_continuous(labels = percent_format(accuracy = 1)) +
  scale_fill_brewer("Variant Type", palette = "Set2") +
  labs(
    title = "Relative Composition of SV Types by Size Bin",
    x     = "SV Size Bin (bp)",
    y     = "Proportion of SVs"
  ) +
  theme_classic() +
  theme(
    axis.text.x    = element_text(angle = 45, hjust = 1),
    legend.position = "bottom"
  )



ggplot(occ, aes(x = bin, y = Occurrence, fill = v2)) +
  geom_col(position = position_dodge(width = 0.8), width = 0.7) +
  scale_fill_brewer("Variant Type", palette = "Set2") +
  labs(
    title = "SV Occurrence by Size Bin and Variant Type",
    x     = "SV Size Bin (bp)",
    y     = "Count of SVs"
  ) +
  theme_classic() +
  theme(
    axis.text.x    = element_text(angle = 45, hjust = 1),
    legend.position = "bottom"
  )
ggplot(occ, aes(x = bin, y = v2, fill = Occurrence)) +
  geom_tile(color = "white") +
  scale_fill_viridis_c(name = "Count", option = "C") +
  labs(
    title = "SV Occurrence by Size Bin and Variant Type",
    x     = "SV Size Bin (bp)",
    y     = "Variant Type"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x    = element_text(angle = 45, hjust = 1),
    panel.grid     = element_blank()
  )
asl2 %>% count(v2)
    
all_sv_lengths %>% ungroup %>% count(rq)

write_csv(all_sv_lengths, "../../../../Documents/all_sv_lengths.csv")

ggplot(all_sv_lengths, aes(x = variant, y = width)) +
  geom_violin(fill = "lightblue", alpha = 0.7, width = 1) +
  geom_boxplot(width = 0.2, outlier.shape = NA) +  # thin boxplot inside the violin
  scale_y_log10() +  # log scale to handle large range
  labs(
    title = "Structural Variant Length Distributions",
    x = "Variant Type",
    y = "SV Length (bp, log10 scale)"
  ) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


sv_width_summary <- 
    all_sv_lengths %>%
  group_by(bin, variant) %>%
  summarise(
    count = n(),
    .groups = "drop"
  )

# Plot: Median SV length by Bin and Variant
ggplot(sv_width_summary, aes(x = bin, y = count, fill = variant)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(
    title = "Distribution of SV Types Across Size Classes",
    x = "SV Size Class (width_bin)",
    y = "Count of SVs",
    fill = "Variant Type"
  ) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(all_sv_lengths, aes(x = bin, fill = bin)) +
  geom_bar(position = "dodge") +
  facet_wrap(~ variant, scales = "free_y") +
  labs(
    title = "Distribution of SVs Across Genome by Size Class",
    x = "Genome Bin",
    y = "Count of SVs",
    fill = "SV Size Bin"
  ) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


sv_per_rq <- 
    all_sv_lengths %>%
  group_by(rq, variant, bin) %>%
  summarise(
    n = n(),
    .groups = "drop"
  )

ggplot(sv_per_rq, aes(x = rq, y = n, fill = variant)) +
  geom_col(position = position_dodge(width = 0.8)) +
  facet_wrap(~ bin, 
             scales = "free_x",   # each bin can re-scale its x axis
             ncol   = 1) +        # stack them vertically
  labs(
    title =    "SV Counts per Ref–Qry, by SV Type and Size Bin",
    x =        "Reference–Query Pair",
    y =        "Count of SVs",
    fill =     "Variant Type"
  ) +
  theme_classic() +
  theme(
    axis.text.x  = element_text(angle = 90, hjust = 1, size = 6),
    strip.text   = element_text(size = 10, face = "bold"),
    legend.position = "bottom"
  )
ggplot(sv_per_rq, aes(x = rq, y = bin, fill = n)) +
  geom_tile(color = "white") +
  facet_wrap(~ variant, nrow = 1, scales = "free_x") +
  scale_fill_viridis_c(name = "SV count", option = "C") +
  labs(
    title =    "SV Counts per Ref–Qry × Size Bin × Variant Type",
    x =        "Reference–Query Pair",
    y =        "Size Bin"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x    = element_text(angle = 90, hjust = 1, size = 6),
    strip.text     = element_text(size = 10, face = "bold"),
    panel.grid     = element_blank()
  )

ggplot(sv_per_rq, aes(x = bin, y = n, fill = variant)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(
    title = "SV Counts per Reference–Query Comparison",
    x = "Reference–Query Pair",
    y = "Number of SVs",
    fill = "Variant Type"
  ) +
  theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, size = 8),
    text = element_text(size = 14)
  ) 


sv2 <- 
    sv_per_rq %>% 
  extract(
    col   = rq,
    into  = c("ref", "qry"),
    regex = "^(.+)_(.+)$"    
  )

ggplot(sv2, aes(x = qry, y = n, fill = variant)) +
  geom_col(position = "dodge") +
  facet_wrap(~ref, scales = "free_x", ncol = 2) +
  labs(
    title = "SV Counts split by Reference Genome",
    x     = "Query Genome",
    y     = "Number of SVs",
    fill  = "Variant"
  ) +
  theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, size = 6),
    strip.text  = element_text(size = 10),
    text        = element_text(size = 14)
  )


```





```{r}
## comparison table
df_tib <- 
    ref_qry_pairs %>%
    mutate(
        expected_basename = str_c(ref, "_v_", qry, "_filtered.delta")
        ) %>% 
    inner_join(
        ., 
        tibble(
            expected_basename = basename(df3),
            fullname = df3
            ), 
        by = "expected_basename"
    ) %>%
    mutate(
        unfiltered_fullname = 
            sprintf(
                "%s/unfiltered/%s.delta",
                dirname(df3[1]),
                str_c(ref, "_v_", qry)
            )
    )

df_inner <- 
    df_tib$fullname
    

row_i <- 
    df_tib[2,]

delta_i <- 
    row_i %>% pull(fullname)

delta_table <- 
    delta_i
plot_delta(delta_i)
uf_i <- 
    row_i %>%
    pull(unfiltered_fullname)

delta_table <- 
    delta_i

### indels 



ggplot(all_indels, aes(x = bin, y = SV_count, fill = bin)) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        text = element_text(size = 22),
        legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = "> 10,000"
    )
  ) 
### duplications
ufd_i <- 
    uf_i %>% read_delta 


ad2 <- 
    all_dups %>% 
    bind_rows %>%
    mutate(
        bin = 
            cut(
                dup_len, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) %>%
    rowwise() %>%
    mutate(rq = paste(ref_name, qry_name, sep ="_")) %>%
    group_by(bin, rq) %>%
    mutate(
        SV_count = n()
    )


saveRDS(
    ad2,
    sprintf(
        "results/%s_all_dups.rds",
        species_name
    )
)

all_dups %>% count(replichores)

ggplot(ad2, aes(x = bin, y = SV_count, fill = bin)) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        text = element_text(size = 22)#,
        #legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = "> 10,000"
    )
  ) 


all_dups %>% count(replichores)


all_dups <- all_dups %>% bind_rows()

delta_dups_i <- 
    delta_duplications(ufd_i)






### delta translocations





i=15
row_i <- 
    df_tib[i,]
delta_i <- 
    row_i %>%
    pull(fullname)

ttt <- 
    dg_tib %>% 
    filter(rid==read_delta(delta_i)$rid[1], qid==read_delta(delta_i)$qid[1]) %>% 
    transmute(rid , refmid, qrymid)
delta_i %>% 
    plot_delta +
      geom_point(
    data        = ttt,
    aes(x = refmid, y = qrymid),
    inherit.aes = FALSE,
    color       = "black",
    size        = 3
  )



at2 <- 
    all_tlocs %>% 
    bind_rows %>%
    mutate(
        bin = 
            cut(
                width, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) %>%
    rowwise() %>%
    mutate(rq = paste(rid, qid, sep ="_")) %>%
    group_by(bin, rq) %>%
    mutate(
        SV_count = n()
    )

dg_tib


rep_counts <- 
    at2 %>%
  count(replichores, name = "n") %>%
  ungroup() %>%
      mutate(
    replichores = fct_reorder(replichores, n, .fun = median)
  )%>% 
  mutate(
    replichores = as.character(replichores),                       # if it’s a factor
    replichores = sapply(
      strsplit(replichores, "-"),                                  # split into list of length-2 vectors
      function(x) paste(sort(x), collapse = "-")                   # sort and recombine
    ),
    replichores = factor(replichores)                              # back to factor if you need
  )


# 3) plot
ggplot(rep_counts, aes(x = replichores, y = n)) +
  geom_boxplot(outlier.alpha = 0.3) +
  labs(
    x = "Replichores combination",
    y = "Number of SVs per pair",
    title = "Distribution of SV counts by replichores"
  ) +
  theme_classic(base_size = 14)

ggplot(at2, aes(x = bin, y = SV_count, fill = bin)) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        text = element_text(size = 22)#,
        #legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = "> 10,000"
    )
  ) 



dup_genes <- 
    list()
for (i in 1:nrow(at2)){
    tl_i <- 
        at2[i,]
    
    tl_ir <- 
        IRanges(
            tl_i$start,
            tl_i$end
        )
    
    tl_seq <- 
        sprintf(
            "%s/%s.txt",
            sync_dir, tl_i$rid
        ) %>% 
        readDNAStringSet() %>%
        subseq(tl_i$start, tl_i$end)
    ref_prok <- 
        all_prokka %>% 
        filter(
            accession==tl_i$rid
        ) %>%
        mutate(
            strand = 
                ifelse(
                    strand==".",
                    "*", strand
                )
        )
    
    gr_genes <- 
        GRanges(
            seqnames = ref_prok$accession,
            ranges   = IRanges(start = ref_prok$start, end = ref_prok$end),
            strand   = ref_prok$strand,
            gene     = ref_prok$gene,
            product  = ref_prok$product
          )

    gr_svs <- 
        GRanges(
            seqnames = tl_i$rid,
            ranges   = IRanges(start = tl_i$start, end = tl_i$end),
            dup_type = tl_i$replichores
            )
    svg_hits <- 
        findOverlaps(gr_svs, gr_genes)
    hits_df <- 
        as.data.frame(svg_hits) %>%
        mutate(
        overlap_len = width(
          pintersect(
            gr_svs[ queryHits ],
            gr_genes[ subjectHits ]
          )
        )
      )
    
    sv_gene_overlaps <- 
        hits_df %>%
        transmute(
            sv_index    = queryHits,
            gene_index  = subjectHits,
            overlap_len
            ) %>%
        left_join(
            ., 
            tl_i %>% 
            transmute(
                svstart = start,
                svstart = end,
                refmid, qrymid, rid, 
                qid, variant, replichores, bin, species,
                sv_index = row_number()
            ), 
            by = "sv_index"
            ) %>%
        left_join(
            ., 
            ref_prok %>% 
                transmute(
                    rid = accession,
                    type, 
                    gs = start, 
                    ge = end, strand, 
                    gwidth = width, gene, product,
                    gene_index = row_number()
                ),
            by = c("rid","gene_index")
            ) %>%
        mutate(comp = i) %>%
        filter(overlap_len==max(overlap_len))
    dup_genes[[i]] <- 
        sv_gene_overlaps
    print(i)
    
}

dg_tib <- 
    dup_genes %>%  bind_rows %>% as_tibble()

rm = 
    dg_tib %>% 
    filter(product=="IS3 family transposase IS987") %>% 
    pull(refmid)
qm = 
    dg_tib %>% 
    filter(product=="IS3 family transposase IS987") %>% 
    pull(qrymid)


plot(rm-qm)

dg_tib %>%
    filter(product=="IS3 family transposase IS987")


saveRDS(
    at2,
    sprintf(
        "results/%s_all_tlocs.rds",
        species_name
    )
)

ggplot(at2, aes(x = bin, y = SV_count, fill = replichores)) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        text = element_text(size = 22)#,
        #legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = "> 10,000"
    )
  ) 





### delta inversions


delta_i %>% plot_delta
Chromosome_level_SVs(delta_i)

delta_table <- 
    delta_i

all_SRs <- 
    pbmclapply(
        1:nrow(df_tib),
        function(i){
            #for (i in 1:nrow(df_tib)){
            row_i <- 
                df_tib[i,]
            delta_i <- 
                row_i %>%
                pull(fullname)
            delta_SR_i <- 
                delta_structural_inversions(delta_i)
            # delta_SR <- 
            #     delta_SR_i
            delta_subSR_i <- 
                delta_substructural_inversions(delta_i, delta_SR_i)
            
            plot_delta(delta_i)
            
            both_types <- 
                list(
                    "structural_invs" = 
                        delta_SR_i,
                    "substructural_invs" =
                        delta_subSR_i
                )
            #print(i)}
            
            return(both_types)
        }
    )


struct_invs <- 
    lapply(
        all_SRs,
        function(x){
            x[[1]]
        }
    ) %>% 
    bind_rows()

substruct_invs <- 
    lapply(
        all_SRs,
        function(x){
            x[[2]]
        }
    ) %>% 
    bind_rows() %>%
    mutate(
        width= meanlen,
        bin = 
            cut(
                width, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, Inf
                        ), 
                right = FALSE),
        species = species_name
        ) %>%
    rowwise() %>%
    mutate(rq = paste(rid, qid, sep ="_")) %>%
    group_by(bin, rq) %>%
    mutate(
        SV_count = n()
    )


# 3) plot
ggplot(substruct_invs, aes(x = replichores, y = n)) +
  geom_boxplot(outlier.alpha = 0.3) +
  labs(
    x = "Replichores combination",
    y = "Number of SVs per pair",
    title = "Distribution of SV counts by replichores"
  ) +
  theme_classic(base_size = 14)

ggplot(substruct_invs, aes(x = bin, y = SV_count, fill = bin)) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        text = element_text(size = 22)#,
        #legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = "> 10,000"
    )
  ) 



```



```{r}
funcs <- tibble::tibble(
  Function      = c(
    "read_delta()", 
    "delta_duplications()", 
    "delta_indels()", 
    "delta_structural_rearrangements()", 
    "delta_structural_rearrangements()"
  ),
  Description   = c(
    "Parse a nucmer .delta file into table of raw coords",
    "Detect and classify tandem/dispersed/inverted duplications",
    "Identify insertions & deletions in reference-centric coords",
    "Call large (>10 kb) chromosome-scale inversions",
    "Aggregate all major SV types (dup, indel, inv, trans) into one summary"
  ),
  Key_Arguments = c(
    "`path` or `tibble`", 
    "`delta_table`, `minoverlap`", 
    "`delta_table`, `rlen`, `qlen`", 
    "`delta_table`, `size_thresh`", 
    "`delta_table`"
  ),
  Returns       = c(
    "Tibble of raw alignment coords", 
    "Tibble of duplication calls", 
    "Tibble of indel events", 
    "Tibble of inversion events", 
    "Tibble of SV event summaries"
  )
)
```




```{r}

df2 <- 
    list.files(delta_dir2, full.names = T)

df3 <- 
    list.files(delta_dir3, full.names = T)


df_tib <- 
    ref_qry_pairs %>%
    mutate(
        expected_basename = str_c(ref, "_v_", qry, "_filtered.delta")
        ) %>% 
    inner_join(
        ., 
        tibble(
            expected_basename = basename(df3),
            fullname = df3
            ), 
        by = "expected_basename"
    )

df_inner <- 
    df_tib$fullname
    
# rq_deltas <- 
#     lapply(
#         1:length(df2),
#         function(i){
#             read_delta(df2[i]) %>% pull(X_dist) %>% plot()
#         }
#     )


df2[3] %>% plot_delta

# chrom_SVs <- 
#     Csv[[3]]
Chrom_SVs <- 
    list()
local_SVs <- 
    list()
pb = txtProgressBar(min = 0, max = length(df_inner), initial = 1) 
for (i in 1:length(df_inner)){
    chr_sv <- 
        Chromosome_level_SVs(df_inner[i])
    
    ### Chromosome level SVs
    dt_i <- 
        read_delta(df_inner[i])
    
    #plot_delta(dt_i)
    
    ref_len <- 
        dt_i$rlen[1]
    qry_len <- 
        dt_i$qlen[1]
        
    Chrom_SVs[[i]] <-
        chr_sv
    #Chromosome_level_SVs(delta_files[i])
    LSV_i <- 
        delta_SVs(
            delta_table = 
                df_inner[i],
            chrom_SVs = 
                chr_sv
            ) %>%
        mutate(
            ref = 
                basename(df_inner[i]) %>% sub("_v_.*", "", .)
        )
    # local_SVs[[i]] <- 
    #     LSV_i
    
    unique(all_prokka$accession)
    ### GFF files
    ref_prokka <- 
        all_prokka %>%
        filter(
            accession == 
                unique(dt_i$rid)
        )
    qry_prokka <- 
        all_prokka %>%
        filter(
            accession == 
                unique(dt_i$qid)
        )
    
    ref_gr <- 
        GRanges(
            seqnames = ref_prokka$accession,
            ranges = IRanges(
                start = ref_prokka$start, 
                end = ref_prokka$end
                ),
            gene = ref_prokka$gene,
            product = ref_prokka$product
            )

    # Convert query Prokka to GRanges
    qry_gr <- 
        GRanges(
            seqnames = qry_prokka$accession,
            ranges = IRanges(start = qry_prokka$start, end = qry_prokka$end),
            gene = qry_prokka$gene,
            product = qry_prokka$product
            )
    
    ### Check ref granges for genic vs intergenic non insertions
    LSV_ov1 <- 
        LSV_i %>% 
        filter(
            variant != "Insertion"
            ) %>%
        rowwise() %>%
        mutate(
            start_genic = 
                {
                    bp <- 
                        GRanges(
                            seqnames = ref, 
                            ranges = IRanges(start = start, width = 1)
                            )
                    hits <- 
                        findOverlaps(bp, ref_gr)
                    if (length(hits) > 0) {
                        subjectHits(hits)[1]  # index in ref_gr
                        } else {
                            NA_integer_
                            }
                    },
            end_genic = 
                {
                    bp <- 
                        GRanges(
                            seqnames = ref, 
                            ranges = IRanges(start = end, width = 1)
                            )
                    hits <- 
                        findOverlaps(bp, ref_gr)
                    if (length(hits) > 0) {
                        subjectHits(hits)[1]  # index in ref_gr
                        } else {
                            NA_integer_
                            }
                },
            both_genic = 
                ((!is.na(start_genic)) & (!is.na(end_genic))),
            one_genic = 
                ((!is.na(start_genic)) | (!is.na(end_genic))) & (!both_genic),
            neither_genic = 
                ((is.na(start_genic)) & (is.na(end_genic))),
            start_hit_prod = 
                ifelse(
                    (!is.na(start_genic)),
                    ref_prokka$product[start_genic],
                    NA_character_
                ),
            end_hit_prod = 
                ifelse(
                    (!is.na(end_genic)),
                    ref_prokka$product[end_genic],
                    NA_character_
                )
            ) %>%
        ungroup()
    
    ## qry gff ranges
    
    LSV_ov2 <- 
        LSV_i %>% 
        filter(
            variant == "Insertion"
            ) %>%
        rowwise() %>%
        mutate(
            start_genic = 
                {
                    bp <- 
                        GRanges(
                            seqnames = acc, 
                            ranges = IRanges(start = start, width = 1)
                            )
                    hits <- 
                        findOverlaps(bp, qry_gr)
                    if (length(hits) > 0) {
                        subjectHits(hits)[1]  # index in ref_gr
                        } else {
                            NA_integer_
                            }
                    },
            end_genic = 
                {
                    bp <- 
                        GRanges(
                            seqnames = acc, 
                            ranges = IRanges(start = end, width = 1)
                            )
                    hits <- 
                        findOverlaps(bp, qry_gr)
                    if (length(hits) > 0) {
                        subjectHits(hits)[1]  # index in ref_gr
                        } else {
                            NA_integer_
                            }
                },
            both_genic = 
                ((!is.na(start_genic)) & (!is.na(end_genic))),
            one_genic = 
                ((!is.na(start_genic)) | (!is.na(end_genic))) & (!both_genic),
            neither_genic = 
                ((is.na(start_genic)) & (is.na(end_genic))),
            start_hit_prod = 
                ifelse(
                    (!is.na(start_genic)),
                    qry_prokka$product[start_genic],
                    NA_character_
                ) %>% as.character(),
            end_hit_prod = 
                ifelse(
                    (!is.na(end_genic)),
                    qry_prokka$product[end_genic],
                    NA_character_
                ) %>% as.character()
            ) %>%
        ungroup() 
    
    LSV_full <- 
        bind_rows(
            LSV_ov1,
            LSV_ov2
            ) %>%
        mutate(
            ref_len = ref_len,
            qry_len = qry_len,
        )
    
    local_SVs[[i]] <- 
        LSV_full
    # LSV_ov1 %>%
    #     count(both_genic, one_genic)
    
    if (nrow(chr_sv)>0){
        
        chr_sv_ov <- 
            chr_sv %>%
            dplyr::select(-c(ms,me, rlen, qlen)) %>%
            rowwise %>%
            mutate(
                start_genic_ref = 
                    {
                        bp <- 
                            GRanges(
                                seqnames = 
                                    unique(ref_prokka$accession), 
                                ranges = 
                                    IRanges(start = rs, width = 1)
                                )
                        hits <- 
                            findOverlaps(bp, ref_gr)
                        if (length(hits) > 0) {
                            subjectHits(hits)[1]  # index in ref_gr
                            } else {
                                NA_integer_
                                }
                        },
                end_genic_ref = 
                    {
                        bp <- 
                            GRanges(
                                seqnames = 
                                    unique(ref_prokka$accession), 
                                ranges = 
                                    IRanges(start = re, width = 1)
                                )
                        hits <- 
                            findOverlaps(bp, ref_gr)
                        if (length(hits) > 0) {
                            subjectHits(hits)[1]  # index in ref_gr
                            } else {
                                NA_integer_
                                }
                    },
                start_genic_qry = 
                    {
                        bp <- 
                            GRanges(
                                seqnames = 
                                    unique(qry_prokka$accession), 
                                ranges = 
                                    IRanges(start = qe, width = 1)
                                )
                        hits <- 
                            findOverlaps(bp, qry_gr)
                        if (length(hits) > 0) {
                            subjectHits(hits)[1]  # index in ref_gr
                            } else {
                                NA_integer_
                                }
                        },
                end_genic_qry = 
                    {
                        bp <- 
                            GRanges(
                                seqnames = 
                                    unique(qry_prokka$accession), 
                                ranges = 
                                    IRanges(start = qs, width = 1)
                                )
                        hits <- 
                            findOverlaps(bp, qry_gr)
                        if (length(hits) > 0) {
                            subjectHits(hits)[1]  # index in ref_gr
                            } else {
                                NA_integer_
                                }
                    },
                ,
            both_genic_ref = 
                ((!is.na(start_genic_ref)) & (!is.na(end_genic_ref))),
            one_genic_ref = 
                ((!is.na(start_genic_ref)) | (!is.na(end_genic_ref))) & 
                (!both_genic_ref),
            neither_genic_ref = 
                ((is.na(start_genic_ref)) & (is.na(end_genic_ref))),
            both_genic_qry = 
                ((!is.na(start_genic_qry)) & (!is.na(end_genic_qry))),
            one_genic_qry = 
                ((!is.na(start_genic_qry)) | (!is.na(end_genic_qry))) & 
                (!both_genic_qry),
            neither_genic_qry = 
                ((is.na(start_genic_qry)) & (is.na(end_genic_qry))),
            ref_start_hit_prod = 
                ifelse(
                    (!is.na(start_genic_ref)),
                    ref_prokka$product[start_genic_ref],
                    NA
                ),
            ref_end_hit_prod = 
                ifelse(
                    (!is.na(start_genic_ref)),
                    ref_prokka$product[start_genic_ref],
                    NA
                ),
            qry_start_hit_prod = 
                ifelse(
                    (!is.na(start_genic_qry)),
                    qry_prokka$product[start_genic_qry],
                    NA
                    ),
            qry_end_hit_prod = 
                ifelse(
                    (!is.na(start_genic_qry)),
                    qry_prokka$product[start_genic_qry],
                    NA
                    )
            ) %>%
            ungroup
        
        Chrom_SVs[[i]] <- 
            chr_sv_ov
    }
    
    setTxtProgressBar(pb,i)
}

close(pb)



gene_regions_tib

## note insertions are coordinates of the query
c(LSVs$acc%>% unique,
LSVs$ref%>% unique) %>% length


all_prokka %>% filter(accession=="NZ_LR882500.1") %>%
    dplyr::slice(1207)


LSVs <- 
    bind_rows(local_SVs)# %>%



LSVs %>% count(start_hit_prod, end_hit_prod, sort = T)

CSVs <- 
    bind_rows(Chrom_SVs)



LSV_summary <- 
    LSVs %>%
    rowwise() %>%
    mutate(
        genic_n = 
            ifelse(
                both_genic,
                2,
                ifelse(
                    one_genic,
                    1, 
                    0
                    )
                ),
        midpoint = 
            round((end+start)/2),
        prods = paste(start_hit_prod, end_hit_prod, collapse = ",")
        ) %>%
    ungroup %>%
    dplyr::select(
        c(ref, acc, variant, width, midpoint, genic_n, prods)
    ) %>%
    mutate(
        level = "localized"
    ) %>%
    `colnames<-`(
        c(
            "ref", "qry", "variant", "length", 
            "midpoint", "genic_n", "prods", "level"
            )
        )
    

LSVs %>%
    filter(variant=="Insertion") %>%
    count(start_hit_prod, end_hit_prod, sort = T)


LSV_summary %>%
    group_by(variant) %>%
    summarise(
        genics = sum(genic_n),
        denom = 2*n()
    )



asv 

CSV_summary <- 
    CSVs %>%
    rowwise() %>%
    mutate(
        genic_n = 
             sum(
                 !is.na(
                     c_across(
                         c(
                             start_genic_ref, end_genic_ref, 
                             start_genic_qry, end_genic_qry
                             )
                         )
                     )
                 ),
        midpoint = 
            mean(
                c(
                    c(
                        (rs+re)/2
                        ),
                    c(
                        (qs+qe)/2
                        )
                    )
                ) %>% 
            round(),
        prods = 
            paste(
                ref_start_hit_prod, ref_end_hit_prod,
                qry_start_hit_prod, qry_end_hit_prod,
                collapse = ","
                )
        ) %>%
    ungroup %>%
    dplyr::select(
        c(rid, qid, meanlen, midpoint, genic_n, prods)
    ) %>%
    mutate(
        variant = "RASR",
        level = "chrom"
    ) %>%
    `colnames<-`(
        c(
            "ref", "qry", "length", "midpoint",
            "genic_n", "prods", "variant", "level"
            )
        )



    



    


all_svs <- 
    bind_rows(
        LSV_summary,
        CSV_summary
    )


all_svs %>% count(variant)


asv <- 
    all_svs %>%
    filter(length >=50) %>%
    left_join(
        ., 
        clustered_genomes %>%
            dplyr::select(c(cluster_id, accessions, len)) %>%
            `colnames<-`(
                c(
                    "ref_clust", "ref", "ref_len" 
                )
            ),
        by = "ref"
    ) %>%
    left_join(
        ., 
        clustered_genomes %>%
            dplyr::select(c(cluster_id, accessions, len)) %>%
            `colnames<-`(
                c(
                    "qry_clust", "qry", "qry_len" 
                )
            ),
        by = "qry"
    ) %>%
    mutate(
        bin = 
            cut(
                length, 
                breaks = 
                    c(
                        50, 500, 1000, 
                        2000, 10000, 50000, Inf
                        ), 
                right = FALSE)
        ) %>%
    rowwise() %>%
    mutate(
        genic_prop = 
            ifelse(
                variant=="RASR",
                round(genic_n/4, 2),
                round(genic_n/2, 2)
            )
    ) %>% 
    ungroup() %>%
    mutate(
        start = 
            midpoint - floor((length - 1) / 2),
        end = 
            midpoint + ceiling((length - 1) / 2)
    )

sv_breaks_gr <- 
    GRanges(
        seqnames = asv$ref,
        ranges   = 
            IRanges(
                start = asv$start,
                end   = asv$end
                )
        ) 

genomes <- 
    unique(asv$ref)


sv_gr_all <- 
    GRanges(
        seqnames = sv_df2$gr_seq,
        ranges   = IRanges(start = sv_df2$gr_start, end = sv_df2$gr_end),
        strand   = "*",
  event_id  = sv_df2$variant
  ) 

sv_gr_unique <- 
    sv_gr_all %>% reduce
sv_gr_all
sv_gr_unique

hits <- findOverlaps(sv_gr_unique, sv_gr_all)
mapping_df <- 
    data.frame(
  SV_type = sv_gr_unique$SV_type[queryHits(hits)],
  genome   = as.character(seqnames(sv_gr_all))[subjectHits(hits)],
  stringsAsFactors = FALSE
) %>% 
  distinct()  # one row per event × genome

# 3) Build a presence/absence matrix (events × genomes)
pres_mat <- mapping_df %>%
  mutate(present = 1) %>%
  pivot_wider(
    names_from  = genome,
    values_from = present,
    values_fill = 0
  ) %>%
  column_to_rownames("event_id") %>%
  as.matrix()

# 4) Ancestral‐state reconstruction on your SNP tree
#    (assumes you have parsnp_tree of class "phylo")
root_state <- apply(pres_mat, 1, function(ev_states) {
  fit <- ace(ev_states, snp_tree, type="discrete", method="ML")
  # compare likelihoods for state 0 vs 1 at the root
  ifelse(fit$lik.anc[1,"0"] >= fit$lik.anc[1,"1"], 0, 1)
})

# 5) For each event, the genomes where pres_mat[event,] != root_state[event]
#    are the ones that actually carry the *derived* SV.
derived_genomes <- apply(pres_mat, 1, function(ev_states, rs) {
  names(ev_states)[which(ev_states != rs)]
}, rs = root_state)

# 6) Build a final table
sv_table <- as.data.frame(sv_gr_unique)[,c("seqnames","start","end","event_id")] %>%
  mutate(
    ancestral = root_state[event_id],
    derived   = sapply(derived_genomes, paste, collapse=";")
  )

head(sv_table)


library(ape)



custom_round <- 
    function(x, resolution) {
        round(x / resolution) * resolution
        }

asv_col2 <- 
    asv %>%
    mutate(
        resolution = 
            case_when(
                length < 500          ~ 50,   
                length < 1000         ~ 100,  
                length < 2000         ~ 200,  
                length < 10000        ~ 1000, 
                length < 50000        ~ 5000, 
                TRUE                  ~ 1e5   
        ),
        start_group = custom_round(start, resolution),
        end_group   = custom_round(end, resolution)
        ) %>%
    group_by(variant, start_group, end_group) %>%
    summarise(
        n_collapsed = n(),                
        start       = min(start),         
        end         = max(end),
        width       = max(end) - min(start),
        gs          = paste(unique(c(ref, qry)), collapse = ","),  
        mid         = round((start + end)/2),
        .groups     = "drop",
        genic_prop = mean(genic_prop),
        prods = paste(prods, collapse = ","),
        bin = unique(bin),
        resolution = unique(resolution),
        meanlen = mean(length)
        ) %>%
    arrange(desc(n_collapsed))

asv2 <- 
    asv %>%
    mutate(
        resolution = 
            case_when(
                length < 500          ~ 50,   
                length < 1000         ~ 100,  
                length < 2000         ~ 200,  
                length < 10000        ~ 1000, 
                length < 50000        ~ 5000, 
                TRUE                  ~ 1e5   
        ),
        start_group = custom_round(start, resolution),
        end_group   = custom_round(end, resolution)
        ) %>%
    group_by(variant, start_group, end_group) %>%
    mutate(
        n_collapsed = n(),                
        start       = min(start),         
        end         = max(end),
        width       = max(end) - min(start),
        gs          = paste(unique(c(ref, qry)), collapse = ","),  
        mid         = round((start + end)/2),
        .groups     = "drop",
        genic_prop = mean(genic_prop),
        prods = paste(prods, collapse = ","),
        bin = unique(bin),
        resolution = unique(resolution),
        meanlen = mean(length)
        ) %>%
    arrange(desc(n_collapsed)) 



events <- 
    asv_col2 %>%
  # give each event an ID
  mutate(event_id = row_number()) %>%
  
  # turn gs into a list-column of unique genome names
  mutate(genomes = strsplit(gs, ",")) %>%
  mutate(genomes = lapply(genomes, unique)) %>%
  dplyr::select(event_id, genomes)

# 2) Build presence/absence matrix
pa <- events %>%
  unnest(genomes) %>%
  mutate(present = 1L) %>%
  pivot_wider(
    names_from  = genomes,
    values_from = present,
    values_fill = 0L
  ) %>%
  arrange(event_id) %>%
  column_to_rownames("event_id") %>%
  as.matrix()

library(phytools)

snp_tree_rooted <- midpoint_root(snp_tree)

snp_binary   <- multi2di(snp_tree_rooted)


snp_binary$tip.label
snp_binary$tip.label


snp_labels_clean <- 
    gsub("\\.txt$|\\.ref$|\\.fa$|\\.txt$", "", snp_binary$tip.label) %>%
    gsub(".txt", "", .)

snp_binary$tip.label <- 
    snp_labels_clean
colnames(pa)
snp_binary$tip.label

tips <- snp_binary$tip.label
pa2 <- pa[, intersect(colnames(pa), tips), drop = FALSE]

# 4) now reorder the columns so they match the exact tip order
pa2 <- pa2[, tips]

# sanity check
identical(colnames(pa2), tips)  # should be TRUE
ncol(pa2) == length(tips)       # should be TRUE

# 5) run ace() on each row of this well‑formed matrix
library(ape)
root_state <- apply(pa2, 1, function(ev_states) {
  fit <- ace(ev_states, snp_binary, type="discrete", method="ML")
  ifelse(fit$lik.anc[1,"0"] >= fit$lik.anc[1,"1"], 0L, 1L)
})



for(i in seq_len(n_evt)) {
  ev  <- pa2[i, ]        # 0/1 vector for event i
  anc <- root_state[i]   # scalar for event i
  derived_list[i] <- paste(
    names(ev)[ev != anc], 
    collapse = ";"
  )
}

# attach it back
asv_col2$derived <- derived_list


anc <- 
    asv_col2 %>%
  mutate(
    ancestral = root_state,
    derived   = sapply(derived, paste, collapse = ";")
  )

pa_ann <- pa_long %>%
  mutate(event_id = as.integer(sub("^evt", "", event))) %>%
  left_join(sv_table %>% select(event_id, ancestral), by = "event_id")

# 2) plot
p_tree <- ggtree(snp_binary) + geom_tiplab(size=3)

sv_table %>%
  mutate(type = if_else(ancestral==0, "gain", "loss")) %>%
  ggplot(aes(x=n_collapsed, fill=type)) +
    geom_histogram(binwidth=1, position="dodge") +
    labs(x="Number of Genomes Sharing the Event",
         y="Count of SV Loci",
         fill="Event type") +
    theme_minimal()

sv_counts <- 
    sv_table %>%
    filter(width>5e3) %>%
  mutate(event_type = if_else(ancestral == 0, "gain", "loss")) %>%
  count(variant, event_type) %>%
  pivot_wider(names_from = event_type, values_from = n, values_fill = 0)

sv_long <- sv_counts %>%
  pivot_longer(-variant, names_to="event_type", values_to="n")

# 3) Barplot: one bar per variant, split by gain vs loss
ggplot(sv_long, aes(x=variant, y=n, fill=event_type)) +
  geom_col(position = "dodge") +
  labs(x = "SV class", y = "Number of events",
       fill = "Event type",
       title = "Gains vs. Losses per Structural‑Variant Class") +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle=45, hjust=1))


sv_table %>%
  count(ancestral) %>%
  mutate(type = if_else(ancestral==0, "gain", "loss")) %>%
  ggplot(aes(x=type, y=n, fill=type)) +
    geom_col() +
    geom_text(aes(label=n), vjust=-0.5) +
    labs(x="", y="Number of SV loci",
         title="Ancestral vs. Derived SV Events") +
    theme_minimal()


sv_plot_df <- sv_table %>%
  # label each event as a Gain (ancestral=0) or Loss (ancestral=1)
  mutate(event_type = if_else(ancestral == 0, "Gain", "Loss")) %>%
  # ensure variant ordering
  mutate(variant = factor(variant, levels=c("Deletion","Insertion",
                                            "Ori inversion","local inversion",
                                            "RASR","Translocation")))

# Panel A: Gains vs Losses per SV class
sv_counts <- sv_plot_df %>%
  count(variant, event_type)

p1 <- ggplot(sv_counts, aes(x=variant, y=n, fill=event_type)) +
  geom_col(position="dodge") +
  scale_fill_manual(values=c(Gain="#4C9F70", Loss="#D95F02")) +
  labs(x="SV Class", y="Number of SV Loci",
       fill="", title="A) Gains vs. Losses by SV Class") +
  theme_minimal(base_size=14) +
  theme(axis.text.x = element_text(angle=45, hjust=1))

# Panel B: Recurrence histogram colored by Gain/Loss
p2 <- ggplot(sv_plot_df, aes(x=n_collapsed, fill=event_type)) +
  geom_histogram(binwidth=1, alpha=0.8, position="identity") +
  scale_fill_manual(values=c(Gain="#4C9F70", Loss="#D95F02")) +
  labs(x="Number of Genomes Sharing Event",
       y="Count of SV Loci",
       fill="", title="B) Recurrence of SVs (Gain vs. Loss)") +
  theme_minimal(base_size=14)

# Combine as a two‑panel figure
fig <- plot_grid(p1, p2, labels = c("", ""), ncol = 2, rel_widths = c(0.5,0.5))
print(fig)


pa2  # your matrix with rows = event_id, cols = snp_binary$tip.label
# convert to long for ggplot
colnames(pa2) <- make.unique(colnames(pa2))

# 2) Turn the rownames into a proper “event” column
pa2_df <- pa2 %>% 
  as.data.frame() %>%
  rownames_to_column(var = "event")

# 3) Pivot longer, keeping “event” and gathering all the genome columns
pa_long <- pa2_df %>%
  pivot_longer(
    cols        = -event, 
    names_to    = "genome",
    values_to   = "present"
  )



keep <- 
    which(n_derived >= 2)


drows <- dist(pa_filt, method="binary")
hrows <- hclust(drows, method="average")
row_order <- hrows$labels[hrows$order]

# reorder pa_filt
pa_filt <- pa_filt[row_order, , drop=FALSE]

# turn into long form for ggplot
pa_long <- as.data.frame(pa_filt) %>%
  rownames_to_column("event") %>%
  pivot_longer(-event, names_to="genome", values_to="present")

# ensure genome factor matches tree tip order
pa_long$genome <- factor(pa_long$genome, levels = snp_binary$tip.label)

# --- 2) Plot tree -------------------------------------------------------

p_tree <- ggtree(snp_binary) +
  geom_tiplab(size=3) +
  theme_tree2()

evt_names   <- rownames(pa_filt)               # e.g. "evt17", "evt42", …
evt_ids     <- as.integer(sub("^evt", "", evt_names))

# 2) subset your collapsed_ranges to those events
sv_top      <- collapsed_ranges[evt_ids, ]

# 3) optionally reorder so it matches the order in pa_filt
sv_top      <- sv_top %>% slice(match(evt_ids, row_number()))

# 4) inspect
sv_top %>% 
  select(variant, start, end, n_collapsed, gs, mid)


p_heat <- ggplot(pa_long, aes(x=genome, y=factor(event, levels=row_order), fill=factor(present))) +
  geom_tile() +
  scale_fill_manual(values = c("0"="white","1"="steelblue"), name="SV present") +
  theme_minimal(base_size=12) +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x = element_text(angle=90, vjust=0.5, size=8),
    panel.grid = element_blank()
  ) +
  labs(x=NULL, y="Event (clustered)")

# --- 4) Combine with cowplot --------------------------------------------

plot_grid(
  p_tree + theme(legend.position="none"),
  p_heat + theme(legend.position="right"),
  ncol = 2,
  rel_widths = c(0.3, 0.7)
)

## here2
install.packages("UpSetR")
library(UpSetR)

# get the top‑20 most shared events
top20 <- order(n_derived, decreasing=TRUE)[1:20]
m_top20 <- pa2[top20, , drop=FALSE]

# convert to list-of-genomes
top20 <- order(n_derived, decreasing=TRUE)[1:20]

# 2) Build a named list of the genomes carrying each of those events
list_top20 <- lapply(top20, function(i) {
  names(pa2)[ pa2[i, ] == 1 ]
})
names(list_top20) <- paste0("evt", top20)   # give each element a name

# 3) Convert to the binary data.frame
up_df <- fromList(list_top20)

# 4) Inspect the column names
print(colnames(up_df))
# [1] "evt1" "evt2" ... "evt20"  # these are the set names we need

# 5) Call upset(), specifying *those* columns
upset(
  up_df,
  sets     = colnames(up_df),
  order.by = "freq",
  keep.order = TRUE
)

ggplot(data.frame(n = n_derived), aes(x = n)) +
  geom_bar(fill = "steelblue") +
  scale_x_continuous(breaks = 1: max(n_derived)) +
  labs(
    x = "Number of Genomes Sharing an SV Event",
    y = "Count of SV Loci",
    title = "Recurrence Spectrum of Structural Variants"
  ) +
  theme_minimal()

p_tree <- ggtree(snp_binary) + geom_tiplab(size=3)

p_heat <- ggplot(pa_long, aes(x = genome, y = event, fill = factor(present))) +
  geom_tile() +
  scale_fill_manual(values = c("0"="white","1"="steelblue")) +
  theme_minimal() +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x = element_text(angle=90, vjust=0.5)
  ) +
  labs(fill = "SV present")


n_derived <- lengths(strsplit(sv_table$derived, ";"))
keep_evt  <- which(n_derived >= 2)
pa_filt   <- pa2[keep_evt, ]
rownames(pa_filt) <- paste0("evt", keep_evt)

sv_table <- sv_table %>% mutate(event = (row_number()))
pa_long  <- as.data.frame(pa2) %>%
  rownames_to_column("event") %>%
  pivot_longer(-event, names_to="genome", values_to="present") %>%
  left_join(sv_table, by="event")

ggplot(pa_long %>% filter(event_id %in% keep_evt),
       aes(x = genome, y = factor(event_id), fill = factor(present))) +
  geom_tile() +
  scale_fill_manual(values = c("0"="white","1"="steelblue")) +
  facet_wrap(~variant, scales="free_y", ncol=1) +
  theme_minimal() +
  theme(
    axis.text.y=element_blank(),
    axis.ticks.y=element_blank(),
    axis.text.x=element_text(angle=90, vjust=0.5)
  )
cowplot::plot_grid(p_tree, p_heat, ncol=2, rel_widths=c(0.3,0.7))

### HERE

xmfa_gr <- 
    jt_gr
sv_gr$block_id     <- NA_integer_
sv_gr$seq_in_xmfa  <- NA_character_

hits <- findOverlaps(sv_gr, xmfa_gr, ignore.strand=TRUE)

# now these should be non‐empty
stopifnot(length(hits) > 0)

q <- queryHits(hits)
s <- subjectHits(hits)

xmfa_gr$block_id      <- seq_along(xmfa_gr)
 sv_gr$seq_in_xmfa[q]   <- as.character(seqnames(xmfa_gr))[s]

unique(as.character(seqnames(xmfa_gr)))
unique(as.character(seqnames(sv_gr)))


sv_status <- 
    asv2 %>%
  mutate(status = if_else(n_collapsed == 1, "singleton", "recurrent")) %>%
  #count(variant, status) %>%
  group_by(variant, status, bin) %>%
  mutate(
      occurance = n(),
      prop = occurance / sum(occurance)
      ) %>%
    dplyr::select(occurance, prop)

sv_status
# Stacked bar plot
ggplot(sv_status, aes(x = bin, y = prop, fill = status)) +
  geom_col() +
  scale_y_continuous(labels = scales::percent_format()) +
  labs(title = "Singleton vs. Recurrent SVs by Variant Type",
       x = "", y = "Proportion of SVs",
       fill = "") +
  theme_minimal(base_size = 14) +
    facet_wrap(~variant)



asv_col2 %>%
    count(variant, n_collapsed)

sv_df <- 
    asv_col2 %>% filter(bin == bins[1])



permutation_test_bin <- function(sv_df, num_perm = 1000) {
  
  
  # Count the number of recurrent groups (n_collapsed > 1)
  obs_recurrent <- 
      sv_df %>%
    filter(n_collapsed > 1) %>%
    pull(n_collapsed) %>% sum
  
  # Prepare vector to store the permuted recurrent counts
  perm_counts <- numeric(num_perm)
  
  # Permutation loop
  for(i in 1:num_perm) {
    sv_perm <- 
        sv_df %>%
      mutate(
        perm_start = sample(start),            # Shuffle start positions
        perm_end   = perm_start + round(meanlen)
      ) %>%
      mutate(
        perm_start_group = custom_round(perm_start, resolution),
        perm_end_group   = custom_round(perm_end, resolution)
      )
    
    perm_grouped <- sv_perm %>%
      group_by(variant, perm_start_group, perm_end_group) %>%
      summarise(n_collapsed_perm = n(), .groups = "drop")
    
    perm_counts[i] <- perm_grouped %>%
      filter(n_collapsed_perm > 1) %>%
      nrow()
  }
  
  list(observed = obs_recurrent, perm_counts = perm_counts)
}


permutation_test_bin <- function(sv_data, num_perm = 1000) {
  # Compute the observed grouping from raw data in this bin
  obs_grouped <- 
      sv_data %>%
    mutate(
      start_group = custom_round(start, resolution),
      end_group   = custom_round(end, resolution)
    ) %>%
    group_by(variant, start_group, end_group) %>%
    summarise(n_collapsed = n(), .groups = "drop")
  
  obs_recurrent <- obs_grouped %>% filter(n_collapsed > 1) %>% nrow()
  
  perm_counts <- numeric(num_perm)
  
  for (i in 1:num_perm) {
    sv_perm <- 
        sv_data %>%
      mutate(perm_start = sample(start)) %>%  # randomize start positions
      mutate(perm_end = perm_start + length) %>%   # recalc end using fixed length
      mutate(
        perm_start_group = custom_round(perm_start, resolution),
        perm_end_group   = custom_round(perm_end, resolution)
      )
    
    perm_grouped <- sv_perm %>%
      group_by(variant, perm_start_group, perm_end_group) %>%
      summarise(n_collapsed_perm = n(), .groups = "drop")
    
    perm_counts[i] <- perm_grouped %>% 
      filter(n_collapsed_perm > 1) %>% 
      nrow()
  }
  
  list(observed = obs_recurrent, perm_counts = perm_counts)
}

asv <- asv %>%
  mutate(
    resolution = case_when(
      length < 500          ~ 50,
      length < 1000         ~ 100,
      length < 2000         ~ 200,
      length < 10000        ~ 1000,
      length < 50000        ~ 5000,
      TRUE                  ~ 1e5
    )
  )

set.seed(123)  # for reproducibility
num_perm <- 1000
bins <- unique(asv$bin)

# Initialize pooled counters.
pooled_observed <- 0
perm_matrix <- matrix(NA, nrow = num_perm, ncol = length(bins))
colnames(perm_matrix) <- bins

for (j in seq_along(bins)) {
  bin_data <- asv %>% filter(bin == bins[j])
  res <- permutation_test_bin(bin_data, num_perm = num_perm)
  pooled_observed <- pooled_observed + res$observed  # sum observed counts across bins
  perm_matrix[, j] <- res$perm_counts               # store vector for this bin
}

# Pool the permutation counts by summing over bins for each replicate.
pooled_perm_counts <- rowSums(perm_matrix, na.rm = TRUE)

# Compute an empirical p-value (the proportion of permuted replicates with count >= observed)
p_value <- mean(pooled_perm_counts >= pooled_observed)

cat("Pooled Observed Recurrent Groups:", pooled_observed, "\n")
cat("Mean of Pooled Permuted Recurrent Groups:", mean(pooled_perm_counts), "\n")
cat("Empirical P-value:", p_value, "\n")

# Visualization: histogram of pooled permutation counts with observed value marked
perm_results_df <- data.frame(perm_count = pooled_perm_counts)
ggplot(perm_results_df, aes(x = perm_count)) +
  geom_histogram(bins = 30, fill = "lightblue", color = "black") +
  geom_vline(xintercept = pooled_observed, color = "red", size = 1.2) +
  labs(title = "Pooled Permutation Test for Recurrent SV Groups",
       subtitle = paste("Empirical p-value =", round(p_value, 3)),
       x = "Pooled Number of Recurrent SV Groups (n_collapsed > 1)",
       y = "Frequency")
# Visualization: histogram of pooled permutation counts with observed value overlaid.
perm_results_df <- data.frame(perm_count = pooled_perm_counts)


ggplot(perm_results_df, aes(x = perm_count)) +
  geom_histogram(bins = 30, fill = "lightblue", color = "black") +
  geom_vline(xintercept = pooled_observed, color = "red", size = 1.2) +
  labs(title = "Pooled Permutation Test for Recurrent SV Groups Across Bins",
       subtitle = paste("Empirical p-value =", round(p_value, 3)),
       x = "Pooled Number of Recurrent SV Groups (n_collapsed > 1)",
       y = "Frequency")


set.seed(123)  # for reproducibility
num_perm <- 1000
bins <- unique(asv$bin)

# Create a data frame to store the results per bin.
bin_results <- lapply(bins, function(b){
  bin_data <- asv %>% filter(bin == b)
  res <- permutation_test_bin(bin_data, num_perm = num_perm)
  observed <- res$observed
  perm_mean <- mean(res$perm_counts)
  p_val <- mean(res$perm_counts >= observed)
  data.frame(bin = b, observed = observed, perm_mean = perm_mean, p_value = p_val)
})
bin_results_df <- bind_rows(bin_results)
print(bin_results_df)

# Visualization: Create individual histograms for each bin showing the null distribution
# with the observed value marked as a vertical red line.
bin_plots <- lapply(bins, function(b) {
  bin_data <- asv %>% filter(bin == b)
  res <- permutation_test_bin(bin_data, num_perm = num_perm)
  observed <- res$observed
  perm_df <- data.frame(perm_count = res$perm_counts)
  ggplot(perm_df, aes(x = perm_count)) +
    geom_histogram(bins = 30, fill = "lightblue", color = "black") +
    geom_vline(xintercept = observed, color = "red", size = 1.2) +
    labs(title = paste("SV Bin:", b),
         subtitle = paste("p =", round(mean(res$perm_counts >= observed), 3)),
         x = "Recurrent SV Group Count",
         y = "Frequency")
})
# Example: Print the plots for the first two bins
grid.arrange(bin_plots)
do.call(
     "grid.arrange",
     c(
         bin_plots,
         ncol=3
         )
     )

asv_col2 %>% 
  group_by(bin) %>% 
  summarise(total_collapsed = sum(n_collapsed))

asv_col2 %>% 
  count(bin, n_collapsed)

asv_col2 %>% group_by(bin) %>%
    summarise(
        nc = n
    )


asv_col2 %>%
    count(n_collapsed)
asv_col2 %>%
    filter(
        n_collapsed>2,
        genic_prop >0
        ) %>% pull(prods
                   )
asv_col2 %>% count(variant)

    asv_col2 %>% count(variant, n_collapsed, bin) %>% arrange(desc(n_collapsed))
    

saveRDS(
    all_svs,
    sprintf(
        "%s/all_svs.rds",
        r_vars_dir
    )
)

saveRDS(
    asv,
    sprintf(
        "%s/asv.rds",
        r_vars_dir
    )
)




#asv dist 
sv_df2 <- 
    asv%>%
  mutate(
    gr_seq   = if_else(variant == "Insertion", qry, ref),
    gr_start = start,
    gr_end   = end
  )


subset_len_tib <- 
    clustered_genomes %>%
    filter(
        accessions %in% unique(sv_df2$gr_seq)
    ) %>% 
    dplyr::select(accessions, len)


full_genome_gr <- 
    GRanges(
        seqnames = subset_len_tib$accessions,
        ranges   = IRanges(start = 1, end = subset_len_tib$len)
        )

sv_gr_all <- 
    GRanges(
        seqnames = sv_df2$gr_seq,
        ranges   = IRanges(start = sv_df2$gr_start, end = sv_df2$gr_end),
        strand   = "*",
  # carry over whatever metadata you like:
  SV_type  = sv_df2$variant# if you have an ID column
  # … add more metadata columns as neede
  ) 


sv_gr_unique <- 
    sv_gr_all %>% reduce


struct_core_gr <- 
    setdiff(full_genome_gr, sv_gr)


length(struct_core_gr) 

struct_core_df <- 
    as.data.frame(struct_core_gr) %>%
  as_tibble() %>%
  dplyr::rename(
    seqname = seqnames,
    start   = start,
    end     = end,
    width   = width
  )

# 2) Global summary of interval widths
struct_core_df %>%
  summarise(
    n_intervals = n(),
    total_bp    = sum(width),
    mean_bp     = mean(width),
    median_bp   = median(width),
    min_bp      = min(width),
    max_bp      = max(width)
  )

# 3) Quantiles of width
struct_core_df %>%
  pull(width) %>%
  quantile(probs = c(0, .1, .25, .5, .75, .9, 1))

# 4) Per‐chromosome (per‐seqname) summary
per_chr_stats <- struct_core_df %>%
  group_by(seqname) %>%
  summarise(
    n_intervals = n(),
    total_bp    = sum(width),
    mean_bp     = mean(width),
    median_bp   = median(width)
  ) %>%
  arrange(desc(total_bp))

print(per_chr_stats)

# 5) Histogram of interval widths (log scale) overall
ggplot(struct_core_df, aes(x = width)) +
  geom_histogram(bins = 50, fill = "steelblue", color = "white", alpha = 0.8) +
  scale_x_log10() +
  labs(title = "Distribution of Structural‐Core Interval Widths",
       x = "Interval width (bp, log10 scale)",
       y = "Count") +
  theme_minimal()

# 6) Faceted histogram by seqname
ggplot(struct_core_df, aes(x = width)) +
  geom_histogram(bins = 30, fill = "steelblue", color = "white", alpha = 0.7) +
  scale_x_log10() +
  facet_wrap(~ seqname, scales = "free_x") +
  labs(title = "Structural‐Core Interval Widths by Chromosome",
       x = "Width (bp, log10)",
       y = "Count") +
  theme_minimal()

# 7) Scatter of interval start vs. width
ggplot(struct_core_df, aes(x = start, y = width)) +
  geom_point(alpha = 0.5, size = 0.7) +
  facet_wrap(~ seqname, scales = "free") +
  labs(title = "Interval Start vs. Width by Chromosome",
       x = "Start position (bp)",
       y = "Width (bp)") +
  theme_minimal()



sv_df2$gr_seq %>% unique


full_genome_gr <- 
    GRanges(
        seqnames = ref_lengths$ref,
        ranges   = IRanges(start = 1, end = ref_lengths$ref_len)
        )

snp_gr <- 
    jt_gr

mcols(snp_gr)$block_id <- seq_along(snp_gr)

# 2) Make sure you have a “taxon” column (the sample / contig name)
#    If your seqnames are exactly the sample IDs, do:
mcols(snp_gr)$taxon    <- as.character(seqnames(snp_gr))
#    Otherwise, replace with whichever metadata column holds your genome ID.

# 3) Count how many distinct taxa each block appears in
n_taxa <- length(unique(mcols(snp_gr)$taxon))
block_counts <- table(mcols(snp_gr)$block_id)

# 4) Define “core” block IDs = those seen in every genome
core_block_ids <- as.integer(names(block_counts)[block_counts == n_taxa])

# 5) Subset your GRanges to just those core blocks
core_gr <- snp_gr[mcols(snp_gr)$block_id %in% core_block_ids]
    core_SVs <- subsetByOverlaps(sv_gr, core_gr, ignore.strand=TRUE)

    sv_gr$in_core <- FALSE

# find overlaps
hits <- findOverlaps(sv_gr, core_gr, ignore.strand=TRUE)

# mark those that overlap
sv_gr$in_core[queryHits(hits)] <- TRUE

# now you can do:
table(sv_gr$in_core)
    
    GRanges(
        seqnames = asv$ref, 
        ranges = 
            IRanges(
                start = asv$midpoint-round(asv$length/2), 
                end = asv$midpoint +round(asv$length/2)
                ),
        acc = asv$ref,
        variant = asv$variant
        )

sv_clusters <- 
    reduce(sv_gr, min.gapwidth = 1000) 

overlaps <- 
    findOverlaps(sv_gr, sv_clusters)
asv$cluster_id <- 
    subjectHits(overlaps)

binary_matrix <- 
    asv %>%
    distinct(ref, cluster_id) %>%        
    mutate(present = 1) %>%              
    pivot_wider(                         
        names_from = ref, 
        values_from = present, 
        values_fill = 0
        ) %>%
    column_to_rownames(var = "cluster_id")

binary_matrix_t <- 
    t(binary_matrix)

# Step 2: Jaccard distance calculation
jaccard_dist <- 
    vegdist(binary_matrix_t, method = "jaccard", binary = TRUE)



chr_sv_ov %>% 
            dplyr::select(c(rs, re, rid, start_genic_ref, end_genic_ref))



csv_breaks <- 
    bind_rows(
        CSVs %>% 
            dplyr::select(c(rs, re, rid, start_genic_ref, end_genic_ref)) %>%
            mutate(
                width = re-rs+1,
                variant = "RASR",
                s_genic = (!is.na(start_genic_ref)),
                e_genic = (!is.na(end_genic_ref))
            ) %>%
            dplyr::select(
                c(
                    rs, re, rid, width, variant, start_genic_ref, end_genic_ref
                    )
                ) %>%
            `colnames<-`(
                c(
                    "start", "end", "source",
                    "width", "variant", "s_genic", "e_genic" 
                )
            ),
        CSVs %>% 
            dplyr::select(c(qs, qe, qid, start_genic_qry, end_genic_qry)) %>%
            rowwise() %>%
            mutate(
                width = abs(qe-qs)+1,
                variant = "RASR",
                start = min(c(qs, qe)),
                end = max(c(qs, qe))
            ) %>%
            dplyr::select(c(start, end, qid, width, variant)) %>%
            `colnames<-`(
                c(
                    "start", "end", "source",
                    "width", "variant"
                )
            )
    )


sampled_asv <- 
    asv %>%
    group_by(ref) %>%
    filter(qry %in% sample(unique(qry), 3)) %>% 
    ungroup()


sampled_asv %>% count(variant, bin)


summary_table <- 
    asv %>%
    rowwise %>%
    mutate(
        comparison = paste(ref, qry, sep = "_v_")
    ) %>%
  group_by( bin, comparison) %>%
  summarise(
    count = n(),
    mean_length = mean(length),
    median_length = median(length),
    mean_midpoint = mean(midpoint),
    .groups = "drop"
  )


sampled_asv %>%
  count(variant, bin) %>%
  ggplot(aes(x = bin, y = n, fill = variant)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.8)) +
  labs(
    title = "Counts of Structural Variants by Size Bin",
    x = "Size Bin",
    y = "Count",
    fill = "Variant"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

length_summary_tables[[species_name]] <- 
    summary_table
length_tables_plot <- 
    ggplot(
        summary_table, aes(x = bin, y = count, fill = bin)) +
      #geom_boxplot(width = 0.4, size = 2) +
        geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
      scale_fill_brewer(palette = "Set1") +
      labs(
        x = "SV length (bp)",
        y = "SV Count",
        title = ""
      ) +
        theme_classic() +
        theme(
            # axis.text.x = 
            #     element_text(
            #         angle = 15, 
            #         hjust = 1
            #         ),
            text = element_text(size = 22),
            legend.position = "none",
            #axis.title.x = element_text(margin = margin(t = 15))
            ) +
        scale_x_discrete(
        labels = c(
          "[50,500)" = "50–500",
          "[500,1e+03)" = "500–1,000",
          "[1e+03,2e+03)" = "1,000–2,000",
          "[2e+03,1e+04)" = "2,000–10,000",
          "[1e+04,5e+04)" = "10,000–50,000",
          "[5e+04,Inf)" = ">50,000"
        )
      ) 
  
saveRDS(
    length_tables_plot,
    sprintf(
        "%s/length_tables_plot.rds",
        r_vars_dir
    )
)




breakpoint_locations <- 
    bind_rows(
        LSVs %>%
            mutate(
                source = 
                    ifelse(
                        variant=="Insertion",
                        unique(acc), unique(ref)
                    ),
                type = "local"
                ) %>%
            dplyr::select(
                start, end, width, variant, source
            ),
        csv_breaks
    ) %>%
    mutate(
        accessions = source
    ) %>% 
    left_join(
        ., 
        clustered_genomes %>% dplyr::select(accessions, len),
        by = "accessions"
    )

clustered_genomes


bpl_long <- 
    breakpoint_locations %>%
    pivot_longer(
        cols = c(start, end),
        names_to = "side",       
        values_to = "position"   
  ) %>% 
    filter(width>50)


ggplot(bpl_long, aes(x = position, fill = variant)) +
  geom_histogram(bins = 400, color = "black") +
  #facet_wrap(~ source, scales = "free_x") +
  # log scale if widths vary over several orders of magnitude
  labs(title = "Distribution of Breakpoint Widths by Genome Source",
       x = "Breakpoint position",
       y = "Count") +
    geom_vline(data = all_box_tibs, aes(xintercept = midpt), linetype = "dashed", color = "blue", alpha = 0.5)


ggplot(breakpoint_locations, aes(x = start, fill = variant)) +
  geom_histogram(bins = 30, color = "black") +
  facet_wrap(~ source, scales = "free_x") +
  scale_x_log10() +  # log scale if widths vary over several orders of magnitude
  labs(title = "Distribution of Breakpoint Widths by Genome Source",
       x = "Breakpoint Width (bp, log10 scale)",
       y = "Count")


ggplot(breakpoint_locations, aes(x = start, xend = end, y = source, yend = source)) +
  geom_segment(color = "purple", size = 1, alpha = 0.5) +
  labs(title = "SV Regions Across Genomes",
       x = "Genomic Position (bp)",
       y = "Genome Source")



asv

bp_sumry <- 
    breakpoint_locations %>%
    filter(
        width>1e4,
        #!(variant %in% c("Insertion", "Deletion"))
        )

collapsed_ranges <- 
    bp_sumry %>%
  # Assuming you're interested in a specific variant type, e.g., "Insertion"
  #filter(variant == "Insertion") %>%
  mutate(
    start_group = round(start, -3),  # round to nearest 1,000 bp
    end_group   = round(end, -3)
  ) %>%
  group_by(variant, start_group, end_group) %>%
  summarise(
    n_collapsed = n(),                # count of similar insertions
    start = min(start),               # or you can use median, or min
    end   = max(end),                 # or median or max, depending on your criteria
    width = max(end) - min(start),
    gs = paste(source, collapse = ","),
    mid = round((start+end)/2),
    .groups = "drop"
  ) %>% 
    arrange(desc(n_collapsed))# %>%
    # filter(
    #     (n_collapsed>1 ) | variant != "Insertion"
    #     )


asv %>% filter(variant=="Translocation")


cap_seqs <- 
    lapply(
        1:nrow(
            asv %>% 
                filter(
                    variant=="Translocation",
                    length<2e3, length>1e3
                    ) %>%
                dplyr::slice(1:10)
            ),
        function(i){
            l_i <- 
                (asv %>% filter(variant=="Translocation"))[i,]
            seq_out <- 
                sprintf("%s/%s.txt", sync_dir, l_i$ref) %>% 
                readDNAStringSet() %>%
                subseq(
                    l_i$midpoint-round((l_i$length)/2),
                    l_i$midpoint+round((l_i$length)/2)
                )
        }
    ) %>%
    do.call(c, .)


writeXStringSet(cap_seqs, filepath = "../data/processing/extra/mt_tlocs.txt")

sprintf("%s/%s.txt", sync_dir, "NZ_LR882500.1") %>% subseq(1076447)

collapsed_ranges %>% filter(variant=="Translocation")


ggplot(
    breakpoint_locations %>% filter(width>1e3),
    aes(x = start, y = width, color = variant)) +
  geom_point(alpha = 0.5) +
  facet_wrap(~ source, scales = "free_x") +
  labs(
    title = "Breakpoints and Their Widths by Variant Type",
    x = "Breakpoint Midpoint (bp)",
    y = "SV Width (bp)",
    color = "Variant Type"
  ) +
  theme_minimal(base_size = 14)


bind_rows(
    tibble(
        loc =
            c(
                bp_sumry %>% filter(width>1e4) %>% pull(start),
                bp_sumry %>% filter(width>1e4) %>% pull(end)
            ),
            # collapsed_ranges %>%
            # pull(mid), 
        type = "SV"
    ),
    tibble(
        loc = 
            all_box_tibs$midpt,
        type = "dnaA_box"
    )
) %>%
    ggplot(., aes(x = loc, fill = type)) +
  geom_density(
      alpha = 0.6,
       adjust = 0.01
      ) +
  labs(
    title = "Distribution of loc by Type",
    x = "Location",
    y = "Density",
    fill = "Type"
  ) +
  theme_minimal(base_size = 14)
    





ggplot(all_box_tibs, aes(x = midpt)) +
  geom_density(
      alpha = 0.6,
       adjust = 0.01
      ) +
  labs(
    title = "Distribution of loc by Type",
    x = "Location",
    y = "Density",
    fill = "Type"
  ) +
  theme_minimal(base_size = 14)



```



## small indel
```{r}


ASV %>% count(bin, variant)

observed_genic2 <- 
    ASV %>% 
    filter(
        bin==
            "[50,500)",
        variant=="Indel"
            ) %>%
    
    left_join(
        ., 
        clustered_genomes %>%
            dplyr::select(c(cluster_id, accessions, len)) %>%
            `colnames<-`(
                c(
                    "ref_clust", "ref", "ref_len" 
                )
            ),
        by = "ref"
    ) %>%
    left_join(
        ., 
        clustered_genomes %>%
            dplyr::select(c(cluster_id, accessions, len)) %>%
            `colnames<-`(
                c(
                    "qry_clust", "qry", "qry_len" 
                )
            ),
        by = "qry"
    )
    mutate(
        denom = 
            2
    ) %>%
    group_by(ref, qry, ref_len, qry_len) %>%
    summarise(
        observed = sum(genic_n),
        n_total = 
            sum(denom),
        .groups = "keep"
      )

gr_summary <-
    gene_regions_tib %>%
  group_by(region) %>%
  summarize(
    mean_prop = mean(proportion),
    se = sd(proportion) / sqrt(n()),
    .groups = 'drop'
  )



```




Intergenic enrichment
```{r}

observed_genic <- 
    asv %>%
    mutate(
        denom = 
            ifelse(
                variant== "RASR",
                4, 2
                )
    ) %>%
    group_by(ref, qry, ref_len, qry_len) %>%
    summarise(
        observed = sum(genic_n),
        n_total = 
            sum(denom),
        .groups = "keep"
      )

gr_summary <-
    gene_regions_tib %>%
  group_by(region) %>%
  summarize(
    mean_prop = mean(proportion),
    se = sd(proportion) / sqrt(n()),
    .groups = 'drop'
  )

## overall genic portions


# ggplot(gr_summary, aes(x = 1, y = mean_prop, fill = region)) +
#   geom_bar(stat = "identity", position = "fill") +
#   labs(x = "Genome", y = "Proportion of Genome", fill = "Region",
#        title = "Stacked Bar: Genic vs. Intergenic per Genome") +
#   theme_minimal() +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1))

gr_plot <- 
    ggplot(
        gr_summary,
        aes(x = region, y = mean_prop, fill = region)) +
    geom_bar(stat = "identity") +
    geom_errorbar(
        aes(ymin = mean_prop - se, ymax = mean_prop + se), 
        width = 0.2, color = "black"
        ) +
      scale_fill_manual(
        values = 
            c(
                "Genic" = "#4F84C4",
                "Intergenic" = "#E69F00"
                )
      ) +
      scale_y_continuous(
        labels = scales::percent_format(accuracy = 1),
        limits = c(0, 1),
        expand = expansion(mult = c(0, 0.05)),
        breaks = seq(0, 1, 0.1)
      ) +
      labs(
        title = 
            sprintf(
                "Complete genome composition"#, 
                #gsub("_", " ", species_name)
                ),
        x = NULL,
        y = "Percent of complete genome sequence"
      ) +
      theme_classic(base_size = 22) +
      theme(
        legend.position = "none",
        plot.title = element_text(face = "bold", size = 20, hjust = 0.5), 
        axis.text.x = element_text(size = 20),  # Adjust axis text size
        axis.text.y = element_text(size = 18),
        axis.title.y = element_text(size = 16),  # Slightly smaller y-axis title
        text = element_text(size = 20),
        panel.grid.major.y = element_line(color = "grey70", size = 0.5),  
        panel.grid.minor.y = element_blank(),
        axis.ticks = element_line(color = "black", linewidth = 0.5)  # Ticks on the axes
      )



#gr_plot
## simulated genic hits 
genic_props <- 
    gene_regions_tib %>%
    filter(region == "Genic") %>%
    dplyr::select(acc, genic_prop = proportion)



og2 <- 
  observed_genic %>%
  left_join(genic_props, by = c("ref" = "acc")) %>%
  dplyr::rename(ref_genic = genic_prop) %>%
  left_join(genic_props, by = c("qry" = "acc")) %>%
  dplyr::rename(qry_genic = genic_prop) %>%
  mutate(
    total_len = ref_len + qry_len,
    genic_fraction = (ref_genic * ref_len + qry_genic * qry_len) / total_len,
    observed_genic = observed,
    observed_intergenic = n_total - observed
  )


saveRDS(
    og2,
    sprintf(
        "%s/og2.rds",
        r_vars_dir
    )
)


set.seed(1234)
n_sim <- 1000

sim_summary <- 
  og2 %>%
  rowwise() %>%
  mutate(
    simulated_genic = list(rbinom(n_sim, size = n_total, prob = genic_fraction)),
    simulated_intergenic = list(n_total - unlist(simulated_genic))
  ) %>%
  ungroup()


sim_long <- 
  sim_summary %>%
  select(ref, qry, simulated_genic, simulated_intergenic, observed_genic, observed_intergenic) %>%
  mutate(pair_id = paste(ref, qry, sep = "_v_")) %>%
  relocate(pair_id) %>%
  unnest(cols = c(simulated_genic, simulated_intergenic))


proportion_summary <- 
  sim_summary %>%
  mutate(
    genic_expected = map_dbl(simulated_genic, mean),
    intergenic_expected = map_dbl(simulated_intergenic, mean),
    
    observed_genic_prop = observed_genic / n_total,
    observed_intergenic_prop = observed_intergenic / n_total,
    
    expected_genic_prop = genic_expected / n_total,
    expected_intergenic_prop = intergenic_expected / n_total
  ) %>%
  select(ref, qry, observed_genic_prop, expected_genic_prop,
         observed_intergenic_prop, expected_intergenic_prop) %>%
  mutate(pair_id = paste(ref, qry, sep = "_v_"))

proportion_long <- 
  proportion_summary %>%
  pivot_longer(
    cols = c(observed_genic_prop, expected_genic_prop,
             observed_intergenic_prop, expected_intergenic_prop),
    names_to = c("type", "region"),
    names_pattern = "(observed|expected)_(genic|intergenic)_prop",
    values_to = "proportion"
  ) %>%
  mutate(
    region = recode(region, genic = "Genic", intergenic = "Intergenic"),
    type = recode(type, observed = "Observed", expected = "Expected")
  )


proportion_diff <- 
  proportion_long %>%
  pivot_wider(
    names_from = type,
    values_from = proportion
  ) %>%
  mutate(
    diff = Observed - Expected
  )

saveRDS(
    proportion_diff,
    sprintf(
        "%s/proportion_diff.rds",
        r_vars_dir
    )
)

enrichment_plot <- 
    ggplot(
        proportion_diff, 
        aes(x = diff, fill = factor(region, levels = c("Genic", "Intergenic")))
        ) +
      geom_histogram(
        binwidth = 0.025,
        alpha = 0.7,
        color = "black",
        position = "identity"
      ) +
      geom_vline(
        xintercept = 0,
        linetype = "dashed",
        color = "grey30"
      ) +
      scale_fill_manual(
        values = 
            c(
                "Intergenic" = "#E69F00",
                "Genic" = "#4F84C4"
                )
      ) +
      labs(
        title = "Structural variant breakpoint locus distribution",
        #subtitle = "Positive = Enrichment, Negative = Depletion",
        x = "Observed - expected SV loci overlap",
        y = "Genome pair count",
        fill = NULL
      ) +
    annotate("text", x = -0.45, y = Inf, label = "← Depletion", hjust = 0, vjust = 1.5, size = 5) +
  annotate("text", x = 0.45, y = Inf, label = "Enrichment →", hjust = 1, vjust = 1.5, size = 5) +
      theme_classic(base_size = 18) +
      theme(
        legend.position = "top",
        plot.title = element_text(face = "bold", size = 18),
        
      ) +
        scale_y_continuous(
      expand = expansion(mult = c(0, 0.05))
    )

combined_plot <- 
    arrangeGrob(
        gr_plot,
        enrichment_plot,
        ncol = 2
        )

grid::grid.draw(combined_plot)
# Save as .rds file
saveRDS(
    combined_plot, 
    file = 
        sprintf(
            "%s/combined_breakpoint_plot.rds",
            r_vars_dir
            )
)





```





SV patterns and genes
```{r}

list.files(r_vars_dir)
SV_files <- 
    list.files(
      path = "../data/processing/r_vars",        # root directory
      pattern = "asv\\.rds$",       # regex pattern
      recursive = TRUE,                          # search subdirectories
      full.names = TRUE                          # return full paths
    )


SV_tibs <- 
    lapply(
        1:length(SV_files),
        function(i){
            out_tib <- 
                readRDS(
                    SV_files[i]
                    ) %>%
                mutate(
                    species = 
                        SV_files[i] %>%
                        dirname() %>% 
                        basename()
                )
        }
    ) %>%
    bind_rows() %>%
    mutate(
        mid_norm = 
            ifelse(
                variant !="Insertion",
                midpoint/ref_len,
                midpoint/qry_len
            )
            )



SV_unique <- 
    SV_tibs %>%
    group_by(species) %>%
    summarise(
        unique_ref = unique(ref),
        unique_qry = unique(qry)[1:length(unique_ref)]
    )


SV_tibs_i <- 
    SV_tibs %>%
    filter(
        species=="Bordetella_pertussis",
        ref %in% SV_unique$unique_ref,
        qry %in% SV_unique$unique_qry
    ) %>%
    mutate(
        start=midpoint- length/2
    )




SV_tibs_i %>%
    mutate


SV_tibs %>%
    filter(species=="Bordetella_pertussis")


asv2 %>% filter(variant=="local inversion") %>% 
    filter(genic_prop==1)


SV_tibs %>%
    filter(variant=="local inversion") %>%
    filter(grichic_prop==1)



ggplot(
    SV_tibs %>%
        filter(variant =="Deletion", species=="Bordetella_pertussis"), 
    aes(x = midpoint, fill = variant)
    ) +
  geom_density(alpha = 0.5, adjust = 0.1) +
    #geom_histogram(fill = "#1f78b4", bins = 60, color = "black") +
  #facet_wrap(~ variant + species, scales = "free_y") +
  labs(x = "Normalized Genome Position", y = "SV Density", title = "") +
  theme_classic() +
    theme(
        text = element_text(size = 22),
        legend.position = "none"
    )


ggplot(
    SV_tibs, 
        #filter(variant=="local inversion"
               #), 
    aes(x = mid_norm, fill = variant)
    ) +
  geom_density(alpha = 0.5, adjust = 0.1) +
    #geom_histogram(fill = "#1f78b4", bins = 60, color = "black") +
  facet_wrap(~ variant + species, scales = "free_y") +
  labs(x = "Normalized Genome Position", y = "SV Density", title = "") +
  theme_classic() +
    theme(
        text = element_text(size = 22),
        legend.position = "none"
    )



bp_tib <- 
    SV_tibs %>%
    filter(
        width>1e4,
        #!(variant %in% c("Insertion", "Deletion"))
        )

collapsed_ranges <- 
    bp_sumry %>%
  # Assuming you're interested in a specific variant type, e.g., "Insertion"
  #filter(variant == "Insertion") %>%
  mutate(
    start_group = round(start, -3),  # round to nearest 1,000 bp
    end_group   = round(end, -3)
  ) %>%
  group_by(variant, start_group, end_group) %>%
  summarise(
    n_collapsed = n(),                # count of similar insertions
    start = min(start),               # or you can use median, or min
    end   = max(end),                 # or median or max, depending on your criteria
    width = max(end) - min(start),
    gs = paste(source, collapse = ","),
    mid = round((start+end)/2),
    .groups = "drop"
  ) %>% 
    arrange(desc(n_collapsed))



```



Interspecies gathering
```{r}
### SV_lengths

readRDS(
    ""
)

saveRDS(
    asv,
    sprintf(
        "%s/asv.rds",
        r_vars_dir
    )
)



csv_breaks <- 
    bind_rows(
        CSVs %>% 
            dplyr::select(c(rs, re, rid)) %>%
            mutate(
                width = re-rs+1,
                variant = "RASR"
            ) %>%
            `colnames<-`(
                c(
                    "start", "end", "source",
                    "width", "variant"
                )
            ),
        CSVs %>% 
            dplyr::select(c(qs, qe, qid)) %>%
            rowwise() %>%
            mutate(
                width = abs(qe-qs)+1,
                variant = "RASR",
                start = min(c(qs, qe)),
                end = max(c(qs, qe))
            ) %>%
            dplyr::select(c(start, end, qid, width, variant)) %>%
            `colnames<-`(
                c(
                    "start", "end", "source",
                    "width", "variant"
                )
            )
    )


sampled_asv <- 
    asv %>%
    group_by(ref) %>%
    filter(qry %in% sample(unique(qry), 3)) %>% 
    ungroup()


sampled_asv %>% count(variant, bin)


summary_table <- 
    asv %>%
    rowwise %>%
    mutate(
        comparison = paste(ref, qry, sep = "_v_")
    ) %>%
  group_by( bin, comparison) %>%
  summarise(
    count = n(),
    mean_length = mean(length),
    median_length = median(length),
    mean_midpoint = mean(midpoint),
    .groups = "drop"
  )


length_summary_tables[[species_name]]
length_summary
saveRDS(
    length_tables_plot,
    sprintf(
        "%s/length_tables_plot.rds",
        r_vars_dir
    )
)

### gene regions

gr_files <- 
    list.files(
      path = "../data/processing/r_vars",        # root directory
      pattern = "gene_regions_tib\\.rds$",       # regex pattern
      recursive = TRUE,                          # search subdirectories
      full.names = TRUE                          # return full paths
    )


gr_tibs <- 
    lapply(
        1:length(gr_files),
        function(i){
            out_tib <- 
                readRDS(
                    gr_files[i]
                    ) %>%
                mutate(
                    species = 
                        gr_files[i] %>%
                        dirname() %>% 
                        basename()
                )
        }
    ) %>%
    bind_rows()


spec_names <- 
    tibble(
        species = c(top_species$Species_[1:5]),
        s2 = 
            sapply(
                species, 
                function(x) {
                    parts <- strsplit(x, "_")[[1]]
                    genus <- parts[1]
                    species <- parts[2]
                    genus_abbrev <- paste0(substr(genus, 1, 1), ".")
                    # Capitalize species
                    paste(genus_abbrev, species)
                    }
                )
        )
    unique(gr_tibs$species) %>%
    c(top_species$Species_[1:5]
gr_tibs 

gr_summaries <- 
    gr_tibs %>%
    group_by(species, region) %>%
    summarise(
        mean_prop = mean(proportion),
        sd_prop   = sd(proportion),
        n         = n(),
        se_prop   = sd_prop / sqrt(n),  # standard error
        .groups   = "drop"
    ) %>% 
    mutate(
        region = factor(region, levels = c( "Intergenic", "Genic")),
        # species = factor(species, levels = c(top_species$Species_[1:5])),
        # species = 
    ) %>%
    left_join(
        ., 
        spec_names, by = "species"
    )

## 2 plor options
gr_summaries

gr_stacked <- 
    ggplot(gr_summaries, aes(x = s2, y = mean_prop, fill = region)) +
  geom_col(
    position = "stack",
    colour = "black",  # Outline color
    size = 0.3,         # Outline thickness,
    alpha = 0.7
  ) +
  scale_fill_manual(
    # 3) Remove the legend title by setting name = NULL.
    #    Also control the order by specifying breaks in the desired sequence.
    name = NULL,
    breaks = c("Genic", "Intergenic"),
    values = c(
      "Genic" = "#4F84C4",
      "Intergenic" = "#E69F00"
    )
  )+
  scale_y_continuous(
    labels = scales::percent_format(accuracy = 1),
    limits = c(0, 1),
    expand = expansion(mult = c(0, 0.05)),
    breaks = seq(0, 1, 0.1)
  ) +
  labs(
    title = NULL,# "Sequence composition",
    x = NULL,
    y = "Sequence composition"
  ) +
  theme_classic(base_size = 22) +
  theme(
    legend.position = "none",
    plot.title = element_text(face = "bold", size = 20, hjust = 0.5),
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 18),
    axis.title.y = element_text(size = 16),
    text = element_text(size = 20),
    panel.grid.major.x = element_line(color = "grey85", size = 0.5),
    panel.grid.minor.x = element_blank(),
    axis.ticks = element_line(color = "black", linewidth = 0.5)
  )+
  coord_flip()


gr_boxplot_df <- 
    gr_tibs %>%
    mutate(
        region = 
            factor(
                region,
                levels = c("Genic", "Intergenic")  
                )
        )

gr_boxplot <- 
    ggplot(gr_tibs, aes(x = region, y = proportion, fill = region)) +
  geom_boxplot(
    colour = "black", # outline of the box
    alpha = 0.7
  ) +
  scale_fill_manual(
    name = NULL,  # Removes legend title
    values = c(
      "Genic" = "#4F84C4",
      "Intergenic" = "#E69F00"
    )
  ) +
  scale_y_continuous(
    labels = scales::percent_format(accuracy = 1),
    limits = c(0, 1),
    expand = expansion(mult = c(0, 0.05)),
    breaks = seq(0, 1, 0.1)
  ) +
  labs(
    title = "Genome composition",
    x = NULL,
    y = "Proportion of complete genome sequence"
  ) +
  theme_classic(base_size = 22) +
  theme(
    legend.position = "none",
    plot.title = element_text(face = "bold", size = 20, hjust = 0.5),
    axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
    axis.text.y = element_text(size = 18),
    axis.title.y = element_text(size = 16),
    text = element_text(size = 20),
    panel.grid.major.y = element_line(color = "grey85", size = 0.5),
    panel.grid.minor.y = element_blank(),
    axis.ticks = element_line(color = "black", linewidth = 0.5)
  )

### enrichment simulations


list.files(r_vars_dir)
prop_files <- 
    list.files(
      path = "../data/processing/r_vars",        # root directory
      pattern = "proportion_diff\\.rds$",       # regex pattern
      recursive = TRUE,                          # search subdirectories
      full.names = TRUE                          # return full paths
    )


prop_tibs <- 
    lapply(
        1:length(prop_files),
        function(i){
            out_tib <- 
                readRDS(
                    prop_files[i]
                    ) %>%
                mutate(
                    species = 
                        prop_files[i] %>%
                        dirname() %>% 
                        basename()
                )
        }
    ) %>%
    bind_rows() %>%
    mutate(diff2 = pow_expand_trans(diff))

pow_expand_trans <- function(alpha = 0.5) {
  trans_new(
    name = paste0("pow_expand_", alpha),
    transform = function(x) sign(x) * abs(x)^alpha,
    inverse   = function(y) sign(y) * abs(y)^(1/alpha)
  )
}

prop_tibs %>%
    mutate(
        abdiff = abs(diff)
    ) %>%
    arrange(abdiff)
    filter(
        # Observed==0 | (Expected==0)
        diff==0
    )


enrichment_plot <- 
    ggplot(
        prop_tibs, 
        aes(x = diff, fill = factor(region, levels = c("Genic", "Intergenic")))
        ) +
      geom_histogram(
           #binwidth = 0.1,
        binwidth = 0.02,
        alpha = 0.7,
        color = "black",
        position = "identity",#"dodge"
      ) +
      geom_vline(
        xintercept = 0,
        linetype = "dashed",
        color = "grey30"
      ) +
      scale_fill_manual(
        values = 
            c(
                "Intergenic" = "#E69F00",
                "Genic" = "#4F84C4"
                )
      ) +
      labs(
        title = "Small (>500bp) indel region enrichment",
        #subtitle = "Positive = Enrichment, Negative = Depletion",
        x = "Observed - expected loci overlap",
        y = "", #"Genome pair count",
        fill = NULL
      ) +
    scale_x_continuous(
        trans = pow_expand_trans(alpha = 0.5)#,
        #limits = c(-1, 1)
        )+
    coord_cartesian(xlim = c(-1, 1)) +
    annotate(
        "text", x = -0.2, y = Inf, 
        label = "← Depletion", hjust = 0, vjust = 1.5, size = 5
        ) +
    annotate(
        "text", x = 0.2, y = Inf, label = "Enrichment →", 
        hjust = 1, vjust = 1.5, size = 5
        ) +
      theme_classic(base_size = 18) +
      theme(
          plot.title = element_text(face = "bold", size = 20, hjust = 0.5),
        legend.position = "top",
      ) +
        scale_y_continuous(
      expand = expansion(mult = c(0, 0.05))
    )
enrichment_plot

final_combined_plot <- 
    arrangeGrob(
        enrichment_plot,
        gr_stacked +theme(legend.position = "none"),

        ncol = 1
        )




grid::grid.draw(final_combined_plot)

g1 <- ggplotGrob(enrichment_plot)
g2 <- ggplotGrob(gr_stacked)


g1$layout
g2$layout
# 2) Align widths so both y-axes line up
#    We pick columns [2:5] because that usually spans the plot panel + axis area
maxWidth <- grid::unit.pmax(g1$widths[5:7], g2$widths[5:7])
g1$widths[5:7] <- maxWidth
g2$widths[5:7] <- maxWidth


# 3) Arrange them vertically
final_combined_plot <- arrangeGrob(g1, g2, ncol = 1)

# Finally display or save
grid::grid.newpage()
grid::grid.draw(final_combined_plot)



# Save as .rds file
saveRDS(
    combined_plot, 
    file = 
        sprintf(
            "%s/combined_breakpoint_plot.rds",
            r_vars_dir
            )
)




gr_plot


gr_summary <-
    gene_regions_tib %>%
  group_by(region) %>%
  summarize(
    mean_prop = mean(proportion),
    se = sd(proportion) / sqrt(n()),
    .groups = 'drop'
  )


### intergenic enrichment


```
























```{r}


all_svs %>% filter(length==50)


    
    # group_by(bin) %>%
    # summarise(
    #     SV_count = n(),
    #     SV_types = 
    #         paste(unique(variant), collapse = ", "),
    #     .groups = "keep"
    #     )


asv %>% filter(variant=="Translocation")
asv %>% count(variant)

asv %>%
  group_by(variant, bin) %>%
  summarise(mean_genic_prop = mean(genic_prop, na.rm = TRUE)) %>%
  ggplot(aes(x = bin, y = variant, fill = mean_genic_prop)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "#f7f7f7", high = "#2166ac", name = "Genic Proportion") +
  labs(
      x = "SV Size Bin", y = "Variant Type", 
      title = "Mean Genic Proportion by SV Type and Size"
      ) +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))



asv %>% filter(variant=="RASR", length<5e4)
asv %>% filter(variant=="RASR", length>5e4)


asv %>%
  group_by(variant, bin) %>%
  mutate(mean_genic_prop = mean(genic_prop, na.rm = TRUE)) %>% 
    filter(variant=="Insertion") %>%
    filter(genic_prop==1) %>%
    ggplot(., aes(x = midpoint)) +
    geom_density(fill = "#1f78b4", alpha = 0.6, adjust = 0.5) +
  scale_x_log10() +
  labs(
    title = "Density of Insertion Lengths",
    x = "Length (log scale)",
    y = "Density"
  ) +
  theme_classic()


asv %>%
    group_by(variant, bin) %>%
  mutate(mean_genic_prop = mean(genic_prop, na.rm = TRUE)) %>% 
    filter(variant=="Insertion") %>%
    filter(genic_prop==1) %>%
    ungroup %>%
    count(ref, qry, sort = T)

asv %>%
  group_by(variant, bin) %>%
  summarise(mean_genic_prop = mean(genic_prop, na.rm = TRUE)) %>%
  ggplot(aes(x = bin, y = variant, fill = mean_genic_prop)) +
  geom_tile(color = "white", linewidth = 0.5) +
  scale_fill_viridis(
    name = "Genic BP",
    option = "D",       # "D" is a balanced palette
    na.value = "grey30",# Missing values in dark grey
    direction = -1,     # Optional: reverse direction if you prefer light = low
    limits = c(0, 1),   # Fix limits so 0 maps to a known color
    breaks = seq(0, 1, 0.2),
    labels = scales::percent_format(accuracy = 1)
  ) +
  labs(
    x = "SV Size Bin", y = "",
    title = ""
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.grid = element_blank(),
    text = element_text(size = 20)
  )



asv2 <- 
    asv %>%
    mutate(mid_norm = midpoint / ref_len)


ggplot(
    asv2, 
        #filter(variant=="local inversion"
               #), 
    aes(x = mid_norm, fill = variant)
    ) +
  geom_density(alpha = 0.5, adjust = 0.5) +
    #geom_histogram(fill = "#1f78b4", bins = 60, color = "black") +
  facet_wrap(~ variant, scales = "free_y") +
  labs(x = "Normalized Genome Position", y = "SV Density", title = "") +
  theme_classic() +
    theme(
        text = element_text(size = 22),
        legend.position = "none"
    )


ggplot(asv2, aes(x = variant, y = length)) +
  geom_violin(fill = "#cce5ff", color = "black", alpha = 0.7) +
  geom_boxplot(width = 0.2, outlier.size = 0.5) +
  scale_y_log10() +  # SV sizes tend to be skewed
  labs(y = "SV Length (bp)", x = NULL, title = "") +
  theme_classic() +
theme(text = element_text(size = 20))


len_tib <- 
    lapply(
        sv_info_list,
        function(x){
            x[["SV_len_tib"]]
        }
    ) %>%
    bind_rows %>%
    ungroup %>%
    mutate(bin = factor(bin)) %>%
    group_by(acc) %>%
    mutate(ordering = SV_count[bin == "[50,500)"]) %>%
    ungroup() %>%
    mutate(acc = fct_reorder(acc, ordering, .fun = mean, na.rm = TRUE))

mean_svs <- 
    len_tib %>% 
    group_by(acc) %>% 
    summarise(
        avg_svs = 
            mean(sum(SV_count))
    )



ggplot(
    asv,# %>% 
        #filter(!(bin %in% c("[1e+04,5e+04)",   "[5e+04,Inf)"))), 
    aes(x = bin, y = SV_count, fill = bin)) +
  #geom_boxplot(width = 0.4, size = 2) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        # axis.text.x = 
        #     element_text(
        #         angle = 15, 
        #         hjust = 1
        #         ),
        text = element_text(size = 22),
        legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,5e+04)" = "10,000–50,000",
      "[5e+04,Inf)" = ">50,000"
    )
  ) 


```







```{r}

ggplot(
  proportion_long %>% filter(type == "Observed"),
  aes(x = proportion, fill = region)
) +
  geom_histogram(
    aes(y = after_stat(density)),
    position = "identity",
    binwidth = 0.02,
    alpha = 0.6,
    color = "black"
  ) +
  geom_density(
    alpha = 0.3,
    adjust = 1.2,
    linewidth = 1
  ) +
  scale_fill_manual(
    values = c(
      "Genic" = "#4F84C4",
      "Intergenic" = "#E69F00"
    )
  ) +
  labs(
    title = "Breakpoints Are Enriched in Intergenic Regions",
    x = "Proportion of Breakpoints",
    y = "Density",
    fill = NULL
  ) +
  theme_classic(base_size = 15) +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 14),
    axis.text = element_text(size = 13),
    axis.title = element_text(size = 14),
    plot.title = element_text(face = "bold", size = 18),
  )






enrichment_long <- 
  sim_summary %>%
  mutate(
    genic_expected = map_dbl(simulated_genic, mean),
    intergenic_expected = map_dbl(simulated_intergenic, mean),
    genic_ratio = observed_genic / genic_expected,
    intergenic_ratio = observed_intergenic / intergenic_expected,
    intergenic_ratio2 = observed_intergenic - intergenic_expected
  ) %>%
  dplyr::select(ref, qry, genic_ratio, intergenic_ratio) %>%
  pivot_longer(
    cols = c(genic_ratio, intergenic_ratio),
    names_to = "type",
    values_to = "ratio"
  ) %>%
  mutate(
    type = recode(
      type,
      genic_ratio = "Genic",
      intergenic_ratio = "Intergenic"
    )
  )


ggplot(enrichment_long, aes(x = ratio, fill = type)) +
  geom_histogram(
    position = "identity",
    binwidth = 0.1,
    alpha = 0.7,
    color = "black"
  ) +
  geom_vline(
    xintercept = 1,
    linetype = "dashed",
    color = "grey20",
    linewidth = 0.8
  ) +
  scale_fill_manual(
    values = 
        c(
            "Genic" = "#4F84C4",
            "Intergenic" = "#E69F00"
            )
  ) +
  # scale_x_continuous(
  #   limits = c(0, 10),
  #   breaks = seq(0, 10, 1),
  #   expand = expansion(mult = c(0, 0.05))
  # ) +
  labs(
    title = "Breakpoint Enrichment in Genic vs Intergenic Regions",
    x = "Observed / Expected",
    y = "Number of Genome Pairs",
    fill = NULL
  ) +
  theme_minimal(base_size = 16) +
  theme(
  legend.position = "top",
  legend.text = element_text(size = 14),
  axis.text = element_text(size = 13),
  plot.title = element_text(face = "bold", size = 18)#,
  #plot.subtitle = element_text(size = 14, margin = margin(b = 10, unit = "pt"))
)


og2 <- 
    observed_genic %>%
    left_join(genic_props, by = c("ref" = "acc")) %>%
    dplyr::rename(ref_genic = genic_prop) %>%
    left_join(genic_props, by = c("qry" = "acc")) %>%
    dplyr::rename(qry_genic = genic_prop) %>%
    mutate(
        total_len = ref_len + qry_len,
        genic_fraction = (ref_genic * ref_len + qry_genic * qry_len) / total_len
        )














## binomial sims
set.seed(1234)
n_sim <- 1000

sim_summary <- 
    og2 %>%
      rowwise() %>%
      mutate(
          sim_hits = list(rbinom(n_sim, size = n_total, prob = genic_fraction))
          ) %>%
      ungroup()

sim_long <- 
    sim_summary %>%
    dplyr::select(ref, qry, sim_hits, observed) %>%
    mutate(pair_id = paste(ref, qry, sep = "_v_")) %>%
    dplyr::select(pair_id, sim_hits, observed) %>%
    unnest(sim_hits)


sim_summary_with_expected <- 
    sim_summary %>%
    mutate(
        expected_mean = map_dbl(sim_hits, mean),
        expected_lower = map_dbl(sim_hits, ~ quantile(.x, 0.025)),
        expected_upper = map_dbl(sim_hits, ~ quantile(.x, 0.975)),
        pair_id = paste(ref, qry, sep = "_v_")
    )


breakpoint_enrichment_summary <- 
    sim_summary_with_expected %>%
  mutate(
    enrichment_ratio = observed / expected_mean,
    p_value = map2_dbl(sim_hits, observed, ~ mean(.x <= .y)),
    is_significant = ifelse(p_value < 0.05, "Significant", "Not Significant"),
    obs_minus_exp = observed - expected_mean
  )

ggplot(breakpoint_enrichment_summary, aes(x = enrichment_ratio, fill = is_significant)) +
  geom_histogram(binwidth = 0.05, color = "black", alpha = 0.8) +
  geom_vline(xintercept = 1, linetype = "dashed", color = "red") +
  scale_fill_manual(values = c("Significant" = "#e31a1c", "Not Significant" = "grey70")) +
  labs(
    title = "Observed / Expected Genic Breakpoints",
    x = "Enrichment Ratio",
    y = "Genome Pair Count",
    fill = "P < 0.05"
  ) +
  theme_classic(base_size = 14)

breakpoint_enrichment_summary %>%
  mutate(pair_id = reorder(pair_id, enrichment_ratio)) %>%
  ggplot(aes(x = pair_id)) +
  geom_segment(aes(xend = pair_id, y = expected_mean, yend = observed),
               color = "grey60", size = 1) +
  geom_point(aes(y = expected_mean), color = "#1f78b4", size = 3) +
  geom_point(aes(y = observed), color = "#e31a1c", size = 3) +
  coord_flip() +
  labs(
    title = "Observed vs Expected Genic Breakpoints per Genome Pair",
    y = "Breakpoint Count", x = NULL
  ) +
  theme_minimal(base_size = 14)


ggplot(breakpoint_enrichment_summary, aes(x = enrichment_ratio)) +
  stat_ecdf(geom = "step", color = "#e31a1c", size = 1.2) +
  geom_vline(xintercept = 1, linetype = "dashed") +
  labs(
    title = "Cumulative Distribution of Observed / Expected",
    x = "Enrichment Ratio",
    y = "Proportion of Genome Pairs"
  ) +
  theme_classic(base_size = 14)


ggplot(breakpoint_enrichment_summary, aes(x = enrichment_ratio, y = -log10(p_value))) +
  geom_point(size = 3, color = "#e31a1c", alpha = 0.8) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "grey30") +
  geom_vline(xintercept = 1, linetype = "dotted") +
  labs(
    title = "Depletion of Genic Breakpoints Across Genome Pairs",
    x = "Observed / Expected",
    y = "-log10(P-value)"
  ) +
  theme_minimal(base_size = 14)

breakpoint_enrichment_summary %>%
  arrange(enrichment_ratio) %>%
  mutate(pair_id = factor(pair_id, levels = pair_id)) %>%
  ggplot(aes(x = pair_id, y = enrichment_ratio)) +
  geom_col(fill = "#e31a1c") +
  geom_hline(yintercept = 1, linetype = "dashed") +
  coord_flip() +
  labs(
    title = "Genic Breakpoint Enrichment by Genome Pair",
    x = "Genome Pair",
    y = "Observed / Expected"
  ) +
  theme_minimal(base_size = 13)


ggplot(combined_df, aes(x = reorder(pair_id, val), y = val, color = type, group = pair_id)) +
  geom_line(color = "grey60", linewidth = 0.8) +
  geom_point(size = 3) +
  scale_color_manual(values = c("expected" = "#1f78b4", "observed" = "#e31a1c")) +
  coord_flip() +
  labs(
    title = "Observed vs Expected Genic Breakpoints",
    y = "Breakpoint Count",
    x = "Genome Pair",
    color = NULL
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.y = element_text(size = 10),
    legend.position = "top"
  )


# Unnest expected simulations
expected_df <- 
    sim_summary_with_expected %>%
    dplyr::select(pair_id, val = sim_hits) %>%
    unnest(val) %>%
    #dplyr::rename(val = expected) %>%
    mutate(type = "Expected")

# Gather all observed values
observed_df <- 
    sim_long %>%
    select(pair_id, val = observed) %>%
    #dplyr::rename(val = expected)
    mutate(type = "Observed")# %>%
    #distinct()

# Combine
combined_df <- 
    bind_rows(expected_df, observed_df)


enrichment_results <- 
    sim_summary_with_expected %>%
  mutate(
    enrichment_ratio = observed / expected_mean,
    p_value = map2_dbl(sim_hits, observed, ~ mean(.x >= .y)),
    significant = ifelse(p_value < 0.05, "Significant", "Not Significant"),
    obs_minus_exp = observed - expected_mean
  )

ggplot(enrichment_results, aes(x = enrichment_ratio, fill = significant)) +
  geom_histogram(binwidth = 0.05, color = "black", alpha = 0.8) +
  geom_vline(xintercept = 1, linetype = "dashed", color = "red") +
  scale_fill_manual(values = c("Significant" = "#e31a1c", "Not Significant" = "grey70")) +
  labs(
    title = "Observed / Expected Genic Breakpoints",
    x = "Enrichment Ratio (Observed / Expected)",
    y = "Count of Genome Pairs",
    fill = "Significance"
  ) +
  theme_classic()


combined_df$type %>% unique
ggplot(combined_df, aes(x = val, fill = factor(type))) +
  geom_histogram(
    aes(y = ..density..),
    binwidth = 10,
    color = "black",
    alpha = 0.6
  ) +
  geom_density(
    aes(y = ..density..),
    binwidth = 10,
    color = "black",
    alpha = 0.6
  ) +
  labs(
    title = "Distribution of Genic Breakpoints: Observed vs Expected",
    x = "Genic Breakpoint Count",
    y = "Density",
    fill = NULL
  ) +
  theme_classic(base_size = 14)





enrichment_summary <- 
    sim_summary_with_expected %>%
  mutate(
    mean_enrichment = observed / expected_mean
  )

ggplot(enrichment_summary, aes(x = mean_enrichment)) +
  geom_histogram(binwidth = 0.05, fill = "#fb9a99", color = "black", alpha = 0.7) +
  geom_vline(xintercept = 1, linetype = "dashed", color = "red") +
  labs(
    title = "Distribution of Observed / Expected Genic Breakpoints",
    x = "Observed / Expected",
    y = "Count"
  ) +
  theme_classic()

sim_summary_with_expected <- sim_summary_with_expected %>%
  mutate(
    enrichment = observed / expected_mean,
    p_value = map2_dbl(sim_hits, observed, ~ mean(.x >= .y)),
    sig = ifelse(p_value < 0.05, "Significant", "Not Significant")
  )


ss3 <- 
    sim_summary_with_expected %>%
  mutate(
    enrichment = observed / expected_mean,
    p_value = map2_dbl(sim_hits, observed, ~ mean(.x >= .y)),
    sig = ifelse(p_value < 0.05, "Significant", "Not Significant")
  )

ggplot(ss3, aes(x = enrichment, fill = sig)) +
  geom_histogram(binwidth = 0.05, color = "black", alpha = 0.8) +
  geom_vline(xintercept = 1, linetype = "dashed", color = "red") +
  scale_fill_manual(values = c("Significant" = "#e31a1c", "Not Significant" = "grey70")) +
  labs(
    title = "Observed / Expected Genic Breakpoints Across Pairs",
    x = "Observed / Expected",
    y = "Number of Genome Pairs",
    fill = "Enrichment P < 0.05"
  ) +
  theme_classic()



# Plot
paired_plot <- 
    ggplot(sim_summary_with_expected, aes(x = reorder(pair_id, -observed))) +
  geom_col(aes(y = observed), fill = "#1f78b4") +
  geom_point(aes(y = expected_mean), color = "#ffb347", size = 3) +
  geom_errorbar(aes(ymin = expected_lower, ymax = expected_upper), width = 0.2) +
  coord_flip() +
  labs(
    title = "Observed vs Expected Genic Breakpoints per Genome Pair",
    x = NULL,
    y = "Genic Breakpoints"
  ) +
  theme_minimal(base_size = 14)


## all individual plots
sim_long_joined <- 
    sim_long

ggplot(sim_long_joined, aes(x = sim_hits)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  geom_vline(aes(xintercept = observed), color = "red", linewidth = 0.6) +
  facet_wrap(~ pair_id, scales = "free") +
  theme_minimal(base_size = 14) +
  labs(
    title = "Observed vs Simulated Genic Breakpoints",
    x = "Simulated Genic Breakpoints",
    y = "Density"
  ) +
    scale_x_continuous(
        limits = 
            c(
                min(c(sim_long_joined$sim_hits, sim_long_joined$observed)),
                max(c(sim_long_joined$sim_hits, sim_long_joined$observed))
                )
    )

#the distributions
dist_df <- 
    bind_rows(
        sim_long_joined %>%
            dplyr::select(breakpoints = sim_hits) %>%
            mutate(type = "Expected"),
        sim_long_joined %>%
            group_by(pair_id) %>%
            summarise(breakpoints = unique(observed)) %>%
            mutate(type = "Observed")
        )

# Plot both distributions
ggplot(
    dist_df, 
    aes(x = breakpoints, fill = type, color = type)
    ) +
  geom_histogram(aes(y = ..density..), bins = 40, alpha = 0.4, position = "identity") +
  geom_density(alpha = 0.4, linewidth = 1) +
  scale_fill_manual(values = c("Observed" = "#1f78b4", "Expected" = "#ffb347")) +
  scale_color_manual(values = c("Observed" = "#1f78b4", "Expected" = "#ffb347")) +
  labs(
    title = "Distribution of Genic Breakpoints (Observed vs Simulated)",
    x = "Genic Breakpoints",
    y = "Density",
    fill = NULL,
    color = NULL
  ) +
  theme_minimal(base_size = 16)


```










```{r}
ggplot(all_SVs, aes(x = variant, y = width)) +
  geom_violin(fill = "#cce5ff", color = "black", alpha = 0.7) +
  geom_boxplot(width = 0.2, outlier.size = 0.5) +
  scale_y_log10() +  # SV sizes tend to be skewed
  labs(y = "SV Length (bp)", x = NULL, title = "Distribution of SV Lengths by Type") +
  theme_classic()




CSVs %>% 
    count(ref_start_hit_prod, ref_end_hit_prod, qry_start_hit_prod, qry_end_hit_prod)

observed_genic <- 
    LSVs %>%
    group_by(acc, ref) %>%
    mutate(
        genic_hits = 
            ifelse(
                both_genic,
                2,
                ifelse(
                    one_genic,
                    1, 
                    0
                )
            )
      ) %>%
      group_by(ref, acc, ref_len, qry_len) %>%
      summarise(
        observed = sum(genic_hits),
        n_total = 2*n(),
        .groups = "keep"
      )

ggplot(gene_regions_tib, aes(x = region, y = proportion, fill = region)) +
  geom_violin(alpha = 0.3, color = NA, width = 1.0) +
  geom_boxplot(width = 0.4, alpha = 0.5, color = "black", outlier.shape = NA) +
  geom_dotplot(
    binaxis = 'y', stackdir = 'center',
    dotsize = 0.5, binwidth = 0.01,
    color = "black", fill = NA,
    alpha = 0.8, width = 0.15
  ) +
  scale_fill_manual(
    values = c("Genic" = "#4F84C4", "Intergenic" = "#E69F00")
  ) +
  scale_y_continuous(
    labels = percent_format(accuracy = 1),
    limits = c(0, 1),
    expand = expansion(mult = c(0, 0.05)),
    breaks = seq(0, 1, 0.1)
  ) +
  labs(
    title =
        sprintf(
            "Genic vs Intergenic Proportions %s",
            species_name
        ),
    x = NULL,
    y = "Proportion of Genome"
  ) +
  theme_classic(base_size = 22) +
  theme(
    legend.position = "none",
    plot.title = element_text(face = "bold", size = 18, hjust = 0.5),
    axis.text = element_text(size = 13),
    axis.title.y = element_text(size = 14),
    text = element_text(size = 22),
    panel.grid.major.y = element_line(color = "grey85"),
    panel.grid.minor.y = element_blank()
  )


grid.arrange(
  enrichment_plot,
  gr_plot,
  ncol = 2
)

```










```{r}


genic_props <- 
    gene_regions_tib %>%
    filter(region == "Genic") %>%
    dplyr::select(acc, genic_prop = proportion)


og2 <- 
    observed_genic %>%
    left_join(genic_props, by = c("ref" = "acc")) %>%
    dplyr::rename(ref_genic = genic_prop) %>%
    left_join(genic_props, by = c("acc" = "acc")) %>%
    dplyr::rename(acc_genic = genic_prop) %>%
    mutate(
        total_len = ref_len + qry_len,
        genic_fraction = (ref_genic * ref_len + acc_genic * qry_len) / total_len
        )


## binomial sims
set.seed(1234)
n_sim <- 1000

sim_summary <- 
    og2 %>%
      rowwise() %>%
      mutate(
          sim_hits = list(rbinom(n_sim, size = n_total, prob = genic_fraction))
          ) %>%
      ungroup()

sim_long <- 
    sim_summary %>%
    dplyr::select(ref, acc, sim_hits, observed) %>%
    mutate(pair_id = paste(ref, acc, sep = "_v_")) %>%
    dplyr::select(pair_id, sim_hits, observed) %>%
    unnest(sim_hits)


sim_summary_with_expected <- 
    sim_summary %>%
  mutate(
    expected_mean = map_dbl(sim_hits, mean),
    expected_lower = map_dbl(sim_hits, ~ quantile(.x, 0.025)),
    expected_upper = map_dbl(sim_hits, ~ quantile(.x, 0.975)),
    pair_id = paste(ref, acc, sep = "_v_")
  )

# Plot
paired_plot <- ggplot(sim_summary_with_expected, aes(x = reorder(pair_id, -observed))) +
  geom_col(aes(y = observed), fill = "#1f78b4") +
  geom_point(aes(y = expected_mean), color = "#ffb347", size = 3) +
  geom_errorbar(aes(ymin = expected_lower, ymax = expected_upper), width = 0.2) +
  coord_flip() +
  labs(
    title = "Observed vs Expected Genic Breakpoints per Genome Pair",
    x = NULL,
    y = "Genic Breakpoints"
  ) +
  theme_minimal(base_size = 14)

diff_histogram <- sim_summary_with_expected %>%
  mutate(diff = expected_mean - observed) %>%
  ggplot(aes(x = diff)) +
  geom_histogram(fill = "#1f78b4", bins = 30, color = "black") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  labs(
    title = "Distribution of Genic Breakpoint Differences",
    x = "Expected - Observed",
    y = "Genome Pair Count"
  ) +
  theme_minimal(base_size = 14)

```










```{r}


sim_long_joined <- 
    sim_long

ggplot(sim_long_joined, aes(x = sim_hits)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  geom_vline(aes(xintercept = observed), color = "red", linewidth = 0.6) +
  facet_wrap(~ pair_id, scales = "free") +
  theme_minimal(base_size = 14) +
  labs(
    title = "Observed vs Simulated Genic Breakpoints",
    x = "Simulated Genic Breakpoints",
    y = "Density"
  ) +
    scale_x_continuous(
        limits = 
            c(
                min(c(sim_long_joined$sim_hits, sim_long_joined$observed)),
                max(c(sim_long_joined$sim_hits, sim_long_joined$observed))
                )
    )



ggplot(sim_long_joined, aes(x = sim_hits)) +
  geom_histogram(aes(y = ..density..), bins = 40, fill = "#ffb347", alpha = 0.5, color = "black") +
  geom_density(color = "#ff9900", linewidth = 1) +
  geom_vline(
    aes(xintercept = observed, group = pair_id),
    color = "#1f78b4",
    alpha = 0.4,
    linewidth = 0.6
  ) +
  labs(
    title = "Pooled Observed vs Simulated Genic Breakpoints",
    x = "Genic Breakpoints",
    y = "Density"
  ) +
  theme_minimal(base_size = 16)

dist_df <- bind_rows(
  sim_long_joined %>%
    select(breakpoints = sim_hits) %>%
    mutate(type = "Expected"),
  
  sim_long_joined %>%
    group_by(pair_id) %>%
    summarise(breakpoints = unique(observed)) %>%
    mutate(type = "Observed")
)

# Plot both distributions
ggplot(dist_df, aes(x = breakpoints, fill = type, color = type)) +
  geom_histogram(aes(y = ..density..), bins = 40, alpha = 0.4, position = "identity") +
  geom_density(alpha = 0.4, linewidth = 1) +
  scale_fill_manual(values = c("Observed" = "#1f78b4", "Expected" = "#ffb347")) +
  scale_color_manual(values = c("Observed" = "#1f78b4", "Expected" = "#ffb347")) +
  labs(
    title = "Distribution of Genic Breakpoints (Observed vs Simulated)",
    x = "Genic Breakpoints",
    y = "Density",
    fill = NULL,
    color = NULL
  ) +
  theme_minimal(base_size = 16)


volcano_df <- 
    sim_summary %>%
  mutate(
    diff = expected_mean - observed,
    neglog10_p = -log10(p_value)  # You'd need to compute p-value earlier
  )

ggplot(volcano_df, aes(x = diff, y = neglog10_p)) +
  geom_point(color = "#1f78b4") +
  geom_vline(xintercept = 0, linetype = "dashed") +
  labs(
    title = "SV Breakpoint Genic Depletion Across Genome Pairs",
    x = "Expected - Observed Genic Breakpoints",
    y = expression(-log[10](p))
  ) +
  theme_minimal(base_size = 14)






ggplot(sim_long, aes(x = sim_hits)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  geom_vline(data = sim_summary, aes(xintercept = observed), color = "red") +
  facet_wrap(~ pair_id, scales = "free") +
  theme_minimal()


observed_long <- 
    sim_summary %>%
    dplyr::select(ref, acc, sim_hits, observed) %>%
    mutate(pair_id = paste(ref, acc, sep = "_v_")) %>%
  dplyr::slice(rep(1:n(), each = 1000)) %>%  # repeat observed to match sim length
  mutate(type = "Observed", breakpoints = observed) %>%
  dplyr::select(pair_id, type, breakpoints)


expected_long <- 
    sim_long %>%
  mutate(type = "Expected", breakpoints = sim_hits) %>%
  dplyr::select(pair_id, type, breakpoints)

dist_df <- bind_rows(observed_long, expected_long)

# STEP 2: Plot histogram with density overlay
ggplot(dist_df, aes(x = breakpoints, fill = type, color = type)) +
  geom_histogram(aes(y = after_stat(density)), position = "identity", alpha = 0.4, bins = 30) +
  geom_density(alpha = 0.3, size = 1.2) +
  scale_fill_manual(values = c("Expected" = "#ffb347", "Observed" = "#1f78b4")) +
  scale_color_manual(values = c("Expected" = "#ffb347", "Observed" = "#1f78b4")) +
  labs(
    title = "Distribution of Genic Breakpoints",
    x = "Genic Breakpoints",
    y = "Density",
    fill = NULL,
    color = NULL
  ) +
  theme_minimal(base_size = 14)


ss2 <- 
    sim_summary %>%
  mutate(
    expected_mean = map_dbl(sim_hits, mean),
    expected_lower = map_dbl(sim_hits, ~ quantile(.x, 0.025)),
    expected_upper = map_dbl(sim_hits, ~ quantile(.x, 0.975))
  ) %>%
    mutate(pair_id = paste(ref, acc, sep = "_v_"))



# STEP 6: Create plot and save to object
LSV_vis <- 
    ggplot(ss2, aes(x = reorder(pair_id, -observed), y = observed)) +
  geom_col(fill = "#1f78b4") +
  geom_errorbar(aes(ymin = expected_lower, ymax = expected_upper), width = 0.2, color = "black") +
  geom_point(aes(y = expected_mean), shape = 21, fill = "#ffb347", size = 3) +
  coord_flip() +
  labs(
    title = "Observed vs Expected Genic Breakpoints per Genome Pair",
    y = "Genic Breakpoints",
    x = NULL
  ) +
  theme_minimal(base_size = 14)


observed_long <- 
    ss2 %>%
    dplyr::select(pair_id, observed) %>%
    dplyr::slice(rep(1:n(), each = 1000)) %>%  # repeat observed to match sim length
    mutate(type = "Observed", breakpoints = observed) %>%
    dplyr::select(pair_id, type, breakpoints)

expected_long <- 
    sim_long %>%
  mutate(type = "Expected", breakpoints = sim_hits) %>%
  select(pair_id, type, breakpoints)

dist_df <- bind_rows(observed_long, expected_long)

# STEP 2: Plot histogram with density overlay
ggplot(dist_df, aes(x = breakpoints, fill = type, color = type)) +
  geom_histogram(aes(y = ..density..), position = "identity", alpha = 0.4, bins = 30) +
  geom_density(alpha = 0.3, size = 1.2) +
  scale_fill_manual(values = c("Expected" = "#ffb347", "Observed" = "#1f78b4")) +
  scale_color_manual(values = c("Expected" = "#ffb347", "Observed" = "#1f78b4")) +
  labs(
    title = "Distribution of Genic Breakpoints",
    x = "Genic Breakpoints",
    y = "Density",
    fill = NULL,
    color = NULL
  ) +
  theme_minimal(base_size = 14)




LSV_paired <- 
    LSVs %>%
    group_by(acc, ref) %>%
    summarise(
        observed_genic = 
            sum(start_genic, na.rm = TRUE) + sum(end_genic, na.rm = TRUE),
        total = 2 * n()
        )

    # inner_join(
    #     ., 
    #     clustered_genomes[,c(1,2)] %>% 
    #         filter(accessions %in% prokka_accs) %>%
    #         #group_by(cluster_id) %>% dplyr::slice(1:3) %>%
    #         `colnames<-`(c("clust", "acc")),
    #     by = "acc"
    # ) %>%
    # rowwise() %>%
    # mutate(acc_c = paste0(acc, "_", clust))


LSV_t <- 
    LSVs %>%
    ungroup %>%
    dplyr::slice(1:2) %>%
    dplyr::select(c(start, end, width ,acc, variant, ref))

all_prokka <- 
    bind_rows(prokka_list)


all_prokka %>% 
    transmute(w=end-start+1, gene,p = tolower(product), type) %>%
    filter(
        type=="CDS", 
        !grepl("hypothetical", p),
        !grepl("putative", p )) %>% 
    arrange(w)
    pull(w) %>% summary()

all_prokka %>% filter(accession == "CP049377.1")
all_prokka %>% filter(accession == "CP051990.1")

### getting gff overlaps

for (i in 1:nrow(ref_qry_pairs){
    rq <- 
        ref_qry_pairs[i,]
    
    
    ### GFF files
    
    ref_prokka <- 
        all_prokka %>%
        filter(
            accession == rq$ref
        )
    qry_prokka <- 
        all_prokka %>%
        filter(
            accession == rq$qry
        )
    
    ref_gr <- 
        GRanges(
            seqnames = ref_prokka$accession,
            ranges = IRanges(
                start = ref_prokka$start, 
                end = ref_prokka$end
                ),
            gene = ref_prokka$gene,
            product = ref_prokka$product
            )

    # Convert query Prokka to GRanges
    qry_gr <- 
        GRanges(
            seqnames = qry_prokka$accession,
            ranges = IRanges(start = qry_prokka$start, end = qry_prokka$end),
            gene = qry_prokka$gene,
            product = qry_prokka$product
            )
    qry_gr
    ref_gr
    #
    
    sv_breakpoints <- 
        LSVs %>%
        filter(
            acc == rq$qry, 
            ref == rq$ref
            ) %>%
        mutate(
            start_bp = start,
            end_bp = end,
            genome = ifelse(variant == "insertion", acc, ref)
          )
    
    
    bp_start_gr <- 
        GRanges(
            seqnames = sv_breakpoints$genome,
            ranges = IRanges(start = sv_breakpoints$start_bp, width = 1)
            )
    
    bp_end_gr <- 
        GRanges(
            seqnames = sv_breakpoints$genome,
            ranges = IRanges(start = sv_breakpoints$end_bp, width = 1)
            )

    sv_ov <- 
        sv_breakpoints %>%
        rowwise() %>%
        mutate(
        genome = ifelse(variant == "Deletion", ref, acc),
        start_genic = {
          bp <- GRanges(seqnames = genome, ranges = IRanges(start = start, width = 1))
          gff <- if (variant == "Deletion") ref_gr else qry_gr
          length(findOverlaps(bp, gff)) > 0
        },
        end_genic = {
          bp <- GRanges(seqnames = genome, ranges = IRanges(start = end, width = 1))
          gff <- if (variant == "Deletion") ref_gr else qry_gr
          length(findOverlaps(bp, gff)) > 0
        }
        ) %>%
        ungroup()

    
    sv_ov %>%
        count(start_genic, end_genic)
    
    
    ### gene regions tib
    ref_gene_prop <- 
        gene_regions_tib %>% 
        filter(acc == rq$ref)
    
    qry_gene_prop <- 
        gene_regions_tib %>% 
        filter(acc == rq$qry)
    
    ref_gene_prop
    qry_gene_prop
    sv_ov
    
    
    #observed genic breakpoints per genome
    bp_tbl <- 
        sv_ov %>%
        mutate(start_genic = as.logical(start_genic),
         end_genic = as.logical(end_genic)) %>%
          group_by(genome) %>%
          summarise(
            n_total = 2 * n(),  # 2 breakpoints per SV
            n_genic = sum(start_genic, na.rm = TRUE) + sum(end_genic, na.rm = TRUE)
          )
    
    
    gene_prop <- 
        bind_rows(ref_gene_prop, qry_gene_prop) %>%
        filter(region == "Genic") %>%
        select(genome = acc, genic_fraction = proportion)

    # Step 3: Join with breakpoint counts
    bp_summary <- bp_tbl %>%
      left_join(gene_prop, by = "genome")
    
    # Step 4: Run binomial test for each genome
    bp_summary <- 
        bp_summary %>%
      rowwise() %>%
      mutate(
        binom_p = binom.test(n_genic, n_total, p = genic_fraction, alternative = "less")$p.value
      ) %>%
      ungroup()
    
    
    n_total <- 2 * nrow(sv_ov)  # 2 breakpoints per SV
    n_genic <- sum(sv_ov$start_genic, na.rm = TRUE) + sum(sv_ov$end_genic, na.rm = TRUE)
    
    ref_len <- 
        ref_gene_prop %>% filter(acc == "NZ_CP134047.1") %>% summarise(bp = sum(bp)) %>% pull(bp)
    qry_len <- 
        qry_gene_prop %>% filter(acc == "NZ_CP084878.1") %>% summarise(bp = sum(bp)) %>% pull(bp)
    
    expected_genic_frac <- (ref_frac * ref_len + qry_frac * qry_len) / total_len

    
    seqlevels(ref_gr) <- "NZ_CP134047.1"
    seqnames(ref_gr) <- Rle("NZ_CP134047.1", length(ref_gr))

    seqlevels(qry_gr) <- "NZ_CP084878.1"
    seqnames(qry_gr) <- Rle("NZ_CP084878.1", length(qry_gr))

# Combine
    cds_gr <- c(ref_gr, qry_gr)

# Total genome length
    genome_length <- ref_len + qry_len
    
    
    set.seed(123)
        n_intergenic <- n_total - n_genic
        sim_intergenic <- n_total - sim_hits
            sim_hits <- 
                replicate(
                    1000, 
                    {
                        rand_pos <- sample(1:genome_length, size = n_total)
                        rand_genome <- sample(c("NZ_CP134047.1", "NZ_CP084878.1"), size = n_total, replace = TRUE,
                                              prob = c(ref_len, qry_len) / genome_length)
                        rand_gr <- GRanges(seqnames = rand_genome, ranges = IRanges(rand_pos, width = 1))
                        sum(countOverlaps(rand_gr, cds_gr) > 0)
                        }
                    )
            
            
            
            ggplot(data.frame(simulated = sim_intergenic), aes(x = simulated)) +
          geom_histogram(bins = 30, fill = "skyblue", color = "black") +
          geom_vline(xintercept = n_intergenic, color = "red", size = 1.2) +
          annotate("text", x = n_intergenic, y = max(table(sim_intergenic)) * 0.9,
                   label = paste("Observed =", n_intergenic), color = "red", hjust = -0.1, size = 4) +
          labs(
            title = "Permutation Test: Intergenic Breakpoint Enrichment",
            x = "Simulated Intergenic Breakpoints",
            y = "Frequency"
          ) +
          theme_minimal(base_size = 14)
            
            
    # Combine proportions
    total_bp <- sum(ref_gene_prop$bp) + sum(qry_gene_prop$bp)
    genic_bp <- sum(ref_gene_prop$bp[ref_gene_prop$region == "Genic"]) +
                sum(qry_gene_prop$bp[qry_gene_prop$region == "Genic"])
    intergenic_bp <- total_bp - genic_bp
    
    genome_prop <- tibble(
      source = "Genome",
      category = c("Genic", "Intergenic"),
      proportion = c(genic_bp, intergenic_bp) / total_bp
    )
    
    # Step 2: SV breakpoint genic/intergenic annotation (pooled)
    sv_bp_genic <- sv_ov %>%
      mutate(
        start = ifelse(start_genic, "Genic", "Intergenic"),
        end   = ifelse(end_genic, "Genic", "Intergenic")
      ) %>%
      pivot_longer(cols = c(start, end), names_to = "bp", values_to = "category") %>%
      count(category) %>%
      mutate(source = "SV Breakpoints", proportion = n / sum(n)) %>%
      select(source, category, proportion)
    
    # Step 3: Combine
    sv_gff_prop <- bind_rows(genome_prop, sv_bp_genic)
    
    # Step 4: Plot
    ggplot(sv_gff_prop, aes(x = source, y = proportion, fill = category)) +
      geom_col(position = position_dodge(width = 0.7), width = 0.6) +
      scale_y_continuous(labels = scales::percent_format()) +
      scale_fill_manual(
          values = c("Genic" = "#1f78b4", "Intergenic" = "#ffb347")
          ) +
      labs(
        title = "",
        x = NULL,
        y = "Proportion",
        fill = NULL
      ) +
      theme_classic(base_size = 22) +
      theme(legend.position = "bottom")
      
            
        
        }
    
    
    
    
    
    LSVs %>% pull(ref) %>% unique

filter(LSVs, width >1e4)

gene_regions

    
CSVs <- 
    bind_rows(Chrom_SVs) %>%
    ungroup %>%
    inner_join(
        ., 
        clustered_genomes[,c(1,2)] %>% 
            filter(accessions %in% prokka_accs) %>%
            #group_by(cluster_id) %>% dplyr::slice(1:3) %>%
            `colnames<-`(c("clust", "qid")),
        by = "qid"
    ) %>%
    rowwise() %>%
    mutate(qid_c = paste0(qid, "_", clust))
```



### permutation test
```{r}
set.seed(123)

# Create a big simulation matrix: rows = ref–qry pairs, cols = simulations
n_sim <- 1000
sim_table <- tibble()

# This assumes you already calculated genome lengths per pair
for (i in 1:nrow(ref_qry_pairs)) {
  # Extract current ref/qry
rq <- ref_qry_pairs[i,]

# GFFs
ref_prokka <- all_prokka %>% filter(accession == rq$ref)
qry_prokka <- all_prokka %>% filter(accession == rq$qry)

ref_gr <- GRanges(
  seqnames = ref_prokka$accession,
  ranges = IRanges(start = ref_prokka$start, end = ref_prokka$end),
  gene = ref_prokka$gene,
  product = ref_prokka$product
)

qry_gr <- GRanges(
  seqnames = qry_prokka$accession,
  ranges = IRanges(start = qry_prokka$start, end = qry_prokka$end),
  gene = qry_prokka$gene,
  product = qry_prokka$product
)

# SVs for this pair
sv_ov <- LSVs %>%
  filter(acc == rq$qry, ref == rq$ref) %>%
  rowwise() %>%
  mutate(
    genome = ifelse(variant == "Deletion", ref, acc),
    start_genic = {
      bp <- GRanges(seqnames = genome, ranges = IRanges(start = start, width = 1))
      gff <- if (variant == "Deletion") ref_gr else qry_gr
      length(findOverlaps(bp, gff)) > 0
    },
    end_genic = {
      bp <- GRanges(seqnames = genome, ranges = IRanges(start = end, width = 1))
      gff <- if (variant == "Deletion") ref_gr else qry_gr
      length(findOverlaps(bp, gff)) > 0
    }
  ) %>%
  ungroup()

# Count observed genic breakpoints
n_genic <- sum(sv_ov$start_genic, na.rm = TRUE) + sum(sv_ov$end_genic, na.rm = TRUE)
n_total <- 2 * nrow(sv_ov)
n_intergenic <- n_total - n_genic

# Genome space proportions
ref_bp <- gene_regions_tib %>% filter(acc == rq$ref) %>% summarise(bp = sum(bp)) %>% pull()
qry_bp <- gene_regions_tib %>% filter(acc == rq$qry) %>% summarise(bp = sum(bp)) %>% pull()
ref_genic <- gene_regions_tib %>% filter(acc == rq$ref, region == "Genic") %>% pull(bp)
qry_genic <- gene_regions_tib %>% filter(acc == rq$qry, region == "Genic") %>% pull(bp)
total_bp <- ref_bp + qry_bp
genic_bp <- ref_genic + qry_genic

# Save per-pair observed results
results_list[[i]] <- tibble(
  pair_id = paste(rq$ref, rq$qry, sep = "_v_"),
  observed_genic = n_genic,
  observed_intergenic = n_intergenic,
  total = n_total,
  genome_genic_frac = genic_bp / total_bp,
  genome_intergenic_frac = 1 - (genic_bp / total_bp)
)

  
  # Get lengths
  ref_len <- gene_regions_tib %>% filter(acc == rq$ref) %>% summarise(bp = sum(bp)) %>% pull()
  qry_len <- gene_regions_tib %>% filter(acc == rq$qry) %>% summarise(bp = sum(bp)) %>% pull()
  genome_length <- ref_len + qry_len
  
  # Save for later use
  sim_hits <- replicate(n_sim, {
    rand_pos <- sample(1:genome_length, size = 480)  # placeholder size
    rand_genome <- sample(c(rq$ref, rq$qry), size = 480, replace = TRUE,
                          prob = c(ref_len, qry_len) / genome_length)
    rand_gr <- GRanges(seqnames = rand_genome, ranges = IRanges(rand_pos, width = 1))
    sum(countOverlaps(rand_gr, cds_gr))  # placeholder cds_gr
  })
  
  sim_table <- bind_rows(
    sim_table,
    tibble(
      pair_id = paste(rq$ref, rq$qry, sep = "_v_"),
      sim_id = 1:n_sim,
      sim_genic = sim_hits
    )
  )
}

```




```{r}
results_list <- list()

for (i in 1:nrow(ref_qry_pairs)) {
  # Extract current ref/qry
    rq <- ref_qry_pairs[i,]
    
    # GFFs
    ref_prokka <- all_prokka %>% filter(accession == rq$ref)
    qry_prokka <- all_prokka %>% filter(accession == rq$qry)
    
    ref_gr <- GRanges(
      seqnames = ref_prokka$accession,
      ranges = IRanges(start = ref_prokka$start, end = ref_prokka$end),
      gene = ref_prokka$gene,
      product = ref_prokka$product
    )
    
    qry_gr <- GRanges(
      seqnames = qry_prokka$accession,
      ranges = IRanges(start = qry_prokka$start, end = qry_prokka$end),
      gene = qry_prokka$gene,
      product = qry_prokka$product
    )
    
    # SVs for this pair
    sv_ov <- LSVs %>%
      filter(acc == rq$qry, ref == rq$ref) %>%
      rowwise() %>%
      mutate(
        genome = ifelse(variant == "Deletion", ref, acc),
        start_genic = {
          bp <- GRanges(seqnames = genome, ranges = IRanges(start = start, width = 1))
          gff <- if (variant == "Deletion") ref_gr else qry_gr
          length(findOverlaps(bp, gff)) > 0
        },
        end_genic = {
          bp <- GRanges(seqnames = genome, ranges = IRanges(start = end, width = 1))
          gff <- if (variant == "Deletion") ref_gr else qry_gr
          length(findOverlaps(bp, gff)) > 0
        }
      ) %>%
      ungroup()
    
    # Count observed genic breakpoints
    n_genic <- sum(sv_ov$start_genic, na.rm = TRUE) + sum(sv_ov$end_genic, na.rm = TRUE)
    n_total <- 2 * nrow(sv_ov)
    n_intergenic <- n_total - n_genic
    
    # Genome space proportions
    ref_bp <- gene_regions_tib %>% filter(acc == rq$ref) %>% summarise(bp = sum(bp)) %>% pull()
    qry_bp <- gene_regions_tib %>% filter(acc == rq$qry) %>% summarise(bp = sum(bp)) %>% pull()
    ref_genic <- gene_regions_tib %>% filter(acc == rq$ref, region == "Genic") %>% pull(bp)
    qry_genic <- gene_regions_tib %>% filter(acc == rq$qry, region == "Genic") %>% pull(bp)
    total_bp <- ref_bp + qry_bp
    genic_bp <- ref_genic + qry_genic
    
    # Save per-pair observed results
    results_list[[i]] <- tibble(
      pair_id = paste(rq$ref, rq$qry, sep = "_v_"),
      observed_genic = n_genic,
      observed_intergenic = n_intergenic,
      total = n_total,
      genome_genic_frac = genic_bp / total_bp,
      genome_intergenic_frac = 1 - (genic_bp / total_bp)
    )

 
}

results_df <- bind_rows(results_list)
```



```{r}

n_sim <- 1000
sim_table <- tibble()
results_list <- list()

for (i in 1:10){# seq_len(nrow(ref_qry_pairs))) {
  rq <- ref_qry_pairs[i, ]

  # Get reference and query gene regions (Genic only)
  ref_prokka <- all_prokka %>% filter(accession == rq$ref)
    qry_prokka <- all_prokka %>% filter(accession == rq$qry)

  # Build GRanges for coding regions
  ref_gr <- GRanges(seqnames = ref_prokka$accession, ranges = IRanges(start = ref_prokka$start, end = ref_prokka$end))
  qry_gr <- GRanges(seqnames = qry_prokka$accession, ranges = IRanges(start = qry_prokka$start, end = qry_prokka$end))
  
  cds_gr <- c(ref_gr, qry_gr)

  # Get SVs for this pair
  sv_ov <- LSVs %>%
    filter(acc == rq$qry, ref == rq$ref) %>%
    rowwise() %>%
    mutate(
      genome = ifelse(variant == "Deletion", ref, acc),
      start_genic = {
        bp <- GRanges(seqnames = genome, ranges = IRanges(start = start, width = 1))
        gff <- if (variant == "Deletion") ref_gr else qry_gr
        length(findOverlaps(bp, gff)) > 0
      },
      end_genic = {
        bp <- GRanges(seqnames = genome, ranges = IRanges(start = end, width = 1))
        gff <- if (variant == "Deletion") ref_gr else qry_gr
        length(findOverlaps(bp, gff)) > 0
      }
    ) %>%
    ungroup()

  n_genic <- sum(sv_ov$start_genic, na.rm = TRUE) + sum(sv_ov$end_genic, na.rm = TRUE)
  n_total <- 2 * nrow(sv_ov)

  # Genome lengths
  ref_len <- gene_regions_tib %>% filter(acc == rq$ref) %>% summarise(bp = sum(bp)) %>% pull()
  qry_len <- gene_regions_tib %>% filter(acc == rq$qry) %>% summarise(bp = sum(bp)) %>% pull()
  genome_length <- ref_len + qry_len

  # Run simulation
  sim_hits <- replicate(n_sim, {
    rand_pos <- sample(1:genome_length, size = n_total)
    rand_genome <- sample(c(rq$ref, rq$qry), size = n_total, replace = TRUE,
                          prob = c(ref_len, qry_len) / genome_length)
    rand_gr <- GRanges(seqnames = rand_genome, ranges = IRanges(rand_pos, width = 1))
    #seqinfo(rand_gr) <- NULL
    sum(countOverlaps(rand_gr, cds_gr) > 0)
  })

  # Store results
  sim_table <- bind_rows(
    sim_table,
    tibble(
      pair_id = paste(rq$ref, rq$qry, sep = "_v_"),
      sim_id = 1:n_sim,
      sim_genic = sim_hits
    )
  )

  results_list[[i]] <- tibble(
    pair_id = paste(rq$ref, rq$qry, sep = "_v_"),
    observed_genic = n_genic
  )
  print(i)
}

# Combine results
results_df <- 
    bind_rows(results_list)

# Prepare data for plotting
observed <- 
    results_df %>%
  dplyr::select(pair_id, breakpoints = observed_genic) %>%
  dplyr::slice(rep(1:n(), each = 100)) %>%
  mutate(type = "Observed")

expected <- sim_table %>%
  select(pair_id, breakpoints = sim_genic) %>%
  mutate(type = "Expected")

dist_df <- bind_rows(observed, expected)

# Plot distributions
ggplot(dist_df, aes(x = breakpoints, fill = type)) +
  geom_density(alpha = 0.5) +
  scale_fill_manual(values = c("Observed" = "#1f78b4", "Expected" = "#ffb347")) +
  labs(
    title = "Distribution of Genic Breakpoints Across Genome Pairs",
    x = "Genic Breakpoints",
    y = "Density",
    fill = NULL
  ) +
  theme_minimal(base_size = 14)

```






```{r}
results_list <- list()

for (i in 1:nrow(ref_qry_pairs)) {
  # Extract current ref/qry
    rq <- ref_qry_pairs[i,]
    
    # GFFs
    ref_prokka <- all_prokka %>% filter(accession == rq$ref)
    qry_prokka <- all_prokka %>% filter(accession == rq$qry)
    
    ref_gr <- GRanges(
      seqnames = ref_prokka$accession,
      ranges = IRanges(start = ref_prokka$start, end = ref_prokka$end),
      gene = ref_prokka$gene,
      product = ref_prokka$product
    )
    
    qry_gr <- GRanges(
      seqnames = qry_prokka$accession,
      ranges = IRanges(start = qry_prokka$start, end = qry_prokka$end),
      gene = qry_prokka$gene,
      product = qry_prokka$product
    )
    
    # SVs for this pair
    sv_ov <- LSVs %>%
      filter(acc == rq$qry, ref == rq$ref) %>%
      rowwise() %>%
      mutate(
        genome = ifelse(variant == "Deletion", ref, acc),
        start_genic = {
          bp <- GRanges(seqnames = genome, ranges = IRanges(start = start, width = 1))
          gff <- if (variant == "Deletion") ref_gr else qry_gr
          length(findOverlaps(bp, gff)) > 0
        },
        end_genic = {
          bp <- GRanges(seqnames = genome, ranges = IRanges(start = end, width = 1))
          gff <- if (variant == "Deletion") ref_gr else qry_gr
          length(findOverlaps(bp, gff)) > 0
        }
      ) %>%
      ungroup()
    
    # Count observed genic breakpoints
    n_genic <- sum(sv_ov$start_genic, na.rm = TRUE) + sum(sv_ov$end_genic, na.rm = TRUE)
    n_total <- 2 * nrow(sv_ov)
    n_intergenic <- n_total - n_genic
    
    # Genome space proportions
    ref_bp <- gene_regions_tib %>% filter(acc == rq$ref) %>% summarise(bp = sum(bp)) %>% pull()
    qry_bp <- gene_regions_tib %>% filter(acc == rq$qry) %>% summarise(bp = sum(bp)) %>% pull()
    ref_genic <- gene_regions_tib %>% filter(acc == rq$ref, region == "Genic") %>% pull(bp)
    qry_genic <- gene_regions_tib %>% filter(acc == rq$qry, region == "Genic") %>% pull(bp)
    total_bp <- ref_bp + qry_bp
    genic_bp <- ref_genic + qry_genic
    
    # Save per-pair observed results
    results_list[[i]] <- tibble(
      pair_id = paste(rq$ref, rq$qry, sep = "_v_"),
      observed_genic = n_genic,
      observed_intergenic = n_intergenic,
      total = n_total,
      genome_genic_frac = genic_bp / total_bp,
      genome_intergenic_frac = 1 - (genic_bp / total_bp)
    )

 
}

results_df <- bind_rows(results_list)





# assumes you have:
# - results_df with observed_genic per pair
# - sim_table with all simulations

# combine into long format
observed <- results_df %>%
  select(pair_id, breakpoints = observed_genic) %>%
  mutate(type = "Observed")

expected <- sim_table %>%
  select(pair_id, breakpoints = sim_genic) %>%
  mutate(type = "Expected")

dist_df <- bind_rows(observed, expected)

ggplot(dist_df, aes(x = breakpoints, fill = type)) +
  geom_density(alpha = 0.5) +
  scale_fill_manual(values = c("Observed" = "#1f78b4", "Expected" = "#ffb347")) +
  labs(
    title = "Distribution of Genic Breakpoints Across Genome Pairs",
    x = "Genic Breakpoints",
    y = "Density",
    fill = NULL
  ) +
  theme_minimal(base_size = 14)






# Join simulation and observation
plot_df <- 
    sim_table %>%
  group_by(pair_id) %>%
  summarise(
    expected_mean = mean(sim_genic),
    expected_sd = sd(sim_genic)
  ) %>%
  left_join(results_df, by = "pair_id") %>%
  mutate(
    expected_lower = expected_mean - expected_sd,
    expected_upper = expected_mean + expected_sd
  )

plot_long <- plot_df %>%
  select(pair_id, observed_genic, expected_mean) %>%
  pivot_longer(
    cols = c(observed_genic, expected_mean),
    names_to = "type", values_to = "breakpoints"
  )%>%
  mutate(breakpoints = breakpoints + 1)

ggplot(plot_long, aes(x = reorder(pair_id, -breakpoints), y = breakpoints, fill = type)) +
  geom_col(position = position_dodge(width = 0.7), width = 0.6, color = "black") +
  scale_fill_manual(
    values = c("observed_genic" = "#1f78b4", "expected_mean" = "#ffb347"),
    labels = c("Observed", "Expected")
  ) +
  coord_flip() +
  labs(
    title = "Observed vs Expected Genic Breakpoints (per genome pair)",
    x = NULL,
    y = "Genic Breakpoints",
    fill = NULL
  ) +
  theme_minimal(base_size = 14) +
  theme(legend.position = "bottom")+ scale_y_log10()


ggplot(sv_gff_prop, aes(x = source, y = proportion, fill = category)) +
  geom_col(position = position_dodge(width = 0.7), width = 0.6, color = "black") +
  geom_errorbar(
    data = filter(sv_gff_prop, source == "SV Breakpoints"),
    aes(ymin = proportion - 0.05, ymax = proportion + 0.05),
    width = 0.2, position = position_dodge(width = 0.7)
  ) +
  scale_fill_manual(values = c("Genic" = "#1f78b4", "Intergenic" = "#ffb347")) +
  labs(title = "", x = NULL, y = "Proportion", fill = NULL) +
  theme_classic(base_size = 22) +
  theme(legend.position = "bottom")



```



















## Identification
```{r}

delta_files <- 
    list.files(
        sprintf("%s/ref_v_all", delta_dir), 
        full.names = T
        )

# delta_file <- 
#     delta_files[
#         grepl(
#             bis$accessions[]
#             #"CP049420.1", 
#             delta_files
#             )
#         ]
which(
    grepl(
        "CP049420.1",
        delta_files
        )
    )
plot_delta(delta_file)


chrom_SVs <- 
    Csv[[3]]
Chrom_SVs <- 
    list()
local_SVs <- 
    list()

for (i in 1:length(delta_files)){
    chr_sv <- 
        Chromosome_level_SVs(delta_files[i])
    
    Chrom_SVs[[i]] <- 
        chr_sv
    #Chromosome_level_SVs(delta_files[i])
    local_SVs[[i]] <- 
        delta_SVs(
            delta_table = 
                delta_files[i],
            chrom_SVs = 
                chr_sv
            )
    print(i)
}


LSVs <- 
    bind_rows(local_SVs) %>%
    inner_join(
        ., 
        clustered_genomes[,c(1,2)] %>% 
            filter(accessions %in% prokka_accs) %>%
            #group_by(cluster_id) %>% dplyr::slice(1:3) %>%
            `colnames<-`(c("clust", "acc")),
        by = "acc"
    ) %>%
    rowwise() %>%
    mutate(acc_c = paste0(acc, "_", clust))

filter(LSVs, width >1e4)

    
CSVs <- 
    bind_rows(Chrom_SVs) %>%
    ungroup %>%
    inner_join(
        ., 
        clustered_genomes[,c(1,2)] %>% 
            filter(accessions %in% prokka_accs) %>%
            #group_by(cluster_id) %>% dplyr::slice(1:3) %>%
            `colnames<-`(c("clust", "qid")),
        by = "qid"
    ) %>%
    rowwise() %>%
    mutate(qid_c = paste0(qid, "_", clust))

head(Chrom_SVs)


delta_files[grepl("CP135613.1", delta_files)] %>% read_delta

ggplot(
    LSVs %>% 
        filter(
            width <2000, width>50), #%>% 
        #filter(acc %in%  bis$accessions[1:10]),
        #filter(acc =="CP135613.1"# %in% bis$accessions[1:10]), 
    aes(x = width, y = acc, fill = variant
        )
    ) +
  geom_density_ridges(scale = 1.2, alpha = 0.7, bandwidth = 30) +
  theme_minimal() +
  labs(
    x = "SV width (bp)",
    y = "",
    title = ""
  ) +
    theme_classic() +
    theme(text = element_text(size = 22))



clustered_genomes
Chromosome_level_SVs(delta_file)

local_SVs[[i]] %>% count(variant)


#delta_files[grepl(most_related, delta_files)]

delta_SVs(delta_table = delta_files[1])

SV_table <- 
    pbmclapply(
        delta_files,
        delta_SVs
    ) %>%
    bind_rows


delta_files[grepl("CP049420.1", delta_files)] %>% delta_SVs

delta_SVs(delta_files[3])



```


## Location
```{r}
SV_tib2 <- 
    SV_table %>%
    filter(width>=50) %>%
    mutate(
        midpt = round((start+end)/2)
    ) %>%
    #filter(width<500, width>50, variant=="Insertion" | variant=="Translocation") %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )


SV_range_smallest <- 
    IRanges(
       start = 
           SV_tib2$s2,
       end = 
           SV_tib2$e2
   )
     

```


### Prokka annotation
```{r}
prokka_commands2 <- 
    annotate_ind(
        fasta_file, species_name, prokka_dir
        #species_name, sync_dir, prokka_dir, accessions = c01$accessions
    )



writeLines(
    text = 
        prokka_commands2$comms,
    con = 
        prokka_commands2$file
    )


gff_i <- 
   read.delim(
       sprintf(
           '%s/%s.gff', prokka_dir, most_related
       ), header=F, comment.char = "#")  %>%
   as_tibble() %>%
   `colnames<-`(
       c(
           "accession", "source", "type", "start", 
           "end", "score", "strand", "phase", "attributes")
       ) %>%
   filter(!is.na(start)) %>%
    mutate(
        width = end-start+1,
        gene = str_extract(attributes, "(?<=gene=)[^;]+"),
        product = str_extract(attributes, "(?<=product=)[^;]+"),
        category = 
            case_when(
            grepl("hypothetical", product, ignore.case = TRUE) ~
                "Hypothetical",
            grepl(
                "transposase|insertion sequence", 
                product, ignore.case = TRUE) ~ "Mobile Element",
            grepl(
                "ribosomal|rRNA|ribosome", 
                product, ignore.case = TRUE) ~ "Ribosomal",
            grepl("tRNA", product, ignore.case = TRUE) ~ "tRNA",
            grepl(
                "DnaA|dnaC|replication initiator", 
                product, ignore.case = TRUE) ~ "Replication initiation",
            grepl(
                "recA|recF|mutS|uvrA|uvrB|uvrC|mismatch repair|DNA repair", 
                product, ignore.case = TRUE) ~ "DNA Repair",
            grepl(
                "Dna[BENG]|DNA polymerase|helicase|ligase|gyrase",
                product, ignore.case = TRUE) ~ "DNA Replication",
            grepl(
                "PTS system|transferase|ABC transporter|permease|efflux|transporter", 
                product, ignore.case = TRUE) ~ "Transporter",
            grepl(
                "dehydrogenase|kinase|isomerase|synthetase|transferase|hydratase|lyase|oxidase",
                product, ignore.case = TRUE) ~ "Metabolism",
            grepl(
                "heat shock|chaperone|stress response|oxidative stress|dnaJ|groEL|hsp", 
                product, ignore.case = TRUE) ~ "Stress Response",
            grepl(
                "beta-lactamase|resistance|antibiotic", 
                product, ignore.case = TRUE) ~ "Antibiotic Resistance",
            grepl(
                "lipoprotein|membrane", product, ignore.case = TRUE) ~ "Membrane Protein",
            grepl(
                "regulator|two-component|response regulator|sensor kinase|YycH", 
                product, ignore.case = TRUE) ~ "Regulator",
            TRUE ~ "Other"
            )
    ) %>%
    dplyr::select(-c(attributes)) %>%
    mutate(
        gff_index = row_number()
    )

ref_len <- 
    readDNAStringSet(
        sprintf(
            "%s/%s.txt", sync_dir, most_related
        ), 
    ) %>% width

gff_i %>%
    filter(grepl("repeat", attributes))

gff_i %>% count(category)

gff_i %>% 
    arrange(width)
    filter(
        width <100
        #grepl("hypothetical", product)
        ) 

gff_ranges <- 
   IRanges(
       start = 
           gff_i$start,
       end = 
           gff_i$end
   )

gff_ranges %>% width %>% mean


gff_ranges


intron_regions <- 
   IRanges(
       start = 
           gff_i$start,
       end = 
           gff_i$end
   ) %>% gaps(., start = 1, end = ref_len)

```


```{r}
LSVs %>% 
    pull(acc) %>% unique

SV2 <- 
    #SV_table %>% 
    LSVs %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )# %>%
    #ungroup %>%
    # mutate(
    #     sv_index = row_number()
    # )



SV2 %>% filter(variant=="Local inversion")

SV2 %>% count(variant)

filtered_combined_delta %>% filter(qid=="CP049420.1") %>% plot_delta

SV2 %>% filter(width >15e2, width<2e3)

SV_range <- 
    IRanges(
       start = 
           SV2$s2,
       end = 
           SV2$e2
   )

qids <- 
    unique(
        SV2$acc
    )


gff_ranges <- 
    IRanges(
        start = 
            gff_i$start,
        end = 
            gff_i$end
    )


#### for each QID
## inv, gaps, ins, tlocs, total length, total counts
## variant distributions
## Geenic SVs 

ref_len

sprintf(
    "%s/%s.txt",
    sync_dir, "CP049377.1"
) %>% readDNAStringSet()

sv_info_list <- 
    lapply(
        1:length(qids),
        function(i){
            qid_i <- 
                qids[i]
            
            SV_i <- 
                SV2 %>%
                filter(
                    acc ==qid_i
                    ) %>% 
                ungroup %>%
                arrange(s2) %>%
                mutate(
                    sv_index = row_number()
                    )
            
            ## which columns hit genes
            SV_range <- 
                IRanges(
                   start = 
                       SV_i %>% 
                       # filter(
                       #     acc ==qid_i
                       # ) %>% 
                       pull(s2),
                   end = 
                       SV_i %>% 
                       # filter(
                       #     acc ==qid_i
                       # ) %>% 
                       pull(e2)
               ) 
            
            
            sv_gr <- 
                sort(SV_range)
            
            hits <- 
                findOverlaps(sv_gr, gff_ranges)
            
            SV_gene
            
            
            overlap_ranges <-
                pintersect(
                    sv_gr[queryHits(hits)], 
                    gff_ranges[subjectHits(hits)]
                    )
            
            SV_gene <-
                SV_i %>%
                filter(
                    acc==qid_i
                ) %>%
                ungroup %>%
                filter(width >50) %>%
                mutate(
                    genic = 
                        (sv_index %in% unique(queryHits(hits)))
                )
            
            SV_gene %>% filter(!genic, width >1e3)
            
            SV_gene %>% arrange(sv_index)
            length(unique(queryHits(hits)))/
                nrow(SV_gene)
            
            SV_gene
            
            
            SV_i_summary <- 
                SV_gene %>% 
                group_by(acc, variant) %>%
                summarise(
                    variant_count = 
                        n(),
                    variant_bp = 
                        sum(width),
                    genic_prop = 
                        sum(genic)/variant_count,
                    intergenic_prop = 
                        sum((!genic)/variant_count),
                    .groups = "keep"
                ) 
            
            sv_lens <- 
                SV_gene %>%
                mutate(
                    bin = 
                        cut(
                            width, 
                            breaks = 
                                c(50, 500, 1000, 2000, 10000, Inf), 
                            right = FALSE)
                    ) %>%
                group_by(acc,bin) %>%
                summarise(
                    SV_count = n(),
                    SV_coverage = sum(width),
                    SV_types = 
                        paste(unique(variant), collapse = ", "),
                    .groups = "keep"
                    #intergenic_count = sum(!genic),
                    #genic_cov = ifelse(genic, sum(width))
                )
            
            sv_lens_genic <- 
                SV_gene %>%
                mutate(
                    bin = 
                        cut(
                            width, 
                            breaks = 
                                c(50, 500, 1000, 2000, 10000, Inf), 
                            right = FALSE)
                    ) %>%
                group_by(acc, bin, genic) %>%
                summarise(
                    SV_count = n(),
                    SV_coverage = sum(width),
                    .groups = "keep"
                    #intergenic_count = sum(!genic),
                    #genic_cov = ifelse(genic, sum(width))
                ) %>%
                mutate(
                    bp_region = 
                        ifelse(
                            genic,
                            filter(
                                genome_coverage_df, 
                                region =="Genic regions"
                                ) %>% pull(bp),
                            filter(
                                genome_coverage_df, 
                                region =="Intergenic regions"
                                ) %>% pull(bp)
                            ),
                    cov_prop = 
                        round(100*SV_coverage/bp_region, 2)
                    )
                    
            out_list <- 
                list(
                    SV_tib = 
                        SV_gene,
                    SV_len_tib = 
                        sv_lens,
                    genic_tib = 
                        sv_lens_genic
                )
            
            return(out_list)
        }
        )


sv_info_list[1]
qids

len_tib <- 
    lapply(
        sv_info_list,
        function(x){
            x[["SV_len_tib"]]
        }
    ) %>%
    bind_rows %>%
    ungroup %>%
    mutate(bin = factor(bin)) %>%
    group_by(acc) %>%
    mutate(ordering = SV_count[bin == "[50,500)"]) %>%
    ungroup() %>%
    mutate(acc = fct_reorder(acc, ordering, .fun = mean, na.rm = TRUE))

mean_svs <- 
    len_tib %>% 
    group_by(acc) %>% 
    summarise(
        avg_svs = 
            mean(sum(SV_count))
    )



ggplot(len_tib, aes(x = bin, y = SV_count, fill = bin)) +
  #geom_boxplot(width = 0.4, size = 2) +
    geom_boxplot(width = 0.4, size = 0.8, alpha = 0.5, outlier.shape = NA) +
  scale_fill_brewer(palette = "Set1") +
    # scale_fill_manual(
    #     values = 
    #         c(
    #             "#440154FF", "#31688EFF", "#35B779FF", "#FDE725FF", "#21908CFF"
    #           )
    #     ) +
  labs(
    x = "SV length (bp)",
    y = "SV Count",
    title = ""
  ) +
    theme_classic() +
    theme(
        # axis.text.x = 
        #     element_text(
        #         angle = 15, 
        #         hjust = 1
        #         ),
        text = element_text(size = 22),
        legend.position = "none",
        #axis.title.x = element_text(margin = margin(t = 15))
        ) +
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = ">10,000"
    )
  ) 


genic_tib %>% filter(bin=="[2e+03,1e+04)")


sv_info_list[1]


full_SV_tib %>% 
    filter(!genic, width>1e4)

full_SV_tib %>%
    filter(acc =="CP049421.1", width >1e4)


genic_tib <- 
    lapply(
        sv_info_list,
        function(x){
            x[["genic_tib"]]
        }
    ) %>%
    bind_rows %>%
    ungroup %>%
    mutate(
        Region = 
            ifelse(
                genic,
                "Genic",
                "Intergenic"
            ),
        prop = 
            100*SV_coverage / bp_region,
        uc_prop = 
            100 - prop
    )
    # mutate(bin = factor(bin)) %>%
    # group_by(acc) %>%
    # mutate(ordering = SV_count[bin == "[50,500)"]) %>%
    # ungroup() %>%
    # mutate(acc = fct_reorder(acc, ordering, .fun = mean, na.rm = TRUE))


ggplot(
    genic_tib, #%>% filter(acc =="CP049377.1"),
    aes(x = bin, y = cov_prop, fill = Region)) +
  stat_summary(fun = mean, geom = "bar", position = "dodge") +
    scale_fill_manual(values = c("#440154FF", "#21908CFF")) +
    #scale_fill_brewer(palette = "Dark1", labels = c("Intergenic", "Genic")) +
  theme_classic() +
  labs(
    x = "SV length (bp)",
    y = "SV coverage (%)",
    #fill = "Region",
    title = ""
  ) +
  theme(
    axis.title.x = element_text(margin = margin(t = 15)),
    text = element_text(size = 22)
  )+
    scale_x_discrete(
    labels = c(
      "[50,500)" = "50–500",
      "[500,1e+03)" = "500–1,000",
      "[1e+03,2e+03)" = "1,000–2,000",
      "[2e+03,1e+04)" = "2,000–10,000",
      "[1e+04,Inf)" = ">10,000"
    )
  ) #+
    #facet_wrap(~acc)




full_SV_tib <- 
    lapply(
        sv_info_list,
        function(x){
            x[["SV_tib"]]
        }
    ) %>%
    bind_rows %>%
    ungroup %>%
    mutate(
        Region = 
            ifelse(
                genic,
                "Genic",
                "Intergenic"
            )
    )

library(GenomicRanges)
sv_gr <- 
    GRanges(
        seqnames = full_SV_tib$acc, 
        ranges = IRanges(start = full_SV_tib$start, end = full_SV_tib$end),
        acc = full_SV_tib$acc,
        variant = full_SV_tib$variant
        )

sv_clusters <- 
    reduce(sv_gr, min.gapwidth = 1000) 

overlaps <- 
    findOverlaps(sv_gr, sv_clusters)
full_SV_tib$cluster_id <- 
    subjectHits(overlaps)

binary_matrix <- 
    full_SV_tib %>%
    distinct(acc, cluster_id) %>%        
    mutate(present = 1) %>%              
    pivot_wider(                         
        names_from = acc, 
        values_from = present, 
        values_fill = 0
        ) %>%
    column_to_rownames(var = "cluster_id")

binary_matrix_t <- 
    t(binary_matrix)

# Step 2: Jaccard distance calculation
jaccard_dist <- 
    vegdist(binary_matrix_t, method = "jaccard", binary = TRUE)


hclust(jaccard_dist) %>% plot
```


```{r}

## snp subset 1
snp_dir_ss1 <- 
    sprintf(
        "%s/ss1",
        snp_dir
    )
if (!dir.exists(snp_dir_ss1)){dir.create(snp_dir_ss1)}

file.copy(
    sprintf(
        "%s/%s.txt",
        sync_dir, 
        most_related
        #qids
    ),
    to = 
        snp_dir_ss1,
    overwrite = T
)

list.files(snp_dir_ss1)


parsnp  -p 6 \
    -r ../data/processing/snp/Bacillota/Staphylococcus/Staphylococcus_aureus/ss1/NZ_CP084878.1.txt \
    -c \
    -d ../data/processing/snp/Bacillota/Staphylococcus/Staphylococcus_aureus/ss1 \
    -o ../data/processing/snp/Bacillota/Staphylococcus/Staphylococcus_aureus/ss1/out







```




```{r}
full_SV_tib %>% 
    dplyr::select(start, end, width, variant, acc, genic) %>%
    count(start, end) %>% arrange(desc(n))






sv_clusters <- 
    reduce(sv_gr, min.gapwidth = 1000) 


sv_id <- 
    full_SV_tib %>% 
    # group_by(acc) %>%
    # summarise(
    #     delta_id = unique(delta_identity)
    # ) %>% 
    left_join(
        ., 
        species_identities$id_dists[c(1,2)] %>%
        `colnames<-`(c("acc", "Id_dist")),
        by = "acc"
    ) %>%
    ungroup %>%
    mutate(
        MASH_identity = 
            100-Id_dist
    ) %>%
    dplyr::select(c(acc, width, variant, Region, MASH_identity))

sv_id %>%
  group_by(acc, MASH_identity) %>%
  summarise(
    mean_width = mean(width),
    n_SVs = n(),
    prop_genic = mean(Region == "Genic"),
    variant_types = n_distinct(variant)
  )



library(lme4)
model <- 
    lmer(MASH_identity ~ width + region + variant + (1 | acc), data = df)


library(randomForest)
rf_sv <- 
    randomForest(MASH_identity ~ width + Region + variant, data = sv_id)


rf_sv$predicted <- 
    predict(rf_sv, newdata = sv_id)

ggplot(sv_id, aes(x = rf_sv$predicted, y = MASH_identity)) +
  geom_point(alpha = 0.6) +
  geom_smooth(method = "lm", color = "red", se = FALSE) +
  theme_classic() +
  labs(
    x = "Predicted MASH Identity",
    y = "Observed MASH Identity",
    title = "Random Forest Predictions vs Observed"
  )


print(rf_sv)


varImpPlot(rf_sv)



plot(sv_id$delta_id, sv_id$MASH_identity)

full_SV_tib %>% count(acc, delta_identity)


```


```{r}
sv_counts <- 
    colSums(binary_matrix)


snp_tree <- 
    read.tree(
        "../data/processing/snp/Bacillota/Staphylococcus/Staphylococcus_aureus/ss1/out/parsnp.tree"
        )


sv_tree <- hclust(jaccard_dist)
names(sv_tree)

 snp_labels<- snp_tree$tip.label
sv_labels <- sv_tree$labels

snp_labels_clean <- 
    gsub("\\.txt$|\\.ref$|\\.fa$|\\.txt$", "", snp_labels) %>%
    gsub(".txt", "", .)

snp_tree$tip.label <- 
    snp_labels_clean

setequal(snp_labels_clean, sv_labels)


common_labels <- 
    intersect(snp_labels, sv_labels)
snp_tree_pruned <- keep.tip(snp_tree, common_labels)
sv_tree_phylo <- as.phylo(sv_tree)
sv_tree_pruned <- keep.tip(sv_tree_phylo, common_labels)



snp_dend <- as.dendrogram(as.hclust(chronos(snp_tree_pruned)))
sv_dend  <- as.dendrogram(sv_tree_pruned)

# Match tip order (optional but improves clarity)
sv_dend <- rotate(sv_dend, order.dendrogram(snp_dend))

# Plot tanglegram
tanglegram(snp_dend, sv_dend,
           main_left = "SNP Tree (Parsnp)",
           main_right = "SV Tree (Jaccard+hclust)",
           highlight_distinct_edges = TRUE,
           common_subtrees_color_lines = TRUE,
           margin_inner = 7,
           lab.cex = 0.8)

dist.topo(snp_tree_pruned, sv_tree_pruned)


setequal(snp_tree_pruned$tip.label, sv_tree_pruned$tip.label)


ggtree(tt) +
  geom_tiplab() +
  geom_tippoint(aes(color = sv_counts[tt$tip.label]))

ggtree(tt) %<+% 
  tibble(label = names(sv_counts), sv_count = as.numeric(sv_counts)) +
  geom_tiplab() +
  geom_tippoint(aes(color = sv_count), size = 3) +
  scale_color_viridis_c() +
  theme_tree2()

sv_counts <- colSums(binary_matrix_df)

ggtree(tree) +
  geom_tiplab() +
  geom_tippoint(aes(color = sv_counts[tree$tip.label]))


```







```{r}    
    
        small_svs_gene_hits <- 
            SV_GFF %>%
            filter(
                width>50, width<1000
                ) %>% 
            arrange(sv_index)
        
        (small_svs_gene_hits %>% pull(width) %>% sum)/2258449
        
        small_svs_outside <- 
            SV2 %>% 
           filter(
               acc ==qid_i,
               width>50, width<300,
               !(sv_index %in% queryHits(hits))
           )
        (small_svs_outside %>% pull(width) %>% sum)/414774
        
        
        SV2 %>% 
            filter(
               acc ==qid_i,
               !(sv_index %in% queryHits(hits))
           ) %>% arrange(desc(width))
        
        
        SV2 %>% 
            filter(
               acc ==qid_i,
           )  %>%
            mutate(
                in_gene = 
                    (sv_index %in% queryHits(hits))
            )
        
        
        
        
        
        
ggplot(gga %>% filter(width>50, width<1000), aes(x = width, fill = in_gene)) +
  geom_density(alpha = 0.6, bw = 5) +
  scale_fill_brewer(palette = "Set1") +
  scale_x_continuous() +
  theme_minimal() +
  labs(
    x = "SV width (bp)",
    y = "Density",
    fill = "Inside gene?",
    title = "Distribution of SV widths by in_gene status"
  )
            
        
    }
)


genome_coverage_df


    
sv_gr <- 
    SV_range
gene_gr <- 
    gff_ranges  

hits <- findOverlaps(sv_gr, gff_ranges)

# Get the actual overlap regions
overlap_ranges <-
    pintersect(sv_gr[queryHits(hits)], gff_ranges[subjectHits(hits)])
top_species$Species[1:5] %>% paste(., collapse = ", ")
SV_GFF <- 
    tibble(
        sv_index = queryHits(hits),
        gff_index = subjectHits(hits),
        overlap_bp = width(overlap_ranges),
        ) %>%
    left_join(
        ., SV2, by = "sv_index"
    ) %>%
    left_join(
        ., 
        gff_i %>%
            dplyr::select(
                -c(
                    "accession", "type", "source", "phase", "score"
                    )
                ) %>%
            `colnames<-`(
                c(
                    "gs", "ge", "str", "gw", "gene", "prod", "cat", "gff_index"
                    )
                ),
        by = "gff_index"
    )

SV_GFF %>%
    filter(
        width>50, width <200
    ) %>%
    filter(is.na(sv_index))


SV_GFF %>% filter()


clustered_genomes

## which SVs don't hit a gene
SV_GFF_summary <- 
    SV2 %>%
    mutate(
        overlaps_gene = 
            sv_index %in% queryHits(hits)
        # gene_hit = 
        #     ifelse(
        #         overlaps_gene, 
        #         SV_GFF %>% filter(sv_index==sv_index, gff_index==)
        #     )
    )
    # filter(
    #     width>50, width<200,
    #    !( sv_index %in% queryHits(hits))
    # )


SV_GFF_summary %>% filter(!overlaps_gene, width >10000)

small_svs <- 
    SV_GFF_summary %>% filter(width>50, width<200)

genes_merged <- reduce(gff_ranges)

# Base-pair level coverage
gene_coverage <- sum(width(genes_merged))
intergenic_coverage <- 
    ref_len - gene_coverage

genome_coverage_df <- 
    tibble(
        region = c("Gene regions", "Intergenic regions"),
        bp = c(gene_coverage, intergenic_coverage)
        ) %>%
  mutate(proportion = bp / sum(bp))

SV_Ols_small_gene <- 
    IRanges(
        start = 
            (small_svs %>% filter(overlaps_gene) %>% pull(s2)),
        end = 
            (small_svs %>% filter(overlaps_gene) %>% pull(e2))
    ) %>% reduce() %>% width %>% sum

SV_NO_overlap_small_gene <- 
    IRanges(
        start = 
            (small_svs %>% filter(!overlaps_gene) %>% pull(s2)),
        end = 
            (small_svs %>% filter(!overlaps_gene) %>% pull(e2))
    ) %>% reduce() %>% width %>% sum
    
gcd <- 
    genome_coverage_df %>%
    mutate(
        SV = c(
            SV_Ols_small_gene, 
            SV_NO_overlap_small_gene
        ),
        SV_prop = 
            round(100*SV/bp, 2)
    ) 
    


######




### getting the overlap of sv_index

        
        # SV_GFF <- 
        #     tibble(
        #         sv_index = queryHits(hits),
        #         gff_index = subjectHits(hits),
        #         overlap_bp = width(overlap_ranges),
        #         ) %>%
        #     left_join(
        #         ., SV2 %>% filter(acc ==qid_i ), by = "sv_index"
        #     ) %>%
        #     left_join(
        #         ., 
        #         gff_i %>%
        #             dplyr::select(
        #                 -c(
        #                     "accession", "type", "source", "phase", "score"
        #                     )
        #                 ) %>%
        #             `colnames<-`(
        #                 c(
        #                     "gs", "ge", "str", "gw", 
        #                     "gene", "prod", "cat", "gff_index"
        #                     )
        #                 ),
        #         by = "gff_index"
        #     )












# Calculate % overlap relative to SV and Gene
merged_df <- bind_cols(
  SV2[queryHits(hits), ],
  gff_i[subjectHits(hits), ]
) %>%
  mutate(
    overlap_bp = width(overlap_ranges),
    percent_overlap_sv = overlap_bp / width(sv_ranges[queryHits(hits)]),
    percent_overlap_gene = overlap_bp / width(gff_ranges[subjectHits(hits)])
  )


# SV-level overlap
hits <- findOverlaps(SV_range, gff_ranges)

# Mark SVs with overlap or not
sv_df <- 
    SV2 %>%
    mutate(
        overlaps_gene = 
            ifelse(1:length(SV_range) %in% queryHits(hits), "Overlaps gene", "No overlap")
        ) %>%
    filter(width >15e2, width<2e3)


sv_summary <- 
    sv_df %>%
    group_by(overlaps_gene) %>%
 summarise(
    small_invs = sum(width > 50 & width<300),
    sum_low = sum(width[width > 50 & width<300]),
    count_mid = sum(width > 300 & width<600),
    count_high = sum(width > 600 & width<1000)
  ) %>%
    arrange(desc(sum_low)) %>%
    dplyr::select(overlaps_gene, sum_low) %>% bind_cols(., genome_coverage_df) %>%
    mutate(
        Inside_gene_prop = 
            c(0.56, 0.65)
    )  %>% dplyr::select(overlaps_gene, sum_low, Inside_gene_prop)


ggplot(sv_summary, aes(x = overlaps_gene, y = bp, fill = Inside_gene_prop)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent) +
  labs(
    x = "SV overlaps gene",
    y = "Proportion of bp",
    fill = "Proportion covered"
  ) +
  theme_classic() +
    theme(text = element_text(size = 22))
   

sv_df %>% 
    #filter(overlaps_gene=="No overlap") %>% 
    count(variant)

sv_overlap_summary <- 
    sv_df %>%
    filter(width>=50, width<=500) %>%
  count(overlaps_gene) %>%
  mutate(proportion = n / sum(n))




combined_df <- 
    bind_rows(
        sv_overlap_summary %>%
            mutate(type = "SV Overlaps", group = overlaps_gene) %>%
            dplyr::select(type, group, proportion),
  genome_coverage_df %>%
    mutate(type = "Genome Coverage", group = region) %>%
    dplyr::select(type, group, proportion)
)

ggplot(combined_df, aes(x = type, y = proportion, fill = group)) +
  geom_bar(stat = "identity", width = 0.6) +
  theme_minimal() +
  labs(
    title = "Proportions of SV overlaps vs. genome coverage",
    x = NULL,
    y = "Proportion",
    fill = "Category"
  ) +
  scale_y_continuous(labels = scales::percent) +
  theme(axis.text.x = element_text(size = 12))


combined_df <- tibble(
  category = c("SV overlaps gene", "SV no overlap"),
  sv_proportion = sv_overlap_summary$proportion,
  genome_proportion = genome_coverage_df$proportion
) %>%
  mutate(
    group = case_when(
      category == "SV overlaps gene" ~ "Gene region",
      category == "SV no overlap" ~ "Intergenic region"
    )
  )
plot_df <- combined_df %>%
  pivot_longer(
    cols = c(sv_proportion, genome_proportion),
    names_to = "type",
    values_to = "proportion"
  )


final_df <- tibble(
  category = c("SV overlaps gene in gene regions", "SV no overlap in intergenic regions"),
  proportion_sv = sv_overlap_summary$proportion,
  proportion_genome = genome_coverage_df$proportion
)

plot_df <- final_df %>%
  pivot_longer(
    cols = starts_with("proportion"),
    names_to = "type",
    values_to = "proportion"
  )

ggplot(plot_df, aes(x = category, y = proportion, fill = type)) +
  geom_bar(stat = "identity", position = "dodge", width = 0.6) +
  theme_minimal() +
  labs(
    title = "SV overlaps aligned with gene/intergenic regions",
    x = NULL,
    y = "Proportion",
    fill = "Source"
  ) +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = c("proportion_sv" = "#1f78b4", "proportion_genome" = "#33a02c"),
                    labels = c("SV dataset", "Genome regions")) +
  theme(axis.text.x = element_text(size = 11, angle = 20, hjust = 1))


ggplot(plot_df, aes(x = type, y = proportion, fill = group)) +
  geom_bar(stat = "identity", width = 0.6) +
  theme_minimal() +
  labs(
    title = "SV overlaps and genome region proportions",
    x = NULL,
    y = "Proportion",
    fill = "Region type"
  ) +
  scale_x_discrete(labels = c("sv_proportion" = "SV Overlaps", "genome_proportion" = "Genome Coverage")) +
  scale_y_continuous(labels = scales::percent) +
  theme(axis.text.x = element_text(size = 12))


genome_coverage_df %>%
  ggplot(aes(x = region, y = proportion, fill = region)) +
  geom_bar(stat = "identity", width = 0.6) +
  theme_minimal() +
  labs(title = "Genome coverage by region", x = "Genome region", y = "Proportion") +
  scale_y_continuous(labels = scales::percent) +
  guides(fill = "none")


sv_overlap_summary %>%
  ggplot(aes(x = overlaps_gene, y = proportion, fill = overlaps_gene)) +
  geom_bar(stat = "identity", width = 0.6) +
  theme_minimal() +
  labs(title = "Proportion of SVs overlapping genes", x = "SV category", y = "Proportion") +
  scale_y_continuous(labels = scales::percent) +
  guides(fill = "none")




# Summarize SV-level proportions
sv_overlap_summary <- sv_df %>%
  count(overlaps_gene) %>%
  mutate(proportion = n / sum(n))




SV_tib2_vs <- 
    SV2 %>%
    filter(width>=50, width<1000) %>%
    mutate(
        midpt = round((start+end)/2)
    ) %>%
    #filter(width<500, width>50, variant=="Insertion" | variant=="Translocation") %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )

SV_tib2_vs %>% pull(width) %>% density %>% plot

SV_range_smallest <- 
    IRanges(
       start = 
           SV_tib2_vs$s2,
       end = 
           SV_tib2_vs$e2
   )

overlaps <- findOverlaps(sv_gr, gene_gr)
sv_table_df <- 
    as.data.frame(SV_range_smallest) %>%
  mutate(overlaps_gene = ifelse(seqnames %in% seqnames(gff_ranges[subjectHits(overlaps)]), "Overlaps gene", "No overlap"))

sv_overlap_summary <- sv_table_df %>%
  count(overlaps_gene) %>%
  mutate(proportion = n / sum(n))

     
sv_ol <- 
    findOverlaps(SV_range_smallest, gff_ranges)


SV_tib2_vs2 <- 
    SV_tib2_vs %>%
    mutate(queryHits = row_number())

gff_tib <- 
    gff_i %>%
    mutate(subjectHits = row_number())


SV_gff_vs <- 
    sv_ol %>%
    as.data.frame() %>% 
    as_tibble %>%
    full_join(., SV_tib2_vs2, by = "queryHits") %>% #filter((queryHits==1))
    left_join(., gff_tib, by = "subjectHits") %>% arrange(s2)

SV_gff_vs %>% count(category)

overlap_widths <- 
    pintersect(SV_range_smallest[queryHits(sv_ol)], gff_ranges[subjectHits(sv_ol)])

percent_overlap <-
    tapply(width(overlap_widths), queryHits(sv_ol), sum) / width(SV_range_smallest)[unique(queryHits(sv_ol))] * 100

sum(width(overlap_widths))/sum(width(SV_range_smallest))

SV_tib2_vs <- 
    SV_tib %>%
    filter(width<500, width>50, variant=="Insertion" | variant=="Translocation") %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )

sv_ol<- 
    findOverlaps(SV_range_smallest, gff_ranges)


overlap_widths <- 
    pintersect(SV_range_smallest[queryHits(sv_ol)], gff_ranges[subjectHits(sv_ol)])

percent_overlap <-
    tapply(width(overlap_widths), queryHits(sv_ol), sum) / width(SV_range_smallest)[unique(queryHits(sv_ol))] * 100

sum(width(overlap_widths))/sum(width(SV_range_smallest))

SV_tib2_vs <- 
    SV_tib %>%
    filter(
        width<500, width>50, 
        variant=="Insertion" | variant=="Translocation"
        ) %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )

small_invs <- 
    SV_tib %>%
    filter(
        width<500, width>50, 
        variant!="Insertion",
        variant!="Translocation",
        variant!="Gap"
        ) %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )



SV_range_smallest <- 
    IRanges(
       start = 
           SV_tib2_vs$s2,
       end = 
           SV_tib2_vs$e2
   )
     
sv_ol<- 
    findOverlaps(SV_range_smallest, gff_ranges)

SV_range_smallest[3]
gff_ranges[172]

sv_vs <- 
    SV_range_smallest %>%
    as.data.frame() %>%
    as_tibble %>%
    mutate(
        overlap = 
            ifelse(
                seq_along(SV_range_smallest) %in% queryHits(sv_ol),
                "Inside Gene", "Outside Gene"
                )

    ) %>%
    add_count(overlap)



in_out_50_500 <- 
    sv_vs %>%
    group_by(overlap) %>%
    summarise(
        
    )
    count(overlap)

    
ggplot(sv_vs, aes(x = overlap, fill = overlap)) +
  geom_bar() +  # Simple bar plot with blue bars
  labs(x = "", y = "Count", title = " SV >50nt, SV <500 nt") +
  theme_classic()+
    theme(
        text = element_text(size = 22),
        legend.position = "none"
        ) 

```







```{r}


delta_table <- 
    read_delta(
        delta_files[
            grepl(qrys[1], delta_files) &
                grepl(most_related, delta_files)
            ]
        )
Capture_SVs <- 
    function(
      qrys, 
      delta_files,
      fcd
    ){
        all_svs <- 
            list()
        
        event_list <- 
            list()
        
        overlap_list <- 
            list()
        
        ins_list <- 
            list()
        
        gap_list <- 
            list()
        
        
        inv_list <- 
            list()
        
        tloc_list <- 
            list()
        
        pb <- txtProgressBar(min = 0, max = length(qrys), style = 3)

# which(qrys=="CP049108.1")
# 
# i=1
        for (i in 1:length(qrys)){#qrys))
            ft <- 
                read_delta(
                    delta_files[
                        grepl(qrys[i], delta_files) &
                            grepl(most_related, delta_files)
                        ]
                    )
            
            fcdi <- 
                fcd %>%
                filter(qid==qrys[i])
                
            clust_i <- 
                best_identity %>%
                filter(accessions==qrys[i]) %>%
                pull(cluster_id)
                
            d2 <- 
              ft %>%
              rowwise() %>%
              mutate(
                qs2 = min(c(qs, qe)),
                qe2 = max(c(qs, qe))
              ) %>% 
              ungroup
            
        ## gap insertion positions
        # insertions
            ref_ranges <- 
                IRanges(
                    start = 
                    ft$rs,
                    end = 
                    ft$re
                    )#
            qry_ranges <- 
              IRanges(
                start = 
                  d2$qs2,
                end = 
                  d2$qe2
              ) 
            
            ins <- 
                ref_ranges %>%
                reduce() %>%
                gaps(start = 1, end = ft$rlen[1]) 
            ins_df <- 
                ins %>%
                as.data.frame() %>% 
                mutate(
                    acc = ft$qid[1],
                    clust = clust_i
                    ) %>%
                mutate(
                    variant = "Insertion"
                )
        
         gaps <- 
             qry_ranges %>%
             reduce() %>%
             gaps(start = 1, end = ft$qlen[1])
         
         gaps_df <- 
             gaps %>%
             as.data.frame() %>% 
             mutate(
                 acc = ft$qid[1],
                 clust = clust_i
                 ) %>%
            mutate(
                variant = "Gap"
            )
         
     qlength <- 
         d2$qlen[1]
     
     rlength <- 
         d2$rlen[1]
     
     ref_covered <- 
         reduce(ref_ranges) %>% 
         width() %>% sum
     
     ref_uncovered <- 
        rlength-ref_covered
     
     qry_covered <- 
         reduce(qry_ranges) %>% 
         width() %>% sum
     
     qry_uncovered <- 
        qlength-qry_covered
     
     error_perc <- 
         sum(
             ft$error
             )/
         mean(c(rlength, qlength))
     
     length_dif_perc <- 
         abs(rlength-qlength)/
         mean(c(rlength, qlength))
     
     delta_identity <- 
         (mean(c(ref_covered, qry_covered))/
              mean(c(rlength, qlength))) -
         error_perc - length_dif_perc
         
     
     
     ## overlaps 
     qols <-
         findOverlaps(qry_ranges, type = "any")
     qunique <- queryHits(qols) != subjectHits(qols)

     qor <- 
         pintersect(
             qry_ranges[queryHits(qols)][qunique], 
             qry_ranges[subjectHits(qols)][qunique]
             ) %>%
         unique
     
     # readDNAStringSet(
     #     sprintf(
     #         "%s/%s.txt", 
     #         sync_dir,
     #         qrys[i]
     #         )
     #     ) %>%
     #     subseq(920006, 940006)
     # qrys[i]
     # qor2 <- 
     #     qor[width(qor) > 0]
     
     qor_tib <- 
         qor %>%
         as.data.frame()
         
     overlap_list[[qrys[i]]] <-
        qor_tib
     
     
     
    #      
    #  ttt <-
    #      ref_ranges
    #  IRanges(
    #      start = c(1,5,8),
    #      end = c(7,9,12)
    #  )
    # 
    #  pintersect(
    #      ttt[queryHits(findOverlaps(ttt))],
    #      ttt[subjectHits(findOverlaps(ttt))]
    #      ) %>%
    #      unique()
    # 
    #  IRanges(
    #      start = c(5,8)
    #      end = c(7,9)
    #  )
    # 
    # pintersect(ß
    #     qry_ranges[queryHits(findOverlaps(qry_ranges))],
    #     qry_ranges[subjectHits(findOverlaps(qry_ranges))]
    #     )
    # print(overlapping_pairs)

    ins_list[[qrys[i]]] <-
        ins_df

    gap_list[[qrys[i]]] <-
        gaps_df
    
    invs1 <- 
        fcdi  %>%
        filter(
            strand=="-",
            meanlen >5e4
        ) %>%
        mutate(
            RASR = 
                #abs((rlen-re) - rs ) < 2e5 & 
            (X_dist < 2e5) & (strand=="-"),
            cluster = clust_i
        ) %>%
        dplyr::select(
            c(
                "rid", qid, "strand", rs, re, qs, qe,
                meanlen, rlen, invs, cluster, RASR
                )
            ) %>%
        dplyr::rename(
           "acc" =  "qid",
           "clust" = "cluster"
        ) %>%
        rowwise() %>%
        mutate(
            start = round(mean(c(rs, qe))),
            end = round(mean(c(re, qs)))
            )  %>%
        ungroup()
            
        # mutate(
        #     abs((rlen-re) - rs ) < 2e5,
        #     X_dist < 2e5,
        #     strand=="-"
        #     ) %>%
        # mutate(cluster = za$cluster_id[i]) %>%
        # dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
    inv_summary <- 
        invs1
    
    ### inner_inversions
    
    
    if (nrow(invs1)>1){
        inner_invs1 <- 
            fcdi %>%
            # filtered_combined_delta %>% 
            #             ungroup %>%
            #             filter(qid==zt1$qid[i]) %>%
            filter(
                strand=="+",
                meanlen >5e4,
                lag(strand=="-") & lead(strand=="-")  
            ) %>%
            mutate(
                RASR = 
                    #abs((rlen-re) - rs ) < 2e5 & 
                (X_dist < 2e5) & (strand=="+"),
                cluster = clust_i
            )%>%
            dplyr::select(
                c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster, RASR)
                ) %>%
            dplyr::rename(
               "acc" =  "qid",
               "clust" = "cluster"
            ) %>%
            rowwise() %>%
            mutate(
                start = round(mean(c(rs, qs))),
                end = round(mean(c(re, qe)))
                )  %>%
            ungroup()
        inv_summary <- 
            bind_rows(invs1, inner_invs1) %>% arrange(rs)
        
        # plot(
        #     tibble(
        #         starts = 
        #             c(invs1$rs, invs1$qe),
        #         ends = 
        #             c(invs1$re, invs1$qs)
        #         )
        #     )
        
        outer_invs <- 
            invs1
        
        
    }
    
    # inv_list[[qrys[i]]] <- 
    #     inv_summary
    # 
    # gaps_df
    
    invs2 <- 
        inv_summary %>%
        mutate(
            width = abs(start-end+1),
            variant = 
                ifelse(
                    RASR, 
                    "RASR",
                    "Local inversion"
                    )
                
            ) %>%
        dplyr::select(c(start, end, width, acc, clust, variant)) %>%
        mutate(
            variant =
                as.character(variant)
        )
    
    inv_list[[qrys[i]]] <-
        invs2
    
        
    tlocs <- 
        ft %>% 
        filter(
            X_dist >2e5#,
            #strand=="+"
            ) %>%
        mutate(
            clust = clust_i,
            acc = qid
        )%>%
        rowwise() %>%
        mutate(
            start = round(mean(c(rs, qe))),
            end = round(mean(c(re, qs))),
            width =  abs(start-end+1),
            variant = "Translocation"
            )  %>%
        ungroup() %>%
        dplyr::select(c(start, end, width, acc, clust, variant))

    
    tloc_list[[qrys[i]]] <-
        tlocs
    
    if (
        nrow(invs2)==0 &
        nrow(gaps_df)==0 &
        nrow(ins_df)==0 &
        nrow(tlocs)==0 
    ){
        all_events <- 
            tibble(
                start = NA,
                end = NA,
                width = NA,
                acc = qrys[i],
                clust = clust_i,
                variant = NA
            ) %>%
            mutate(
                delta_identity = delta_identity
                )
    } else {
        
    all_events <- 
        bind_rows(
            invs2, 
            gaps_df,
            ins_df,
            tlocs
        ) %>%
        arrange(start) %>%
        mutate(
            delta_identity = delta_identity
        )
    }
    
   
    #     filter(
    #         abs((rlen-re) - rs ) < 2e5,
    #         X_dist < 2e5,
    #         strand=="-"
    #         ) %>%
    #     mutate(cluster = za$cluster_id[i]) %>%
    #     dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
    # 
    # ### ## flanking inversions
    # 
    # filtered_combined_delta %>% 
    #                 ungroup %>%
    #                 filter(qid==zt1$qid[i]) %>%
    #     mutate(
    #         inv_flanked = 
    #         lag(strand=="-") & lead(strand=="-")  
    #         )
    
    event_list[[qrys[i]]] <- 
        all_events
        
    
    setTxtProgressBar(pb,i)
        
    #print(i)

    }
        
    }

all_svs <- 
    list()

event_list <- 
    list()

overlap_list <- 
    list()

ins_list <- 
    list()

gap_list <- 
    list()


inv_list <- 
    list()

tloc_list <- 
    list()



which(qrys=="NZ_CP041815.1")


delta_files <- 
    list.files(
        sprintf("%s/ref_v_all", delta_dir), 
        full.names = T
        )


clustered_genomes %>% filter(cluster_id==1)

pb <- txtProgressBar(min = 0, max = length(qrys), style = 3)

# which(qrys=="CP049108.1")
# 
# i=1
for (i in 1:length(qrys)){#qrys))
    
    ft <- 
        read_delta(
            delta_files[
                grepl(qrys[i], delta_files) &
                    grepl(most_related, delta_files)
                ]
                    )
    
    fcdi <- 
        fcd %>%
        filter(qid==qrys[i])
        
    clust_i <- 
        best_identity %>%
        filter(accessions==qrys[i]) %>%
        pull(cluster_id)
        
    d2 <- 
      ft %>%
      rowwise() %>%
      mutate(
        qs2 = min(c(qs, qe)),
        qe2 = max(c(qs, qe))
      ) %>% 
      ungroup
    
    ## gap insertion positions
    # insertions
    ref_ranges <- 
        IRanges(
            start = 
            ft$rs,
            end = 
            ft$re
            )#
    qry_ranges <- 
      IRanges(
        start = 
          d2$qs2,
        end = 
          d2$qe2
      ) 
    
    ins <- 
        ref_ranges %>%
        reduce() %>%
        gaps(start = 1, end = ft$rlen[1]) 
    ins_df <- 
        ins %>%
        as.data.frame() %>% 
        mutate(
            acc = ft$qid[1],
            clust = clust_i
            ) %>%
        mutate(
            variant = "Insertion"
        )
    
     gaps <- 
         qry_ranges %>%
         reduce() %>%
         gaps(start = 1, end = ft$qlen[1])
     
     gaps_df <- 
         gaps %>%
         as.data.frame() %>% 
         mutate(
             acc = ft$qid[1],
             clust = clust_i
             ) %>%
        mutate(
            variant = "Gap"
        )
     
     qlength <- 
         d2$qlen[1]
     
     rlength <- 
         d2$rlen[1]
     
     ref_covered <- 
         reduce(ref_ranges) %>% 
         width() %>% sum
     
     ref_uncovered <- 
        rlength-ref_covered
     
     qry_covered <- 
         reduce(qry_ranges) %>% 
         width() %>% sum
     
     qry_uncovered <- 
        qlength-qry_covered
     
     error_perc <- 
         sum(
             ft$error
             )/
         mean(c(rlength, qlength))
     
     length_dif_perc <- 
         abs(rlength-qlength)/
         mean(c(rlength, qlength))
     
     delta_identity <- 
         (mean(c(ref_covered, qry_covered))/
              mean(c(rlength, qlength))) -
         error_perc - length_dif_perc
         
     
     
     ## overlaps 
     qols <-
         findOverlaps(qry_ranges, type = "any")
     qunique <- queryHits(qols) != subjectHits(qols)

     qor <- 
         pintersect(
             qry_ranges[queryHits(qols)][qunique], 
             qry_ranges[subjectHits(qols)][qunique]
             ) %>%
         unique
     
     # readDNAStringSet(
     #     sprintf(
     #         "%s/%s.txt", 
     #         sync_dir,
     #         qrys[i]
     #         )
     #     ) %>%
     #     subseq(920006, 940006)
     # qrys[i]
     # qor2 <- 
     #     qor[width(qor) > 0]
     
     qor_tib <- 
         qor %>%
         as.data.frame()
         
     overlap_list[[qrys[i]]] <-
        qor_tib
     
     
     
    #      
    #  ttt <-
    #      ref_ranges
    #  IRanges(
    #      start = c(1,5,8),
    #      end = c(7,9,12)
    #  )
    # 
    #  pintersect(
    #      ttt[queryHits(findOverlaps(ttt))],
    #      ttt[subjectHits(findOverlaps(ttt))]
    #      ) %>%
    #      unique()
    # 
    #  IRanges(
    #      start = c(5,8)
    #      end = c(7,9)
    #  )
    # 
    # pintersect(ß
    #     qry_ranges[queryHits(findOverlaps(qry_ranges))],
    #     qry_ranges[subjectHits(findOverlaps(qry_ranges))]
    #     )
    # print(overlapping_pairs)

    ins_list[[qrys[i]]] <-
        ins_df

    gap_list[[qrys[i]]] <-
        gaps_df
    
    invs1 <- 
        fcdi  %>%
        filter(
            strand=="-",
            meanlen >5e4
        ) %>%
        mutate(
            RASR = 
                #abs((rlen-re) - rs ) < 2e5 & 
            (X_dist < 2e5) & (strand=="-"),
            cluster = clust_i
        ) %>%
        dplyr::select(
            c(
                "rid", qid, "strand", rs, re, qs, qe,
                meanlen, rlen, invs, cluster, RASR
                )
            ) %>%
        dplyr::rename(
           "acc" =  "qid",
           "clust" = "cluster"
        ) %>%
        rowwise() %>%
        mutate(
            start = round(mean(c(rs, qe))),
            end = round(mean(c(re, qs)))
            )  %>%
        ungroup()
            
        # mutate(
        #     abs((rlen-re) - rs ) < 2e5,
        #     X_dist < 2e5,
        #     strand=="-"
        #     ) %>%
        # mutate(cluster = za$cluster_id[i]) %>%
        # dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
    inv_summary <- 
        invs1
    
    ### inner_inversions
    
    
    if (nrow(invs1)>1){
        inner_invs1 <- 
            fcdi %>%
            # filtered_combined_delta %>% 
            #             ungroup %>%
            #             filter(qid==zt1$qid[i]) %>%
            filter(
                strand=="+",
                meanlen >5e4,
                lag(strand=="-") & lead(strand=="-")  
            ) %>%
            mutate(
                RASR = 
                    #abs((rlen-re) - rs ) < 2e5 & 
                (X_dist < 2e5) & (strand=="+"),
                cluster = clust_i
            )%>%
            dplyr::select(
                c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster, RASR)
                ) %>%
            dplyr::rename(
               "acc" =  "qid",
               "clust" = "cluster"
            ) %>%
            rowwise() %>%
            mutate(
                start = round(mean(c(rs, qs))),
                end = round(mean(c(re, qe)))
                )  %>%
            ungroup()
        inv_summary <- 
            bind_rows(invs1, inner_invs1) %>% arrange(rs)
        
        # plot(
        #     tibble(
        #         starts = 
        #             c(invs1$rs, invs1$qe),
        #         ends = 
        #             c(invs1$re, invs1$qs)
        #         )
        #     )
        
        outer_invs <- 
            invs1
        
        
    }
    
    # inv_list[[qrys[i]]] <- 
    #     inv_summary
    # 
    # gaps_df
    
    invs2 <- 
        inv_summary %>%
        mutate(
            width = abs(start-end+1),
            variant = 
                ifelse(
                    RASR, 
                    "RASR",
                    "Local inversion"
                    )
                
            ) %>%
        dplyr::select(c(start, end, width, acc, clust, variant)) %>%
        mutate(
            variant =
                as.character(variant)
        )
    
    inv_list[[qrys[i]]] <-
        invs2
    
        
    tlocs <- 
        ft %>% 
        filter(
            X_dist >2e5#,
            #strand=="+"
            ) %>%
        mutate(
            clust = clust_i,
            acc = qid
        )%>%
        rowwise() %>%
        mutate(
            start = round(mean(c(rs, qe))),
            end = round(mean(c(re, qs))),
            width =  abs(start-end+1),
            variant = "Translocation"
            )  %>%
        ungroup() %>%
        dplyr::select(c(start, end, width, acc, clust, variant))

    
    tloc_list[[qrys[i]]] <-
        tlocs
    
    if (
        nrow(invs2)==0 &
        nrow(gaps_df)==0 &
        nrow(ins_df)==0 &
        nrow(tlocs)==0 
    ){
        all_events <- 
            tibble(
                start = NA,
                end = NA,
                width = NA,
                acc = qrys[i],
                clust = clust_i,
                variant = NA
            ) %>%
            mutate(
                delta_identity = delta_identity
                )
    } else {
        
    all_events <- 
        bind_rows(
            invs2, 
            gaps_df,
            ins_df,
            tlocs
        ) %>%
        arrange(start) %>%
        mutate(
            delta_identity = delta_identity
        )
    }
    
   
    #     filter(
    #         abs((rlen-re) - rs ) < 2e5,
    #         X_dist < 2e5,
    #         strand=="-"
    #         ) %>%
    #     mutate(cluster = za$cluster_id[i]) %>%
    #     dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
    # 
    # ### ## flanking inversions
    # 
    # filtered_combined_delta %>% 
    #                 ungroup %>%
    #                 filter(qid==zt1$qid[i]) %>%
    #     mutate(
    #         inv_flanked = 
    #         lag(strand=="-") & lead(strand=="-")  
    #         )
    
    event_list[[qrys[i]]] <- 
        all_events
        
    
    setTxtProgressBar(pb,i)
        
    #print(i)

}

close(pb)

clustered_genomes %>%count(cluster_id)

event_list[[149]]

SV_tib <- 
    bind_rows(event_list) %>%
    filter(width>=50) %>%
    mutate(
        midpt = round((start+end)/2)
    )

SV_V_small <- 
    SV_tib %>%
    filter(width<1e3)


SV_small <- 
    SV_tib %>%
    filter(
        width>1e3,
        width<2e3
        )


ggplot(
    SV_tib %>% 
        filter(
            #variant %in% c("Gap", "Insertion", "Translocation")
            #width>=50, width < 1e3
            width>=1e3, width < 2e3
            ), 
    aes(x = midpt, fill = variant, y = variant)) +
  geom_density_ridges(alpha = 0.7, scale = 1.2, bandwidth = 0.2) +
  #scale_x_log10() +  # Log scale for better visibility
  labs(x = "Structural variant location (bp)", y = "",
       title = "") +
  theme_classic() +
  theme(text = element_text(size = 18), legend.position = "none")


SV_tib %>%
    group_by(acc) 


SV_tib %>%
    filter(variant=="Local inversion")

# SV_filtered <- 
#     SV_tib 

find_density_peaks <- function(tib, x_var) {
    dens <- density(tib[[x_var]], na.rm = TRUE, bw = "SJ")  # Use adaptive bandwidth
  
  # Filter out negative x-values
    valid <- dens$x >= min(tib[[x_var]])
    dens$x <- dens$x[valid]
    dens$y <- dens$y[valid]

  # Find local maxima (modes)
    peaks <- which(diff(sign(diff(dens$y))) == -2)

  # Keep only the **strongest** peaks (top 3 by height
    peak_heights <- dens$y[peaks]
    top_peaks <- order(peak_heights, decreasing = TRUE)[1:min(2, length(peaks))]  # Top 3 peaks
    data.frame(x = dens$x[peaks[top_peaks]], y = dens$y[peaks[top_peaks]])
}


SV_tib %>%
    filter(variant=="Translocation") %>% arrange(width) %>% filter(width>50)

SV_tib %>% 
        filter(
            variant, #%in% c("Gap", "Insertion", "Translocation"),
            #width<5e3, 
            width >50
            ) %>%
    group_by(variant) %>%
  do(find_density_peaks(., "width"))


ggplot(
    SV_tib %>% 
        filter(
            variant %in% c("Gap", "Insertion", "Translocation"),
            width<5e3, width>50
            ), 
    aes(x = width, fill = variant, y = variant)) +
  geom_density_ridges(alpha = 0.7, scale = 1.2) +
  #scale_x_log10() +  # Log scale for better visibility
  labs(x = "Structural Variant Size (bp)", y = "Count",
       title = "Distribution of Structural Variant Sizes by Type") +
  theme_classic() +
  theme(text = element_text(size = 18))


ggplot(
    SV_tib %>% 
        filter(
            variant %in% c("Gap", "Insertion", "Translocation"),
            width<5e3
            ), 
    aes(y = width, fill = variant, x = variant)) +
  geom_boxplot(alpha = 0.7, scale = 1.2) +
  #scale_x_log10() +  # Log scale for better visibility
  labs(x = "Structural Variant Size (bp)", y = "Count",
       title = "Distribution of Structural Variant Sizes by Type") +
  theme_classic() +
  theme(text = element_text(size = 18))



SV_tib %>%
    group_by(acc) %>%
    summarise()



id_SV <- 
    SV_tib %>%
    group_by(acc, variant) %>%
    summarise(
        SV_count = 
            n(),
        delta_id = unique(delta_identity)
    ) %>%
    # dplyr::count(acc, variant) %>%
    # pivot_wider(
    #     names_from = variant,
    #     values_from = n, values_fill = list(n =0)
    #     ) %>%
    # dplyr::select(-c(`NA`)) %>%
    # right_join(
    #     .,
    #     SV_tib[c("acc", "delta_identity")],
    #     by = "acc"
    # ) %>%
    right_join(
        ., 
        species_identities$id_dists[c(1,2)] %>%
        `colnames<-`(c("acc", "Id_dist")),
        by = "acc"
    ) %>%
    ungroup %>%
    mutate(
        MASH_identity = 
            100-Id_dist
    ) %>%
    group_by(
        acc
    ) %>%
    mutate(
        Variant_types = 
            length(unique(variant[!is.na(variant)])),
    )
    #dplyr::select(-c(acc)) %>%
    # pivot_longer(
    #     cols = 
    #         -c(acc, delta_identity, "Id_dist"), 
    #     names_to = "SV", values_to = "SV_count"
    #     ) %>%
    # group_by(acc) %>%
    #mutate(SVs = sum(SV_count))


saveRDS(
    id_SV, 
    sprintf(
        "%s/id_SV.rds",
        r_vars_dir
    )
)

plot(id_SV$delta_id, id_SV$MASH_identity)


saveRDS(
    SV_tib, 
    sprintf(
        "%s/SV_tib.rds",
        r_vars_dir
    )
)



id_SV %>%
    filter(
        variant=="RASR"
    )




id_SV %>%
    filter(!is.na(SV_count)) %>%
    arrange(desc(`Identity distance`))

ggplot(
    id_SV, 
    aes(x = SV_count, y = Id_dist)) +
  geom_point(size = 4) +  # Adjust point size
  geom_line(aes(group = SV_count), linewidth = 1) + 
  theme_classic() +
  theme(text = element_text(size = 18)) +  
  labs(x = "SVs", y = "Identity distance")


most_related

fasta_file <- 
    sprintf("%s/%s.txt", sync_dir, most_related)

```



```{r}
prokka_commands2 <- 
    annotate_ind(
        fasta_file, species_name, prokka_dir
        #species_name, sync_dir, prokka_dir, accessions = c01$accessions
    )

writeLines(
    text = 
        prokka_commands2$comms,
    con = 
        prokka_commands2$file
    )


gff_i <- 
   read.delim(
       sprintf(
           '%s/%s.gff', prokka_dir, most_related
       ), header=F, comment.char = "#")  %>%
   as_tibble() %>%
   `colnames<-`(
       c(
           "accession", "source", "type", "start", 
           "end", "score", "strand", "phase", "attributes")
       ) %>%
   filter(!is.na(start))


gff_ranges <- 
   IRanges(
       start = 
           gff_i$start,
       end = 
           gff_i$end
   )

gff_ranges <- 
   IRanges(
       start = 
           gff_i$start,
       end = 
           gff_i$end
   ) %>% gaps()

## which ranges

gff_ranges %>%
    width %>% summary


 

overlaps_i <- 
   findOverlaps(
       query = bp_range, subject = gff_ranges_i, type = "any"
   ) %>% subjectHits()

SV_tib2_vs <- 
    SV_tib %>%
    filter(width>=50) %>%
    mutate(
        midpt = round((start+end)/2)
    )
    #filter(width<500, width>50, variant=="Insertion" | variant=="Translocation") %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )

SV_range_smallest <- 
    IRanges(
       start = 
           SV_tib2_vs$s2,
       end = 
           SV_tib2_vs$e2
   )
     
sv_ol<- 
    findOverlaps(SV_range_smallest, gff_ranges)


overlap_widths <- 
    pintersect(SV_range_smallest[queryHits(sv_ol)], gff_ranges[subjectHits(sv_ol)])

percent_overlap <-
    tapply(width(overlap_widths), queryHits(sv_ol), sum) / width(SV_range_smallest)[unique(queryHits(sv_ol))] * 100

sum(width(overlap_widths))/sum(width(SV_range_smallest))

SV_tib2_vs <- 
    SV_tib %>%
    filter(width<500, width>50, variant=="Insertion" | variant=="Translocation") %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )

SV_range_smallest <- 
    IRanges(
       start = 
           SV_tib2_vs$s2,
       end = 
           SV_tib2_vs$e2
   )
     
sv_ol<- 
    findOverlaps(SV_range_smallest, gff_ranges)

sv_vs <- 
    SV_range_smallest %>%
    as.data.frame() %>%
    as_tibble %>%
    mutate(
        overlap = 
            ifelse(
                seq_along(SV_range_smallest) %in% queryHits(sv_ol), "Inside Gene", "Outside Gene"
                )

    ) %>%
    add_count(overlap)
    
ggplot(sv_vs, aes(x = overlap, fill = overlap)) +
  geom_bar() +  # Simple bar plot with blue bars
  labs(x = "", y = "Count", title = " SV >50nt, SV <500 nt") +
  theme_classic()+
    theme(
        text = element_text(size = 22),
        legend.position = "none"
        ) 


SV_tib2_vs <- 
    SV_tib %>%
    filter(width<500, width>50, variant=="Insertion" | variant=="Translocation") %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end),
    )








####
SV_tib2_sm <- 
    SV_tib %>%
    filter(width<2000, width>1000, variant=="Insertion" | variant=="Translocation") %>%
    rowwise %>%
    mutate(
        s2 = ifelse(end<start, end, start),
        e2 = ifelse(end<start, start, end)
    ) %>% filter(e2>s2)

SV_range <- 
    IRanges(
       start = 
           SV_tib2_sm$s2,
       width = 
           SV_tib2_sm$width
   )
     
sv_ol<- 
    findOverlaps(SV_range_smallest, gff_ranges)
SV_range_smal <- 
    IRanges(
       start = 
           SV_tib2_sm$s2,
       end = 
           SV_tib2_sm$e2
   )
     
sv_ol<- 
    findOverlaps(
        gff_ranges,
        SV_range_smal, type = "within"
        )

sv_sm <- 
    SV_range_smal%>%
    as.data.frame() %>%
    as_tibble %>%
    mutate(
        overlap = 
            ifelse(
                seq_along(SV_range_smal) %in% subjectHits(sv_ol), "Encompassing Gene", "Unencompassing Gene"
                )

    ) %>%
    add_count(overlap)
    
ggplot(sv_sm, aes(x = overlap, fill = overlap)) +
  geom_bar() +  # Simple bar plot with blue bars
  labs(x = "", y = "Count", title = " SVs >1kb, SV <2 Kb") +
  theme_classic() +
    theme(
        text = element_text(size = 22),
        legend.position = "none"
        ) 
    



overlap_widths <- 
    pintersect(SV_range_smallest[queryHits(sv_ol)], gff_ranges[subjectHits(sv_ol)])

percent_overlap <-
    tapply(width(overlap_widths), queryHits(sv_ol), sum) / width(SV_range_smallest)[unique(queryHits(sv_ol))] * 100




#system(prokka_com)
```




```{r}
clustered_genomes_list$cluster_summary %>% count(cluster_id)


best_identity <- 
    inner_join(
        SV_i %>% `colnames<-`(c("accessions", "identity")),
        clustered_genomes[,-5],
        by = "accessions"
    ) %>%
    mutate(
        identity = 
            ifelse(identity>1,  1, identity),
        identity = 
            ifelse(accessions==most_related,  1.01, identity),
    ) %>%
    arrange(cluster_id)

    ## Anova for all others
    
    # best_identity %>% filter(cluster_id!=0) %>%
    #     arrange(cluster_id) %>%
    #     count(cluster_id)
    # best_identity_ss
mt1 <- 
    species_identities$mash_tib %>%
    filter(
        ref %in% clustered_genomes$accessions,
        qry %in% clustered_genomes$accessions
    ) %>%
    left_join(
        .,
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("ref_clust", "ref")),
        by = "ref"
    ) %>%
    left_join(
        .,
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("qry_clust", "qry")),
        by = "qry"
    )
    
mt2 <- 
    mt1%>%
    filter(
        (ref_clust %in% 0:5),
        (qry_clust %in% 0:5)
        )

ggplot(
            data = mt2,
            aes(
                x= ref,#factor(ref, seq_order),
                y= qry,#factor(qry, seq_order),
                fill = identity
            )
        ) +
        geom_tile() +
        theme_classic() +
        theme(
            #axis.text.x = element_text(angle=45, hjust = 1),
            panel.background=element_rect(fill="white"),
            plot.title = element_text(size=20,face="bold", hjust = 0.5),
            axis.title = element_blank(),
            axis.text = element_blank()
        ) +
        ggtitle(
            sprintf(
                "%s \n Sequence pairwise MASH similarity",
                sub("_", " ", species_name)
            )
        ) +
        theme(
            legend.title=element_text( size=18),
            axis.text=element_blank(),
                #element_text(size=14),
            legend.text = element_text(size=18)
        ) +
        scale_fill_viridis(
            option = "B", limits = c(99.5,100)
        ) +
    facet_grid(
    rows = vars(qry_clust), cols = vars(ref_clust),
    scales = "free", space = "free", switch = "y")
    
bis <- 
    best_identity %>% 
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>% 
    filter(!is.na(cluster_id)) %>%
    dplyr::slice(1)


bi_ss <- 
    best_identity %>%
    filter(grouped_sequences>=3) %>%
    group_by(cluster_id) %>%
    dplyr::slice(1:3)




bis_f <- 
    sprintf(
        "%s/%s.txt",
        sync_dir, bi_ss$accessions
    )


file.copy(
    bis_f,
    file.path(
        sprintf(
            "%s",
            subset_dir
        ),
        basename(bis_f)
    )
)

list.files(subset_dir)


sprintf(
    "poppunk profile -i %s/ -o %s/kmer_profiles/",
    subset_dir,
    test_dir
) %>% system()

    

saveRDS(
    bis, 
    sprintf(
        "%s/bis.rds",
        results_dir
    )
    )

bis_1 <- 
    bis[-1,]

if (nrow(bis_1)==0) next
mpis <- 
    lapply(
        1:nrow(bis_1),
        function(j){
            dfj <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==pull(bis_1, accessions)[j]) 
            # if (j==5){
            #     pj <- 
            #         plot_delta(dfj)  +
            #         xlab("Cluster 0")
            # } else {
                pj <- 
                    plot_delta(dfj)  +
                    xlab("")
            #}
            pjo <- 
                pj +
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 22)
                ) +
                ylab(sprintf("Cluster %s", j)) 
            return(pjo)
        }
    )

saveRDS(
    mpis, 
    sprintf(
        "%s/mpis.rds",
        results_dir
    )
    )

if (length(mpis)>15){
    mpis <- 
        mpis[1:15]
}

plots_dir[[i]] <- 
    sprintf("%s/delta_plots.png", results_dir)

png(sprintf("%s/delta_plots.png", results_dir), width = 800, height = 600)
do.call(
     "grid.arrange",
     c(
         mpis,
         ncol=3
         )
     )
    # Close the PNG device to save the image
    dev.off()
```


