---
title: "final"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r}

```

#Final
Getting the NCBI table
Load the full table, save it as an rds table or load the predownloaded dataset from github.
```{r}
# Loading the package
source("BSPN.R", verbose = F, print.eval = F)

database_dir <- 
    "../data/ncbi_db"

options(timeout = 300)

ncbi_table_list_2 <- 
  get_ncbi_table(db_dir = database_dir) 

saveRDS(ncbi_table_list, sprintf("%s/ncbi_table_list_v2.rds", database_dir))
#saveRDS(ncbi_table_list2, sprintf("%s/ncbi_table_list.rds", database_dir))
ncbi_table_list <- 
    readRDS(sprintf("%s/ncbi_table_list.rds", database_dir))
ncbi_table <- 
    ncbi_table_list$ncbi_table

species_summary %>%
    pull(technology)


ncbi_table 

```



```{r}

fread(genbank_file, nrows = 4)

refseq_count <- 
    fread(
        refseq_file, 
        stringsAsFactors = F, 
        quote = "", 
        select = c(1,3,5,6,7,8,11,12,15,16,17,18,19,20)
    ) %>% 
    filter(
        #assembly_level =="Complete Genome" &
            version_status =="latest"
    ) %>%
    `colnames<-`(c("assembly", colnames(.)[-1])) %>%
    as_tibble %>%
    mutate(db = "refseq")


 big_gen_count <- 
     fread(
            genbank_file, 
            stringsAsFactors = F, 
            quote = "", 
            select = c(1,3,5,6,7,8,11,12,15,16,17,18,19,20)
        ) %>% 
        filter(
            #assembly_level =="Complete Genome" &
                #version_status =="latest"
        ) %>%
        `colnames<-`(c("assembly", colnames(.)[-1])) %>%
        as_tibble %>%
        #filter(paired_asm_comp != "identical") %>%
        mutate(db = "genbank") #%>%
         # bind_rows(
         #     refseq_count
         #     )
 
 big_gen_count%>% count(version_status)

big_gen_redone_assemblies <- 
    big_gen_count %>%
    `colnames<-`(c("assembly", colnames(.)[-1])) %>%
    add_count(biosample, name = "ac") %>%
    filter(ac>1)
big_gen_count %>%
    count(paired_asm_comp)

big_gen_count %>%
    filter(asm_name =="GCA_005206335.2")

# asm_tib <- 
#     big_gen_count
big_gen_count$asm_name[1:100][
    which(!big_gen_count$assembly[1:100] %in% asm_info$asm_name)
]
asm_info$asm_name
big_gen_count %>%
    dplyr::select(assembly, biosample, seq_rel_date, assembly_level, db) %>%
    filter(db=="refseq")

bgc <- 
    big_gen_count %>%
    dplyr::select(assembly, biosample, seq_rel_date, assembly_level, ftp_path) %>%
    arrange(seq_rel_date) %>%
    filter(!is.na(seq_rel_date)) %>%
    mutate(
        year = as.integer(format(seq_rel_date, "%Y")),
        assembly_level = 
            factor(
                assembly_level, 
                levels = 
                    rev(c("Contig", "Scaffold", "Chromosome", "Complete Genome"))
                ),
         year_collapsed = 
            ifelse(year >= 1999 & year <= 2014, "1999-2014", as.character(year)),
        assembly = 
            ifelse(
                assembly_level == "Complete Genome",
                "Complete", "Incomplete"
            )
        ) %>%
    filter(
        !(year_collapsed %in% c("1999-2014", "2025"))
            )

bgc %>% 
    filter(year=="2018") %>% 
    filter(assembly =="Complete")

big_gen_count %>%
    dplyr::select(assembly, biosample, seq_rel_date, assembly_level) %>%
    arrange(seq_rel_date) %>%
    filter(!is.na(seq_rel_date)) %>%
    mutate(
        year = as.integer(format(seq_rel_date, "%Y")),
        assembly_level = 
            factor(
                assembly_level, 
                levels = 
                    rev(c("Contig", "Scaffold", "Chromosome", "Complete Genome"))
                ),
         year_collapsed = 
            ifelse(year >= 1999 & year <= 2014, "1999-2014", as.character(year)),
        assembly = 
            ifelse(
                assembly_level == "Complete Genome",
                "Complete", "Incomplete"
            )
        ) %>%
    filter(
        (year_collapsed %in% c("1999-2014")),
        assembly == "Complete"
            )


df_counts <-
    bgc %>%
  group_by(seq_rel_date, assembly) %>%
  summarise(
      count = n(), 
      .groups = 'drop',
      cul = cumsum(count)
      ) 

ggplot(
    df_counts %>% 
        filter(assembly=="Complete"), 
    aes(x = seq_rel_date, y = count, color = assembly)
    ) +
 geom_point(alpha = 0.5) +       # Optional: show individual points
  geom_smooth(se = FALSE, method = "loess", span = 0.3) +  # Smooth curve
  labs(
    title = "Assembly completeness over time",
    x = "Date",
    y = "Assembly count"
  ) +
  theme_classic() +
    theme(
        axis.text = element_text(size = 18),
        axis.title = element_text(size = 22),
        legend.title = element_text(size = 18),
        legend.text = element_text(size = 16)
        ) +
  scale_y_continuous(
    labels = function(x) x / 10000  # Convert counts to (y × 10^4)
  )+
  labs(
    x = "Year",
    y = expression("Assembly count" ~ (x ~ 10^4)),  
    fill = "Assembly Level"
  )


ggplot(
    bgc %>% 
        filter(assembly =="Complete") %>%
        count(year_collapsed), 
    aes(x = as.factor(year_collapsed), y = n)
    ) +
  geom_bar(position = "stack", colour = "black", fill = "red", stat = "identity") +
    scale_fill_brewer(palette = "Set1") +
  theme_classic() +
    theme(
        axis.text = element_text(size = 18),
        axis.title = element_text(size = 22),
        legend.title = element_text(size = 18),
        legend.text = element_text(size = 16)
        ) +
  #scale_y_continuous(
   # labels = function(x) x / 10000  # Convert counts to (y × 10^4)
  #)+
  labs(
    x = "Year",
    y ="Assembly count", # expression("Assembly count" ~ (x ~ 10^4)),  
    fill = "Assembly Level"
  )

bgc_complete <- 
    big_gen_count %>%
    dplyr::select(assembly, asm_name, biosample, seq_rel_date, assembly_level, ftp_path, db) %>%
    arrange(seq_rel_date) %>%
    filter(assembly_level == "Complete Genome")
    
bgc_complete %>%
    filter(db=="refseq")

bgc %>% count(assembly_level)


reports <- 
    pbmclapply(
        1:nrow(bgc_complete),
        function(i){
            f1 <- 
                bgc_complete %>%
                filter(db=="refseq") %>%
                dplyr::slice(1) %>%
                pull(ftp_path)
                #bgc_complete$ftp_path[i]
            
            report_file <- 
                sprintf(
                    "%s/%s_assembly_report.txt",
                    f1, gsub(".*/", "", f1)
                ) 
            
            if (!url.exists(report_file)){
                report_file <- 
                    sprintf(
                        "%s/%s_%s/%s_%s_assembly_report.txt",
                        dirname(f1), 
                        pull(bgc_complete[i,], assembly),
                        pull(bgc_complete[i,], asm_name),
                        pull(bgc_complete[i,], assembly),
                        pull(bgc_complete[i,], asm_name)
                    ) 
                }
            
            
            # if (!url.exists(report_file)){
            #     report_file <- 
            #         sprintf(
            #             "%s/%s_%s/%s_%s_assembly_report.txt",
            #             dirname(f1), 
            #             pull(bgc_complete[i,], assembly),
            #             pull(bgc_complete[i,], asm_name),
            #             pull(bgc_complete[i,], assembly),
            #             pull(bgc_complete[i,], asm_name)
            #         ) 
            #     }
            # 
            # rep_read <-
            #     tryCatch(readLines(report_file))
            # 
            # technology <- 
            #     rep_read[grepl("# Sequencing", rep_read)] %>% 
            #     sub(".*technology:", "", .) %>% 
            #     trimws
            # if (length(technology) ==0){
            #    technology <- NA 
            # }
            # tech_out <- 
            #     bgc_complete %>%
            #     dplyr::slice(i) %>%
            #     mutate(tech = technology)
            
            return(report_file)
        }, mc.cores = 6
    ) %>% unlist


t2 <- 
    lapply(
        1:10,
        function(i){
            
            report_file <- 
                reports[i]
            rep_read <-
                tryCatch(readLines(report_file))
            
            technology <- 
                rep_read[grepl("# Sequencing", rep_read)] %>% 
                sub(".*technology:", "", .) %>% 
                trimws
            if (length(technology) ==0){
               technology <- NA 
            }
            
            
            out_tib <- 
                tibble(
                    asm_name = 
                        big_gen_count$assembly[i], 
                    tech = technology,
                    date = big_gen_count$seq_rel_date[i],
                    level = big_gen_count$assembly_level[i],
                    biosample = big_gen_count$biosample[i]
                )

            return(out_tib)
        }
    )

writeLines(reports, "../data/processing/test/reports.txt")
length(reports)

tech_count <- 
    fread("../data/processing/test/biosample_sequencing_info.tsv", fill=TRUE) %>% 
    as_tibble%>%
    filter(!is.na(BioSample))


tc

tc <- 
    tech_count %>%
    filter(!is.na(Sequencing_Technology)) %>%
    mutate(
        t1 = 
            if_else(
                str_detect(Sequencing_Technology, "Pac"), 
                "PacBio", 
                ifelse(
                    str_detect(Sequencing_Technology, "Illumina") |
                        str_detect(Sequencing_Technology, "illumina"),
                    "Illumina",
                    # ifelse(
                    #     str_detect(Sequencing_Technology, "454"),
                    #     "454",
                        ifelse(
                            str_detect(Sequencing_Technology, "Oxford"),
                            "Oxford nanopore",
                        #         ifelse(
                        #             str_detect(Sequencing_Technology, "Nanopore"),
                        #             "Nanopore",
                        #             ifelse(
                        #                 str_detect(Sequencing_Technology, "Ion"),
                        #                 "IonTorrent/Ion PGM",
                                        "Other")
                    #                 )
                         )
                    # ))
                ),
        biosample = BioSample,
        t2 = factor(t1, levels = rev(c("Illumina", "PacBio", "Oxford nanopore", "Other")))
        ) %>%
    filter(!duplicated(biosample)) %>%
    add_count(t1) %>%
    arrange(desc(n)) %>%
    left_join(., bgc_complete[,c(2, 3)], by = "biosample") %>%
    mutate(
        year = as.integer(format(seq_rel_date, "%Y")),
         year_collapsed = 
            ifelse(year >= 1999 & year <= 2014, "1999-2014", as.character(year))
        )


ggplot(tc, aes(x = as.factor(year_collapsed), fill = t2)) +
  geom_bar(position = "stack", colour = "black") +
    scale_fill_brewer(palette = "Dark2") +
  theme_classic() +
    theme(
        axis.text = element_text(size = 18),
        axis.title = element_text(size = 22),
        legend.title = element_text(size = 18),
        legend.text = element_text(size = 16)
        ) +
  # scale_y_continuous(
  #   labels = function(x) x / 10000  # Convert counts to (y × 10^4)
  # )+
  labs(
    x = "Year",
    y = "Assembly count",#expression("Assembly count" ~ (x ~ 10^4)),  
    fill = "Assembly Level"
  )


tech_count %>% count(Sequencing_Technology)



asm_info <- 
    pbmclapply(
        1:15,#nrow(big_gen_count),
        function(i){
            f1 <- 
                big_gen_count$ftp_path[i]
            
            report_file <- 
                sprintf(
                    "%s/%s_assembly_report.txt",
                    f1, gsub(".*/", "", f1)
                ) 
            
            if (!url.exists(report_file)){
                report_file <- 
                    sprintf(
                        "%s/%s_%s/%s_%s_assembly_report.txt",
                        dirname(f1), 
                        pull(big_gen_count[i,], assembly),
                        pull(big_gen_count[i,], asm_name),
                        pull(big_gen_count[i,], assembly),
                        pull(big_gen_count[i,], asm_name)
                    ) 
                }
            
            rep_read <-
                tryCatch(readLines(report_file))
            
            technology <- 
                rep_read[grepl("# Sequencing", rep_read)] %>% 
                sub(".*technology:", "", .) %>% 
                trimws
            if (length(technology) ==0){
               technology <- NA 
            }
            
            
            out_tib <- 
                tibble(
                    asm_name = 
                        big_gen_count$assembly[i], 
                    tech = technology,
                    date = big_gen_count$seq_rel_date[i],
                    level = big_gen_count$assembly_level[i],
                    biosample = big_gen_count$biosample[i]
                )
            # big_gen_count %>%
            #     filter(ftp_path ==f1)
            # asm_summary <- 
            #     big_gen_count
            # 
            # ftp_path <-
            #     big_gen_count %>% 
            #     pull(ftp_path)
            #     (species_table %>% pull(ftp_path))[i]
            
            return(out_tib)
            
        },
        mc.cores = 4
    ) %>% 
    bind_rows


big_gen %>% count(assembly_level) %>%
    mutate(
        perc = round(100*(n/sum(n)), 2)
    )

            
```


```{r}
ncbi_table_clade_count <- 
  ncbi_table %>% 
  add_count(species, name = "spec_count") %>%
  filter(spec_count>=10) %>%
  group_by(p2, species) %>%
  summarise(spec_count = n(), .groups="keep") %>%
  ungroup %>%
  mutate(genus = sub("([A-Za-z]+).*", "\\1", species)) %>%
  group_by(p2, genus) %>%
  mutate(genus_count = sum(spec_count)) %>%
  arrange(desc(genus_count), desc(spec_count)) %>%
  mutate(species2 = sub(" ", "_", species))


ncbi_table_clade_count %>%
    group_by()


## Setting up species specific directories

ncbi_table %>%
    filter(grepl("salmonella", organism_name, ignore.case = T)) %>%
    count(species)
    


species_list <- 
  ncbi_table_clade_count %>% 
  ungroup %>%
  arrange(desc(spec_count))

species_list_40 <- 
    ncbi_table_clade_count %>% 
    ungroup %>%
    arrange((spec_count)) %>%
    filter(spec_count>=40)

spec2_list_40_100 <- 
    species_list_40 %>%
    filter(spec_count<=100)

spec2_list_100_500 <- 
    species_list %>%
    filter(
        (spec_count>100) &
            (spec_count<500)
        ) %>%
    arrange(spec_count)

spec2_list_500_over <- 
    species_list %>%
    filter(
        (spec_count>=500)
        )
    

spec2_list_100_500 %>% pull(spec_count) %>% sum
spec2_list_40_100 %>% pull(spec_count) %>% sum
spec2_list_40_100 %>% pull(spec_count) %>% sum
spec2_list_500_over %>% pull(spec_count) %>% sum

spec2_list <- 
  species_list_40$species2

species_name <- 
    (ncbi_table_list$ncbi_cc %>% 
    arrange(desc(spec_count)) %>% pull(species_))[i]


# spec2_list < - 
#   species_list$species2

spec2_list %>% grep("Vibrio", .)


i=1
species_list %>% filter(species=="Vibrio parahaemolyticus")

head(spec2_list, 20)

which(spec2_list=="Leptospira_interrogans")
i=123

i
```




```{r}

### choose species
species_name <- 
        (species_list %>% pull(species2))[i]

#set up directories
genus_name <- 
      strsplit(species_name, "_")[[1]][1]
    
clade_name <- 
    filter(ncbi_table_list$ncbi_cc, species_==species_name) %>%
    pull(phylum) %>% gsub(" ", "_", .)

species_variable_names <-
  c(
      "fna_dir",
      "sync_dir", "gff_dir", "identity_dir", "mummer_dir",
      "delta_dir", "r_vars_dir", "results_dir", 
      "repeats_dir", "invs_dir",
      "disparities_dir",# "c01_dir",
      "subset_dir",
      "breakpoint_dir", "prokka_dir",
      "breakpoint_comparisons_dir", "snp_dir",
      "kmer_dir", "bd_dir", "plot_dir", "roary_dir",
      "sync2_dir"
  )

species_dirs <- 
  gsub("_dir", "", species_variable_names)

for (var_name in species_variable_names){
  dir_name <- 
    gsub("_dir", "", var_name)
  full_dir_name <- 
    sprintf(
        "../data/processing/%s/%s/%s/%s", 
        dir_name, clade_name, 
        genus_name, species_name
        )
  
  assign(
    var_name, 
    full_dir_name
  )
  if (!dir.exists(full_dir_name)){dir.create(full_dir_name, recursive = T)}
}

species_table <- 
    ncbi_table %>%
    filter(species== sub("_", " ", species_name))

spec_count <- 
    nrow(species_table)

if(spec_count<250){
    subsample_idx <- 
        1:spec_count
} else {
    set.seed(123)
    subsample_idx <- 
        sample(1:spec_count, 200)
}

st_ss <- 
        species_table %>%
        dplyr::slice(subsample_idx)
    
    the_rest_idx <- 
        species_table %>%
        filter(!row_number() %in% subsample_idx)
        
    species_summary <- #species_table
        get_species_summary(st_ss) 
    
    ssn[[i]] <- 
        tibble(
            spec_ = species_name,
            stn = nrow(species_table),
            ssn = nrow(species_summary)
        )
    if (nrow(species_summary)<40) next
        
    #print(paste(species_name, i))
    saveRDS(
        species_summary, 
        sprintf(
            "%s/species_summary.rds",
            r_vars_dir
        )
    )
    
    ftp_paths <- 
        species_summary$ftp_path
    
    gff_paths <- 
        ftp_paths %>%
                sprintf(
                    "%s/%s_genomic.gff.gz",
                    ., gsub(".*/", "", .)
                )
    # fna_paths <- 
    #     ftp_paths %>%
    #             sprintf(
    #                 "%s/%s_genomic.fna.gz",
    #                 ., gsub(".*/", "", .)
    #             )
    
    # file_exists_at_url(fna_paths[1])
    # file_exists_at_url(gff_paths[1])
    gff_check <- 
        pbmclapply(
            gff_paths, 
            file_exists_at_url
        ) %>% unlist
    
    if (!any(gff_check)) next
    
    # ij <-
    #     2
    # 
    # f_ij <- 
    #     ftp_paths[ij]
    # gff_ij <-
    #     f_ij %>%
    #             sprintf(
    #                 "%s/%s_genomic.gff.gz",
    #                 ., gsub(".*/", "", .)
    #             )
    # 
    # gff_table <-
    #         read.gff(
    #             gzcon(url(gff_path))
    #         )
    # chr2_acc <-
    #     gff_table %>%
    #     as_tibble %>%
    #     filter(
    #         (grepl("chromosome=", attributes, ignore.case = F))
    #         #(grepl("origin", attributes, ignore.case = F))
    #     ) %>%
    #     pull(seqid) %>%
    #     as.character
    # 
    # repAs <- 
    #     list()
    # for (ij in 1:length(ftp_paths)){
    #     f_ij <- 
    #         ftp_paths[ij]
    #     gff_ij <- 
    #         f_ij %>%
    #                 sprintf(
    #                     "%s/%s_genomic.gff.gz",
    #                     ., gsub(".*/", "", .)
    #                 )
    #     
    #     gff_table <- 
    #             read.gff(
    #                 gzcon(url(gff_path))
    #             )
    #     chr2_acc <- 
    #         gff_table %>%
    #         as_tibble %>%
    #         filter(
    #             (grepl("chromosome=II", attributes, ignore.case = F))
    #             #(grepl("origin", attributes, ignore.case = F))
    #         ) %>%
    #         pull(seqid) %>% 
    #         as.character
            
        
    #     gff_table_chr2 <- 
    #         gff_table %>% filter(seqid==chr2_acc)%>% as_tibble %>%
    #         mutate(
    #             gene_value = 
    #                 str_replace(
    #                     str_extract(attributes, "gene=\\w+"), "gene=", ""
    #                     ),
    #             # name_value = 
    #             #     str_replace(
    #             #         str_extract(attributes, "Name=\\w+"), "Name=", ""
    #             #         )
    #             prod_value = 
    #                 str_replace(
    #                     str_extract(attributes, "product=\\w+"), "product=", ""
    #                     )
    #             ) %>%
    #         filter(!(is.na(gene_value))) %>%
    #         pull(gene_value) %>% unique
    #             
    #     gff_table$seqid %>% unique
    #     gff_table %>% as_tibble
    #     gff_table %>% as_tibble %>%
    #         group_by(seqid) %>% 
    #         filter(end==max(end))
    #     
    #     repAs[[ij]] <- 
    #         gff_table %>%
    #         as_tibble %>%
    #         filter(
    #             (grepl("chromosome=II", attributes, ignore.case = F))
    #             #(grepl("origin", attributes, ignore.case = F))
    #         )
    #     repAs[[ij]]
    #     
    # }
    # lapply(repAs, head) %>% unlist
    
    ## check for 2nd chromosomes
    
    
    
    # species_summary <- #species_table
    #     readRDS(
    #         sprintf(
    #            "%s/species_summary.rds",
    #             r_vars_dir
    #            )
    #     )
    closeAllConnections()
    sync_tib <- 
        pbmclapply(
            1:nrow(species_summary),
            function(ii){
                #for (ii in nrow(species_summary):1){
                assembly_path <- 
                    species_summary$ftp_path[ii]
                
                species_summary$accession[ii]
                sync_summary <- 
                    sync_gene(
                        assembly_path = 
                            assembly_path,
                        sync_dir = 
                            sync_dir, 
                        gff_dir = 
                            gff_dir
                        ) %>%
                    mutate(
                        idx = ii
                    )
                return(sync_summary)
            }, mc.cores = 4
        ) %>% bind_rows()
    
    if (all(is.na(sync_tib$ori_pos))){
        saveRDS(
            sync_tib,
            sprintf(
                "%s/sync_tib.rds",
                results_dir
            )
        )
        next
    }
    
    ## put all in the same directory
    
    dirname(sync_tib$out_file)
    
    ##
    
    # which(!(sync_tib %>% filter(chr==1) %>% pull(accession)) %in%
    # (list.files(sdc) %>% sub(".txt", "", .)))
    # 
    # (sync_tib %>% pull(accession))[2]
    # 
    # sync_tib %>%
    #     filter(chr==2)
    # synceing the unsynced
    unsynced_idx <- 
        sync_tib %>%
        filter(is.na(ori_pos)) %>%
        filter(chr==1)
    
    unsynced_idx %>% filter(!is.na(ori_pos))
    
    rs_tibs <- list()
    if (nrow(unsynced_idx)>0){
        
        for (m in 1:length(unsynced_idx$idx)){
        #species_summary
        assembly_path <- 
            species_summary$ftp_path[unsynced_idx$idx[m]]
        
        us_chrom <-
            sync_tib$chr[unsynced_idx$idx[m]]
    
        unsynced_seq <- 
            get_fasta(assembly_path, chromosome = us_chrom)
        fasta_seq <- 
            sprintf(
                "%s/%s.txt",
                sync_dir,
                names(unsynced_seq)
            )
        writeXStringSet(
            unsynced_seq,
            fasta_seq
        )
        
        ref_seq <- 
            list.files(
                sprintf(
                    "%s/chr_%s_gene",
                    sync_dir, us_chrom
                ),
                full.names = T
            )[1]
    
        startpoint_dir <- 
            sprintf("%s/starts", sync_dir)
        if (!dir.exists(startpoint_dir)){dir.create(startpoint_dir)}
        
        rs_synced <- 
            rs_sync(
                ref_seq, fasta_seq, 
                delta_out = 
                    sprintf("%s/%s", startpoint_dir,  names(unsynced_seq)
                            ),
                out_file = 
                    fasta_seq
            ) 
        
        if (!is.null(rs_synced)){
            
            ref_seq <- 
                list.files(
                    sprintf(
                        "%s/chr_%s_gene",
                        sync_dir, us_chrom
                    ),
                    full.names = T
                )[2]
            rs_synced <- 
                rs_sync(
                    ref_seq, fasta_seq, 
                    delta_out = 
                        sprintf("%s/%s_2", startpoint_dir,  names(unsynced_seq)
                                ),
                    out_file = 
                        fasta_seq
                ) 
            if (!is.null(rs_synced)){
                rs_tibs[[m]] <- 
                    rs_synced %>%
                    mutate(
                        asm_name =
                            basename(assembly_path),
                        chr = us_chrom,
                        idx = unsynced_idx$idx[m]
                        )
            } else {
                rs_tibs[[m]] <- 
                    rs_synced 
            }
        } else {
            rs_tibs[[m]] <- 
                rs_synced 
        }
        
        
      
        }
        rs_tib <- 
            bind_rows(rs_tibs)
        
        sync_tib <- 
            bind_rows(
                sync_tib %>% filter(!is.na(ori_pos)),
                rs_tib
            )
    }
    
    
    if (nrow(sync_tib)<10) next
    
    saveRDS(
        sync_tib, 
        sprintf(
            "%s/sync_tib.rds",
            results_dir
        )
        )
    
    ###
    # check chr 2
    if (("chr_1" %in% basename(list.dirs(sync_dir)))){
        sdc <- 
            sprintf(
                "%s/chr_1",
                sync_dir
            )
        ## move any files if necessary
        no_2nd <- 
            (sync_tib %>%
            filter(chr==1) %>%
            pull(accession))[
                !((sync_tib %>% filter(chr==1) %>% pull(accession)) %in%
                  (list.files(sdc) %>% sub(".txt", "", .)))
            ]
        list.files(sdc)
        
        no_2nd <- 
            sprintf(
                "%s/%s.txt", 
                sync_dir, no_2nd
            )
        if (length(no_2nd)>0) {
            for (ij in 1:length(no_2nd)){
                n2 <- 
                    sprintf(
                        "%s/%s.txt", 
                        sync_dir, 
                        no_2nd[ij]
                    )
                    
                if (file.exists(n2)){
                    file.copy(
                        from = n2,
                        to = 
                            sprintf("%s/%s", sdc, basename(n2)) 
                        )
                    
                }
            }
        }
        
        
        
    } else{
        sdc <- 
            sync_dir
    }
    list.files(sdc)
    ## Identities
    species_identities <- 
        get_pairwise_identities(
            sync_dir = sdc, 
            identity_dir = identity_dir, 
            species_table = sync_tib, 
            genus_name, species_name
            )
    
    #species_identities$mash_plot
    saveRDS(
        species_identities, 
        sprintf(
            "%s/species_identities.rds",
            results_dir
        )
        )
    
    # species_identities <- 
    #     readRDS(
    #         sprintf(
    #             "%s/species_identities.rds",
    #             results_dir
    #         )
    #     )
    # 
    species_identities$id_dists
    #read_delta("../data/processing/delta/Pseudomonadota/Escherichia/Escherichia_coli/ref_v_all/NZ_CP010")
    ##
    sampled_id_table <- 
        species_identities$id_dists %>%
        arrange(desc(Sequence_mean_distance))# %>%
        #filter(Sequence_mean_distance<10) 
    
    most_related <- 
        sampled_id_table %>%
        arrange(Sequence_mean_distance) %>%
        dplyr::slice(1) %>% 
        pull(Sequence)
    
    ## pairwise alignemnts
    accessions <- 
        sync_tib %>%
        filter(chr==1) %>%
        pull(accession)
    
    sync_tib %>%filter(chr==1, is.na(ori_pos))
    
    which(
        !(accessions %in%
              sub(".txt", "", list.files(sdc)))
    )
    
    nucmer_commands <- 
        nucmer_ref_v_all(
            most_related, 
            accessions,
            delta_dir =  delta_dir, 
            sync_dir = sdc
            )
    
    minimum_alignment_percentage <- 
        90
    minimum_inversion_length <- 
        5e4
    filtered_combined_delta <- 
        filter_combined_delta(
            sprintf("%s/ref_v_all", delta_dir), 
            minimum_alignment_percentage, 
            minimum_inversion_length
        ) 
    
    likely_missassembled <- 
        filtered_combined_delta %>% 
        group_by(qid) %>%
        mutate(
            ncs = n(),
            X_sum = sum(X_dist)
            ) %>%
        filter(ncs>20 | X_sum>1e7) %>%
        pull(qid) %>% unique
    
    #filtered_combined_delta %>% filter(qid %in%likely_missassembled[2] ) %>% plot_delta
    #filtered_combined_delta %>% filter(qid %in%likely_missassembled[1] ) %>% plot_delta
    
    fcd <- 
        filtered_combined_delta %>%
        filter(!(qid %in% likely_missassembled)) %>%
        filter(qid %in% accessions)
    
    saveRDS(
        fcd, 
        sprintf(
            "%s/fcd.rds",
            results_dir
        )
        )
    
    # fcd <- 
    #     readRDS(
    #         sprintf(
    #         "%s/fcd.rds",
    #         results_dir
    #     )
    #     )

    clustered_genomes_list <- 
        cluster_genomes(fcd)
    
    saveRDS(
        clustered_genomes_list, 
        sprintf(
            "%s/clustered_genomes_list.rds",
            results_dir
        )
        )
    
    out_ds <- 
        sprintf(
            "%s/ref_v_all/%s_%s_filtered.delta",
            delta_dir, most_related, accessions[1]
        ) %>%file.exists()
    
    
    delta_identities <- 
        pbmclapply(
            1:length((list.files(sprintf("%s/ref_v_all", delta_dir)))),
            function(i){
                delta_tib <- 
                    read_delta(
                        list.files(
                            sprintf("%s/ref_v_all", delta_dir), 
                            full.names = T
                            )[i]
                        ) 
                
                nucmer_id <- 
                    sum(
                        delta_tib$meanlen
                        )/
                    (mean(
                        c(
                            delta_tib$rlen[1], 
                            delta_tib$qlen[1])
                         ))
                
                out_tib <- 
                    tibble(
                        accession = 
                            delta_tib$qid[1],
                        identity = 
                            nucmer_id
                    )
                
                return(out_tib)
                
            }, mc.cores = 4
        ) %>% bind_rows() %>%
        filter(accession %in% accessions) 
    
    # delta_identities %>%
    #     add_count(accession)
    #     duplicated(accession)
    #length(accessions)
    #length(unique(accessions))
    
    delta_identities$accession
    
    clustered_genomes <- 
       clustered_genomes_list$cluster_summary %>%
        left_join(
            ., 
            sync_tib[,c("accession", "len")] %>% 
                `colnames<-`(c("accessions", "len")),
            by = "accessions"
            ) 
    
    
    best_identity <- 
        inner_join(
            delta_identities %>% `colnames<-`(c("accessions", "identity")),
            clustered_genomes[,-5],
            by = "accessions"
        ) %>%
        mutate(
            identity = 
                ifelse(identity>1,  1, identity),
            identity = 
                ifelse(accessions==most_related,  1.01, identity),
        ) %>%
        arrange(cluster_id)
    
    ## Anova for all others
    
    # best_identity %>% filter(cluster_id!=0) %>%
    #     arrange(cluster_id) %>%
    #     count(cluster_id)
    # best_identity_ss
    mt1 <- 
        species_identities$mash_tib %>%
        filter(
            ref %in% clustered_genomes$accessions,
            qry %in% clustered_genomes$accessions
        ) %>%
        left_join(
            .,
            clustered_genomes[,c(1,2)] %>%
                `colnames<-`(c("ref_clust", "ref")),
            by = "ref"
        ) %>%
        left_join(
            .,
            clustered_genomes[,c(1,2)] %>%
                `colnames<-`(c("qry_clust", "qry")),
            by = "qry"
        )
    
    mt2 <- 
        mt1%>%
        filter(
            (ref_clust %in% 0:5),
            (qry_clust %in% 0:5)
            )
    ggplot(
                data = mt2,
                aes(
                    x= ref,#factor(ref, seq_order),
                    y= qry,#factor(qry, seq_order),
                    fill = identity
                )
            ) +
            geom_tile() +
            theme_classic() +
            theme(
                #axis.text.x = element_text(angle=45, hjust = 1),
                panel.background=element_rect(fill="white"),
                plot.title = element_text(size=20,face="bold", hjust = 0.5),
                axis.title = element_blank(),
                axis.text = element_blank()
            ) +
            ggtitle(
                sprintf(
                    "%s \n Sequence pairwise MASH similarity",
                    sub("_", " ", species_name)
                )
            ) +
            theme(
                legend.title=element_text( size=18),
                axis.text=element_blank(),
                    #element_text(size=14),
                legend.text = element_text(size=18)
            ) +
            scale_fill_viridis(
                option = "B", limits = c(95,100)
            ) +
        facet_grid(
        rows = vars(qry_clust), cols = vars(ref_clust),
        scales = "free", space = "free", switch = "y")
    
    bis <- 
        best_identity %>% 
        group_by(cluster_id) %>%
        arrange(desc(identity)) %>% 
        filter(!is.na(cluster_id)) %>%
        dplyr::slice(1)
    
    
    bi_ss <- 
        best_identity %>%
        filter(grouped_sequences>=5) %>%
        group_by(cluster_id) %>%
        dplyr::slice(1:3)
    
    bis_f <- 
        sprintf(
            "%s/%s.txt",
            sync_dir, bi_ss$accessions
        )
    
    # lapply(
    #     1:length(bis_f),
    #     function(ij){
    #         file.copy(
    #             bis_f[ij],
    #             sprintf(
    #                 "../data/processing/extra/s_pyogenes3/%s",
    #                 basename(bis_f[ij])
    #             )
    #         )
    #     }
    # )
    
    saveRDS(
        bis, 
        sprintf(
            "%s/bis.rds",
            results_dir
        )
        )
    
    bis_1 <- 
        bis[-1,]
    
    if (nrow(bis_1)==0) next
    mpis <- 
        lapply(
            1:nrow(bis_1),
            function(j){
                dfj <- 
                    filtered_combined_delta %>% 
                    ungroup %>%
                    filter(qid==pull(bis_1, accessions)[j]) 
                # if (j==5){
                #     pj <- 
                #         plot_delta(dfj)  +
                #         xlab("Cluster 0")
                # } else {
                    pj <- 
                        plot_delta(dfj)  +
                        xlab("")
                #}
                pjo <- 
                    pj +
                    theme(
                        axis.text = element_text(size = 18),
                        axis.title = element_text(size = 22)
                    ) +
                    ylab(sprintf("Cluster %s", j)) 
                return(pjo)
            }
        )
    
    saveRDS(
        mpis, 
        sprintf(
            "%s/mpis.rds",
            results_dir
        )
        )
    
    if (length(mpis)>15){
        mpis <- 
            mpis[1:15]
    }
    
    plots_dir[[i]] <- 
        sprintf("%s/delta_plots.png", results_dir)
    
    png(sprintf("%s/delta_plots.png", results_dir), width = 800, height = 600)
    do.call(
         "grid.arrange",
         c(
             mpis,
             ncol=3
             )
         )
    # Close the PNG device to save the image
    dev.off()



```




```{r}
plots_dir <- 
    list()

ssn <- 
    list()

#i = 39, 50, 66, 72
which(species_name == spec2_list)

spec2_list_100_500 %>% arrange(desc(spec_count))


which(spec2_list_100_500$species2=="Citrobacter_freundii")
which(spec2_list_100_500$species2=="Streptococcus_suis")

i=25
for (i in 15){#length(spec2_list)){
    i=1
    species_name <- 
        (species_list %>% pull(species2))[i]
        #spec2_list[i]
    
    ## set up directories for this species
    genus_name <- 
      strsplit(species_name, "_")[[1]][1]
    
    clade_name <- 
      filter(ncbi_table_list$ncbi_cc, species_==species_name) %>%
      pull(phylum) %>% gsub(" ", "_", .)
    
    species_variable_names <-
      c(
          "fna_dir",
          "sync_dir", "gff_dir", "identity_dir", "mummer_dir",
          "delta_dir", "r_vars_dir", "results_dir", 
          "repeats_dir", "invs_dir",
          "disparities_dir", "c01_dir",
          "breakpoint_dir", "prokka_dir",
          "breakpoint_comparisons_dir", "snp_dir",
          "kmer_dir", "bd_dir", "plot_dir", "roary_dir",
          "sync2_dir"
      )
    
    species_dirs <- 
      gsub("_dir", "", species_variable_names)
    
    for (var_name in species_variable_names){
      dir_name <- 
        gsub("_dir", "", var_name)
      full_dir_name <- 
        sprintf(
            "../data/processing/%s/%s/%s/%s", 
            dir_name, clade_name, 
            genus_name, species_name
            )
      
      assign(
        var_name, 
        full_dir_name
      )
      if (!dir.exists(full_dir_name)){dir.create(full_dir_name, recursive = T)}
    }
    
    species_table <- 
        ncbi_table %>%
        filter(species== sub("_", " ", species_name))
    
    spec_count <- 
        nrow(species_table)
    
    if(spec_count<250){
        subsample_idx <- 
            1:spec_count
    } else {
        set.seed(123)
        subsample_idx <- 
            sample(1:spec_count, 200)
    }
    
    #subsample_idx
    
    st_ss <- 
        species_table %>%
        dplyr::slice(subsample_idx)
    
    the_rest_idx <- 
        species_table %>%
        filter(!row_number() %in% subsample_idx)
        
    species_summary <- #species_table
        get_species_summary(st_ss) 
    
    ssn[[i]] <- 
        tibble(
            spec_ = species_name,
            stn = nrow(species_table),
            ssn = nrow(species_summary)
        )
    if (nrow(species_summary)<40) next
        
    #print(paste(species_name, i))
    saveRDS(
        species_summary, 
        sprintf(
            "%s/species_summary.rds",
            r_vars_dir
        )
    )
    
    ftp_paths <- 
        species_summary$ftp_path
    
    gff_paths <- 
        ftp_paths %>%
                sprintf(
                    "%s/%s_genomic.gff.gz",
                    ., gsub(".*/", "", .)
                )
    # fna_paths <- 
    #     ftp_paths %>%
    #             sprintf(
    #                 "%s/%s_genomic.fna.gz",
    #                 ., gsub(".*/", "", .)
    #             )
    
    # file_exists_at_url(fna_paths[1])
    # file_exists_at_url(gff_paths[1])
    gff_check <- 
        pbmclapply(
            gff_paths, 
            file_exists_at_url
        ) %>% unlist
    
    if (!any(gff_check)) next
    
    # ij <-
    #     2
    # 
    # f_ij <- 
    #     ftp_paths[ij]
    # gff_ij <-
    #     f_ij %>%
    #             sprintf(
    #                 "%s/%s_genomic.gff.gz",
    #                 ., gsub(".*/", "", .)
    #             )
    # 
    # gff_table <-
    #         read.gff(
    #             gzcon(url(gff_path))
    #         )
    # chr2_acc <-
    #     gff_table %>%
    #     as_tibble %>%
    #     filter(
    #         (grepl("chromosome=", attributes, ignore.case = F))
    #         #(grepl("origin", attributes, ignore.case = F))
    #     ) %>%
    #     pull(seqid) %>%
    #     as.character
    # 
    # repAs <- 
    #     list()
    # for (ij in 1:length(ftp_paths)){
    #     f_ij <- 
    #         ftp_paths[ij]
    #     gff_ij <- 
    #         f_ij %>%
    #                 sprintf(
    #                     "%s/%s_genomic.gff.gz",
    #                     ., gsub(".*/", "", .)
    #                 )
    #     
    #     gff_table <- 
    #             read.gff(
    #                 gzcon(url(gff_path))
    #             )
    #     chr2_acc <- 
    #         gff_table %>%
    #         as_tibble %>%
    #         filter(
    #             (grepl("chromosome=II", attributes, ignore.case = F))
    #             #(grepl("origin", attributes, ignore.case = F))
    #         ) %>%
    #         pull(seqid) %>% 
    #         as.character
            
        
    #     gff_table_chr2 <- 
    #         gff_table %>% filter(seqid==chr2_acc)%>% as_tibble %>%
    #         mutate(
    #             gene_value = 
    #                 str_replace(
    #                     str_extract(attributes, "gene=\\w+"), "gene=", ""
    #                     ),
    #             # name_value = 
    #             #     str_replace(
    #             #         str_extract(attributes, "Name=\\w+"), "Name=", ""
    #             #         )
    #             prod_value = 
    #                 str_replace(
    #                     str_extract(attributes, "product=\\w+"), "product=", ""
    #                     )
    #             ) %>%
    #         filter(!(is.na(gene_value))) %>%
    #         pull(gene_value) %>% unique
    #             
    #     gff_table$seqid %>% unique
    #     gff_table %>% as_tibble
    #     gff_table %>% as_tibble %>%
    #         group_by(seqid) %>% 
    #         filter(end==max(end))
    #     
    #     repAs[[ij]] <- 
    #         gff_table %>%
    #         as_tibble %>%
    #         filter(
    #             (grepl("chromosome=II", attributes, ignore.case = F))
    #             #(grepl("origin", attributes, ignore.case = F))
    #         )
    #     repAs[[ij]]
    #     
    # }
    # lapply(repAs, head) %>% unlist
    
    ## check for 2nd chromosomes
    
    
    
    # species_summary <- #species_table
    #     readRDS(
    #         sprintf(
    #            "%s/species_summary.rds",
    #             r_vars_dir
    #            )
    #     )
    closeAllConnections()
    sync_tib <- 
        pbmclapply(
            1:nrow(species_summary),
            function(ii){
                #for (ii in nrow(species_summary):1){
                assembly_path <- 
                    species_summary$ftp_path[ii]
                
                species_summary$accession[ii]
                sync_summary <- 
                    sync_gene(
                        assembly_path = 
                            assembly_path,
                        sync_dir = 
                            sync_dir, 
                        gff_dir = 
                            gff_dir
                        ) %>%
                    mutate(
                        idx = ii
                    )
                return(sync_summary)
            }, mc.cores = 4
        ) %>% bind_rows()
    
    if (all(is.na(sync_tib$ori_pos))){
        saveRDS(
            sync_tib,
            sprintf(
                "%s/sync_tib.rds",
                results_dir
            )
        )
        next
    }
    
    ## put all in the same directory
    
    dirname(sync_tib$out_file)
    
    ##
    
    # which(!(sync_tib %>% filter(chr==1) %>% pull(accession)) %in%
    # (list.files(sdc) %>% sub(".txt", "", .)))
    # 
    # (sync_tib %>% pull(accession))[2]
    # 
    # sync_tib %>%
    #     filter(chr==2)
    # synceing the unsynced
    unsynced_idx <- 
        sync_tib %>%
        filter(is.na(ori_pos)) %>%
        filter(chr==1)
    
    unsynced_idx %>% filter(!is.na(ori_pos))
    
    rs_tibs <- list()
    if (nrow(unsynced_idx)>0){
        
        for (m in 1:length(unsynced_idx$idx)){
        #species_summary
        assembly_path <- 
            species_summary$ftp_path[unsynced_idx$idx[m]]
        
        us_chrom <-
            sync_tib$chr[unsynced_idx$idx[m]]
    
        unsynced_seq <- 
            get_fasta(assembly_path, chromosome = us_chrom)
        fasta_seq <- 
            sprintf(
                "%s/%s.txt",
                sync_dir,
                names(unsynced_seq)
            )
        writeXStringSet(
            unsynced_seq,
            fasta_seq
        )
        
        ref_seq <- 
            list.files(
                sprintf(
                    "%s/chr_%s_gene",
                    sync_dir, us_chrom
                ),
                full.names = T
            )[1]
    
        startpoint_dir <- 
            sprintf("%s/starts", sync_dir)
        if (!dir.exists(startpoint_dir)){dir.create(startpoint_dir)}
        
        rs_synced <- 
            rs_sync(
                ref_seq, fasta_seq, 
                delta_out = 
                    sprintf("%s/%s", startpoint_dir,  names(unsynced_seq)
                            ),
                out_file = 
                    fasta_seq
            ) 
        
        if (!is.null(rs_synced)){
            
            ref_seq <- 
                list.files(
                    sprintf(
                        "%s/chr_%s_gene",
                        sync_dir, us_chrom
                    ),
                    full.names = T
                )[2]
            rs_synced <- 
                rs_sync(
                    ref_seq, fasta_seq, 
                    delta_out = 
                        sprintf("%s/%s_2", startpoint_dir,  names(unsynced_seq)
                                ),
                    out_file = 
                        fasta_seq
                ) 
            if (!is.null(rs_synced)){
                rs_tibs[[m]] <- 
                    rs_synced %>%
                    mutate(
                        asm_name =
                            basename(assembly_path),
                        chr = us_chrom,
                        idx = unsynced_idx$idx[m]
                        )
            } else {
                rs_tibs[[m]] <- 
                    rs_synced 
            }
        } else {
            rs_tibs[[m]] <- 
                rs_synced 
        }
        
        
      
        }
        rs_tib <- 
            bind_rows(rs_tibs)
        
        sync_tib <- 
            bind_rows(
                sync_tib %>% filter(!is.na(ori_pos)),
                rs_tib
            )
    }
    
    
    if (nrow(sync_tib)<10) next
    
    saveRDS(
        sync_tib, 
        sprintf(
            "%s/sync_tib.rds",
            results_dir
        )
        )
    
    ###
    # check chr 2
    if (("chr_1" %in% basename(list.dirs(sync_dir)))){
        sdc <- 
            sprintf(
                "%s/chr_1",
                sync_dir
            )
        ## move any files if necessary
        no_2nd <- 
            (sync_tib %>%
            filter(chr==1) %>%
            pull(accession))[
                !((sync_tib %>% filter(chr==1) %>% pull(accession)) %in%
                  (list.files(sdc) %>% sub(".txt", "", .)))
            ]
        list.files(sdc)
        
        no_2nd <- 
            sprintf(
                "%s/%s.txt", 
                sync_dir, no_2nd
            )
        if (length(no_2nd)>0) {
            for (ij in 1:length(no_2nd)){
                n2 <- 
                    sprintf(
                        "%s/%s.txt", 
                        sync_dir, 
                        no_2nd[ij]
                    )
                    
                if (file.exists(n2)){
                    file.copy(
                        from = n2,
                        to = 
                            sprintf("%s/%s", sdc, basename(n2)) 
                        )
                    
                }
            }
        }
        
        
        
    } else{
        sdc <- 
            sync_dir
    }
    list.files(sdc)
    ## Identities
    species_identities <- 
        get_pairwise_identities(
            sync_dir = sdc, 
            identity_dir = identity_dir, 
            species_table = sync_tib, 
            genus_name, species_name
            )
    
    #species_identities$mash_plot
    saveRDS(
        species_identities, 
        sprintf(
            "%s/species_identities.rds",
            results_dir
        )
        )
    
    # species_identities <- 
    #     readRDS(
    #         sprintf(
    #             "%s/species_identities.rds",
    #             results_dir
    #         )
    #     )
    # 
    species_identities$id_dists
    #read_delta("../data/processing/delta/Pseudomonadota/Escherichia/Escherichia_coli/ref_v_all/NZ_CP010")
    ##
    sampled_id_table <- 
        species_identities$id_dists %>%
        arrange(desc(Sequence_mean_distance))# %>%
        #filter(Sequence_mean_distance<10) 
    
    most_related <- 
        sampled_id_table %>%
        arrange(Sequence_mean_distance) %>%
        dplyr::slice(1) %>% 
        pull(Sequence)
    
    ## pairwise alignemnts
    accessions <- 
        sync_tib %>%
        filter(chr==1) %>%
        pull(accession)
    
    sync_tib %>%filter(chr==1, is.na(ori_pos))
    
    which(
        !(accessions %in%
              sub(".txt", "", list.files(sdc)))
    )
    
    nucmer_commands <- 
        nucmer_ref_v_all(
            most_related, 
            accessions,
            delta_dir =  delta_dir, 
            sync_dir = sdc
            )
    
    minimum_alignment_percentage <- 
        90
    minimum_inversion_length <- 
        5e4
    filtered_combined_delta <- 
        filter_combined_delta(
            sprintf("%s/ref_v_all", delta_dir), 
            minimum_alignment_percentage, 
            minimum_inversion_length
        ) 
    
    likely_missassembled <- 
        filtered_combined_delta %>% 
        group_by(qid) %>%
        mutate(
            ncs = n(),
            X_sum = sum(X_dist)
            ) %>%
        filter(ncs>20 | X_sum>1e7) %>%
        pull(qid) %>% unique
    
    #filtered_combined_delta %>% filter(qid %in%likely_missassembled[2] ) %>% plot_delta
    #filtered_combined_delta %>% filter(qid %in%likely_missassembled[1] ) %>% plot_delta
    
    fcd <- 
        filtered_combined_delta %>%
        filter(!(qid %in% likely_missassembled)) %>%
        filter(qid %in% accessions)
    
    saveRDS(
        fcd, 
        sprintf(
            "%s/fcd.rds",
            results_dir
        )
        )
    
    # fcd <- 
    #     readRDS(
    #         sprintf(
    #         "%s/fcd.rds",
    #         results_dir
    #     )
    #     )

    clustered_genomes_list <- 
        cluster_genomes(fcd)
    
    saveRDS(
        clustered_genomes_list, 
        sprintf(
            "%s/clustered_genomes_list.rds",
            results_dir
        )
        )
    
    out_ds <- 
        sprintf(
            "%s/ref_v_all/%s_%s_filtered.delta",
            delta_dir, most_related, accessions[1]
        ) %>%file.exists()
    
    
    delta_identities <- 
        pbmclapply(
            1:length((list.files(sprintf("%s/ref_v_all", delta_dir)))),
            function(i){
                delta_tib <- 
                    read_delta(
                        list.files(
                            sprintf("%s/ref_v_all", delta_dir), 
                            full.names = T
                            )[i]
                        ) 
                
                nucmer_id <- 
                    sum(
                        delta_tib$meanlen
                        )/
                    (mean(
                        c(
                            delta_tib$rlen[1], 
                            delta_tib$qlen[1])
                         ))
                
                out_tib <- 
                    tibble(
                        accession = 
                            delta_tib$qid[1],
                        identity = 
                            nucmer_id
                    )
                
                return(out_tib)
                
            }, mc.cores = 4
        ) %>% bind_rows() %>%
        filter(accession %in% accessions) 
    
    # delta_identities %>%
    #     add_count(accession)
    #     duplicated(accession)
    #length(accessions)
    #length(unique(accessions))
    
    delta_identities$accession
    
    clustered_genomes <- 
       clustered_genomes_list$cluster_summary %>%
        left_join(
            ., 
            sync_tib[,c("accession", "len")] %>% 
                `colnames<-`(c("accessions", "len")),
            by = "accessions"
            ) 
    
    
    best_identity <- 
        inner_join(
            delta_identities %>% `colnames<-`(c("accessions", "identity")),
            clustered_genomes[,-5],
            by = "accessions"
        ) %>%
        mutate(
            identity = 
                ifelse(identity>1,  1, identity),
            identity = 
                ifelse(accessions==most_related,  1.01, identity),
        ) %>%
        arrange(cluster_id)
    
    ## Anova for all others
    
    # best_identity %>% filter(cluster_id!=0) %>%
    #     arrange(cluster_id) %>%
    #     count(cluster_id)
    # best_identity_ss
    mt1 <- 
        species_identities$mash_tib %>%
        filter(
            ref %in% clustered_genomes$accessions,
            qry %in% clustered_genomes$accessions
        ) %>%
        left_join(
            .,
            clustered_genomes[,c(1,2)] %>%
                `colnames<-`(c("ref_clust", "ref")),
            by = "ref"
        ) %>%
        left_join(
            .,
            clustered_genomes[,c(1,2)] %>%
                `colnames<-`(c("qry_clust", "qry")),
            by = "qry"
        )
    
    mt2 <- 
        mt1%>%
        filter(
            (ref_clust %in% 0:5),
            (qry_clust %in% 0:5)
            )
    ggplot(
                data = mt2,
                aes(
                    x= ref,#factor(ref, seq_order),
                    y= qry,#factor(qry, seq_order),
                    fill = identity
                )
            ) +
            geom_tile() +
            theme_classic() +
            theme(
                #axis.text.x = element_text(angle=45, hjust = 1),
                panel.background=element_rect(fill="white"),
                plot.title = element_text(size=20,face="bold", hjust = 0.5),
                axis.title = element_blank(),
                axis.text = element_blank()
            ) +
            ggtitle(
                sprintf(
                    "%s \n Sequence pairwise MASH similarity",
                    sub("_", " ", species_name)
                )
            ) +
            theme(
                legend.title=element_text( size=18),
                axis.text=element_blank(),
                    #element_text(size=14),
                legend.text = element_text(size=18)
            ) +
            scale_fill_viridis(
                option = "B", limits = c(95,100)
            ) +
        facet_grid(
        rows = vars(qry_clust), cols = vars(ref_clust),
        scales = "free", space = "free", switch = "y")
    
    bis <- 
        best_identity %>% 
        group_by(cluster_id) %>%
        arrange(desc(identity)) %>% 
        filter(!is.na(cluster_id)) %>%
        dplyr::slice(1)
    
    
    bi_ss <- 
        best_identity %>%
        filter(grouped_sequences>=5) %>%
        group_by(cluster_id) %>%
        dplyr::slice(1:3)
    
    bis_f <- 
        sprintf(
            "%s/%s.txt",
            sync_dir, bi_ss$accessions
        )
    
    # lapply(
    #     1:length(bis_f),
    #     function(ij){
    #         file.copy(
    #             bis_f[ij],
    #             sprintf(
    #                 "../data/processing/extra/s_pyogenes3/%s",
    #                 basename(bis_f[ij])
    #             )
    #         )
    #     }
    # )
    
    saveRDS(
        bis, 
        sprintf(
            "%s/bis.rds",
            results_dir
        )
        )
    
    bis_1 <- 
        bis[-1,]
    
    if (nrow(bis_1)==0) next
    mpis <- 
        lapply(
            1:nrow(bis_1),
            function(j){
                dfj <- 
                    filtered_combined_delta %>% 
                    ungroup %>%
                    filter(qid==pull(bis_1, accessions)[j]) 
                # if (j==5){
                #     pj <- 
                #         plot_delta(dfj)  +
                #         xlab("Cluster 0")
                # } else {
                    pj <- 
                        plot_delta(dfj)  +
                        xlab("")
                #}
                pjo <- 
                    pj +
                    theme(
                        axis.text = element_text(size = 18),
                        axis.title = element_text(size = 22)
                    ) +
                    ylab(sprintf("Cluster %s", j)) 
                return(pjo)
            }
        )
    
    saveRDS(
        mpis, 
        sprintf(
            "%s/mpis.rds",
            results_dir
        )
        )
    
    if (length(mpis)>15){
        mpis <- 
            mpis[1:15]
    }
    
    plots_dir[[i]] <- 
        sprintf("%s/delta_plots.png", results_dir)
    
    png(sprintf("%s/delta_plots.png", results_dir), width = 800, height = 600)
    do.call(
         "grid.arrange",
         c(
             mpis,
             ncol=3
             )
         )
    # Close the PNG device to save the image
    dev.off()
    
    
}
```

Iterative Synchronize, Align, Remove
```{r}




```




## results 

```{r}

outer_results_dir <- 
    sprintf(
        "../data/processing/%s", 
        "results"
        )

## phyla covered
phyla_results <- 
    list.dirs(
        outer_results_dir, recursive = F
    ) %>% basename

more_gs <- 
    spec2_list_100_500

## check results

species_run <- 
    pbmclapply(
        1:nrow(species_list),
        function(i){
            bis_dir <- 
                list.files(
                    sprintf(
                        "%s/%s/%s/%s",
                        outer_results_dir, 
                        species_list$p2[i],
                        species_list$genus[i],
                        species_list$species2[i]
                        ), full.names = T, 
                    pattern = "bis.rds"
                    )
            if (length(bis_dir)==1){
                bi_tib <- 
                    read_rds(bis_dir)
                out_tib <- 
                    tibble(
                        clusters = 
                            nrow(bi_tib),
                        clusters_5plus = 
                            sum(bi_tib$grouped_sequences>5),
                        species = 
                            more_gs$species2[i]
                        )
            } else{
                out_tib <- 
                    tibble(
                        clusters = NA,
                        clusters_5plus = 
                            sum(bi_tib$grouped_sequences>5),
                        species = 
                            species_list$species2[i]
                    )
            }
            return(out_tib)
        }
    ) %>% 
    bind_rows()
species_run
sr <- 
    species_run %>%
    group_by(species) %>%
    summarise(
        nc = n(),
        run = sum(!(is.na(cluster_id)))
    )

bis_files <- 
    lapply(
        1:nrow(species_list_40),
        function(i){
            bis_dir <- 
                list.files(
                    sprintf(
                        "%s/%s/%s/%s",
                        outer_results_dir, 
                        species_list_40$p2[i],
                        species_list_40$genus[i],
                        species_list_40$species2[i]
                        ), full.names = T, 
                    pattern = "bis.rds"
                    )
            if (length(bis_dir)==1){
                out_tib <- 
                    read_rds(bis_dir) %>%
                    mutate(
                        species = 
                            species_list_40$species2[i]
                    )
                
                
            } else{
                out_tib <- 
                    NULL
            }
            return(out_tib)
        }
    )

bis_files[[26]]

bi_over5 <- 
    bind_rows(bis_files) %>%
    ungroup %>%
    filter(cluster_id %in% c(0,1)) %>%
    group_by(species) %>%
    mutate(
        c1t = 
            sum(
                ifelse(
                    (((cluster_id==1) & (grouped_sequences>=5)) |
                         
                         (cluster_id==0)), 
                    1,0
                   )
            )
        ) %>%
    filter(c1t==2)
 more_gs$species2  
 
 bi_5_2 %>% filter(species == "Mannheimia_haemolytica")
  
 bi_5_2 <- 
     bind_rows(bis_files) %>%
     ungroup %>%
     filter(
         grouped_sequences >=5,
         ) %>%
     group_by(species) %>%
     mutate(
         clusters_5 = n()
     ) #%>%
     # filter(
     #     clusters_5 >=2
     # )
     # 
     
more_gs <- 
    species_list %>%
    filter(
        species2 %in% bi_5_2$species
    )
more_gs$species


out_boots <- 
    pbmclapply(
        1:nrow(more_gs),
        function(i){
            for (i in nrow(more_gs):1){
            dir_files <- 
                list.files(
                    sprintf(
                        "%s/%s/%s/%s",
                        outer_results_dir, 
                        more_gs$p2[i],
                        more_gs$genus[i],
                        more_gs$species2[i]
                        ), full.names = T
                    )
            if (length(dir_files)>2){
                out_tib <- 
                    tibble(
                        dir_files = dir_files,
                        spec = more_gs$species2[i]
                    )
                bi_tib <- 
                    list.files(
                    sprintf(
                        "%s/%s/%s/%s",
                        outer_results_dir, 
                        more_gs$p2[i],
                        more_gs$genus[i],
                        more_gs$species2[i]
                        ), full.names = T, 
                    pattern = "bis.rds"
                    )
                
                cluster_list <- 
                    dir_files[grepl("clustered_genomes", dir_files)] %>%
                    readRDS
                
                clustered_gs <- 
                    cluster_list$cluster_summary %>%
                    filter(grouped_sequences>=5)
                clustered_gs %>% count(cluster_id)
                if (length(unique(clustered_gs$cluster_id))<2){
                    out_tib <- 
                        NULL
                } else {
                    
                
                #clustered_gs %>% count(cluster_id)
                
                # c1_len <-
                #     nrow(
                #         clustered_gs %>% filter(cluster_id==1)
                #     )
                # clustered_gs %>% dplyr::count(cluster_id)
                # 
                # cluster_gs_sub <- 
                #     clustered_gs %>%
                #     group_by(cluster_id) %>%
                #     dplyr::slice(1:c1_len)
                
                    id_list <- 
                        dir_files[grepl("species_identities", dir_files)] %>%
                        readRDS
                    id_list$mash_plot
                    mash_tib <- 
                        id_list$mash_tib%>%
                            left_join(
                                .,
                                clustered_gs[,c(1,2)] %>%
                                    `colnames<-`(c("ref_clust", "ref")),
                                by = "ref"
                            ) %>%
                            left_join(
                                .,
                                clustered_gs[,c(1,2)] %>%
                                    `colnames<-`(c("qry_clust", "qry")),
                                by = "qry"
                            ) 
                    
                    mt2 <- 
                        mash_tib %>% 
                        mutate(
                            type = 
                                ifelse(
                                    ((ref_clust==qry_clust) & (ref_clust==0)), 
                                    "within_0",
                                    ifelse(
                                        ((ref_clust==qry_clust) & (ref_clust==1)),
                                        "within_1",
                                        "between"
                                    )
                                    ),
                            t3 = 
                                ifelse(
                                    ref_clust==qry_clust, 
                                    "within",
                                    "between"
                                )
                        ) %>%
                        filter(ref!=qry)
                        # filter(
                        #     (ref %in% cluster_gs_sub$accessions),
                        #     (qry %in% cluster_gs_sub$accessions),
                        #     ref !=qry
                        #     )
                    
                    ## uniquify the mash tib
                    
                    # mt2 %>% filter(ref_clust==1) %>% pull(ref) %>% unique
                    # mt3 %>% filter(rc_qc=="1-1")
                    mt3 <- 
                        mt2 %>%
                        dplyr::select(
                            c(ref, qry, identity, ref_clust, qry_clust, type, t3)
                            ) %>%
                        filter(
                            ref %in% clustered_gs$accessions,
                            qry %in% clustered_gs$accessions
                        ) %>%
                        mutate(
                            pair_id = 
                                apply(
                                    dplyr::select(., ref, qry), 
                                    1, 
                                    function(x){
                                        paste(sort(x), collapse = "_VS_")
                                        }
                                    ),
                            rc_qc = 
                                apply(
                                    dplyr::select(., ref_clust, qry_clust), 
                                    1, 
                                    function(x){
                                        paste(sort(x), collapse = "_VS_")
                                        }
                                    )
                            ) %>%
                        group_by(
                            pair_id, rc_qc, t3
                        ) %>%
                        summarise(
                            id = 
                                mean(identity),
                            #n=n(),
                            .groups = "keep"
                        ) %>% 
                        ungroup %>%
                        mutate(
                            id_dist = 
                                round(100-id, 2)#,
                            # type2 = 
                            #     gsub("(.+?)(\\_.*)", "\\1", type)
                        ) 
                    
                    ## bootstrap results
                    bootstrap_diff2 <- function(data1, indices1) {
                          sample_data <- data1[indices1, ]
                          within_mean <- mean(sample_data$id_dist[sample_data$t3 == "within"])
                          between_mean <- mean(sample_data$id_dist[sample_data$t3 == "between"])
                          
                          return(between_mean - within_mean)
                    }
                    # bootstrap_within <- function(data1, indices) {
                    #   sample_data <- data1[indices, ]
                    #   mean(sample_data$id_dist[sample_data$type == "within"])
                    # }
                    # 
                    # bootstrap_between <- function(data1, indices) {
                    #   sample_data <- data1[indices, ]
                    #   mean(sample_data$id_dist[sample_data$type == "between"])
                    # }
                    
                    unique_combos <- 
                        mt3 %>%
                        filter(t3 =="between") %>%
                        pull(rc_qc) %>%
                        unique
                    data.frame(unique_combos, 1:length(unique_combos))
                   
                     # ggplot(
                     #    mt3 %>%
                     #        filter(
                     #            (t3 == "within") |
                     #                grepl(0, rc_qc)
                     #        ),
                     #    aes(x = rc_qc, y = id_dist, fill = t3)
                     #    ) +
                     #  geom_violin(trim = FALSE) +
                     #    geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9)) +
                     #     theme_classic() +
                     #     theme(text = element_text(size=22), legend.position = "none") + 
                     #     labs(x = "distance", y = "")
                     
                    #  ggplot(
                    #     mt3 %>%
                    #         filter(
                    #             (t3 == "between") &
                    #                 (!grepl(0, rc_qc))
                    #         ),
                    #     aes(x = rc_qc, y = id_dist, fill = t3)
                    #     ) +
                    #   geom_violin(trim = FALSE) +
                    #     geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9)) +
                    #      theme_classic() +
                    #      theme(text = element_text(size=22), legend.position = "none") + 
                    #      labs(x = "distance", y = "")
                    #      
                    #  
                    #  ggplot(
                    #     mt3 %>%
                    #     filter(
                    #         (t3 == "within") |
                    #             grepl(0, rc_qc)
                    #     ),
                    #     aes(x = rc_qc, y = id_dist, fill = t3)
                    #     ) +
                    #   geom_violin(trim = FALSE) +
                    #     geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9)) +
                    #      theme_classic() +
                    #      theme(text = element_text(size=22)) +
                    #      coord_flip()
                    #  
                    # ggplot(
                    #     mt3 %>%
                    #         filter(
                    #             (t3 == "within") |
                    #                 rc_qc %in% 
                    #                 unique_combos[c( 13, 16, 19, 4,9, 2,17,5, 14)]
                    #         ), 
                    #     aes(x = rc_qc, y = id_dist, fill = t3)
                    #     ) +
                    #   geom_violin(trim = FALSE) +
                    #     geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9))
                    # 
                    # 
                    #mt3$type2 %>% unique
                    set.seed(123)
                    boot_results <- 
                        boot(mt3, bootstrap_diff2, R = 50)
                    
                    boot_ci <- 
                        boot.ci(boot_results, type = "perc")
                    
                    boot_results$t %>% head
                    
                    out_tib <- 
                        more_gs[i,c(-2)] %>%
                        mutate(
                            "Between-within" = 
                                mean(boot_results$t, na.rm = T),
                            "95% CI low" = 
                                boot_ci$percent[4],
                            "95% CI high" = 
                                boot_ci$percent[5],
                            clusters = 
                                length(unique(clustered_gs$cluster_id))
                        )
                        
                }
                # boot_diff_between_df <- 
                #     data.frame(boot_diff = boot_results$t)
                # ggplot(boot_diff_between_df, aes(x = boot_diff)) +
                #   geom_histogram(bins = 30, fill = "blue", alpha = 0.7) +
                #   geom_vline(xintercept = 0, linetype = "dashed", color = "red") +  # Reference line at 0
                #   labs(title = "Bootstrapped Differences in Mean Distances (Between - Within)",
                #        x = "Difference in Mean Distances (Between - Within)",
                #        y = "Frequency")
                # 
                # 
                # bootstrap_diff <- 
                #     function(data1, indices1) {
                #         sample_data <- data1[indices1, ]
                #         between_mean <- mean(sample_data$id_dist[sample_data$type == "between"])
                #         w0_mean <- mean(sample_data$id_dist[sample_data$type == "within_0"])
                #         w1_mean <- mean(sample_data$id_dist[sample_data$type == "within_1"])
                #         # Return differences: (Group 1 within - Between), (Group 2 within - Between)
                #         return(c(w0_mean + w1_mean - between_mean))
                #     }
                # 
                # boot_diff_results <- 
                #     boot(mt3, bootstrap_diff, R = 100)
                # boot_diff_results$t
                # 
                # boot_data <- 
                #     as_tibble(boot_diff_results$t) %>%
                #     `colnames<-`(
                #         c("within_group_0", "within_group_1", "between_1_2")
                #         )
                # 
                # boot_long <- 
                #     boot_data %>%
                #     pivot_longer(
                #         cols = everything(), 
                #         names_to = "Type", values_to = "bootstrapped_mean")
                # 
                # 
                # ggplot(boot_long, aes(x = Type, y = bootstrapped_mean, fill = Type)) +
                #     geom_boxplot() +
                #     theme_minimal() +
                #     labs(title = "Boxplot of Bootstrapped Means", x = "Group", y = "Bootstrapped Mean Similarity")
                #                     
                # mt3 %>% ungroup %>% count(type)
                # 
                # ggplot(boot_long, aes(x = Type, y = bootstrapped_mean, fill = Type)) +
                #   geom_violin(trim = FALSE) +
                #   geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9))
                #     
                    
                
            } else{
                out_tib <- 
                    NULL
            }
            return(out_tib)
        }, 
        mc.cores = 4
    )


library(igraph)
edges <- c(0, 1, 0, 2, 1, 3, 2, 4)
g <- graph(edges, directed = FALSE)

# Set node labels and sizes
V(g)$label <- V(g)$name
V(g)$size <- seq(10, 18, length.out = vcount(g))  # increasing node sizes

# Define edge colors
E(g)$color <- "black"    # Default color
E(g)$color[E(g) %in% c(1, 2)] <- "blue"  # Set color for edges 0-1 and 1-3
E(g)$color[E(g) %in% c(2, 3)] <- "red"   # Set color for edges 1-3 and 2-4

# Plot the graph with custom styles
plot(g,
     vertex.color = "lightgreen",
     vertex.label.color = "black",
     vertex.frame.color = "gray",
     vertex.label.cex = 1.5,
     edge.width = 2,
     edge.color = E(g)$color,
     layout = layout_with_kk)





# Define edges with vertex IDs starting from 1 instead of 0
edges <- c(1, 2, 1, 3, 2, 4, 3, 5)
g <- graph(edges, directed = FALSE)

# Set node labels as 0, 1, 2, 3, 4
V(g)$label <- 0:(vcount(g) - 1)

# Set node sizes to increase with each node
V(g)$size <- seq(10, 18, length.out = vcount(g))  # increasing node sizes

# Define edge colors
E(g)$color <- "black"    # Default color
E(g)$color[c(1, 2)] <- "blue"  # Set color for edges 1-2 and 2-4
E(g)$color[c(3, 4)] <- "red"   # Set color for edges 2-4 and 3-5

# Plot the graph with custom styles
plot(g,
     vertex.color = "lightgreen",
     vertex.label.color = "black",
     vertex.frame.color = "gray",
     vertex.label.cex = 1.5,
     edge.width = 2,
     edge.color = E(g)$color,
     layout = layout_with_kk)

species_boots <- 
    bind_rows(out_boots) %>%
    arrange(desc(`Between-within`)) %>%
    mutate(
        ci_lower = `95% CI low`,
        ci_upper = `95% CI high`,
        s3 = 
            sub("^(.).*_(.*)$", "\\1_\\2", species2) %>%
            sprintf("%s (%s)", ., spec_count) %>%
            factor(., levels = .[order(`Between-within`)]),
        Phylum = p2
    ) %>%
    dplyr::select(c(2,4,5,6,9:13))

ggplot(species_boots, aes(y = s3, x = `Between-within`, colour = Phylum)) +
  geom_point(size = 4) +    
    geom_vline(xintercept = 0, linetype = "dotted", color = "red", size = 1) +
  geom_errorbarh(aes(xmin = ci_lower, xmax = ci_upper), height = 0.2) +  
  labs(
      title = "Mean nucleotide distance between VS within \n Structural clusters",
       y = "", x = "Point Estimate (Between - Within Distance)"
      ) +
    theme_classic() +
    theme(
        plot.title = element_text(hjust = 0.5),
        text = element_text(size = 22)
    )
 

mg25 <- 
    more_gs %>%
    filter(species2 %in% species_boots$species2)
species_boots



out_boots2 <- 
    pbmclapply(
        1:nrow(mg25),
        function(i){
            dir_files <- 
                list.files(
                    sprintf(
                        "%s/%s/%s/%s",
                        outer_results_dir, 
                        mg25$p2[i],
                        mg25$genus[i],
                        mg25$species2[i]
                        ), full.names = T
                    )
            if (length(dir_files)>2){
                out_tib <- 
                    tibble(
                        dir_files = dir_files,
                        spec = more_gs$species2[i]
                    )
                bi_tib <- 
                    list.files(
                    sprintf(
                        "%s/%s/%s/%s",
                        outer_results_dir, 
                        more_gs$p2[i],
                        more_gs$genus[i],
                        more_gs$species2[i]
                        ), full.names = T, 
                    pattern = "bis.rds"
                    )
                
                cluster_list <- 
                    dir_files[grepl("clustered_genomes", dir_files)] %>%
                    readRDS
                
                clustered_gs <- 
                    cluster_list$cluster_summary %>%
                    filter(grouped_sequences>=5)
                cluster_counts <- 
                    clustered_gs %>% count(cluster_id)
                if (length(unique(clustered_gs$cluster_id))==1){
                    out_tib <- 
                        NULL
                } else {
                    

                
                    id_list <- 
                        dir_files[grepl("species_identities", dir_files)] %>%
                        readRDS
                    
                    mash_tib <- 
                        id_list$mash_tib%>%
                            left_join(
                                .,
                                clustered_gs[,c(1,2)] %>%
                                    `colnames<-`(c("ref_clust", "ref")),
                                by = "ref"
                            ) %>%
                            left_join(
                                .,
                                clustered_gs[,c(1,2)] %>%
                                    `colnames<-`(c("qry_clust", "qry")),
                                by = "qry"
                            ) 
                    
                    mt2 <- 
                        mash_tib %>% 
                        mutate(
                            type = 
                                ifelse(
                                    ((ref_clust==qry_clust) & (ref_clust==0)), 
                                    "within_0",
                                    ifelse(
                                        ((ref_clust==qry_clust) & (ref_clust==1)),
                                        "within_1",
                                        "between"
                                    )
                                    )
                        ) %>%
                        filter(ref!=qry)
                       
                    mt3 <- 
                        mt2 %>%
                        dplyr::select(
                            c(ref, qry, identity, ref_clust, qry_clust, type)
                            ) %>%
                        filter(
                            ref %in% clustered_gs$accessions,
                            qry %in% clustered_gs$accessions
                        ) %>%
                        mutate(
                            pair_id = 
                                apply(
                                    dplyr::select(., ref, qry), 
                                    1, 
                                    function(x){
                                        paste(sort(x), collapse = "-")
                                        }
                                    ),
                            rc_qc = 
                                apply(
                                    dplyr::select(., ref_clust, qry_clust), 
                                    1, 
                                    function(x){
                                        paste(sort(x), collapse = "-")
                                        }
                                    )
                            ) %>%
                        group_by(
                            pair_id, rc_qc, type
                        ) %>%
                        summarise(
                            id = 
                                mean(identity),
                            #n=n(),
                            .groups = "keep"
                        ) %>% 
                        ungroup %>%
                        mutate(
                            id_dist = 
                                round(100-id, 2),
                            type2 = 
                                gsub("(.+?)(\\_.*)", "\\1", type)
                        ) 
                    
                    count(mt3,rc_qc)
                    
                    within_clust_diffs <- 
                        list()
                    for (j in 1:nrow(cluster_counts)){
                        
                    }
                    
                    
                    ## bootstrap results
                    bootstrap_diff2 <- function(data1, indices1) {
                          sample_data <- data1[indices1, ]
                          within_mean <- mean(sample_data$id_dist[sample_data$type2 == "within"])
                          between_mean <- mean(sample_data$id_dist[sample_data$type2 == "between"])
                          
                          return(between_mean - within_mean)
                    }
                    bootstrap_within <- function(data1, indices) {
                      sample_data <- data1[indices, ]
                      mean(sample_data$id_dist[sample_data$type == "within"])
                    }
                    
                    bootstrap_between <- function(data1, indices) {
                      sample_data <- data1[indices, ]
                      mean(sample_data$id_dist[sample_data$type == "between"])
                    }
                    
                    #mt3$type2 %>% unique
                    set.seed(123)
    
                    boot_within <- 
                        boot(mt3, bootstrap_within, R = 500)
                    boot_between <- 
                        boot(mt3, bootstrap_between, R = 500)

                ggplot(mt3, aes(x = rc_qc, y = id_dist, fill = type)) +
                  geom_violin(trim = FALSE) +
                  geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9))
                    
                    
                    boot_ci <- 
                        boot.ci(boot_results, type = "perc")
                    
                    boot_results$t %>% head
                    
                    
                    out_tib <- 
                        more_gs[i,c(-2)] %>%
                        mutate(
                            "Between-within" = 
                                mean(boot_results$t, na.rm = T),
                            "95% CI low" = 
                                boot_ci$percent[4],
                            "95% CI high" = 
                                boot_ci$percent[5],
                            clusters = 
                                length(unique(clustered_gs$cluster_id))
                        )
                        
                }
                # boot_diff_between_df <- 
                #     data.frame(boot_diff = boot_results$t)
                # ggplot(boot_diff_between_df, aes(x = boot_diff)) +
                #   geom_histogram(bins = 30, fill = "blue", alpha = 0.7) +
                #   geom_vline(xintercept = 0, linetype = "dashed", color = "red") +  # Reference line at 0
                #   labs(title = "Bootstrapped Differences in Mean Distances (Between - Within)",
                #        x = "Difference in Mean Distances (Between - Within)",
                #        y = "Frequency")
                # 
                # 
                # bootstrap_diff <- 
                #     function(data1, indices1) {
                #         sample_data <- data1[indices1, ]
                #         between_mean <- mean(sample_data$id_dist[sample_data$type == "between"])
                #         w0_mean <- mean(sample_data$id_dist[sample_data$type == "within_0"])
                #         w1_mean <- mean(sample_data$id_dist[sample_data$type == "within_1"])
                #         # Return differences: (Group 1 within - Between), (Group 2 within - Between)
                #         return(c(w0_mean + w1_mean - between_mean))
                #     }
                # 
                # boot_diff_results <- 
                #     boot(mt3, bootstrap_diff, R = 100)
                # boot_diff_results$t
                # 
                # boot_data <- 
                #     as_tibble(boot_diff_results$t) %>%
                #     `colnames<-`(
                #         c("within_group_0", "within_group_1", "between_1_2")
                #         )
                # 
                # boot_long <- 
                #     boot_data %>%
                #     pivot_longer(
                #         cols = everything(), 
                #         names_to = "Type", values_to = "bootstrapped_mean")
                # 
                # 
                # ggplot(boot_long, aes(x = Type, y = bootstrapped_mean, fill = Type)) +
                #     geom_boxplot() +
                #     theme_minimal() +
                #     labs(title = "Boxplot of Bootstrapped Means", x = "Group", y = "Bootstrapped Mean Similarity")
                #                     
                # mt3 %>% ungroup %>% count(type)
                # 
                # ggplot(boot_long, aes(x = Type, y = bootstrapped_mean, fill = Type)) +
                #   geom_violin(trim = FALSE) +
                #   geom_boxplot(width = 0.1, color = "black", position = position_dodge(0.9))
                #     
                    
                
            } else{
                out_tib <- 
                    NULL
            }
            return(out_tib)
        }, 
        mc.cores = 4
    )
















head(dir_files)

more_gs$species2

pl_tib <- list()

for (pl in unique(more_gs$p2)){
    #pl = phyla_results[1]
    pl = more_gs$p2[1]
    #more_gs$p2
    
    genera_dirs <- 
        list.dirs(
            sprintf(
                "%s/%s",
                outer_results_dir, pl
                ), recursive = F
            )
    species_dirs <- 
        lapply(
            genera_dirs, 
            function(x){
                list.dirs(x, full.names = T, recursive = F)
            }
            ) %>% unlist
        
    results_files <- 
        lapply(
            species_dirs, 
            function(x){
                list.files(x, full.names = T)
                }
            )
    pl_tib[[pl]] <-         
        lapply(
            1:length(species_dirs),
            function(i){
                #for (i in 1:length(species_dirs)){
                spec_i <- 
                    basename(species_dirs[i])
                spec_files <- 
                    list.files(species_dirs[i], full.names = T)
                bi_file <- 
                    spec_files[grepl("bis.rds", spec_files)]
                cg_file <- 
                    spec_files[grepl("clustered_genomes", spec_files)]
                    
                if (length(bi_file)==0){
                    spec_out <- 
                        tibble(
                            phylum = 
                                pl,
                            species = 
                                spec_i,
                            all_collinear = 
                                NA,
                            SV_present = 
                                NA,
                            SV_depth = 
                                NA,
                            SV_clusters = 
                                NA,
                            genomes = 
                                NA
                            )
                    if (length(cg_file)!=0){
                        if (grepl("rds", cg_file)){
                            rcsv <- 
                                #fread(cg_file) %>%
                                readRDS(cg_file)$cluster_summary %>%
                                group_by(cluster_id) %>%
                                summarise(
                                    grouped_sequences = 
                                        n()
                                )
                        } else {
                            rcsv <- 
                                fread(cg_file) %>%
                                group_by(cluster_id) %>%
                                summarise(
                                    grouped_sequences = 
                                        n()
                                )
                        }
                        
                        
                        spec_out <- 
                            tibble(
                                phylum = 
                                    pl,
                                species = 
                                    spec_i,
                                all_collinear = 
                                    nrow(rcsv)==1,
                                SV_present = 
                                    nrow(rcsv)>1,
                                SV_depth = 
                                    rcsv$grouped_sequences[2]>5,
                                SV_clusters = 
                                    nrow(rcsv)-1,
                                genomes = 
                                    sum(rcsv$grouped_sequences)
                                )
                                
                    }
                } else {
                    bi_tib <- 
                        readRDS(bi_file)
                    spec_out <- 
                        tibble(
                            phylum = 
                                pl,
                            species = 
                                spec_i,
                            all_collinear = 
                                nrow(bi_tib)==1,
                            SV_present = 
                                nrow(bi_tib)>1,
                            SV_depth = 
                                bi_tib$grouped_sequences[2]>5,
                            SV_clusters = 
                                nrow(bi_tib)-1,
                            genomes = 
                                sum(bi_tib$grouped_sequences)
                            )
                }
                return(spec_out)
            }
        ) %>%
        bind_rows
      print(pl)
    
}


pl_tib <- 
    list()
for (pl in phyla_results){
    #pl = phyla_results[1]
    genera_dirs <- 
        list.dirs(
            sprintf(
                "%s/%s",
                outer_results_dir, pl
                ), recursive = F
            )
    species_dirs <- 
        lapply(
            genera_dirs, 
            function(x){
                list.dirs(x, full.names = T, recursive = F)
            }
            ) %>% unlist
        
    results_files <- 
        lapply(
            species_dirs, 
            function(x){
                list.files(x, full.names = T)
                }
            )
    pl_tib[[pl]] <-         
        lapply(
            1:length(species_dirs),
            function(i){
                spec_i <- 
                    basename(species_dirs[i])
                spec_files <- 
                    list.files(species_dirs[i], full.names = T)
                bi_file <- 
                    spec_files[grepl("bis.rds", spec_files)]
                cg_file <- 
                    spec_files[grepl("clustered_genomes", spec_files)]
                    
                if (length(bi_file)==0){
                    spec_out <- 
                        tibble(
                            phylum = 
                                pl,
                            species = 
                                spec_i,
                            all_collinear = 
                                NA,
                            SV_present = 
                                NA,
                            SV_depth = 
                                NA,
                            SV_clusters = 
                                NA,
                            genomes = 
                                NA
                            )
                    if (length(cg_file)!=0){
                        if (grepl("rds", cg_file)){
                            rcsv <- 
                                #fread(cg_file) %>%
                                readRDS(cg_file)$cluster_summary %>%
                                group_by(cluster_id) %>%
                                summarise(
                                    grouped_sequences = 
                                        n()
                                )
                        } else {
                            rcsv <- 
                                fread(cg_file) %>%
                                group_by(cluster_id) %>%
                                summarise(
                                    grouped_sequences = 
                                        n()
                                )
                        }
                        
                        spec_out <- 
                            tibble(
                                phylum = 
                                    pl,
                                species = 
                                    spec_i,
                                all_collinear = 
                                    nrow(rcsv)==1,
                                SV_present = 
                                    nrow(rcsv)>1,
                                SV_depth = 
                                    rcsv$grouped_sequences[2]>5,
                                SV_clusters = 
                                    nrow(rcsv)-1,
                                genomes = 
                                    sum(rcsv$grouped_sequences)
                                )
                                
                    }
                } else {
                    bi_tib <- 
                        readRDS(bi_file)
                    spec_out <- 
                        tibble(
                            phylum = 
                                pl,
                            species = 
                                spec_i,
                            all_collinear = 
                                nrow(bi_tib)==1,
                            SV_present = 
                                nrow(bi_tib)>1,
                            SV_depth = 
                                bi_tib$grouped_sequences[2]>5,
                            SV_clusters = 
                                nrow(bi_tib)-1,
                            genomes = 
                                sum(bi_tib$grouped_sequences)
                            )
                }
                return(spec_out)
            }
        ) %>%
        bind_rows
      print(pl)
    
}

SV_results <- 
    bind_rows(pl_tib) 

SV_summary <- 
    SV_results %>%
    filter(!is.na(genomes)) %>%
    group_by(phylum) %>%
    summarise(
        species = n(),
        genomes = sum(genomes),
        SV_count = sum(SV_present),
        SV_depth = sum(SV_depth, na.rm = T)
    ) %>%
    arrange(
        desc(genomes),
        desc(species)
    ) %>%
    mutate(
        SV_pathway = 
            c(8, 3, 2, 1, 0,0, 1, 1)
    )

SV_summary %>%
    `colnames<-`(c("Phylum", "Species", "Genomes", "SV count", "SV Depth", "Pathways")) %>%
     kbl() %>%
  kable_classic_2(full_width = F)

ggplot(SV_results, aes(x = interaction(phylum, species), y = Count, fill = LogicalColumn)) +
  geom_bar(stat = "identity", position = "dodge") +  # Use position = "stack" for stacked bars
  labs(x = "Category1 and Category2", y = "Count", fill = "Logical Value") +
  theme_minimal() +
  ggtitle("Counts of TRUE vs FALSE for each Category Combination")
    



```








```{r}

ssnt <- 
        bind_rows(ssn)

#########################


events <- 
    list()

for (j in 1:nrow(bis_1)){
####
    qi <- 
        pull(bis_1, accessions)[j]
    # the delta file filtered
    fdi <- 
        filtered_combined_delta %>% 
        ungroup %>%
        filter(qid==qi) %>%
        mutate(
            rmid = round((rs+re)/2),
            qmid = round((qs+qe)/2)
        ) %>%
        mutate(meanlen = round(meanlen))
        
    #plot_delta(fdi)
    fd_less <- 
        fdi %>%
        dplyr::select(
            qid, strand, X_dist, rs, re, qs, qe, meanlen, rmid, qmid, segment
        ) %>%
        filter(
            segment != min(segment),
            segment != max(segment),
            ) %>%
        mutate(
            inv_type1 = 
                ifelse(
                    X_dist < 1e6,
                    "RASR",
                    "Local"
                    )
            )
    # Checking segments
    # +ve
    fd_less2_p <- 
        fdi %>%
        dplyr::select(
            strand, rmid, qmid, meanlen, rs, re, qs, qe, X_dist
            ) %>%
        mutate(
            mid_diff = 
                abs(rmid-qmid)
        ) %>% filter(strand=="+")
    # positive mid diff 
    pmd <- 
        fd_less2_p %>% pull(mid_diff)
    
    positive_tlocs <- 
        pmd[
            (pmd < (quantile(pmd, 0.25) - 3 * IQR(pmd))) | 
                (pmd > (quantile(pmd, 0.75) + 3 * IQR(pmd)))
            ]
        
    fd_less2_n <- 
        fdi %>%
        dplyr::select(strand, rmid, qmid, meanlen) %>%
        mutate(
            mid_diff = 
                abs(rmid-qmid)
        ) %>% filter(strand=="-")
        
        fd_less_RASR <- 
            fd_less %>%
            group_by(segment) %>%
            summarise(
                qid = unique(qid),
                strand = unique(strand),
                meanlen = round(mean(meanlen)),
                refmid = round(mean(rmid)),
                qrymid = round(mean(qmid)),
                X_dist = round(mean(X_dist)),
                inv_type1 = unique(inv_type1)
                )
        
        segdiff <- 
            fd_less_RASR %>%
            mutate(segdiff = segment-lag(segment, default = 1)) %>%
            pull(segdiff)
        
        if (any(segdiff>1)) stop("seg missing")#next
        
        fd_less_RASR_check <- 
            fd_less %>%
            filter(
                strand=="-" |
                    (
                        (lag(strand, default = "+")=="-") & 
                            (lead(strand, default = "+")=="-")
                    )
            )
        
        inner_seg <- 
            median(unique(fd_less_RASR$segment))
        
        if (nrow(fd_less_RASR)==1){
            middle_seg_row <- 
                fd_less_RASR %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    segment = as.character(segment),
                    X_dist = abs(refmid-qrymid)
                    ) 
            events[[j]] <- 
                middle_seg_row
        } else {
            middle_seg_row <- 
                fd_less_RASR %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    segment = as.character(segment),
                    X_dist = abs(refmid-qrymid)
                    ) 
            
            outer_segs <- 
                fd_less_RASR %>%
                filter(segment != median(segment)) 
                
            possible_outers <- 
                nrow(outer_segs) %/% 2
            
            outer_events <- 
                list()
            for (k in 1:possible_outers){
                outer_events[[k]] <- 
                    fd_less %>% 
                    filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                    summarise(
                        qid = unique(qid),
                        strand= unique(strand),
                        segment = paste(segment, collapse = ","),
                        #X_dist = round(mean(X_dist)),
                        rs = min(rs), 
                        re = max(re),
                        qs = 
                             ifelse(
                                strand=="+",
                                min(qs),
                                min(qe)
                            ),
                        qe =
                            ifelse(
                                strand=="+",
                                max(qs),
                                max(qe)
                            ),
                        meanlen = round(sum(meanlen)),
                        refmid = 
                            round((rs+re)/2),
                        qrymid = 
                            round((qs+qe)/2),
                        inv_type1 = unique(inv_type1),
                        X_dist = abs(refmid-qrymid)
                    ) %>%
                    dplyr::select(
                        c(
                            qid, strand, meanlen, refmid, 
                            qrymid,X_dist, inv_type1, segment
                        )
                    )
            }
            inv_summary <- 
                bind_rows(
                    middle_seg_row,
                    bind_rows(outer_events)
                )
            events[[j]] <- 
                inv_summary
        }
    #plot_delta(fdi)
}



event_summary <- 
    bind_rows(events) %>%
    left_join(
        .,
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("clust", "qid")),
        by = "qid"
    )

bi_f <- 
    lapply(
        list.dirs("../data/processing/results"),
        function(x){
            (list.files(x, full.names = T, pattern = "bis.rds"))
        }
    ) %>% unlist()

lapply(
    bi_f, 
    function(x){
        readRDS(x) %>%
            mutate(
                spec = 
                    basename(dirname(x))
            )
    }
    
    )

basename(dirname(bi_f[1]))
list.dirs(sprintf("%s", results_dir))
species_list_40

bi_tibs <- 
    lapply(
    bi_f, 
    function(x){
        readRDS(x) %>%
            mutate(
                spec = 
                    basename(dirname(x))
            )
    }
    )


"emmtyper -w pcr ../data/processing/extra/s_pyogenes3*.txt --output-format visual"

SV_ids <- 
     c(
                "All colinear",
                "≥1 strucutral variant group",
                "> 1 type of structural variation",
                "≥ 5 sequences in at least 1 noncolinear group",
                "≥ 5 sequences in at least 2 noncolinear groups",
                "≥ 1 reoccurring structural variant",
                "≥ 1 nested structural variants"#,
                #"RASRs",
                #"Local_invs"
            )
    


species_SV_summary <- 
    tibble(
        species = 
            species_name,
        sequences = 
            nrow(sync_tib),
        SV_id = 
            c(
                "All colinear",
                "≥1 strucutral variant group",
                "> 1 type of structural variation",
                "≥ 5 sequences in at least 1 noncolinear group",
                "≥ 5 sequences in at least 2 noncolinear groups",
                "≥ 1 reoccurring structural variant",
                "≥ 1 nested structural variants",
                "RASRs",
                "Local_invs"
            ),
        SV_present = 
            c(
                nrow(bis)==1,
                nrow(bis)>=1,
                nrow(bis)>=2,
                any(bis_1$grouped_sequences>=5),
                sum(bis_1$grouped_sequences>=5)>=2,
                #(length(unique(event_summary$inv_type1)))>1,
                # nrow(filter(bis_1,grouped_sequences>=5))>=1,
                # nrow(filter(bis_1,grouped_sequences>=5))>1,
                0,
                0,
                sum(event_summary$inv_type1=="RASR"),
                sum(event_summary$inv_type1=="Local")
            )
    )

list.dirs("")


sss2 <- 
    bind_rows(
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Actinomycetota",
            sv_sums = 
                c(
                    3,7,6,4,2,5,4
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Bacillota",
            sv_sums = 
                c(
                    8,7,6,4,2,5,4
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Bacteroidota",
            sv_sums = 
                c(
                    5,5,4,3,2,1,1
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Campylobacterota",
            sv_sums = 
                c(
                    0,1,3,3,1,2,1
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Fusobacteriota",
            sv_sums = 
                c(
                    2,1,1,0,0,0, 0
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Mycoplasmatota",
            sv_sums = 
                c(
                    3,7,6,2,1,3,2
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Pseudomonadota",
            sv_sums = 
                c(
                    8,12,6,4,2,5,4
                )
        ),
        tibble(
            SV_id = 
                SV_ids, 
            Phylum = 
                "Spirochaetota",
            sv_sums = 
                c(
                    3,1,1,0,1,1,1
                )
        )
    )
    
ggplot(sss2, aes(x = SV_id, y = sv_sums, fill = SV_id)) +
  geom_bar(stat = "identity") +  # stat = "identity" is used because we're plotting precomputed values
  #facet_wrap(~Phylum) +  # Facet based on the 'group' variable
  labs(title = "", x = "", y = "Count") +
  theme_minimal() +
    scale_fill_brewer(palette = "Dark2")


sss3 <- 
    sss2 %>%
    mutate(
        SV_id = factor(SV_id, levels = SV_ids[-c(8,9)])
    )
SV_ids[1]
sss4 <- 
    sss3 %>%
    mutate(
        sv2 = 
            case_when(
                SV_id %in% c(SV_ids[1]) ~ "Collinear",
                SV_id %in% c(SV_ids[c(2,3)]) ~ "Multiple SV clusters",
                SV_id %in% c(SV_ids[c(4,5)]) ~ "≥ 5 genomes in SV cluster(s)",
                SV_id %in% c(SV_ids[c(6,7)]) ~ "Branching SV pathway",
                )
        ) %>%
    group_by(sv2) %>%
    summarise(
        sv_sum = 
            sum(sv_sums)
        ) %>%
    mutate(
        sv2 = 
            factor(
                sv2, 
                levels = 
                    c(
                        "Collinear","Multiple SV clusters",
                        "≥ 5 genomes in SV cluster(s)",
                        "Branching SV pathway"
                        )
            )
    )

ggplot(sss3, aes(x = SV_id, y = sv_sums, fill = SV_id)) +
  geom_bar(stat = "identity") +  # stat = "identity" is used because we're plotting precomputed values
  #facet_wrap(~Phylum) +  # Facet based on the 'group' variable
  labs(title = "", x = "", y = "Count") +
  theme_classic() +
    scale_fill_brewer(palette = "Set1") +
      theme(
          legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size = 22)
          )
ggplot(sss4, aes(x = sv2, y = sv_sum, fill = sv2)) +
  geom_bar(stat = "identity", colour ="black") +  # stat = "identity" is used because we're plotting precomputed values
  #facet_wrap(~Phylum) +  # Facet based on the 'group' variable
  labs(title = "", x = "", y = "Count") +
  theme_classic() +
    scale_fill_brewer(palette = "Set1") +
      theme(
          legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size = 22)
          )

species_names <- c("e_coli", "s_enterica", "b_subtilis", "l_monocytogenes", 
                   "h_influenzae", "p_aeruginosa", "k_pneumoniae", 
                   "v_cholerae", "m_tuberculosis", "s_aureus")

d_1 <- 
    tibble(
  species = rep(species_names, each = 20),  # 10 species, each repeated 20 times
  group = rep(c("group1", "group2"), times = 100),  # 10 species, 2 groups each
  value = c(
    rnorm(100, mean = 10, sd = 0.02),  # Group 1 with very low standard deviation (sd = 0.1)
    rnorm(100, mean = rep(seq(30, 35, length.out = 10), each = 10), sd = 0.1)  # Group 2 with even larger mean differences and low sd
  )
)

# Perform t-tests for each species
t_test_results <- d_1 %>%
  group_by(species) %>%
  summarise(
    t_test = list(t.test(value ~ group)),  # Perform t-test
    p_value = t_test[[1]]$p.value,  # Extract p-value
    mean_diff = t_test[[1]]$estimate[1] - t_test[[1]]$estimate[2],  # Mean difference
    conf_low = t_test[[1]]$conf.int[1],  # Confidence interval lower bound
    conf_high = t_test[[1]]$conf.int[2]  # Confidence interval upper bound
  )

ggplot(t_test_results, aes(x = species, y = mean_diff)) +
  geom_point() +  # Plot the mean difference
  geom_errorbar(aes(ymin = conf_low, ymax = conf_high), width = 0.2) +  # Add confidence intervals
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +  # Line at zero difference
  labs(title = "Mean Difference with Confidence Intervals for Each Species",
       y = "Mean Difference (Group1 - Group2)",
       x = "Species") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),  text = element_text(size = 22)) +
  coord_flip()

tt2 <- 
    tt1 %>%
    filter(!(SV_id %in% c(SV_ids[8], SV_ids[9]))) %>%
    mutate(
        SV_id = factor(SV_id, levels = SV_ids[-c(8,9)]),
        species =  sub("^(.).*_(.*)$", "\\1_\\2", species)
    )


ggplot(tt2, aes(x = species, y = SV_id, fill = factor(SV_present))) +
  geom_tile(color = "black", linewidth = 0.5) +  # Creates the heatmap tiles
  scale_fill_manual(values = c("0" = "lightgrey", "1" = "darkblue"), name = "Condition Met") +
  labs(x = "", y = "") +
  theme_minimal() +  # Use a minimal theme for a clean look
  theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size = 22)) 





bi_tibs[[1]]
bi_f[2]

tt1 <- 
    bind_rows(
        species_SV_summary,
        tibble(
            species = 
                bi_tibs[[1]]$spec[1],
            sequences = 
                sum(bi_tibs[[1]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[1]])==1,
                    nrow(bi_tibs[[1]])>1,
                    nrow(bi_tibs[[1]])>=2,
                    any(bi_tibs[[1]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[1]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    0,
                    1
                )
        ),
        tibble(
            species = 
                bi_tibs[[2]]$spec[1],
            sequences = 
                sum(bi_tibs[[2]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[2]])==1,
                    nrow(bi_tibs[[2]])>1,
                    nrow(bi_tibs[[2]])>=2,
                    any(bi_tibs[[2]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[2]][-1,]$grouped_sequences>=5)>=2),
                    0,#"≥ 1 reoccurring structural variant",
                    0,#"≥ 1 nested structural variants",
                    2, #"RASRs",
                    0 #"Local_invs"
                )
        ),
        tibble(
            species = 
                bi_tibs[[3]]$spec[1],
            sequences = 
                sum(bi_tibs[[3]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[3]])==1,
                    nrow(bi_tibs[[3]])>1,
                    nrow(bi_tibs[[3]])>=2,
                    any(bi_tibs[[3]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[3]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    1,
                    1
                )
        ),
        tibble(
            species = 
                bi_tibs[[4]]$spec[1],
            sequences = 
                sum(bi_tibs[[4]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[4]])==1,
                    nrow(bi_tibs[[4]])>1,
                    nrow(bi_tibs[[4]])>=2,
                    any(bi_tibs[[4]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[1]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    0,
                    0
                )
        ),
        tibble(
            species = 
                bi_tibs[[5]]$spec[1],
            sequences = 
                sum(bi_tibs[[5]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[5]])==1,
                    nrow(bi_tibs[[5]])>=1,
                    nrow(bi_tibs[[5]])>=2,
                    any(bi_tibs[[5]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[5]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    1,
                    1
                )
        ),
        tibble(
            species = 
                bi_tibs[[6]]$spec[1],
            sequences = 
                sum(bi_tibs[[6]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[6]])==1,
                    nrow(bi_tibs[[6]])>=1,
                    nrow(bi_tibs[[6]])>=2,
                    any(bi_tibs[[6]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[6]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    2,
                    0
                )
        ),
        tibble(
            species = 
                bi_tibs[[7]]$spec[1],
            sequences = 
                sum(bi_tibs[[7]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[7]])==1,
                    nrow(bi_tibs[[7]])>=1,
                    nrow(bi_tibs[[7]])>=2,
                    any(bi_tibs[[7]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[7]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    5,
                    0
                )
        ),
        tibble(
            species = 
                bi_tibs[[8]]$spec[1],
            sequences = 
                sum(bi_tibs[[8]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[8]])==1,
                    nrow(bi_tibs[[8]])>=1,
                    nrow(bi_tibs[[8]])>=2,
                    any(bi_tibs[[8]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[8]][-1,]$grouped_sequences>=5)>=2),
                    1,
                    0,
                    3,
                    0
                )
        ),
        tibble(
            species = 
                bi_tibs[[9]]$spec[1],
            sequences = 
                sum(bi_tibs[[9]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[9]])==1,
                    nrow(bi_tibs[[9]])>=1,
                    nrow(bi_tibs[[9]])>=2,
                    any(bi_tibs[[9]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[9]][-1,]$grouped_sequences>=5)>=2),
                    0,
                    0,
                    3,
                    0
                )
        ),
        tibble(
            species = 
                bi_tibs[[11]]$spec[1],
            sequences = 
                sum(bi_tibs[[11]]$grouped_sequences),
            SV_id = 
                SV_ids,
            SV_present = 
                c(
                    nrow(bi_tibs[[11]])==1,
                    nrow(bi_tibs[[11]])>=1,
                    nrow(bi_tibs[[11]])>=2,
                    any(bi_tibs[[11]][-1,]$grouped_sequences>=5),
                    as.numeric(sum(bi_tibs[[11]][-1,]$grouped_sequences>=5)>=2),
                    1,
                    1,
                    3,
                    0
                )
        )
    )
    

tt2 <- 
    tt1 %>%
    filter(!(SV_id %in% c(SV_ids[8], SV_ids[9]))) %>%
    mutate(
        SV_id = factor(SV_id, levels = SV_ids[-c(8,9)]),
        species =  sub("^(.).*_(.*)$", "\\1_\\2", species)
    )


ggplot(tt2, aes(x = species, y = SV_id, fill = factor(SV_present))) +
  geom_tile(color = "black", linewidth = 0.5) +  # Creates the heatmap tiles
  scale_fill_manual(values = c("0" = "lightgrey", "1" = "darkblue"), name = "Condition Met") +
  labs(x = "", y = "") +
  theme_minimal() +  # Use a minimal theme for a clean look
  theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size = 22)) 



SSV2 <- 
    bind_roes(
        species_SV_summary,
        
        tibble(
            bi_tibs[[1]]
        )
        
    )


sp_sum[[species_name]] <- 
    species_SV_summary


print(species_name)
```

## synchronization
```{r}

## synchronize the first 5 sequences

#species_table <- species_summary

ftps <- 
    species_summary$ftp_path

species_summary %>% pull(accession)

fasta_assemblies <- 
    pbmclapply(
        1:5,#length(ftps),
        function(i){
            assembly_path <- 
                species_summary$ftp_path[i]
            
            fasta_path <- 
                sprintf(
                    "%s/%s_genomic.fna.gz",
                    assembly_path, 
                    gsub(".*/", "", assembly_path)
                )
            
            genome_seq <- 
                readDNAStringSet(fasta_path)
            genome_seq <- 
                genome_seq[
                    !(grepl("plasmid", names(genome_seq),  
                            ignore.case = T))
                ]
            genome_seq <- 
                genome_seq[
                    !(grepl("phage", names(genome_seq),  
                            ignore.case = T)) |
                        grepl(
                            "phage resistant", 
                            names(genome_seq),  
                            ignore.case = T
                        )
                ]
            # take the bigger sequence ie the main one
            genome_seq <- 
                genome_seq[order(width(genome_seq), decreasing = T),]
        }
    )




(species_summary %>% pull(accession))[1]

sync_tibs <- 
    list()
GC_notes <- 
    list()

seq_starts <- 
    list()
accessions <- 
    list()

sync_tibs <- list()
which(species_summary$accession=="CP030403.1")

sync_tib <- 
    pbmclapply(
        1:nrow(species_summary),
        function(i){
            assembly_path <- 
                species_summary$ftp_path[i]
            sync_summary <- 
                sync_gene(
                    assembly_path = 
                        assembly_path,
                    sync_dir = 
                        sync_dir, 
                    gff_dir = 
                        gff_dir
                    ) %>%
                mutate(
                    idx = i
                )
            return(sync_summary)
        }, mc.cores = 4
    ) %>% bind_rows()

which(sync_tib$accession=="CP030403.1")

sync_tib$accession


    
sync_dir


## unsynced

unsynced_idx <- 
    sync_tib %>%
    filter(is.na(ori_pos))

get_fasta <- 
    function(asm_path, chromosome = 1){
        fasta_path <- 
            sprintf(
                "%s/%s_genomic.fna.gz",
                assembly_path, 
                gsub(".*/", "", assembly_path)
            )
        
        genome_seqs <- 
            readDNAStringSet(fasta_path)
        genome_seqs <- 
            genome_seqs[
                !(grepl("plasmid", names(genome_seqs),  
                        ignore.case = T))
            ]
        genome_seqs <- 
            genome_seqs[
                !(grepl("phage", names(genome_seqs),  
                        ignore.case = T)) |
                    grepl(
                        "phage resistant", 
                        names(genome_seqs),  
                        ignore.case = T
                    )
            ]
        # Which wasn't synced
        genome_seq <- 
            genome_seqs[order(width(genome_seqs), decreasing = T),][
                chromosome
            ]
        genome_acc <- 
            strsplit(names(genome_seq), " ")[[1]][1]
        
        names(genome_seq) <- genome_acc
            
        return(genome_seq)
    }

sync_tib %>% filter(accession == "CP099938.1")
which(unsynced_idx$accession=="CP099938.1")

rs_tibs <- list()
which(sync_tib$accession=="NZ_CP119302.1")
sync_tib %>% filter(accession=="NZ_CP119302.1")

for (m in 1:length(unsynced_idx$idx)){
    #species_summary
    
    assembly_path <- 
        species_summary$ftp_path[unsynced_idx$idx[m]]
    
    us_chrom <-
        sync_tib$chr[unsynced_idx$idx[m]]
    # 
    # #rs_sync
    # chrom_dir <- 
    #     sprintf(
    #         "%s/chr_%s",
    #         sync_dir, us_chrom
    #     )
    
    unsynced_seq <- 
        get_fasta(assembly_path, chromosome = us_chrom)
    fasta_seq <- 
        sprintf(
            "%s/%s.txt",
            sync_dir,
            names(unsynced_seq)
        )
    writeXStringSet(
        unsynced_seq,
        fasta_seq
    )
    
    ref_seq <- 
        list.files(
            sprintf(
                "%s/chr_%s_gene",
                sync_dir, us_chrom
            ),
            full.names = T
        )[1]
    # readDNAStringSet(
    #     ref_seq
    # )
    
    
    startpoint_dir <- 
        sprintf("%s/starts", sync_dir)
    if (!dir.exists(startpoint_dir)){dir.create(startpoint_dir)}
    
    rs_tibs[[m]] <- 
        rs_sync(
            ref_seq, fasta_seq, 
            delta_out = 
                sprintf("%s/%s", startpoint_dir,  names(unsynced_seq)
                        ),
            out_file = 
                fasta_seq
            #     sprintf(
            #     "%s/%s.txt",
            #     chrom_dir,
            #     names(unsynced_seq)
            # )
        )
    print(i)
  
}

rs_tibs %>% bind_rows

lapply(
    1:length(list.files(chrom_dir)),
    function(i){
        tibble(
            bn = 
                sub(".txt", "", basename(list.files(chrom_dir)[i])),
            cn = 
                list.files(chrom_dir, full.names = T)[i] %>%
                read.table(file = ., header = F, nrows = 1) %>%
                sub(">", "", .),
            bc = bn==cn
        )
        
    }
) %>% bind_rows() %>% pull(bc) %>% all()
read.table(file = 'file.txt',header = F,nrows = 1)

unsync_tib <- 
    sync_tib %>% filter(is.na(ori_pos))




species_summary


for (m in 1:nrow(unsync_tib)){
    asm <- 
        pull(unsync_tib, asm_name)[m]
    
     assembly_path <- 
                species_summary$ftp_path[m]
            
    fasta_path <- 
        sprintf(
            "%s/%s_genomic.fna.gz",
            assembly_path, 
            gsub(".*/", "", assembly_path)
        )
    
    genome_seq <- 
        readDNAStringSet(fasta_path)
    genome_seq <- 
        genome_seq[
            !(grepl("plasmid", names(genome_seq),  
                    ignore.case = T))
        ]
    genome_seq <- 
        genome_seq[
            !(grepl("phage", names(genome_seq),  
                    ignore.case = T)) |
                grepl(
                    "phage resistant", 
                    names(genome_seq),  
                    ignore.case = T
                )
        ]
    # take the bigger sequence ie the main one
    genome_seq <- 
        genome_seq[order(width(genome_seq), decreasing = T),]
    
}




most_related_start <- 
    readDNAStringSet(
        sprintf("%s/%s.txt", sync_chrom_dir, most_related)
    ) %>% subseq(1,1200)

ref_seq <- 
    sprintf("%s/%s_rs.txt", sync2_dir, most_related)

writeXStringSet(
    most_related_start %>% `names<-`(sprintf("%s_rstart", genome_acc)),
    ref_seq
)

```





```{r}

for (i in 1:10){
    ### Synchronize with GFF
    assembly_path <- 
        species_summary$ftp_path[i]
    sync_tibs[[i]] <- 
        sync_gene(
            assembly_path = 
                assembly_path,
            sync_dir = 
                sync_dir, 
            gff_dir = 
                gff_dir
            )
        if (i %%5==0){closeAllConnections()}
    
    # sync_tibs[[i]] <- 
    #     sync3(
    #         assembly_path = 
    #             assembly_path,
    #         sync_dir = 
    #             sync_dir, 
    #         gff_dir = 
    #             gff_dir
    #         )
    # acc_i <- 
    #     sync_tibs[[i]] %>% pull(accession)
    # 
    # seq_starts[[i]] <- 
    #     readDNAStringSet(
    #         sprintf("%s/%s.txt", sync_dir, acc_i)
    #     ) %>% 
    #     subseq(1,500)
    # accessions[[i]] <- acc_i
    
    ### Verify with GC skew
    # GC_notes[[i]] <- 
    #     GC_skew_sync(
    #         fasta_file = 
    #             sprintf(
    #                 "%s/%s.txt",
    #                 sync_dir, 
    #                 acc_i
    #                 ), 
    #         disparity_file = 
    #             sprintf(
    #                 "%s/%s.txt",
    #                 disparities_dir, 
    #                 acc_i
    #                 )
    #             )
    # 
    
}

sync_summary <- 
    bind_rows(sync_tibs)

GC_summary <- 
    bind_rows(GC_notes)

## checking that the starts are similar

seq_starts <- 
    do.call(
        c,
        seq_starts
    )

ref_start <- 
    readDNAStringSet(
        sprintf("%s/%s.txt", sync2_dir, most_related)
    ) %>% subseq(1,1200)

ref_seq <- 
    sprintf("%s/%s_rs.txt", sync2_dir, names(seq_starts)[1])

writeXStringSet(
    seq_starts[1] %>% `names<-`(sprintf("%s_start", names(seq_starts)[1])),
    ref_seq
)
sync2_dir2 <- 
    sprintf(
        "%s/refsynced",
        sync2_dir
    )

## synchronize other sequences with the start seq

sync_chr2 <- 
    sprintf(
        "%s/chr2", 
        sync_dir
    )

dir.create(sync_chr2, recursive = T)
for (i in 1:length(ftps)){
    
    asm_path <- 
        species_summary$ftp_path[i]
    
    fasta_file <- 
        sprintf(
            "%s/%s_genomic.fna.gz",
            asm_path, 
            gsub(".*/", "", asm_path)
        )
    asm_name <- 
        sub( ".fna.gz", "", basename(fasta_file))
    
    fasta_seq <- 
        sprintf("%s/%s.txt.gz", sync2_dir, asm_name)
    
    download_attempt <- 
        try(
            writeXStringSet(
                try(readDNAStringSet(fasta_file), silent = T),
                fasta_seq
            ), silent = T
        )
    
    if (!is.null(download_attempt)){
        out_tib <- 
                tibble(
                    acc = genome_acc,
                    delta = "Fasta file unavailable",
                    delta_file = delt
                )
        out_tib <- 
            sprintf(
                "%s fasta file unavailable",
                genome_acc
            )
    } else{
        ## synchronize it with the beginning 
    ### of the most_related sequence
        fna_seq <-
            readDNAStringSet(fasta_seq) 
        fna_seq <- 
            fna_seq[order(width(fna_seq), decreasing = T),][2]
        
        genome_acc <- 
            strsplit(names(fna_seq), " ")[[1]][1]
        
        accessions[[i]] <- genome_acc
        
        names(fna_seq) <- 
            genome_acc
        wp <- 
            file.remove(fasta_seq)
        rm(wp)
        
        fasta_seq <- 
            sprintf(
                "%s/%s.txt",
                sync_chr2, genome_acc
            )
            
        writeXStringSet(
            fna_seq, 
            fasta_seq
        )
        
        # delta_out <- 
        #     sprintf(
        #         "%s/%s_startpoint",
        #         sync_chr2, genome_acc
        #     )
        # #genome_acc
        # 
        # out_file <- 
        #     sprintf(
        #         "%s/%s.txt",
        #         sync_chr2, genome_acc
        #     )
        # rs_result <- 
        #     rs_sync(
        #         ref_seq,
        #         fasta_seq,
        #         delta_out,
        #         out_file
        #     )
        # wp <- file.remove(fasta_seq)
        # rm(wp)
        # fasta_seq <- 
        #     sprintf(
        #         "%s/%s.txt",
        #         sync_dir, genome_acc
        #     )
        # writeXStringSet(
        #     fna_seq, 
        #     fasta_seq
        # )
        # 
        
        if (is.null(rs_result)){
            GC_notes[[i]] <- 
                GC_skew_sync(
                    fasta_file =
                        fasta_seq,
                    disparity_file =
                        sprintf(
                            "%s/%s_chr2.txt",
                            disparities_dir,
                            genome_acc
                            )
                        )
            
        }
        
        #readDNAStringSet(out_file)
        #readDNAStringSet(out_file)
        ## remove the original sequence
       
    }

    
}


list.files(disparities_dir)
i



```




##


### starting with a subset
```{r}


ss_accessions <- 
    species_summary$accession
    #unlist(accessions)
    

print(species_name)

delta_dir_200 <- 
    sprintf("%s/200", delta_dir)
if (!dir.exists(delta_dir_200)){dir.create(delta_dir_200)}

subset_sync_dir <-
    sprintf("%s/subset2", sync_dir)

if (!dir.exists(subset_sync_dir)){dir.create(subset_sync_dir)}


# accession_list <-
#     list.files(subset_sync_dir) %>%
#     sub(".txt", "", .)

identity_200 <- 
    sprintf("%s/200", identity_dir)
if (!dir.exists(identity_200)){dir.create(identity_200)}

st_subsampled <-
    species_summary %>%
    filter(accession %in% ss_accessions)

## ###initial synchronization
sync_report <- 
    synchronize(
        st_subsampled, 
        subset_sync_dir,
        gff_dir
        )
unsynced <- 
    which(
        !(
            ss_accessions %in% 
            sub( ".txt", "", list.files(subset_sync_dir))
        )
        )
"NZ_CP062406.1" %in%ss_accessions

GC_notes <- 
    list()
for (i in unsynced){
    asm_path <- 
        (species_table %>% pull(ftp_path))[unsynced[1]]
    
    acc_i <- 
        ss_accessions[i]
    
    genome_file <- 
        sprintf(
            "%s/%s_genomic.fna.gz",
            asm_path, 
            gsub(".*/", "", asm_path)
        )
    genome_seq <- 
        readDNAStringSet(genome_file)
    genome_seq <- 
        genome_seq[
            !(grepl("plasmid", names(genome_seq),  
                    ignore.case = T))
            ]
    genome_seq <- 
        genome_seq[
            !(grepl("phage", names(genome_seq),  
                    ignore.case = T)) |
                grepl(
                    "phage resistant", 
                    names(genome_seq),  
                    ignore.case = T
                )
            ]
    # take the bigger sequence ie the main one
    genome_seq <- 
        genome_seq[
            width(genome_seq)==max(width(genome_seq))
            ]
    writeXStringSet(
        genome_seq, 
        sprintf(
            "%s/%s.txt",
            subset_sync_dir, 
            acc_i
            )
    )
    readDNAStringSet(
        sprintf(
            "%s/%s.txt",
            subset_sync_dir, 
            acc_i
            )
    )
    
    
    GC_notes[[i]] <- 
        GC_skew_sync(
        fasta_file = 
            sprintf(
                "%s/%s.txt",
                subset_sync_dir, 
                acc_i
                ), 
        disparity_file = 
            sprintf(
                "%s/%s.txt",
                disparities_dir, 
                acc_i
                )
            )
    
}

pbmclapply(
    1:length(ss_accessions),
    function(i){
        nuc_disparities <- 
            sprintf(
                "python3 nucleotide_disparities.py %s/%s.txt %s/%s.csv",
                subset_sync_dir, 
                ss_accessions[i],
                disparities_dir, 
                ss_accessions[i]
            )
        system(nuc_disparities)
        print(i)
    }
)

sampled_disparity_tables <- 
    lapply(
        1:length(ss_accessions),
        function(i){
            nt_tib <- 
                fread(
                    sprintf(
                        "%s/%s.csv",
                        disparities_dir, 
                        ss_accessions[i]
                    )
                ) %>%
                 filter(row_number() %% 10 == 0) %>%
                gather(value = "value") %>%
                group_by(key) %>% 
                mutate(position = row_number()) %>% 
                mutate(
                    accession = ss_accessions[i],
                    accession_key = 
                        paste(accession, key, sep = "_")
                    )
        }
    ) %>% 
    bind_rows()

sampled_disparity_tables %>% filter(accession =="CP130783.1")

head(sampled_disparity_tables)

sampled_disparity_tables %>% pull(accession) 
    
sampled_disparity_tables %>% pull(key) %>% unique
ggplot(
    data=sampled_disparity_tables %>% 
        filter(key=="G-C", accession %in% ss_accessions[unsynced]),
    aes(x = position, y=value,colour=accession_key)
    ) +
  geom_line()+ theme(legend.position = "none")


list.files(subset_sync_dir, full.names = T) %>% lapply(., file.size) %>% unlist


## GC table
```



```{r}
data$range_group <- cut(data$value, 
                        breaks = c(40, 100, 400, Inf), 
                        labels = c("40-100", "100-400", "over 400"),
                        right = FALSE)


data_test <- data.frame(
  species = c('Species1', 'Species1', 'Species2', 'Species2', 'Species3', 'Species3'),
  condition = c('Condition1', 'Condition2', 'Condition1', 'Condition2', 'Condition1', 'Condition2'),
  value = c(1, 0, 1, 1, 0, 0)
)

# Create heatmap
ggplot(data_test, aes(x = species, y = condition, fill = factor(value))) +
  geom_tile() +  # Creates the heatmap tiles
  scale_fill_manual(values = c("0" = "red", "1" = "green"), name = "Condition Met") +
  labs(x = "", y = "Condition", title = "Condition Satisfaction by Species") +
  theme_minimal() +  # Use a minimal theme for a clean look
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 


```

```{r}

sp_sum <- 
    list()

sp_sum[[species_name]] <- 
    species_SV_summary
    
```





#############


########################
```{r}

chr_1_dir <- 
    sprintf("%s/chr_1", sync_dir)
chr_1_id_dir <- 
    sprintf("%s/chr_1", identity_dir)
if(!dir.exists(chr_1_id_dir)){dir.create(chr_1_id_dir)}

chr_2_dir <- 
    sprintf("%s/chr_2", sync_dir)
chr_2_id_dir <- 
    sprintf("%s/chr_2", identity_dir)
if(!dir.exists(chr_2_id_dir)){dir.create(chr_2_id_dir)}

sync_tib

list.files(chr_1_dir)
list.files(chr_2_dir)

species_identities_ss <- 
    get_pairwise_identities(
        sync_dir = chr_1_dir, 
        identity_dir = chr_1_id_dir, 
        species_table = sync_tib, 
        genus_name, species_name
        )

species_identities_chr2 <- 
    get_pairwise_identities(
        sync_dir = chr_2_dir, 
        identity_dir = chr_2_id_dir, 
        species_table = sync_tib, 
        genus_name, species_name
        )

identity_dir2 <- 
    sprintf("%s/chr2", identity_dir)

dir.create(identity_dir2)
species_identities_ss2 <- 
    get_pairwise_identities(
        sync_chr2, 
        identity_dir2, 
        species_summary, 
        genus_name, species_name
        )


"../data/processing/sync/Bacillota/Staphylococcus/Staphylococcus_aureus/subset2/NZ_CP062406"

species_identities_ss$mash_plot +
    scale_fill_viridis(
        option = "B", limits = c(99.5,100)
            ) 

species_identities_ss$mash_plot
species_identities_chr2$mash_plot
species_identities_ss$id_dists

boxplot(species_identities_ss$id_dists$Sequence_mean_distance)
    
identity_dist
species_identities_ss$mash_plot
species_identities_ss$mash_tib


sampled_id_table$Sequence_mean_distance %>% unname
sampled_id_table_chr1 <- 
    species_identities_ss$id_dists %>%
    arrange(desc(Sequence_mean_distance)) %>%
    filter(Sequence_mean_distance<5) 

most_related <- 
    sampled_id_table_chr1 %>%
    arrange(Sequence_mean_distance) %>%
    dplyr::slice(1) %>% 
    pull(Sequence)


sampled_id_table_chr2 <- 
    species_identities_chr2$id_dists %>%
    arrange(desc(Sequence_mean_distance)) %>%
    filter(Sequence_mean_distance<5) 

most_related_chr2 <- 
    sampled_id_table_chr2 %>%
    arrange(Sequence_mean_distance) %>%
    dplyr::slice(1) %>% 
    pull(Sequence)
# accession_list <-
#     list.files(sync_dir) %>%
#     sub(".txt", "", .)

## align the most related accession versus all others

list.files(sync_dir)

delta_chr1 <- 
    sprintf(
        "%s/chr_1", delta_dir
    )
if(!dir.exists(delta_chr1)){(dir.create(delta_chr1))}
delta_chr2 <- 
    sprintf(
        "%s/chr_2", delta_dir
    )
if(!dir.exists(delta_chr2)){(dir.create(delta_chr2))}


delta_dir2 <- 
    sprintf("%s/chr2", delta_dir)

ss_accessions <- 
    sub(".txt", "", list.files(chr_1_dir))

nucmer_commands <- 
    nucmer_ref_v_all(
        most_related, 
        ss_accessions,
        delta_dir =  delta_chr1, 
        sync_dir = chr_1_dir
        )
chr2_accessions <- 
    sub(".txt", "", list.files(chr_2_dir))

nc2 <- 
    nucmer_ref_v_all(
        most_related_chr2, 
        chr2_accessions,
        delta_dir =  delta_chr2, 
        sync_dir = chr_2_dir
        )

list.files(subset_sync_dir)

accession_list

#sub(".txt", "", list.files(subset_sync_dir))

minimum_alignment_percentage <- 
    90
minimum_inversion_length <- 
    5e4
filtered_combined_delta_chr1 <- 
    filter_combined_delta(
        sprintf("%s/ref_v_all", delta_chr1), 
        minimum_alignment_percentage, 
        minimum_inversion_length
    )
list.files(sprintf("%s/ref_v_all", delta_chr2))
filtered_combined_delta_chr2 <- 
    filter_combined_delta(
        sprintf("%s/ref_v_all", delta_chr2), 
        minimum_alignment_percentage, 
        minimum_inversion_length
    )
filtered_combined_delta_chr1 %>% filter(qid=="CP099926.1") %>% plot_delta
filtered_combined_delta_chr2 %>% filter(qid=="NZ_CP119302.1") %>% plot_delta

## missassemblies
filtered_combined_delta_ss %>%
    group_by(qid) %>%
    summarise(
        segments = n(),
        mean_X_dist = mean(X_dist)
    ) %>% arrange(desc(mean_X_dist))

filtered_combined_delta_ss %>%
    filter(
        rs !=1, 
        qs !=1
    ) %>%
    group_by(qid) %>%
    summarise(
        ss1 = n(),
        ss2 = unique(segments),
        mean_X_dist = mean(X_dist)
    ) %>% arrange(desc(mean_X_dist))

## more than 20 segments is sypically misassembled
likely_missassembled <- 
    filtered_combined_delta_ss %>% 
    group_by(qid) %>%
    mutate(
        ncs = n(),
        X_sum = sum(X_dist)
        ) %>%
    filter(ncs>20 | X_sum>1e7) %>%
    pull(qid) %>% unique

#filtered_combined_delta_ss %>% filter(qid %in%likely_missassembled[2] ) %>% plot_delta

fcd_ss <- 
    filtered_combined_delta_ss %>%
    filter(!(qid %in% likely_missassembled))


clustered_genomes_list_ss <- 
    cluster_genomes(filtered_combined_delta_chr1)

clustered_genomes_list_chr2 <- 
    cluster_genomes(filtered_combined_delta_chr2)

cg2 <- 
    cluster_genomes(filtered_combined_delta_chr1)

clustered_genomes_chr1 <- 
   clustered_genomes_list_ss$cluster_summary %>%
    left_join(
        ., 
        sync_tib[,c("accession", "len")] %>% 
            `colnames<-`(c("accessions", "len")),
        by = "accessions"
        ) 

clustered_genomes_list_chr2$cluster_summary %>% count(cluster_id)


clustered_genomes_chr2 <- 
   clustered_genomes_list_chr2$cluster_summary %>%
    left_join(
        ., 
        sync_tib[,c("accession", "len")] %>% 
            `colnames<-`(c("accessions", "len")),
        by = "accessions"
        ) 

clustered_genomes_ss2 %>% count(cluster_id)
clustered_genomes_ss %>% count(cluster_id)

if (any(is.na(clustered_genomes_ss$len))){
    
}

## ###initial synchronization

```

```{r}
delta_identities_chr1 <- 
    pbmclapply(
        1:length((list.files(sprintf("%s/ref_v_all", delta_chr1)))),
        function(i){
            delta_tib <- 
                read_delta(
                    list.files(
                        sprintf("%s/ref_v_all", delta_chr1), 
                        full.names = T
                        )[i]
                    ) 
            
            nucmer_id <- 
                sum(
                    delta_tib$meanlen
                    )/
                (mean(
                    c(
                        delta_tib$rlen[1], 
                        delta_tib$qlen[1])
                     ))
            c
            out_tib <- 
                tibble(
                    accession = 
                        delta_tib$qid[1],
                    identity = 
                        nucmer_id
                )
            
            return(out_tib)
            
        }, mc.cores = 4
    ) %>% bind_rows()

delta_identities_chr2 <- 
    pbmclapply(
        1:length((list.files(sprintf("%s/ref_v_all", delta_chr2)))),
        function(i){
            delta_tib <- 
                read_delta(
                    list.files(
                        sprintf("%s/ref_v_all", delta_chr2), 
                        full.names = T
                        )[i]
                    ) 
            
            nucmer_id <- 
                sum(
                    delta_tib$meanlen
                    )/
                (mean(
                    c(
                        delta_tib$rlen[1], 
                        delta_tib$qlen[1])
                     ))
            c
            out_tib <- 
                tibble(
                    accession = 
                        delta_tib$qid[1],
                    identity = 
                        nucmer_id
                )
            
            return(out_tib)
            
        }, mc.cores = 4
    ) %>% bind_rows()

delta_identities_ss %>%
    arrange(identity)


clustered_genomes_ss %>% as.data.frame()
clustered_genomes_ss2 %>% as.data.frame()
```

### capturing the rest of the sequences
```{r}
the_rest <- 
    the_rest_idx

delta_identities

largest_cluster <- 
    clustered_genomes %>% 
    dplyr::count(cluster_id) %>%
    filter(
        n==max(n)
        ) %>%
    dplyr::slice(1) %>%
    pull(cluster_id)

most_related_biggest_cluster <- 
    sampled_id_table %>%
    arrange(Sequence_mean_distance) %>%
    left_join(
        ., 
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("cluster_id", "Sequence")), 
        by = "Sequence"
    ) %>%
    filter(
        cluster_id==largest_cluster
    ) %>%
    dplyr::slice(1) %>% 
    pull(Sequence)



file.copy(
    sprintf(
        "%s/%s.txt",
        sync_dir, most_related_biggest_cluster
        ),
    sprintf(
        "%s/%s.txt",
        sync2_dir, most_related_biggest_cluster
        )
)

list.files(
    most_related_biggest_cluster
)

most_related_start <- 
    readDNAStringSet(
        sprintf("%s/%s.txt", sync2_dir, most_related_biggest_cluster)
    ) %>% subseq(1,1200)

ref_seq <- 
    sprintf("%s/%s_rs.txt", sync2_dir, most_related_biggest_cluster)

genome_acc <-
    (the_rest %>% pull(accession))[i]


writeXStringSet(
    most_related_start %>% `names<-`(sprintf("%s_rstart", most_related_biggest_cluster)),
    ref_seq
)
sync2_dir2 <- 
    sprintf(
        "%s/refsynced",
        sync2_dir
    )

if(!dir.exists(sync2_dir2)){dir.create(sync2_dir2)}
file.copy(
    sprintf(
        "%s/%s.txt",
        sync_dir, most_related_biggest_cluster
        ),
    sprintf(
        "%s/%s.txt",
        sync2_dir2, most_related_biggest_cluster
        )
)


### synchronize, align, discard
out_tibs <- 
    list()
rest_sync_tib <- 
    list()

for (i in 132:nrow(the_rest)){#which(!check_delta)){
            print(i)
            #i = 1754
            # genome_acc <-
            #     (the_rest %>% pull(accession))[i]

            asm_path <- 
                (the_rest %>% pull(ftp_path))[i]
            
            fasta_file <- 
                sprintf(
                    "%s/%s_genomic.fna.gz",
                    asm_path, 
                    gsub(".*/", "", asm_path)
                )
            asm_name <- 
                sub( ".fna.gz", "", basename(fasta_file))
            
            fasta_seq <- 
                sprintf("%s/%s.txt.gz", sync2_dir, asm_name)
            
            download_attempt <- 
                try(
                    writeXStringSet(
                        try(readDNAStringSet(fasta_file)[1], silent = T),
                        fasta_seq
                    ), silent = T
                )
            
            if (!is.null(download_attempt)){
                out_tib <- 
                        tibble(
                            acc = genome_acc,
                            delta = "Fasta file unavailable",
                            delta_file = delt
                        )
                out_tib <- 
                    sprintf(
                        "%s fasta file unavailable",
                        genome_acc
                    )
            } else{
                ## synchronize it with the beginning 
            ### of the most_related sequence
                fna_seq <-
                    readDNAStringSet(fasta_seq) 
                fna_seq <- 
                    fna_seq[order(width(fna_seq), decreasing = T),][1]
                
                genome_acc <- 
                    strsplit(names(fna_seq), " ")[[1]][1]
                
                names(fna_seq) <- 
                    genome_acc
                wp <- 
                    file.remove(fasta_seq)
                rm(wp)
                
                fasta_seq <- 
                    sprintf(
                        "%s/%s.txt",
                        sync2_dir, genome_acc
                    )
                    
                writeXStringSet(
                    fna_seq, 
                    fasta_seq
                )
                
                delta_out <- 
                    sprintf(
                        "%s/%s_startpoint",
                        sync2_dir, genome_acc
                    )
                genome_acc <- 
                    names(fna_seq)
                
                out_file <- 
                    sprintf(
                        "%s/%s.txt",
                        sync2_dir2, genome_acc
                    )
                
                rest_sync_tib[[i]] <- 
                    rs_sync(
                        ref_seq,
                        fasta_seq,
                        delta_out,
                        out_file
                    )
                
                #readDNAStringSet(out_file)
                #readDNAStringSet(out_file)
                ## remove the original sequence
                wp <- file.remove(fasta_seq)
                rm(wp)
                #out_file
                delta_out2 <- 
                    sprintf("%s/the_rest", delta_dir)
                if(!dir.exists(delta_out2)){dir.create(delta_out2)}
                ###
                gnc_txt <- 
                    generate_nucmer_commands(
                        genome_matrix = 
                            matrix(
                                c(
                                    most_related_biggest_cluster, genome_acc
                                ), nrow = 1
                            ),
                            sync_dir = sync2_dir2,
                            delta_dir = delta_out2
                        ) %>% system(., ignore.stdout = T, ignore.stderr = T)
                
                # readDNAStringSet(sprintf("%s/%s.txt", sync_dir, genome_acc))
                # readDNAStringSet(sprintf("%s/%s.txt", sync_dir,most_related ))
                # 
                # ## checking the delta_file 
                # filter(species_summary, accession %in% c(most_related,genome_acc)) %>% View
                delt <- 
                    sprintf(
                            "%s/%s_v_%s_filtered.delta",
                            delta_out2, most_related_biggest_cluster,genome_acc
                            )
                qry_delta <- 
                    read_delta(
                        delt
                        )
                
                if ((qry_delta$qid[1]) != genome_acc) stop(print(i))
                #plot_delta(qry_delta)
                filtered_qd <- 
                    filter_delta(qry_delta)
                #plot_delta(qry_delta)
                fqd2 <- 
                    filtered_qd %>%
                    group_by(rid, qid) %>% 
                    filter(
                        ## reomve the small inversions
                        (meanlen > minimum_inversion_length) | 
                            (rs ==1) |
                            (re==max(re)),
                    ) %>%
                    mutate(
                        segment = rleid(strand),
                        segments = length(rle(strand)$lengths),
                        invs = sum(rle(strand)$values=="-")
                    ) %>% 
                    mutate(
                        ref_al_len =re-rs+1
                    ) %>%
                    ungroup %>%
                    dplyr::select(
                        strand, X_dist, rs, re, qs, qe, qid, rlen, qlen, 
                        meanlen, segment, segments, invs
                    )
                    
                
                nuc_cov <- 
                    sum(
                        qry_delta$meanlen
                        )/
                    mean(
                        c(
                            qry_delta$rlen[1], 
                            qry_delta$qlen[1])
                         )
                
                invs <- 
                    fqd2$invs[1]
                
                if (nuc_cov < .80){
                    out_tib <- 
                        tibble(
                            acc = genome_acc,
                            delta = 
                                sprintf(
                                    "low_coverage (%s)", 
                                    round(100*nuc_cov, 1)),
                            delta_file = delt
                        )
                } else if (invs==0) {
                    out_tib <- 
                        tibble(
                            acc = genome_acc,
                            delta = "collinear to cluster 0",
                            delta_file = delt
                        )
                } else {
                    out_tib <- 
                        tibble(
                            acc = genome_acc,
                            delta = "separated from cluster 0",
                            delta_file = delt
                        )
                }
            }
            out_tibs[[i]] <- 
                out_tib

        }
bind_rows(head(out_tibs))

```







```{r}

filtered_combined_delta_ss2 %>% filter(qid=="CP099940.1") %>% plot_delta


    


best_identity_chr1 <- 
    inner_join(
        delta_identities_chr1 %>% `colnames<-`(c("accessions", "identity")),
        clustered_genomes_chr1[,-5],
        by = "accessions"
    ) %>%
    mutate(
        identity = 
            ifelse(identity>1,  1, identity),
        identity = 
            ifelse(accessions==most_related,  1.01, identity),
    )

best_identity_chr2 <- 
    inner_join(
        delta_identities_chr2 %>% `colnames<-`(c("accessions", "identity")),
        clustered_genomes_chr2[,-5],
        by = "accessions"
    ) %>%
    mutate(
        identity = 
            ifelse(identity>1,  1, identity),
        identity = 
            ifelse(accessions==most_related_chr2,  1.01, identity),
    )
    


best_identity %>% filter(cluster_id==0) %>% arrange(identity)
best_identity %>% filter(cluster_id==0) %>% arrange(identity) %>% dplyr::slice(35:45)

list.files(new_out_deltadir, pattern = "CP073335.1", full.names = T) %>% plot_delta()

fcd2 %>% filter(qid=="NZ_CP075174.1")
## Anova for all others

best_identity %>% filter(cluster_id!=0) %>%
    arrange(cluster_id) %>%
    count(cluster_id)
best_identity_ss

bis_chr1 <- 
    best_identity_chr1 %>% 
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>% 
    filter(!is.na(cluster_id)) %>%
    dplyr::slice(1)

bis_chr2 <- 
    best_identity_chr2 %>% 
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>% 
    filter(!is.na(cluster_id)) %>%
    dplyr::slice(1)

## table of the most common
bis_tib

bis_1_chr1 <- 
    bis_chr1[-1,]
mpis_chr1 <- 
    lapply(
        1:nrow(bis_chr1),
        function(j){
            dfj <- 
                filtered_combined_delta_chr1 %>% 
                ungroup %>%
                filter(qid==pull(bis_1_chr1, accessions)[j]) 
            # if (j==5){
            #     pj <- 
            #         plot_delta(dfj)  +
            #         xlab("Cluster 0")
            # } else {
                pj <- 
                    plot_delta(dfj)  +
                    xlab("")
            #}
            pjo <- 
                pj +
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 22)
                ) +
                ylab(sprintf("Cluster %s", j)) 
            return(pjo)
        }
    )
mpis_chr1[[1]]

filtered_combined_delta_chr2 %>%
    filter(qid=="NZ_CP023473.1") %>% plot_delta()

bis_1_chr2 <- 
    bis_chr2[-1,]
mpis_chr2 <- 
    lapply(
        1:nrow(bis_chr2),
        function(j){
            dfj <- 
                filtered_combined_delta_chr2 %>% 
                ungroup %>%
                filter(qid==pull(bis_1_chr2, accessions)[j]) 
            # if (j==5){
            #     pj <- 
            #         plot_delta(dfj)  +
            #         xlab("Cluster 0")
            # } else {
                pj <- 
                    plot_delta(dfj)  +
                    xlab("")
            #}
            pjo <- 
                pj +
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 22)
                ) +
                ylab(sprintf("Cluster %s", j)) 
            return(pjo)
        }
    )
mpis_chr2[[1]]

bis_1_chr2

do.call(
     "grid.arrange",
     c(
         mpis_chr1[1:12],
         ncol=3
         )
     )

do.call(
     "grid.arrange",
     c(
         mpis_chr2[1:3],
         ncol=2
         )
     )

do.call(
     "grid.arrange",
     c(
         mpis_ss[c(4,15,1,2,3,9, 13, 25)],
         ncol=3
         )
     )
do.call(
     "grid.arrange",
     c(
         mpis_ss[26:50],
         ncol=5
         )
     )

do.call(
     "grid.arrange",
     c(
         mpis_ss[26:50],
         ncol=5
         )
     )

chr_clusts <- 
    sync_tib %>%
    filter(chr==1) %>%
    left_join(
        ., 
        clustered_genomes_chr1[c(1:3)] %>%
            `colnames<-`(c("cluster_id", "accession", "seqs")),
        by = "accession"
    ) %>%
    bind_rows(
        ., 
        sync_tib %>%
            filter(chr==2) %>%
        left_join(
            ., 
            clustered_genomes_chr2[c(1:3)] %>%
                `colnames<-`(c("cluster_id", "accession", "seqs")),
        by = "accession"
        )
    )

clustered_genomes_chr2 %>% count(cluster_id)

chr_clusts %>% filter(chr==2, cluster_id==2) 

chr_clusts %>% filter(chr==2, cluster_id==3) 

chr_clusts %>% filter(asm_name =="GCF_003612695.1_ASM361269v1")
chr_clusts

cc1 <- 
    chr_clusts %>% group_by(asm_name) %>%
    filter(
        !is.na(cluster_id),
        n() ==2
        ) %>%
    summarise(
        cs = n(),
        clusters = paste(cluster_id, collapse = ",")
    ) %>% 
    filter(cs==2) %>%
    add_count(clusters)

ggplot(data=cc1, aes(x=clusters, y=n)) +
  geom_bar(stat="identity") +
    labs(title = "Chrom1 vs Chrom2", y = "Count", x = "Chr1, Chr2") +
  theme_classic() + 
    theme(
        legend.position="none", 
        plot.title = element_text(hjust = 0.5, size = 22),
        text = element_text(size=22)
        ) 
p


clustered_genomes_chr1

clustered_genomes_chr2

mpis_ss[[1]]

best_identity_ss_2 <- 
    best_identity_ss %>%
    group_by(cluster_id) %>% 
    arrange(desc(identity)) %>%
    dplyr::slice(1:20) %>%
    ungroup

snp_fa_dir <- 
    sprintf(
        "%s/fa", snp_dir
    )
if(!(dir.exists(snp_fa_dir)) ){(dir.create(snp_fa_dir))}

pbmclapply(
    1:nrow(best_identity_ss_2),
    function(i){
        file.copy(
            list.files(
                subset_sync_dir, pattern = best_identity_ss_2$accessions[i],
                full.names = T
            ),
            sprintf(
                "%s/%s.fa", snp_fa_dir,  best_identity_ss_2$accessions[i]
            )
        )
        
    }
)
any(best_identity_ss_2$accessions==most_related)
most_related %in% list.files(snp_fa_dir) %>% sub(".fa", "", .)
    
ref_id <- 
    list.files(
        snp_fa_dir, 
        pattern = most_related,
        full.names = T
        )


most_related

```



SNP similarity
```{bash}
parsnp  -p 6 \
    -r ../data/processing/snp/Spirochaetota/Leptospira/Leptospira_interrogans/fa/NZ_CP020414.2.fa \
    -c \
    -d ../data/processing/snp/Spirochaetota/Leptospira/Leptospira_interrogans/fa \
    -o ../data/processing/snp/Spirochaetota/Leptospira/Leptospira_interrogans/phylo

```

```{bash}
parsnp  -p 6 \
    -r ../data/processing/snp/Pseudomonadota/Bordetella/Bordetella_pertussis/fa/NZ_CP026987.1.fa \
    -c \
    -d ../data/processing/snp/Pseudomonadota/Bordetella/Bordetella_pertussis/fa \
    -o ../data/processing/snp/Pseudomonadota/Bordetella/Bordetella_pertussis/phylo

```





```{r}
install.packages("phytools")
library(phytools)

s_tree <- 
    read.tree("../data/processing/snp/Pseudomonadota/Bordetella/Bordetella_pertussis/phylo/parsnp.tree")

s_tree <- 
    read.tree("../data/processing/snp/Spirochaetota/Leptospira/Leptospira_interrogans/phylo/parsnp.tree")

s_tree$tip.label <- 
    sub(".ref", "", s_tree$tip.label)
s_tree$tip.label <- 
    sub(".fa", "", s_tree$tip.label)

plot(s_tree)

s_tree_pruned <- 
    drop.tip(s_tree, "NZ_CP085969.1")
plot(s_tree_pruned)
s_tree_pruned <- 
    drop.tip(s_tree_pruned, "NZ_CP034101.1")

dm_ss <- 
    cophenetic(s_tree_pruned) %>% as.dist 



ffr <- 
    rowSums(dplyr::select(fff, `0.33MB`, `1.58MB`, `3.16MB`, `0.99MB`))
fff2 <- 
    fff %>%
    mutate(RS = ffr) %>%
    filter(RS>0 | cluster_id==0) %>%
    filter(!(accessions %in% c("NZ_CP085969.1", "NZ_CP034101.1"))) %>%
    filter(cluster_id <=45)

mds_ss <- 
    cmdscale(dm_ss, k = 2)  %>%
    as.data.frame() %>%
    rownames_to_column("accessions") %>%
    inner_join(., fff2, by = "accessions") %>%
    filter(V2>-0.002)


mds_ss %>% as_tibble

ggplot(mds_ss, aes(x = V1, y = V2, colour = factor(`0.33MB`))) +
  geom_point(size = 6) +
  labs(title = "0.33MB", x = "MDS Dimension 1", y = "MDS Dimension 2") +
  theme_classic() + 
    theme(legend.position="none", plot.title = element_text(hjust = 0.5, size = 22)) +
    scale_color_manual(values = c("#000", col_vector[1]))

ggplot(mds_ss, aes(x = V1, y = V2, colour = factor(`1.58MB`))) +
  geom_point(size = 6) +
  labs(title = "1.58MB", x = "MDS Dimension 1", y = "MDS Dimension 2") +
  theme_classic() + 
    theme(legend.position="none", plot.title = element_text(hjust = 0.5, size = 22))+
    scale_color_manual(values = c("#000", col_vector[2]))

ggplot(mds_ss, aes(x = V1, y = V2, colour = factor(`0.99MB`))) +
  geom_point(size = 6) +
  labs(title = "0.99MB", x = "MDS Dimension 1", y = "MDS Dimension 2") +
  theme_classic() + 
    theme(legend.position="none", plot.title = element_text(hjust = 0.5, size = 22))+
    scale_color_manual(values = c("#808080", col_vector[4]))

ggplot(mds_ss, aes(x = V1, y = V2, colour = factor(`3.16MB`))) +
  geom_point(size = 6) +
  labs(title = "3.16MB", x = "MDS Dimension 1", y = "MDS Dimension 2") +
  theme_classic() + 
    theme(legend.position="none", plot.title = element_text(hjust = 0.5, size = 22))+
    scale_color_manual(values = c("#808080", col_vector[6]))

do.call(
     "grid.arrange",
     c(
         list(g12, g18, g52, g83),
         ncol=2
         )
     )

i0_33 <- 
    mds_ss$`0.33MB` %>%
    `names<-`(mds_ss$accessions)
i0_99 <- 
    mds_ss$`0.99MB` %>%
    `names<-`(mds_ss$accessions)
sil099_tib <- 
    silhouette(
        i0_99,  dist(mds_ss[, c("V1", "V2")])
    ) %>%
    as_tibble %>%
    mutate(
        Inversion = "0.33MB"
    )%>% 
    mutate(
        idx = 1:nrow(.),
        variant = 
            factor(
                ifelse(cluster==0, "Absence", "Presence"), 
                levels = c("Presence", "Absence")
            ),
        sil_mod = 
            ifelse(
                cluster==0,
                sil_width*(-1),
                sil_width
            )
    ) %>%
    group_by(variant) %>%
    arrange(desc(sil_mod)) %>%
    mutate(
        idx_f = factor(idx, levels = idx)
    )
i1_58 <- 
    mds_ss$`1.58MB` %>%
    `names<-`(mds_ss$accessions)
sil158_tib <- 
    silhouette(
        i1_58,  dist(mds_ss[, c("V1", "V2")])
    ) %>%
    as_tibble %>%
    mutate(
        Inversion = "3.16MB"
    )%>% 
    mutate(
        idx = 1:nrow(.),
        variant = 
            factor(
                ifelse(cluster==0, "Absence", "Presence"), 
                levels = c("Presence", "Absence")
            ),
        sil_mod = 
            ifelse(
                cluster==0,
                sil_width*(-1),
                sil_width
            )
    ) %>%
    group_by(variant) %>%
    arrange(desc(sil_mod)) %>%
    mutate(
        idx_f = factor(idx, levels = idx)
    )

i3_16 <- 
    mds_ss$`3.16MB` %>%
    `names<-`(mds_ss$accessions)
sil316_tib <- 
    silhouette(
        i3_16,  dist(mds_ss[, c("V1", "V2")])
    ) %>%
    as_tibble %>%
    mutate(
        Inversion = "3.16MB"
    )%>% 
    mutate(
        idx = 1:nrow(.),
        variant = 
            factor(
                ifelse(cluster==0, "Absence", "Presence"), 
                levels = c("Presence", "Absence")
            ),
        sil_mod = 
            ifelse(
                cluster==0,
                sil_width*(-1),
                sil_width
            )
    ) %>%
    group_by(variant) %>%
    arrange(desc(sil_mod)) %>%
    mutate(
        idx_f = factor(idx, levels = idx)
    )

mds_ss %>% head
sil033_tib <- 
    silhouette(
        i0_33,  dist(mds_ss[, c("V1", "V2")])
    ) %>%
    as_tibble %>%
    mutate(
        Inversion = "0.33MB"
    )%>% 
    mutate(
        idx = 1:nrow(.),
        variant = 
            factor(
                ifelse(cluster==0, "Absence", "Presence"), 
                levels = c("Presence", "Absence")
            ),
        sil_mod = 
            ifelse(
                cluster==0,
                sil_width*(-1),
                sil_width
            )
    ) %>%
    group_by(variant) %>%
    arrange(desc(sil_mod)) %>%
    mutate(
        idx_f = factor(idx, levels = idx)
    )


ggplot(sil033_tib, aes(x = idx_f, y = sil_mod, fill = variant)) +
  geom_bar(stat = "identity", width=1) +
  labs(
      title = "", 
      x = "Cluster silhouette similarity", y = "Silhouette width"
      ) +
  #scale_fill_discrete(name = "Cluster") +
  theme_classic() +
    theme(
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18),
        plot.title = element_text(hjust = 0.5, size = 24),
      strip.background = element_rect(colour="white", fill="white"),
      legend.position=c(.9,.9), 
      legend.text=element_text(size=20),
      legend.title = element_text(size = 20)
    ) +
    scale_y_continuous(limits = c(-1,1)) +
    scale_fill_manual(values = c(col_vector[1], "wheat3")) +
    ggtitle("0.33MB")

ggplot(sil158_tib, aes(x = idx_f, y = sil_mod, fill = variant)) +
  geom_bar(stat = "identity", width=1) +
  labs(
      title = "", 
      x = "Cluster silhouette similarity", y = "Silhouette width"
      ) +
  #scale_fill_discrete(name = "Cluster") +
  theme_classic() +
    theme(
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18),
        plot.title = element_text(hjust = 0.5, size = 24),
      strip.background = element_rect(colour="white", fill="white"),
      legend.position=c(.9,.9), 
      legend.text=element_text(size=20),
      legend.title = element_text(size = 20)
    ) +
    scale_y_continuous(limits = c(-1,1)) +
    scale_fill_manual(values = c(col_vector[2], "wheat3")) +
    ggtitle("1.58MB")

ggplot(sil316_tib, aes(x = idx_f, y = sil_mod, fill = variant)) +
  geom_bar(stat = "identity", width=1) +
  labs(
      title = "", 
      x = "Cluster silhouette similarity", y = "Silhouette width"
      ) +
  #scale_fill_discrete(name = "Cluster") +
  theme_classic() +
    theme(
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18),
        plot.title = element_text(hjust = 0.5, size = 24),
      strip.background = element_rect(colour="white", fill="white"),
      legend.position=c(.9,.9), 
      legend.text=element_text(size=20),
      legend.title = element_text(size = 20)
    ) +
    scale_y_continuous(limits = c(-1,1)) +
    scale_fill_manual(values = c(col_vector[6], "wheat3")) +
    ggtitle("3.16MB")

ggplot(sil099_tib, aes(x = idx_f, y = sil_mod, fill = variant)) +
  geom_bar(stat = "identity", width=1) +
  labs(
      title = "", 
      x = "Cluster silhouette similarity", y = "Silhouette width"
      ) +
  #scale_fill_discrete(name = "Cluster") +
  theme_classic() +
    theme(
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        text = element_text(size = 18),
        plot.title = element_text(hjust = 0.5, size = 24),
      strip.background = element_rect(colour="white", fill="white"),
      legend.position=c(.9,.9), 
      legend.text=element_text(size=20),
      legend.title = element_text(size = 20)
    ) +
    scale_y_continuous(limits = c(-1,1)) +
    scale_fill_manual(values = c(col_vector[4], "wheat3")) +
    ggtitle("0.99MB")



pheatmap::pheatmap(as.matrix(dm_ss))

s_tree_phy <- midpoint_root(s_tree_pruned)
plot(s_tree_pruned)

s_tree_phy$edge.length <- NULL

plot(
    s_tree_phy,
     show.tip.label = FALSE, type = "f"
    )

s_tree_phy <- 
    as.phylo(s_tree) 

s_tree_pruned$edge.length <- 
    pmax(s_tree_pruned$edge.length, 4e-3)
s_tree_pruned$edge.length <- 
    rep(1, length(s_tree_pruned$edge.length))


tip_labels <- s_tree_pruned$tip.label
group_data <- 
    data.frame(
        accessions = s_tree_pruned$tip.label,
        idx = 1:length( s_tree_pruned$tip.label)
    )
gl1 <- 
    left_join(
        group_data, 
        clustered_genomes_ss[,c(1,2)]
    ) %>% as_tibble %>%
    mutate(
        g1 = factor(cluster_id)
    )


ggtree(s_tree_pruned, layout = "circular", branch.length = "none") +
  geom_tippoint(aes(color = g1), size = 3, data = gl1) +
  geom_tiplab(size = 2, aes(angle = angle)) +
  scale_color_brewer(palette = "Set1") +  # Use a predefined color palette
  theme_tree2()

tree <- rtree(10)

# Create a data frame for group membership
group_data <- data.frame(
  label = tree$tip.label,
  group = factor(sample(c("Group1", "Group2", "Group3"), length(tree$tip.label), replace = TRUE))
)

# Create a basic circular tree plot
p <- ggtree(tree, layout = "circular") +
  geom_tippoint(aes(color = group), size = 3, data = group_data) +  # Add colored dots for group
  geom_tiplab(size = 3, aes(angle = angle))


plot(
    s_tree_pruned, type = "fan", use.edge.length = FALSE,
    show.tip.label = FALSE
)


hclust_order_ss %>% count(cluster_id)

bis_1_ss

fff2

hclust_order_ss <- 
    ge_f2%>%
    filter(accessions %in% s_tree_pruned$tip.label) %>%
    mutate(
        a2 = factor(accessions, levels = unique(s_tree_pruned$tip.label))
    ) %>%
    arrange(a2)
    clustered_genomes_ss %>%
    filter(accessions %in% s_tree_pruned$tip.label) %>%
    mutate(
        a2 = factor(accessions, levels = unique(s_tree_pruned$tip.label))
    ) %>%
    arrange(a2) %>%
    mutate(
        id2 = 
            ifelse(cluster_id==0, 1,0),
        id3 = 
            ifelse(cluster_id==2, 1,0),
        id4 = 
            ifelse(cluster_id==3, 1,0),
        id5 = 
            ifelse(cluster_id==4, 1,0),
        id6 = 
            ifelse(
                cluster_id %in% c(5:13),
                0,1
            )
        )

plot(
    s_tree_pruned, type = "fan", use.edge.length = FALSE,
    show.tip.label = FALSE
)

variant_cols <- 
    c("firebrick1", "darkgoldenrod", "skyblue", "blue", "red")
col_v

tiplabels(pch = 19, col =  ifelse(hclust_order_ss$"2.15MB" == 0, transparent_color, col_vector[1]), cex = 1.4, offset = 0.03)
tiplabels(pch = 19, col =  ifelse(hclust_order_ss$"1.23MB" == 0, transparent_color, col_vector[2]), cex = 1.4, offset = 0.03)
tiplabels(pch = 19, col =  ifelse(fff2$"0.99MB" == 0, transparent_color, col_vector[4]), cex = 1.4, offset = 0.05)
tiplabels(pch = 19, col =  ifelse(fff2$"1.58MB" == 0, transparent_color, col_vector[2]), cex = 1.4, offset = 0.03)
tiplabels(pch = 19, col =  ifelse(fff2$"3.16MB" == 0, transparent_color, col_vector[6]), cex = 1.4, offset = 0.03)

plot(
    s_tree_pruned,
    # tip.color = ifelse(hclust_order_ss$`12` == 0, "blue", "red"), 
    #  edge.color = ifelse(hclust_order_ss$`18` == 0, "blue", "red"),
     show.tip.label = FALSE, type = "f"
    )



#######
ggtree(s_tree_pruned,  branch.length="none", layout = "circular")

s_tree_phy$edge.length <- sqrt(s_tree_phy$edge.length + 1)

s_tree4 <- midpoint(s_tree4)
ggtree(s_tree4) + geom_tiplab() + xlim_tree(1)

ss_phylo <- 
    as.phylo(hc_row) 


ss_phylo$edge.length%>% sort %>% head(20)
ss_phylo$edge.length <- 
    pmax(ss_phylo$edge.length, 4e-3)

s_tree4 <- midpoint(s_tree4)
ggtree(s_tree4) + geom_tiplab() + xlim_tree(1)



s_tree <- midpoint_root(s_tree)
ggtree(s_tree,  branch.length="none") #+ geom_tiplab(size =7) + xlim_tree(1)

```


Structural variant metrics & defininitions
SV index score - Probability of drawing 2 structurally distinct sequences

Reoccurring structural variant - occurs in >1 strucutural subtype

##
SV event matrix, 
each inversion occurrence, number of same cluster occurrence, cross cluster occurrence, 
total occurrence


```{r}
### group sizes
group_sizes <- 
    bis_ss$grouped_sequences

same_group_prob <-
    function(
        group_N, total_N
    ){
        (group_N/total_N)*((group_N-1)/(total_N-1))
    }
total_sg_prob <- 
    lapply(
        1:length(group_sizes),
        function(i){
            (same_group_prob(
                group_sizes[i],
                sum(group_sizes)
            )*100) %>% 
                round(2)
        }
    ) %>% unlist %>%
    sum
different_group_prob <-
    100-total_sg_prob

```


## subset event capture
Inversion classification:

Replication-associated - symmetric to Ori, lays on or near X

Replication-associated, non-mirrored

Reoccurring - Present across strucutral subtypes

Proximal non-contiguous inversions - inversion axis is shifted

Local inversion - inversion occurs on one strand


```{r}




inner_events <-
    filtered_combined_delta %>%
    filter(
        rs !=min(rs),
        re !=max(re)
    ) %>%
    mutate(
        refmid = round((rs+re)/2),
        qrymid = round((qs+qe)/2),
        mid_diff = abs(refmid-qrymid)
        ) %>%
    dplyr::select(
        c(
            rid, qid, strand, 
            rs, re, qs, qe, 
            qrymid, refmid, mid_diff, meanlen
            )
        ) %>%
    inner_join(
        bis_1[,c(1,3, 4)] %>% `colnames<-`(c("qid", "cluster")),
        #clustered_genomes_ss[,c(1,2)] %>% `colnames<-`(c("cluster", "qid")),
        by = "qid"
    ) %>% arrange(cluster)


dbs_ref <- 
    fpc::dbscan(
        inner_events %>% ungroup %>% dplyr::select(rs), 
        eps = 1, MinPts = 1, scale = F, method = "raw"
        )
ie2 <- 
    inner_events %>% 
    ungroup %>%
    mutate(rs_clus = dbs_ref$cluster) 

shared <-
    ie2 %>% add_count(rs_clus, name = "occ") %>%
    filter(occ>1) %>%
    group_by(rs_clus) %>%
    mutate(clusts = paste(cluster, collapse = ",")) %>%
    distinct(clusts, .keep_all = T)

bis_1_ss$accessions[1]
filtered_combined_delta_ss %>% filter(qid==bis_1_ss$accessions[1])
    


ie2 %>%
    filter(rs_clus==2)

events <- 
    inner_events %>% 
    filter(meanlen<320000, meanlen>312000)


variants <- 
    tibble(
        
    )


inversion_axis_p <-
    inner_events %>%
    filter(strand =="+")

filtered_combined_delta_ss %>%
    filter(qid =="NZ_CP010964.1") %>% plot_delta


inversion_axis_p %>% pull(mid_diff) %>% mean()
inversion_axis_p %>% 
    filter(mid_diff>1e6) %>%
    pull(mid_diff) %>% boxplot
    
    

events_ss <- 
    list()

events_ss[c(1,2)]
evd_ss
bie2_ss$meanlen
bie2_ss %>% filter(cluster_id==2)


se <- 
    bie2_ss %>% arrange(cluster_id) %>%
    mutate(
        esize = round(meansize*len*0.01),
        mp = round(midpoint*len*0.01)
        )

dbs_ref <- 
    fpc::dbscan(
        bie2_ss %>% dplyr::select(meanlen, refmid) %>% filter(!is.na(meanlen)), 
        eps = 1, MinPts = 1, scale = F, method = "raw"
        )

dbs_ref <- 
    fpc::dbscan(
        se %>% dplyr::select(meansize, midpoint) %>% filter(!is.na(meansize)), 
        eps = 1, MinPts = 1, scale = F, method = "raw"
        )
ge <- 
    se %>%
    dplyr::select(qid, cluster_id, meanlen, refmid) %>%#, esize, mp) %>%
    mutate(event = dbs_ref$cluster) %>%
    add_count(event, name = "ec")

ge_final <- 
    ge %>%
    filter(ec>5) %>%
    mutate(value = 1) %>%  # Add a column to indicate presence
    pivot_wider(names_from = event, values_from = value, values_fill = 0) %>%
    filter(!is.na(esize))

ge_f2 <- 
    ge_final %>%
    group_by(cluster_id) %>%
    summarise(
        "0.33MB" = sum(`1`),
        "1.58MB" = sum(`2`),
        "3.16MB" = sum(`3`),
        "0.99MB" = sum(`4`)
    )

bie2_ss


ge_f2 <- 
    clustered_genomes_ss %>%
    mutate(
        "2.15MB"  = ifelse(cluster_id %in% c(1,2,3,5,7,8,9,10,12,13), 1,0),
        "1.23MB"  = ifelse(cluster_id %in% c(5,8,10,11), 1,0)
    )
 

filtered_combined_delta_ss%>%
    inner_join(
        bis_1_ss[,c(1,3)] %>% `colnames<-`(c("qid", "cluster")),
        #clustered_genomes_ss[,c(1,2)] %>% `colnames<-`(c("cluster", "qid")),
        by = "qid"
    ) %>% arrange(cluster) %>%
    filter(cluster%in% c(5,8,10,11))



fff <- 
    clustered_genomes_ss[c(1,2)] %>%
    left_join(., ge_f2, by = "cluster_id") %>%
    mutate(
        `0.33MB` = ifelse(is.na(`0.33MB`), 0, `0.33MB`),
        `1.58MB` = ifelse(is.na(`1.58MB`), 0, `1.58MB`),
        `3.16MB` = ifelse(is.na(`3.16MB`), 0, `3.16MB`),
        `0.99MB`= ifelse(is.na(`0.99MB`), 0, `0.99MB`)
    )









ge %>% filter(ec>5)

ge$ec
table(ge$cluster_id, ge$event) %>% 
    as.data.frame()
    as_tibble()


bis_1_ss %>% arrange(identity)
events_ss <- list()

for (i in 1:nrow(bis_1_ss)){
####
    qi <- 
        pull(bis_1_ss, accessions)[i]
    # the delta file filtered
    fdi <- 
        filtered_combined_delta_ss %>% 
        ungroup %>%
        filter(qid==qi) %>%
        mutate(
            rmid = round((rs+re)/2),
            qmid = round((qs+qe)/2)
        ) %>%
        mutate(meanlen = round(meanlen))
        
    plot_delta(fdi)
    fd_less <- 
        fdi %>%
        dplyr::select(
            qid, strand, X_dist, rs, re, qs, qe, meanlen, rmid, qmid, segment
        ) %>%
        filter(
            segment != min(segment),
            segment != max(segment),
            ) %>%
        mutate(
            inv_type1 = 
                ifelse(
                    X_dist < 1e6,
                    "RASR",
                    "Local"
                    )
            )
    # Checking segments
    # +ve
    fd_less2_p <- 
        fdi %>%
        dplyr::select(
            strand, rmid, qmid, meanlen, rs, re, qs, qe, X_dist
            ) %>%
        mutate(
            mid_diff = 
                abs(rmid-qmid)
        ) %>% filter(strand=="+")
    # positive mid diff 
    pmd <- 
        fd_less2_p %>% pull(mid_diff)
    
    positive_tlocs <- 
        pmd[
            (pmd < (quantile(pmd, 0.25) - 3 * IQR(pmd))) | 
                (pmd > (quantile(pmd, 0.75) + 3 * IQR(pmd)))
            ]
        
    fd_less2_n <- 
        fdi %>%
        dplyr::select(strand, rmid, qmid, meanlen) %>%
        mutate(
            mid_diff = 
                abs(rmid-qmid)
        ) %>% filter(strand=="-")
        
        fd_less_RASR <- 
            fd_less %>%
            group_by(segment) %>%
            summarise(
                qid = unique(qid),
                strand = unique(strand),
                meanlen = round(mean(meanlen)),
                refmid = round(mean(rmid)),
                qrymid = round(mean(qmid)),
                X_dist = round(mean(X_dist)),
                inv_type1 = unique(inv_type1)
                )
        
        segdiff <- 
            fd_less_RASR %>%
            mutate(segdiff = segment-lag(segment, default = 1)) %>%
            pull(segdiff)
        
        if (any(segdiff>1)) stop("seg missing")#next
        
        fd_less_RASR_check <- 
            fd_less %>%
            filter(
                strand=="-" |
                    (
                        (lag(strand, default = "+")=="-") & 
                            (lead(strand, default = "+")=="-")
                    )
            )
        
        inner_seg <- 
            median(unique(fd_less_RASR$segment))
        
        if (nrow(fd_less_RASR)==1){
            middle_seg_row <- 
                fd_less_RASR %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    segment = as.character(segment),
                    X_dist = abs(refmid-qrymid)
                    ) 
            events_ss[[i]] <- 
                middle_seg_row
        } else {
            middle_seg_row <- 
                fd_less_RASR %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    segment = as.character(segment),
                    X_dist = abs(refmid-qrymid)
                    ) 
            
            outer_segs <- 
                fd_less_RASR %>%
                filter(segment != median(segment)) 
                
            possible_outers <- 
                nrow(outer_segs) %/% 2
            
            outer_events <- 
                list()
            for (j in 1:possible_outers){
                outer_events[[j]] <- 
                    fd_less %>% 
                    filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                    summarise(
                        qid = unique(qid),
                        strand= unique(strand),
                        segment = paste(segment, collapse = ","),
                        #X_dist = round(mean(X_dist)),
                        rs = min(rs), 
                        re = max(re),
                        qs = 
                             ifelse(
                                strand=="+",
                                min(qs),
                                min(qe)
                            ),
                        qe =
                            ifelse(
                                strand=="+",
                                max(qs),
                                max(qe)
                            ),
                        meanlen = round(sum(meanlen)),
                        refmid = 
                            round((rs+re)/2),
                        qrymid = 
                            round((qs+qe)/2),
                        inv_type1 = unique(inv_type1),
                        X_dist = abs(refmid-qrymid)
                    ) %>%
                    dplyr::select(
                        c(
                            qid, strand, meanlen, refmid, 
                            qrymid,X_dist, inv_type1, segment
                        )
                    )
            }
            inv_summary <- 
                bind_rows(
                    middle_seg_row,
                    bind_rows(outer_events)
                )
            events_ss[[i]] <- 
                inv_summary
        }
        
    
    #plot_delta(fdi)
    
    print(i)
    
}
```


```{r}

i=10
i=2
list.files(
    sprintf("%s/ref_v_all", delta_dir_200), full.names = T, 
    pattern = bis_1_ss[12,]$accessions
) %>% plot_delta

## clustering similar variants



###### prev
for (i in 1:nrow(bis_1_ss)){
####
    qi <- 
        pull(bis_1_ss, accessions)[i]
    # the delta file filtered
    fdi <- 
        filtered_combined_delta_ss %>% 
        ungroup %>%
        filter(qid==qi) %>%
        mutate(
            rmid = round((rs+re)/2),
            qmid = round((qs+qe)/2)
        )
    fd_less <- 
        fdi %>%
        dplyr::select(
            strand, X_dist, rs, re, qs, qe, meanlen, segment
        ) %>%
        filter(
            segment != min(segment),
            segment != max(segment),
            )
    
    # plot_delta(fdi)
    fd2 <- 
        fdi %>%
        filter( ## not the first or last alignment, 
            # rs !=min(rs), re !=max(re), 
            # qs !=1
            segment != min(segment),
            segment != max(segment)
            )%>%
        filter(X_dist<1e6)
    
    if (nrow(fd2)==0){
        events_ss[[i]] <- 
            fdi %>%
            filter( ## not the first or last alignment, 
                # rs !=min(rs), re !=max(re), 
                # qs !=1
                segment != min(segment),
                segment != max(segment)
                ) %>%
            rowwise() %>%
            mutate(
                rsp = 
                    ((100*rs)/rlen) %>% round(.),
                rep = 
                    ((100*re)/rlen) %>% round(.),
                qsp = 
                    ((100*qs)/qlen) %>% round(.),
                qep = 
                    ((100*qe)/qlen) %>% round(.),
                rmp = round((rs+re)/2),
                qmp = round((qs+qe)/2),
                rmprop = round(100*rmp/rlen),
                qmprop = round(100*qmp/qlen),
                meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                midpoint = mean(rmprop, qmprop),
                segment = as.character(segment)
            ) %>%
            dplyr::select(
                qid,# rsp, rep, qsp, qep, 
                strand, segment,
                X_dist, meansize, midpoint
                ) %>%
            ungroup()
            
    } else{
        fd3 <- 
            fdi %>%
            filter( ## not the first or last alignment, 
                # rs !=min(rs), re !=max(re), 
                # qs !=1
                segment != min(segment),
                segment != max(segment)
                ) %>%
            filter(X_dist<1e6) %>%
            group_by(segment) %>%
            summarise(
                strand = unique(strand),
                rs = min(rs), 
                re = max(re),
                qs = ifelse(strand=="+", min(qs), max(qs)),
                qe = ifelse(strand=="+", max(qe), min(qe)),
                X_dist = round(mean(X_dist)),
                rlen = unique(rlen),
                qlen = unique(qlen),
                meanlen = round(sum(meanlen)),
                qid = unique(qid)
                )
        
        segdiff <- 
            fd3 %>%
            mutate(segdiff = segment-lag(segment, default = 1)) %>%
            pull(segdiff)
        
        if (any(segdiff>1)) next
        
        fd2_props <- 
            fd3 %>%
            rowwise() %>%
            mutate(
                rsp = 
                    ((100*rs)/rlen) %>% round(.),
                rep = 
                    ((100*re)/rlen) %>% round(.),
                qsp = 
                    ((100*qs)/qlen) %>% round(.),
                qep = 
                    ((100*qe)/qlen) %>% round(.),
                rmp = round((rs+re)/2),
                qmp = round((qs+qe)/2),
                rmprop = round(100*rmp/rlen),
                qmprop = round(100*qmp/qlen),
                meanlen_prop = round((100*meanlen)/mean(c(rlen, qlen))),
                midpoint_prop = mean(rmprop, qmprop)
            ) %>%
            dplyr::select(
                qid, rsp, rep, qsp, qep, strand, segment,
                X_dist, meanlen, midpoint, meanlen_prop, midpoint_prop
                ) %>%
            ungroup() %>%
            filter(
                strand=="-" |
                    (
                        (lag(strand, default = "+")=="-") & 
                            (lead(strand, default = "+")=="-")
                    )
            )
        
        inner_seg <- 
            median(unique(fd2_props$segment))
        
        if (nrow(fd2_props)==1){
            middle_seg_row <- 
                fd2_props %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        ),
                    segment = as.character(segment)
                    ) %>%
                dplyr::select(
                    qid, strand, segment, X_dist, meansize, midpoint, ms, me
                )
            events_ss[[i]] <- 
                middle_seg_row
        } else {
            
            middle_seg_row <- 
                fd2_props %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        ),
                    segment = as.character(segment)
                    ) %>%
                dplyr::select(
                    qid, strand, segment, X_dist, meansize, midpoint, ms, me
                )
            outer_segs <- 
                fd2_props %>%
                filter(segment != median(segment)) 
                
            possible_outers <- 
                nrow(outer_segs) %/% 2
            outer_events <- 
                list()
            for (j in 1:possible_outers){
                outer_events[[j]] <- 
                    fd2_props %>% 
                    filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                    summarise(
                        qid = unique(qid),
                        strand= unique(strand),
                        segment = paste(segment, collapse = ","),
                        X_dist = mean(X_dist),
                        ms =
                            ifelse(
                                strand=="+",
                                mean(c(min(rsp), min(qsp))),
                                mean(c(min(rsp), min(qep)))
                            ),
                        me =
                            ifelse(
                                strand=="+",
                                mean(c(max(rep), max(qep))),
                                mean(c(max(rep), max(qsp)))
                            ),
                        meansize = 
                            me-ms,
                        midpoint = 
                            (ms+me)/2
                    )
            }
            inv_summary <- 
                bind_rows(
                    middle_seg_row,
                    bind_rows(outer_events)
                )
            events_ss[[i]] <- 
                inv_summary
        }
        
    }
    #plot_delta(fdi)
    
    print(i)
    
}
events_ss %>% head
bind_rows(events) %>% filter(!is.na(rsp))
evd_ss <- 
    bind_rows(events_ss) %>% arrange(meanlen)

e2_ss <- 
    evd_ss %>%
    filter(midpoint>40 & midpoint<60)

bie2_ss <- 
    left_join(
        evd_ss,
        (bis_1_ss %>% dplyr::rename(qid = accessions)),
          by = "qid"
    ) %>% 
    arrange(meanlen) %>%
    add_count(qid, name = "i2") %>%
    filter(inversions==i2)

list.files(new_out_deltadir, pattern = "NZ_CP067078.1", full.names = T) %>% plot_delta


bie3_n <- 
    pbmclapply(
        1:length(unique(bie2$qid)),
        function(i){
            qi <- 
                unique(bie2$qid)[i]
            q_events <- 
                bie2_ss %>%
                filter(qid==qi) %>% 
                dplyr::select(qid, meansize) %>%
                mutate(
                    mlow = 
                        meansize-1,
                    mhigh = 
                        meansize+1
                )
            #which(bie2$qid=="NZ_LR590082.1")
            
            #bie2 %>% filter()
            
            shared_events <- 
                bie2_ss %>%#[which(bie2$qid=="NZ_LR590082.1"),] %>%
                mutate(
                    shared_event = 
                        sapply(
                            # bie2[which(bie2$qid=="NZ_LR590082.1"),] %>%
                            #     pull(meansize),
                            bie2_ss$meansize, 
                            function(value){
                              any(
                                  (value >= q_events$mlow) & 
                                      (value <=  q_events$mhigh)
                                  )  
                                } 
                            )
                ) %>%
                filter(
                    qid != qi,
                    shared_event
                )

            #shared_events %>% filter(qid == qi)
            #which(shared_events$cluster_id==47)
            
            event_tib <- 
                lapply(
                    unique(shared_events$qid),
                    function(x){
                        x <-
                         unique(shared_events$qid[1])
                        #for (x in unique(shared_events$qid)){
                        x_shared <- 
                            shared_events %>%
                            filter(qid==x) %>%
                            dplyr::select(-c("shared_event"))
                        x_events <- 
                            bie2_ss %>%
                            filter(qid==x)
                        
                        xdiff <- 
                            bind_rows(x_shared, x_events) %>% 
                            dplyr::add_count(meansize, name = "occ") %>%
                            filter(occ==1)
                   
                        qdiff <- 
                            q_events %>%
                            mutate(
                                qd = 
                                    lapply(
                                        q_events$meansize,
                                        function(qe){
                                            unlist(
                                                lapply(
                                                    x_events$meansize,
                                                    function(xe){
                                                        (qe >= xe-1) &
                                                            (qe <=  xe+1)
                                                    }
                                                )
                                            ) %>% any
                                        }
                                    ) %>% unlist()
                            ) %>%
                            filter(
                                !qd
                            ) #
                            
                        out_tib <- 
                            tibble(
                                qid = x,
                                se = nrow(x_shared),
                                q_events = nrow(q_events),
                                x_events = nrow(x_events)
                            ) %>%
                            mutate(
                                des = abs(se-q_events),
                                event = 
                                    ifelse(
                                        length(xdiff$meansize)==1,
                                        xdiff$meansize,
                                        ifelse(
                                            length(qdiff$meansize)==1,
                                            qdiff$meansize,
                                            NA
                                        )
                                    )
                            )
                        
                        return(out_tib)
                    }
                ) %>% 
                bind_rows()
        }
    )


## identify shared events
bie3 <- 
    pbmclapply(
        1:length(unique(bie2_ss$qid)),
        function(i){
            qi <- 
                unique(bie2_ss$qid)[i]
            q_events <- 
                bie2_ss %>%
                filter(qid==qi) %>% 
                dplyr::select(qid, meansize) %>%
                mutate(
                    mlow = 
                        meansize-1,
                    mhigh = 
                        meansize+1
                )
            #which(bie2$qid=="NZ_LR590082.1")
            
            #bie2 %>% filter()
            
            shared_events <- 
                bie2_ss %>%#[which(bie2$qid=="NZ_LR590082.1"),] %>%
                mutate(
                    shared_event = 
                        sapply(
                            # bie2[which(bie2$qid=="NZ_LR590082.1"),] %>%
                            #     pull(meansize),
                            bie2_ss$meansize, 
                            function(value){
                              any(
                                  (value >= q_events$mlow) & 
                                      (value <=  q_events$mhigh)
                                  )  
                                } 
                            )
                ) %>%
                filter(
                    qid != qi,
                    shared_event
                )

            #shared_events %>% filter(qid == qi)
            #which(shared_events$cluster_id==47)
            
            event_tib <- 
                lapply(
                    unique(shared_events$qid),
                    function(x){
                        # x <-
                        #  unique(shared_events$qid[38])
                        #for (x in unique(shared_events$qid)){
                        x_shared <- 
                            shared_events %>%
                            filter(qid==x) %>%
                            dplyr::select(-c("shared_event"))
                        x_events <- 
                            bie2_ss %>%
                            filter(qid==x)
                        
                        xdiff <- 
                            bind_rows(x_shared, x_events) %>% 
                            dplyr::add_count(meansize, name = "occ") %>%
                            filter(occ==1)
                   
                        qdiff <- 
                            q_events %>%
                            mutate(
                                qd = 
                                    lapply(
                                        q_events$meansize,
                                        function(qe){
                                            unlist(
                                                lapply(
                                                    x_events$meansize,
                                                    function(xe){
                                                        (qe >= xe-1) &
                                                            (qe <=  xe+1)
                                                    }
                                                )
                                            ) %>% any
                                        }
                                    ) %>% unlist()
                            ) %>%
                            filter(
                                !qd
                            ) #%<%
                            # x_events %>% 
                            # dplyr::select(qid, meansize) %>%
                            # mutate(
                            #     mlow = 
                            #         meansize-1,
                            #     mhigh = 
                            #         meansize+1
                            # ) %>%
                            # bind_rows(
                            #     ., 
                            #     q_events
                            # )
                            
                        out_tib <- 
                            tibble(
                                qid = x,
                                se = nrow(x_shared),
                                q_events = nrow(q_events),
                                x_events = nrow(x_events)
                            ) %>%
                            mutate(
                                des = abs(se-q_events),
                                event = 
                                    ifelse(
                                        length(xdiff$meansize)==1,
                                        xdiff$meansize,
                                        ifelse(
                                            length(qdiff$meansize)==1,
                                            qdiff$meansize,
                                            NA
                                        )
                                    )
                            )
                        
                        return(out_tib)
                    }
                ) %>% 
                bind_rows()
            #shared_events %>%
             #   group_by(qid)
            # same_cluster <- 
            #     event_tib %>%
            #     filter(des==0)
            # adjacent_cluster <- 
            #      event_tib %>%
            #      filter(des==1) 
            
            if (nrow(event_tib)>0){
                out_tib <- 
                    event_tib %>%
                    filter(
                        des %in% c(0,1)
                    ) %>%
                    mutate(
                        rid = qi,
                        rclust = 
                            bie2_ss$cluster_id[i]
                    ) %>%
                    left_join(
                        ., 
                        bie2_ss[,c("qid", "cluster_id")] %>%
                            `colnames<-`(c("qid", "qclust")),
                        by = "qid"
                    )
            } else{
                out_tib <- NULL
            }
            
            
            return(out_tib)
                
        }
        )

qi
list.files(new_out_deltadir, pattern = "CP090240.1", full.names = T) %>%
    plot_delta()


###

bie4 <- 
    bind_rows(bie3) %>%
    dplyr::select(-des) %>%
    # filter(
    #     !(rclust %in% redun),
    #     !(qclust %in% redun)
    #     ) %>%
    mutate(
        qs = q_events-se,
        xs = x_events-se,
        qx = qs+xs
    ) %>%
    #filter(qx==1) %>%
    rowwise() %>%
         mutate(
            rq = paste(sort(c(rid, qid)), collapse = ",")
            ) %>%
    ungroup %>%
    distinct(rq, .keep_all = T) %>%
    add_count(event) %>%
    arrange(event)  %>%
    dplyr::select(qid, rid, rq, event, rclust, qclust, n)


filtered_combined_delta_ss %>%
    filter(qid=="CP046995.1") %>% plot_delta



species_identities$mash_tib %>% 
    filter(qry == "NZ_LT905063.1")

bie4q <- 
    bie4_ss$qid %>% unique
bie4r <- 
    bie4_ss$rid %>% unique

species_identities_ss$mash_tib



```



```{r}
mash_tib2 <-
    species_identities_ss$mash_tib %>%
    `colnames<-`(
        c(
            "rid", "qid",  "identity", 
            "reflen", "qrylen", "ld", 
            "sld", "identity_sld"
            )
        ) %>%
    mutate(
        rid = as.character(rid),
        qid = as.character(qid)
    ) %>%
    filter(rid !=qid) %>%
    rowwise() %>%
         mutate(
            rq = paste(sort(c(rid, qid)), collapse = ",")
            ) %>%
    ungroup %>%
    mutate(
        dist = 100-identity
    ) %>%
    group_by(rq) %>%
    summarise(
        ref = rid[1],
        qry = qid[1],
        mean_dist = mean(dist),
        dist_diff = abs(diff(dist)),
        mean_id = mean(identity),
        id_diff = abs(diff(identity))
    ) 


mash_tib2_wide <- 
    #        mash_tib2 %>%
    species_identities_ss$mash_tib %>%
    dplyr::select(c(ref, qry, identity)) %>%
    spread(., qry, identity) %>%
    as.data.frame() %>%
    column_to_rownames("ref")

species_identities_ss$mash_plot

clust_identity_list$clust_identity_diff 
    

acc_rowsums <- 
    mash_tib2_wide %>%
    as.data.frame() %>%
    as_tibble() %>%
    mutate(
        mean_id = rowMeans(.),
        accession = rownames(mash_tib2_wide)
    ) %>%
    dplyr::select(accession, mean_id) %>% arrange(mean_id)



list.files(new_out_deltadir, pattern = "CP073326.1", full.names = T) %>% 
    plot_delta
list.files(new_out_deltadir, pattern = "NZ_CP079838.1", full.names = T) %>% 
    plot_delta

filtered_combined_delta_ss %>% filter(qid == "CP073326.1")

mash_tib2_wide_dist <- 
    100-as.matrix(mash_tib2_wide) 
mash_tib2_wide_dist[upper.tri(mash_tib2_wide_dist, diag = T)]<-NA
        
install.packages("pheatmap")
pheatmap(mash_tib2_wide_dist)
library(pheatmap)

nzs <- 
    clustered_genomes_ss %>% 
    filter(cluster_id != 0) %>%
    pull(accessions)

mash_dist2 <- 
    as.dist(
        mash_tib2_wide_dist[
            (rownames(mash_tib2_wide_dist) %in% nzs),
            (colnames(mash_tib2_wide_dist) %in% nzs)
        ]
        )
mash_clust2 <- 
    hclust(mash_dist2, method = "complete")

mash_clust2$height <- 
    mash_clust2$height *2
hclust_dendrogram <- 
    as.dendrogram(mash_clust2)

hclust_dendrogram

cophenetic(hclust_dendrogram)
dend_data <- 
    dendro_data(hclust_dendrogram)

dd_clust <- 
    dend_data$labels %>%
    left_join(
        ., 
        clustered_genomes_ss[,c(1,2)] %>%
            `colnames<-`(c("Strucutral_cluster", "label"))
        , by = "label"
        ) %>%
    filter(!is.na(Strucutral_cluster)) %>%
    filter(Strucutral_cluster!=0) %>%
    mutate(Strucutral_cluster=factor(Strucutral_cluster)) %>%
    add_count(Strucutral_cluster) 

dend_df <- 
    dend_data$segments %>% 
    mutate(struct_clust = dd_clust$Strucutral_cluster) 

label(dend_data)

clustered_genomes_ss$cluster_id

labels_colors(hclust_dendrogram) <- 
    as.factor(
        clustered_genomes_ss$cluster_id
        )[order.dendrogram(hclust_dendrogram)]

dend_data
segment_data <- 
    segment(dend_data)
labels_df <- 
    label(dend_data) %>% 
    

segment_data$group <- labels_df$group[match(segment_data$x, labels_df$x)]

ggplot(segment_data) +
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend, color = group)) +
  theme_minimal() +
  labs(title = "Colored Dendrogram Based on Groups",
       x = "Sample",
       y = "Height")

segment(dend_data)

"d"
plot(mash_clust2)
```

```{r}
ggplot() +
  geom_segment(
      data = 
          segment(dend_data), 
      aes(x = x, y = y, xend = xend, yend = yend)) +
  geom_text(
      data = dd_clust,
      aes(x = x, y = y, label = label, colour = Strucutral_cluster), hjust = 0
      ) +
  theme_minimal() +
  labs(
      title = "Dendrogram of Hierarchical Clustering",
      x = "Sample",
      y = "Height"
      )+
  coord_flip() + scale_y_reverse(expand=c(0.2, 0)) +
    scale_color_manual(values = col_vector[1:18])
  #scale_y_reverse(expand = c(0.3, 0)) +
  #coord_polar(theta="x")


col_vector<-
    c(
        '#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', 
        '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', 
        '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', 
        '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', 
        '#ffffff', '#000000'
        )




ggplot(
    pca_tib,
    aes(x = PC1, y = PC2, colour = Structural_cluster)
) +
    geom_point(size=4) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      )  + 
    scale_color_manual(values = col_vector[1:18])

cluster

"d"





```






```{r}

clust_order <-
    mash_clust2$labels[mash_clust2$order]

mash_dist2
outliers <- 
    acc_rowsums %>% filter(mean_id<97) %>% pull(accession)

pca_result <- 
    prcomp(mash_dist2, scale = TRUE, center = T)

summary(pca_result)

pca_tib <- 
    pca_result$x[,c(1,2)] %>%
    as.matrix() %>%
    as.data.frame() %>%
    rownames_to_column(var = "accessions") %>%
    as_tibble() %>%
    arrange(PC1) %>%
    filter(!(accessions %in% outliers)) %>%
    left_join(
        .,
        clustered_genomes_ss[,c(1,2)], by = "accessions"
    ) %>%
    mutate(
        Structural_cluster = factor(cluster_id)
    ) %>% filter(Structural_cluster !=0)

cc1 <- 
    pca_tib %>% filter(Structural_cluster %in% c(2,4,16,17,18)) %>%
    dplyr::group_by(Structural_cluster) %>% 
    dplyr::slice(1) %>% 
    ungroup %>% 
    pull(accessions)

cc1p <- 
    lapply(
        cc1, 
        function(x){
            list.files(new_out_deltadir, pattern = x, full.names = T) %>% 
                plot_delta
        }
    )

do.call(
     "grid.arrange",
     c(
         cc1p,
         ncol=3
         )
     )

cc2 <- 
    pca_tib %>% filter(Structural_cluster %in% c(5,9,13,14)) %>%
    dplyr::group_by(Structural_cluster) %>% 
    dplyr::slice(1) %>% 
    ungroup %>% 
    pull(accessions)

cc2p <- 
    lapply(
        cc2, 
        function(x){
            list.files(new_out_deltadir, pattern = x, full.names = T) %>% 
                plot_delta
        }
    )

do.call(
     "grid.arrange",
     c(
         cc2p,
         ncol=3
         )
     )

cc3 <- 
    pca_tib %>% filter(Structural_cluster %in% c(1,3,6,8,10,15)) %>%
    dplyr::group_by(Structural_cluster) %>% 
    dplyr::slice(1) %>% 
    ungroup %>% 
    pull(accessions)




cc3p <- 
    lapply(
        cc3, 
        function(x){
            list.files(new_out_deltadir, pattern = x, full.names = T) %>% 
                plot_delta
        }
    )

do.call(
     "grid.arrange",
     c(
         cc3p,
         ncol=3
         )
     )

cc1p[[1]]



clustered_genomes_ss

col_vector<-
    c(
        '#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', 
        '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', 
        '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', 
        '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', 
        '#ffffff', '#000000'
        )




ggplot(
    pca_tib,
    aes(x = PC1, y = PC2, colour = Structural_cluster)
) +
    geom_point(size=4) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      )  + 
    scale_color_manual(values = col_vector[1:18])


ggplot(
    pca_tib %>% filter(PC1<=-4),
    aes(x = PC1, y = PC2, colour = Structural_cluster)
) +
    geom_point(size=6) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      )  + 
    scale_color_manual(values = col_vector[1:18])



ggplot(
    pca_tib %>% filter(PC2>=4),
    aes(x = PC1, y = PC2, colour = Structural_cluster)
) +
    geom_point(size=8) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      )  + 
    scale_color_manual(values = col_vector[1:18])

ggplot(
    pca_tib %>% filter(PC1>=0, PC1 <=3),
    aes(x = PC1, y = PC2, colour = Structural_cluster)
) +
    geom_point(size=8) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      )  + 
    scale_color_manual(values = col_vector[1:18])


pca_tib

ggplot(
    aes(PC1)
)


plot(
    pca_tib$PC1, 
    pca_tib$PC2, col = pca_tib$cluster_id,
    xlab = "PC1", ylab = "PC2", main = "PCA Plot"
    )
pca_result$x[,1] %>% arrange(x)

#distinct(rq, .keep_all = T)

mash_tib2 %>% 
    filter(
        rid %in% (bie4_ss$qid %>% unique),
        qid %in% (bie4_ss$rid %>% unique)
        )


bie5_ss <- 
    mash_tib2 %>%
    left_join(., bie4_ss , by = c("rq"))

### ani



in_file <- 
    ani_inlines

ani_dir <- 
    sprintf(
        "%s/ani/ac1_ss",
        identity_dir
        )

if (!dir.exists(ani_dir)){dir.create(ani_dir)}
out_dir <- 
    ani_dir

ani_inlines_1 <- 
    sprintf(
        "%s/%s.txt", 
        sync_dir, 
        bie4r
    )

ani_inlines_2 <- 
    sprintf(
        "%s/%s.txt", 
        sync_dir, 
        bie4q
    )

ani_infile_ss_1 <- 
    sprintf(
        "%s/ac_ss_1.txt",
        ani_dir
    )
writeLines(
    ani_inlines_1,
    ani_infile_ss_1
)

ani_infile_ss_2 <- 
    sprintf(
        "%s/ac_ss_2.txt",
        ani_dir
    )
writeLines(
    ani_inlines_2,
    ani_infile_ss_2
)

ani_out_ss <- 
    sprintf(
        "%s/ac_ss_out.txt",
        out_dir  
    )

ani_commands2 <- 
    function(
        in_file1, in_file2,
        out_file
    ){
        ani_com <- 
            sprintf(
                "%s --ql %s --rl %s -o %s",
                "fastANI -t 7",
                in_file1, in_file2,
                out_file
            )
        if (!file.exists(out_file)){
            system(ani_com)
        }
        
        ani_tib <- 
            fread(out_file) %>%
            mutate(V1 = sub(".txt", "", basename(V1))) %>%
            mutate(V2 = sub(".txt", "", basename(V2))) %>%
            `colnames<-`(c("rid", "qid", "ani", "matched", "total")) %>%
            mutate(ani = as.numeric(ani)) %>%
            as_tibble %>%
            mutate(
                ani_dist = 100-ani
            ) %>%
            rowwise() %>%
            #filter(ref !=qry) %>%
            mutate(
                rq = paste(sort(c(rid, qid)), collapse = ",")
                ) %>%
            #filter(rq == "1.5,CP074215.1") %>%
            ungroup
        return(ani_tib)
        
    }

out_file <- 
    ani_out_ss1

ani_commands2(ani_infile_ss_1, ani_infile_ss_2, ani_out_ss)
ani_table_ss <-
    fread(ani_out_ss) %>%
    mutate(V1 = sub(".txt", "", basename(V1))) %>%
    mutate(V2 = sub(".txt", "", basename(V2))) %>%
    `colnames<-`(c("rid", "qid", "ani", "matched", "total")) %>%
    mutate(ani = as.numeric(ani)) %>%
    as_tibble %>%
    mutate(
        ani_dist = 100-ani
    ) %>%
    rowwise() %>%
    #filter(ref !=qry) %>%
    mutate(
        rq = paste(sort(c(ref, qry)), collapse = ",")
        ) %>%
    #filter(rq == "1.5,CP074215.1") %>%
    ungroup# %>%
    # group_by(rq) %>%
    # summarise(
    #     ref = ref[1],
    #     qry = qry[1],
    #     mean_dist = mean(ani_dist),
    #     dist_diff = abs(diff(ani_dist)),
    #     mean_ani = mean(ani),
    #     ani_diff = abs(diff(ani))
    # ) #%>% 
    #dplyr::select(-rq)

bie5_ss <- 
    bie4_ss %>%
    left_join(
        .,
        (ani_table_ss %>% dplyr::select(ani, ani_dist, rq)),
        by = "rq"
    ) %>%
    
    mutate(
        pclus = as.factor(dbs3_ss$cluster)
        ) %>%
    group_by(pclus) %>%
    filter(qid != "CP093081.1") %>%
    mutate(
        e2 = round(mean(event)),
        e3 = as.factor(e2)
        )%>% 
            ungroup
        # mutate(
        #     eventm = factor(round(mean(event)), levels = sort((round(mean(event)))))
        # )
    #summarise(ani_dist = mean(ani_dist))

dbs3_ss <- 
    fpc::dbscan(bie5_ss$event, eps = 1, MinPts = 1, scale = F, method = "raw")

list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP032393.1") %>% plot_delta
list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP038591.1") %>% plot_delta

filtered_combined_delta_ss %>%
    filter(qid=="CP093081.1")

ggplot(
    bie5_ss,
    aes(x = e3, y = ani_dist)
) +
    geom_boxplot(width=0.1) +
    geom_dotplot(binaxis='y', stackdir='center', dotsize=0.1) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      ) 
   
fcd
 ggplot(bie5_ss, aes(x = event, y = ani_dist)) +
      #geom_violin()+ 
     
      #geom_boxplot(width=0.1) +
      geom_dotplot(binaxis='y', stackdir='center', dotsize=0.3) +
      theme_classic() +
      theme(
        axis.text = element_text(size=16),
        axis.title = element_text(size=24)
      ) 


id_sld_wide <- 
            ani_table %>%
            dplyr::select(c(ref, qry, ani)) %>%
            spread(., qry, ani) %>%
            as.data.frame() %>%
            column_to_rownames("ref")
        
        id_sld_wide_dist <- 
            100 - id_sld_wide
        
        
        distance_matrix <- 
            id_sld_wide_dist
        
        id_dists <- 
            get_dists(distance_matrix) %>%
            arrange(Sequence_mean_distance)
        
        
        id_sld_dists <-
            get_dists(id_sld_wide_dist) %>%
            arrange(Sequence_mean_distance)
        
        id_dist_lt <- 
            as.dist(distance_matrix)
        
        seq_order <- 
            colnames(distance_matrix)[
                hclust( id_dist_lt, method = "ward.D" )$order
            ]


ggplot(
    data = ani_table, 
    aes(
        x= factor(ref, seq_order),
        y= factor(qry, seq_order),
        fill = ani#_sld#identity
    )
) + 
geom_tile() + 
theme_classic() +
theme(
    axis.text.x = element_text(angle=45, hjust = 1),
    panel.background=element_rect(fill="white"),
    plot.title = element_text(size=20,face="bold", hjust = 0.5),
    axis.title = element_blank()
) + 
ggtitle(
    sprintf(
        "%s \n Sequence pairwise MASH similarity", 
        sub("_", " ", species_name)
    )
) + 
theme(
    legend.title=element_text( size=18),
    #axis.text=element_blank(),
       # element_text(size=14), 
    legend.text = element_text(size=18)
) + 
scale_fill_viridis(
    option = "B", , limits = c(95,100)
) 

    


```


##############3
### major inversion breakpoint clustering
also get the bp_seqs
Using the filtered delta
```{r}
minimum_alignment_percentage <- 
    90
minimum_inversion_length <- 
    5e4
filtered_combined_delta <- 
    filter_combined_delta(
        sprintf("%s/all", delta_dir), 
        minimum_alignment_percentage, 
        minimum_inversion_length
    )
saveRDS(
    filtered_combined_delta,
    sprintf(
        "%s/filtered_combined_delta.rds",
        r_vars_dir
    )
)
filtered_combined_delta <-
    readRDS(
        sprintf(
            "%s/filtered_combined_delta.rds",
            r_vars_dir
        )
    )

#delta_file <- 
#    list.files(sprintf("%s/ref_v_all", delta_dir_200), full.names = T)[1]
clustered_genomes_list$cluster_coords
clustered_genomes_list <- 
    readRDS(
        sprintf(
           "%s/clustered_genomes_list.rds",
            r_vars_dir
        )
    )

clustered_genomes_list <- 
    cluster_genomes(filtered_combined_delta)

clustered_genomes <- 
   clustered_genomes_list$cluster_summary %>%
    left_join(
        ., 
        species_summary[,c("accession", "len")] %>% 
            `colnames<-`(c("accessions", "len")),
        by = "accessions"
        )
saveRDS(
    clustered_genomes_list, 
    sprintf(
        "%s/clustered_genomes_list.rds",
        r_vars_dir
    )
    )

clustered_genomes_list <- 
    readRDS(
        sprintf(
           "%s/clustered_genomes_list.rds",
            r_vars_dir
           )
    )
###
all_bps <-
    filtered_combined_delta %>%
    filter( ## not the first or last alignment, 
        rs !=min(rs), re !=max(re), qs !=1
        ) %>%
    filter( # inverted strand or positive strand 
        (lag(strand, default = "+") != strand) & 
            (lead(strand, default = "+") != strand) 
    ) %>%
    dplyr::select(rid, qid, rs, re, qs, qe, strand, meanlen) %>%
    mutate(midpoint = round((rs+re)/2), meanlen = round(meanlen)) %>%
    ungroup() %>% 
    arrange(rs) %>%
    left_join(
        .,
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("cluster", "qid"))
    )

all_bps %>% filter(qid=="NZ_CP117290.1")

filtered_combined_delta

delta_identities <- 
    pbmclapply(
        1:length((list.files(new_out_deltadir))),
        function(i){
            delta_tib <- 
                read_delta(list.files(new_out_deltadir, full.names = T)[i]) 
            
            nucmer_id <- 
                sum(
                    delta_tib$meanlen
                    )/
                (mean(
                    c(
                        delta_tib$rlen[1], 
                        delta_tib$qlen[1])
                     ))
            c
            out_tib <- 
                tibble(
                    accession = 
                        delta_tib$qid[1],
                    identity = 
                        nucmer_id
                )
            
            return(out_tib)
            
        }, mc.cores = 4
    ) %>% bind_rows()

best_identity <- 
    left_join(
        delta_identities %>% `colnames<-`(c("accessions", "identity")),
        clustered_genomes,
        by = "accessions"
    )

best_identity %>% filter(cluster_id==0) %>% arrange(identity)
best_identity %>% filter(cluster_id==0) %>% arrange(identity) %>% dplyr::slice(35:45)

list.files(new_out_deltadir, pattern = "CP073335.1", full.names = T) %>% plot_delta()

fcd2 %>% filter(qid=="NZ_CP075174.1")
## Anova for all others

best_identity %>% filter(cluster_id!=0) %>%
    arrange(cluster_id) %>%
    count(cluster_id)

bis <- 
    best_identity %>% 
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>% 
    filter(!is.na(cluster_id)) %>%
    dplyr::slice(1)

## table of the most common
bis_tib

mpis

bis_1 <- 
    bis[-1,]

bis %>% filter(cluster_id %in% c(2,4))

list.files(sync_dir) %>% head

file.copy(
    from = sprintf("%s/%s.txt", sync_dir, bis$accessions[1]),
    to = sprintf("../../Ken/Cluster_0_2_4/%s.txt", bis$accessions[1])
)

file.copy(
    from = sprintf("%s/%s.txt", sync_dir, bis$accessions[3]),
    to = sprintf("../../Ken/Cluster_0_2_4/%s.txt", bis$accessions[3])
)

file.copy(
    from = sprintf("%s/%s.txt", sync_dir, bis$accessions[5]),
    to = sprintf("../../Ken/Cluster_0_2_4/%s.txt", bis$accessions[5])
)


mpis <- 
    lapply(
        1:5,
        function(j){
            dfj <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==pull(bis_1, accessions)[j]) 
            if (j==5){
                pj <- 
                    plot_delta(dfj)  +
                    xlab("Cluster 0")
            } else {
                pj <- 
                    plot_delta(dfj)  +
                    xlab("")
            }
            pjo <- 
                pj +
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 22)
                ) +
                ylab(sprintf("Cluster %s", j)) 
            return(pjo)
        }
    )


do.call(
     "grid.arrange",
     c(
         mpis,
         ncol=1
         )
     )
```

Capturing all events
```{r}


all_events <- 
    list()
i=10

events <- 
    list()
#events[[10]]
which(bis_1$accessions=="NZ_CP110931.1")
list.files(new_out_deltadir, full.names = T, pattern = "CP092307.1") %>%
    plot_delta


id_table <- 
    bis_1
combined_delta <- 
    filtered_combined_delta



tloc_seg <- 
    lapply(
        1:length(bis_1),
        function(i){
            qi <- 
                pull(id_table, accessions)[i]
            fdi <- 
                combined_delta %>% 
                ungroup %>%
                filter(qid==qi)  %>%
                    filter( ## not the first or last alignment, 
                        segment != min(segment),
                        segment != max(segment)
                        )
            
            fdi_tloc <- 
                fdi %>%
                filter(X_dist>1e6)
            
            return(fdi_tloc)
        }
    )

filtered_combined_delta$rid %>% unique

list.files(new_out_deltadir, pattern = "_1.3_", full.names = T) %>%
    read_delta %>%
    plot_delta

filtered_combined_delta %>%
    ungroup %>%
    filter(qid == "1.3") %>% 
    plot_delta


tloc_tib <- 
    filtered_combined_delta %>%
    #filter(rs !=1, re !=max(re)) %>%
    ungroup %>%
    mutate(
        offset = 
            ifelse(
                strand == "-",
                ifelse( # if the midpoint is higher than the middle
                    ((rs+re)/2)>((rlen/2)),
                    "right",
                    ifelse(
                        ((rs+re)/2)==((rlen/2)),
                        "center",
                        "left"
                    )
                ),
                ifelse(
                    ((rs+re)/2) > ((qs+qe)/2),
                    "right", 
                    ifelse(
                        (((rs+re)/2) ==((qs+qe)/2)),
                        "center",
                        "left"
                    )
                )
            ),
        sync_check = 
            (rs ==1 ) & (X_dist >3e6 )
    ) %>%
    left_join(
        ., 
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("qclust", "qid")),
        by = "qid"
    ) %>%
    filter(
        !sync_check
    )

## Segemnts between the first and last alignments
event_potential <- 
    tloc_tib %>%
    group_by(qid) %>%
    filter(
        rs !=1, 
        qs !=1,
        re != max(re),
        ) %>%
    filter(X_dist <5e5) %>%
    mutate(
        nested = 
            (
                (lag(strand, default = "+")=="-") &
                    (lead(strand, default = "+")=="-") &
                    strand=="+"
            ) |
            
            ,
        outer = 
            lag(nested, default = F)
    )


 tloc_tib %>%
    group_by(qid) %>%
    filter(
        rs !=1, 
        qs !=1,
        re != max(re),
        ) %>%
     filter(X_dist <5e5)


tloc_tib %>% filter(X_dist > 1e6 | ) %>% plot_delta

tloc_tib %>% filter(X_dist > 1e5, X_dist <2e5) %>% plot_delta

tloc_tib %>% 
    filter(X_dist > 1e5, X_dist <2e5) %>% 
    pull(X_dist) %>% density %>% plot

tloc_tib %>% filter(X_dist < 1e5) %>% pull(X_dist) %>% density %>% plot

tloc_counts 


tloc_tib$X_dist %>% density %>% plot

tloc_tib %>%
    filter(
        rs==1
    ) %>%
    arrange(desc(X_dist))

tloc_tib %>%
    mutate(
        sync_check = 
            (rs ==1 ) & (X_dist >3e6 )
    ) %>%
    filter(sync_check)
    
    filter(
        !((rs !=1 ) & (X_dist <3e6 ))
        )

ls() %>% head

head(sort( sapply(ls(),function(x){object.size(get(x))}), decreasing = T) , 12)


#rm(cm_dist_summary, cm_dist_filt, cm_dist, mash_mto, id_wide_dist, between_dists, mash_mto2_dist, id_dist_lt)
rm(
    mmo2, mash_mto2, id_wide, bie3tt, group_j, cds_h
)

tloc_rmid <- 
    filtered_combined_delta %>%
    ungroup %>%
    filter(X_dist>5e5, rs !=1) %>%
    mutate(
        rmid = round((rs+re)/2),
        qmid = round((qs+qe)/2)
        ) %>%
    dplyr::select(qid, strand, X_dist, rmid, qmid, meanlen) %>%
  pivot_longer(cols = c(rmid, qmid), names_to = "variable", values_to = "value")


tloc_ids <- 
    tloc_rmid %>%
    pull(qid) %>%
    unique()

filtered_combined_delta %>% filter(qid %in% tloc_ids[1]) %>% plot_delta 

filtered_combined_delta %>% filter(!(qid %in% tloc_ids)) %>% plot_delta

tloc_ids


ggplot(tloc_rmid, aes(x = value, fill = variable, color = variable)) +
  geom_density(alpha = 0.5) +
  geom_jitter(aes(y = 0), width = 0, height = 0.01, alpha = 0.5) +
  labs(title = "Density Plot of col1 and col2 with Data Points",
       x = "Value",
       y = "Density") +
  theme_minimal() +
  scale_fill_manual(values = c("col1" = "orangered", "col2" = "lightskyblue")) +
  scale_color_manual(values = c("col1" = "orangered", "col2" = "lightskyblue"))

ggplot(tloc_rmid, aes(x = variable, fill = variable, y= value)) + 
    geom_violin(trim=FALSE)+
  labs(title = "translocations",
       x = "Value",
       y = "Density") +
  theme_classic()





bind_rows(tloc_seg)

## coordinates of inner and outer inversions
inversion_capture <- 
    function(
        id_table,
        combined_delta
    ){
      inv_list <- 
          list()
      
      for (i in 1:nrow(id_table)){
            qi <- 
                pull(id_table, accessions)[i]
            # the delta file filtered
            fdi <- 
                combiend_delta %>% 
                ungroup %>%
                filter(qid==qi) 
            plot_delta(fdi)
            
            fd2 <- 
                fdi %>%
                filter( ## not the first or last alignment, 
                    segment != min(segment),
                    segment != max(segment)
                    )
            inv_row <- 
                fdi %>%
                filter( ## not the first or last alignment, 
                    segment != min(segment),
                    segment != max(segment)
                    ) %>%
                rowwise() %>%
                mutate(
                    rsp = 
                        ((100*rs)/rlen) %>% round(.),
                    rep = 
                        ((100*re)/rlen) %>% round(.),
                    qsp = 
                        ((100*qs)/qlen) %>% round(.),
                    qep = 
                        ((100*qe)/qlen) %>% round(.),
                    rmp = round((rs+re)/2),
                    qmp = round((qs+qe)/2),
                    r_mid = round(100*rmp/rlen),
                    q_mid = round(100*qmp/qlen),
                    meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                    rmprop = round(100*rmp/rlen),
                    qmprop = round(100*qmp/qlen),
                    midpoint = mean(rmprop, qmprop),
                    segment = as.character(segment),
                     ms =
                            ifelse(
                                strand=="+",
                                mean(c(rsp, qsp)),
                                mean(c(rsp, qep))
                            ),
                        me = 
                            ifelse(
                                strand=="+",
                                mean(c(rep, qep)),
                                mean(c(rep, qsp))
                            )
                ) %>%
                dplyr::select(
                    qid, rsp, rep, qsp, qep, 
                    strand, segment,
                    X_dist, meansize, 
                    midpoint, ms, me,
                    rmprop, qmprop
                    ) %>%
                ungroup()
            
            if (nrow(inv_row)==1){
                inv_list[[i]] <- 
                    inv_row
            } else {
                
                
                
                
            }
        
    }

      
      
      
      
      
inner_outer <- 
    function(
        delta_table
    ){
         fd3 <- 
                    fdi %>%
                    filter( 
                        segment != min(segment),
                        segment != max(segment)
                        )%>%
                    filter(X_dist<1e6) %>%
                    group_by(segment) %>%
                    summarise(
                        strand = unique(strand),
                        rs = min(rs), 
                        re = max(re),
                        qs = ifelse(strand=="+", min(qs), max(qs)),
                        qe = ifelse(strand=="+", max(qe), min(qe)),
                        X_dist = round(mean(X_dist)),
                        rlen = unique(rlen),
                        qlen = unique(qlen),
                        meanlen = round(sum(meanlen)),
                        qid = unique(qid)
                        )
                
    }
      
      

event_capture <- 
    funciton(
        id_table, combined_delta
    ){
        
        events <- 
            list()
        for (i in 1:nrow(id_table)){
            qi <- 
                pull(id_table, accessions)[i]
            # the delta file filtered
            fdi <- 
                combiend_delta %>% 
                ungroup %>%
                filter(qid==qi) 
            plot_delta(fdi)
            
            fd2 <- 
                fdi %>%
                filter( ## not the first or last alignment, 
                    segment != min(segment),
                    segment != max(segment)
                    )%>%
                filter(X_dist<1e6)
        
            if (nrow(fd2)==0){
                events[[i]] <- 
                    fdi %>%
                    filter( ## not the first or last alignment, 
                        # rs !=min(rs), re !=max(re), 
                        # qs !=1
                        segment != min(segment),
                        segment != max(segment)
                        ) %>%
                    rowwise() %>%
                    mutate(
                        rsp = 
                            ((100*rs)/rlen) %>% round(.),
                        rep = 
                            ((100*re)/rlen) %>% round(.),
                        qsp = 
                            ((100*qs)/qlen) %>% round(.),
                        qep = 
                            ((100*qe)/qlen) %>% round(.),
                        rmp = round((rs+re)/2),
                        qmp = round((qs+qe)/2),
                        r_mid = round(100*rmp/rlen),
                        q_mid = round(100*qmp/qlen),
                        meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                        rmprop = round(100*rmp/rlen),
                        qmprop = round(100*qmp/qlen),
                        midpoint = mean(rmprop, qmprop),
                        segment = as.character(segment),
                         ms =
                                ifelse(
                                    strand=="+",
                                    mean(c(rsp, qsp)),
                                    mean(c(rsp, qep))
                                ),
                            me = 
                                ifelse(
                                    strand=="+",
                                    mean(c(rep, qep)),
                                    mean(c(rep, qsp))
                                )
                    ) %>%
                    dplyr::select(
                        qid,# rsp, rep, qsp, qep, 
                        strand, segment,
                        X_dist, meansize, 
                        midpoint, ms, me
                        #r_midprop, q_midprop
                        ) %>%
                    ungroup()
                    
            } else{
                fd3 <- 
                    fdi %>%
                    filter( 
                        segment != min(segment),
                        segment != max(segment)
                        )%>%
                    filter(X_dist<1e6) %>%
                    group_by(segment) %>%
                    summarise(
                        strand = unique(strand),
                        rs = min(rs), 
                        re = max(re),
                        qs = ifelse(strand=="+", min(qs), max(qs)),
                        qe = ifelse(strand=="+", max(qe), min(qe)),
                        X_dist = round(mean(X_dist)),
                        rlen = unique(rlen),
                        qlen = unique(qlen),
                        meanlen = round(sum(meanlen)),
                        qid = unique(qid)
                        )
                
                segdiff <- 
                    fd3 %>%
                    mutate(segdiff = segment-lag(segment, default = 1)) %>%
                    pull(segdiff)
                
                if (any(segdiff>1)) next
                
                fd2_props <- 
                    fd3 %>%
                    rowwise() %>%
                    mutate(
                        rsp = 
                            ((100*rs)/rlen) %>% round(.),
                        rep = 
                            ((100*re)/rlen) %>% round(.),
                        qsp = 
                            ((100*qs)/qlen) %>% round(.),
                        qep = 
                            ((100*qe)/qlen) %>% round(.),
                        rmp = round((rs+re)/2),
                        qmp = round((qs+qe)/2),
                        rmprop = round(100*rmp/rlen),
                        qmprop = round(100*qmp/qlen),
                        meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                        midpoint = mean(rmprop, qmprop)
                    ) %>%
                    dplyr::select(
                        qid, rsp, rep, qsp, qep, strand, segment,
                        X_dist, meansize, midpoint
                        ) %>%
                    ungroup() %>%
                    filter(
                        strand=="-" |
                            (
                                (lag(strand, default = "+")=="-") & 
                                    (lead(strand, default = "+")=="-")
                            )
                    )
                
                inner_seg <- 
                    median(unique(fd2_props$segment))
                
                if (nrow(fd2_props)==1){
                    middle_seg_row <- 
                        fd2_props %>%
                        filter(
                            segment==median(segment)
                        ) %>%
                        mutate(
                            ms =
                                ifelse(
                                    strand=="+",
                                    mean(c(rsp, qsp)),
                                    mean(c(rsp, qep))
                                ),
                            me = 
                                ifelse(
                                    strand=="+",
                                    mean(c(rep, qep)),
                                    mean(c(rep, qsp))
                                ),
                            segment = as.character(segment)
                            ) %>%
                        dplyr::select(
                            qid, strand, segment, X_dist, 
                            meansize, midpoint, ms, me
                        )
                    events[[i]] <- 
                        middle_seg_row
                } else {
                    
                    middle_seg_row <- 
                        fd2_props %>%
                        filter(
                            segment==median(segment)
                        ) %>%
                        mutate(
                            ms =
                                ifelse(
                                    strand=="+",
                                    mean(c(rsp, qsp)),
                                    mean(c(rsp, qep))
                                ),
                            me = 
                                ifelse(
                                    strand=="+",
                                    mean(c(rep, qep)),
                                    mean(c(rep, qsp))
                                ),
                            segment = as.character(segment)
                            ) %>%
                        dplyr::select(
                            qid, strand, segment, X_dist, 
                            meansize, midpoint, ms, me
                        )
                    outer_segs <- 
                        fd2_props %>%
                        filter(segment != median(segment)) 
                        
                    possible_outers <- 
                        nrow(outer_segs) %/% 2
                    outer_events <- 
                        list()
                    for (j in 1:possible_outers){
                        outer_events[[j]] <- 
                            fd2_props %>% 
                            filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                            summarise(
                                qid = unique(qid),
                                strand= unique(strand),
                                segment = paste(segment, collapse = ","),
                                X_dist = mean(X_dist),
                                ms =
                                    ifelse(
                                        strand=="+",
                                        mean(c(min(rsp), min(qsp))),
                                        mean(c(min(rsp), min(qep)))
                                    ),
                                me =
                                    ifelse(
                                        strand=="+",
                                        mean(c(max(rep), max(qep))),
                                        mean(c(max(rep), max(qsp)))
                                    ),
                                meansize = 
                                    me-ms,
                                midpoint = 
                                    (ms+me)/2
                            )
                    }
                    inv_summary <- 
                        bind_rows(
                            middle_seg_row,
                            bind_rows(outer_events)
                        )
                    events[[i]] <- 
                        inv_summary
                }
                
            }
            #plot_delta(fdi)
            
            print(i)
            
        }
        
        evd <- 
            bind_rows(events) %>% arrange(meansize) 
        
        ### all inversions included
        evd2 <- 
            left_join(
                evd,
                (id_table %>% dplyr::rename(qid = accessions)),
                by = "qid"
            ) %>% 
            arrange(meansize) %>%
            add_count(qid, name = "i2")# %>%
            #filter(inversions==i2)
        
         # left_join(
         #        evd,
         #        (id_table %>% dplyr::rename(qid = accessions)),
         #        by = "qid"
         #    ) %>% 
         #    arrange(meansize) %>%
         #    add_count(qid, name = "i2") %>%
         #    filter(inversions!=i2)
         
         
         ## shared events
         
         
         
        
    }




###########
for (i in 1:nrow(bis_1)){
####
    qi <- 
        pull(bis_1, accessions)[i]
    # the delta file filtered
    fdi <- 
        filtered_combined_delta %>% 
        ungroup %>%
        filter(qid==qi) 
    # plot_delta(fdi)
    fd2 <- 
        fdi %>%
        filter( ## not the first or last alignment, 
            # rs !=min(rs), re !=max(re), 
            # qs !=1
            segment != min(segment),
            segment != max(segment)
            )%>%
        filter(X_dist<1e6)
    
    if (nrow(fd2)==0){
        events[[i]] <- 
            fdi %>%
            filter( ## not the first or last alignment, 
                # rs !=min(rs), re !=max(re), 
                # qs !=1
                segment != min(segment),
                segment != max(segment)
                ) %>%
            rowwise() %>%
            mutate(
                rsp = 
                    ((100*rs)/rlen) %>% round(.),
                rep = 
                    ((100*re)/rlen) %>% round(.),
                qsp = 
                    ((100*qs)/qlen) %>% round(.),
                qep = 
                    ((100*qe)/qlen) %>% round(.),
                rmp = round((rs+re)/2),
                qmp = round((qs+qe)/2),
                r_midprop = round(100*rmp/rlen),
                q_midprop = round(100*qmp/qlen),
                meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                rmprop = round(100*rmp/rlen),
                qmprop = round(100*qmp/qlen),
                midpoint = mean(rmprop, qmprop),
                segment = as.character(segment),
                 ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        )
            ) %>%
            dplyr::select(
                qid,# rsp, rep, qsp, qep, 
                strand, segment,
                X_dist, meansize, 
                midpoint, ms, me
                #r_midprop, q_midprop
                ) %>%
            ungroup()
            
    } else{
        fd3 <- 
            fdi %>%
            filter( ## not the first or last alignment, 
                # rs !=min(rs), re !=max(re), 
                # qs !=1
                segment != min(segment),
                segment != max(segment)
                )%>%
            filter(X_dist<1e6) %>%
            group_by(segment) %>%
            summarise(
                strand = unique(strand),
                rs = min(rs), 
                re = max(re),
                qs = ifelse(strand=="+", min(qs), max(qs)),
                qe = ifelse(strand=="+", max(qe), min(qe)),
                X_dist = round(mean(X_dist)),
                rlen = unique(rlen),
                qlen = unique(qlen),
                meanlen = round(sum(meanlen)),
                qid = unique(qid)
                )
        
        segdiff <- 
            fd3 %>%
            mutate(segdiff = segment-lag(segment, default = 1)) %>%
            pull(segdiff)
        
        if (any(segdiff>1)) next
        
        fd2_props <- 
            fd3 %>%
            rowwise() %>%
            mutate(
                rsp = 
                    ((100*rs)/rlen) %>% round(.),
                rep = 
                    ((100*re)/rlen) %>% round(.),
                qsp = 
                    ((100*qs)/qlen) %>% round(.),
                qep = 
                    ((100*qe)/qlen) %>% round(.),
                rmp = round((rs+re)/2),
                qmp = round((qs+qe)/2),
                rmprop = round(100*rmp/rlen),
                qmprop = round(100*qmp/qlen),
                meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                midpoint = mean(rmprop, qmprop)
            ) %>%
            dplyr::select(
                qid, rsp, rep, qsp, qep, strand, segment,
                X_dist, meansize, midpoint
                ) %>%
            ungroup() %>%
            filter(
                strand=="-" |
                    (
                        (lag(strand, default = "+")=="-") & 
                            (lead(strand, default = "+")=="-")
                    )
            )
        
        inner_seg <- 
            median(unique(fd2_props$segment))
        
        if (nrow(fd2_props)==1){
            middle_seg_row <- 
                fd2_props %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        ),
                    segment = as.character(segment)
                    ) %>%
                dplyr::select(
                    qid, strand, segment, X_dist, meansize, midpoint, ms, me
                )
            events[[i]] <- 
                middle_seg_row
        } else {
            
            middle_seg_row <- 
                fd2_props %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        ),
                    segment = as.character(segment)
                    ) %>%
                dplyr::select(
                    qid, strand, segment, X_dist, meansize, midpoint, ms, me
                )
            outer_segs <- 
                fd2_props %>%
                filter(segment != median(segment)) 
                
            possible_outers <- 
                nrow(outer_segs) %/% 2
            outer_events <- 
                list()
            for (j in 1:possible_outers){
                outer_events[[j]] <- 
                    fd2_props %>% 
                    filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                    summarise(
                        qid = unique(qid),
                        strand= unique(strand),
                        segment = paste(segment, collapse = ","),
                        X_dist = mean(X_dist),
                        ms =
                            ifelse(
                                strand=="+",
                                mean(c(min(rsp), min(qsp))),
                                mean(c(min(rsp), min(qep)))
                            ),
                        me =
                            ifelse(
                                strand=="+",
                                mean(c(max(rep), max(qep))),
                                mean(c(max(rep), max(qsp)))
                            ),
                        meansize = 
                            me-ms,
                        midpoint = 
                            (ms+me)/2
                    )
            }
            inv_summary <- 
                bind_rows(
                    middle_seg_row,
                    bind_rows(outer_events)
                )
            events[[i]] <- 
                inv_summary
        }
        
    }
    #plot_delta(fdi)
    
    print(i)
    
}




bis_1 %>%
    filter(inversions==1)



events %>% head
bind_rows(events) %>% filter(!is.na(rsp))
evd <- 
    bind_rows(events) %>% arrange(meansize)


sei <- 
    evd %>%
    filter(
        qid %in%
            (bis_1 %>%filter(inversions==1) %>% pull(accessions))
            )


eff <- 
    evd %>% 
    #add_count(midpoint, meansize) %>% 
    arrange(meansize, midpoint) %>%
    left_join(
        ., 
        (bis_1 %>% dplyr::rename(qid = accessions)),
          by = "qid"
    ) %>%
    filter(grouped_sequences>1)
    ##filter(n>1 | grouped_sequences>1)

bie4
miniset <- 
    eff %>%
    filter(meansize<=15, meansize>10) %>%
    pull(qid)

mini_tib <- 
    clustered_mash %>% 
                    filter(
                        (ref %in% miniset) &
                            (qry %in% miniset) 
                        # (ref_clust %in% 0:6) &
                        #     (qry_clust %in% 0:6)
                        )

eff %>%
    filter(meansize<=15, meansize>10) %>%
    filter(cluster_id %in% c(11,19, 77))

mini_tib %>% 
    filter(
        ref_clust %in% 11, 19, 77
    )


ggplot(
                data = 
                    mini_tib, 
                aes(
                    x= factor(ref, seq_order),
                    y= factor(qry, seq_order),
                    fill = identity#_sld#identity
                )
            ) + 
            geom_tile() + 
            theme_classic() +
            theme(
                #axis.text.x = element_text(angle=45, hjust = 1),
                panel.background=element_rect(fill="white"),
                plot.title = element_text(size=20,face="bold", hjust = 0.5),
                axis.title = element_blank(),
                axis.text = element_blank()
            ) + 
            ggtitle(
                sprintf(
                    "%s \n Sequence pairwise MASH similarity", 
                    sub("_", " ", species_name)
                )
            ) + 
            theme(
                legend.title=element_text( size=18),
                axis.text=element_blank(),
                    #element_text(size=14), 
                legend.text = element_text(size=18)
            ) + 
            scale_fill_viridis(
                option = "B", limits = c(95,100)
            ) +
      facet_grid(
        rows = vars(qry_clust), cols = vars(ref_clust), 
        scales = "free", space = "free", switch = "y")

bie2 <- 
    left_join(
        e2,
        (bis_1 %>% dplyr::rename(qid = accessions)),
          by = "qid"
    ) %>% 
    arrange(meansize) %>%
    add_count(qid, name = "i2") %>%
    filter(inversions==i2)


for (i in 1:nrow(bie2)){
    i=1
    
    qi <- 
        bie2$qid[i]
    
    qni <- 
        bie2$qid[-i]
    
    q_events <- 
        e2 %>% 
        filter(qid==qi)
    
    
    
}


list.files(new_out_deltadir, pattern = "NZ_CP067078.1", full.names = T) %>%
    plot_delta()



left_join(
        e2,
        (bis_1 %>% dplyr::rename(qid = accessions)),
          by = "qid"
    ) %>% 
    arrange(meansize) %>%
    add_count(qid, name = "i2") %>%
    filter(inversions!=i2)

fcd2 %>% filter(qid =="NZ_CP117293.1") %>% plot_delta

NZ_CP100670.1

list.files(new_out_deltadir, pattern = "NZ_CP067078.1", full.names = T) %>%
    plot_delta()
list.files(new_out_deltadir, pattern = "NZ_CP117290.1", full.names = T) %>%
    plot_delta()
fcd2 %>% filter(qid =="NZ_CP100670.1") %>% plot_delta
fcd2 %>% filter(qid =="CP082704.1") %>% plot_delta


bie2

ani_dir <- 
    sprintf(
        "%s/ani/ac1_ss",
        identity_dir
        )

if (!dir.exists(ani_dir)){dir.create(ani_dir)}

all_connected <- 
    bie2$qid
    # c(
    #     bie4$qid,
    #     bie4$rid
    # ) %>% unique


best_identity %>%
    filter(cluster_id!=0) %>%
    group_by(cluster_id) %>%
    arrange((identity)) %>%
    dplyr::slice(1:3) %>%
    ungroup %>%
    arrange(cluster_id)

list.files(
        sprintf(
            "%s/all",
            delta_dir
            ),
        pattern = "NZ_CP038604.1",
        full.names = T
        ) %>% plot_delta


all_separated <- 
    best_identity %>%
    filter(cluster_id!=0) %>%
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>%
    top_n(identity, n=15) %>%
    ungroup %>%
    arrange(cluster_id) 



fcd2 %>% filter(qid =="NZ_CP084532.1") %>% plot_delta

lapply(ani_inlines, file.exists) %>% unlist %>% all

in_file <- 
    ani_inlines
out_dir <- 
    ani_dir

ani_inlines <- 
    sprintf(
        "%s/%s.txt", 
        sync_dir, 
        all_separated$accessions
    )

ani_infile <- 
    sprintf(
        "%s/ac.txt",
        ani_dir
    )
writeLines(
    ani_inlines,
    ani_infile
)

ani_out <- 
    sprintf(
        "%s/ac_15_out.txt",
        out_dir  
    )

ani_commands <- 
    function(
        in_file,
        out_file
    ){
        ani_com <- 
            sprintf(
                "%s --ql %s --rl %s -o %s",
                "fastANI -t 7",
                in_file, in_file,
                out_file
            )
        system(ani_com)
    }

ani_commands(ani_infile, ani_out)
ani_table <-
    fread(ani_out) %>%
    mutate(V1 = sub(".txt", "", basename(V1))) %>%
    mutate(V2 = sub(".txt", "", basename(V2))) %>%
    `colnames<-`(c("ref", "qry", "ani", "matched", "total")) %>%
    mutate(ani = as.numeric(ani)) %>%
    as_tibble %>%
    mutate(
        ani_dist = 100-ani
    ) %>%
    rowwise() %>%
    #filter(ref !=qry) %>%
    mutate(
        rq = paste(sort(c(ref, qry)), collapse = ",")
        ) %>%
    #filter(rq == "1.5,CP074215.1") %>%
    ungroup# %>%
    # group_by(rq) %>%
    # summarise(
    #     ref = ref[1],
    #     qry = qry[1],
    #     mean_dist = mean(ani_dist),
    #     dist_diff = abs(diff(ani_dist)),
    #     mean_ani = mean(ani),
    #     ani_diff = abs(diff(ani))
    # ) #%>% 
    #dplyr::select(-rq)


id_sld_wide <- 
            ani_table %>%
            dplyr::select(c(ref, qry, ani)) %>%
            spread(., qry, ani) %>%
            as.data.frame() %>%
            column_to_rownames("ref")
        
        id_sld_wide_dist <- 
            100 - id_sld_wide
        
        
        distance_matrix <- 
            id_sld_wide_dist
        
        id_dists <- 
            get_dists(distance_matrix) %>%
            arrange(Sequence_mean_distance)
        
        
        id_sld_dists <-
            get_dists(id_sld_wide_dist) %>%
            arrange(Sequence_mean_distance)
        
        id_dist_lt <- 
            as.dist(distance_matrix)
        
        seq_order <- 
            colnames(distance_matrix)[
                hclust( id_dist_lt, method = "ward.D" )$order
            ]


ggplot(
    data = ani_table, 
    aes(
        x= factor(ref, seq_order),
        y= factor(qry, seq_order),
        fill = ani#_sld#identity
    )
) + 
geom_tile() + 
theme_classic() +
theme(
    axis.text.x = element_text(angle=45, hjust = 1),
    panel.background=element_rect(fill="white"),
    plot.title = element_text(size=20,face="bold", hjust = 0.5),
    axis.title = element_blank()
) + 
ggtitle(
    sprintf(
        "%s \n Sequence pairwise MASH similarity", 
        sub("_", " ", species_name)
    )
) + 
theme(
    legend.title=element_text( size=18),
    #axis.text=element_blank(),
       # element_text(size=14), 
    legend.text = element_text(size=18)
) + 
scale_fill_viridis(
    option = "B", , limits = c(95,100)
) 


bie2 %>% add_count(cluster_id) %>% filter(n>1)

bie2


bie3a <- 
    pbmclapply(
        1:length(unique(bie2$qid)),
        function(i){
            #for(i in 1:length(unique(bie2$qid))){
            qi <- 
                unique(bie2$qid)[i]
            q_events <- 
                bie2 %>%
                filter(qid==qi) %>% 
                dplyr::select(qid, meansize) %>%
                mutate(
                    mlow = 
                        meansize-1,
                    mhigh = 
                        meansize+1
                )
            ##
            qi_id <- 
                ani_table %>%
                filter(
                    (ref==qi) |
                        (qry==qi),
                    ref != qry
                ) %>%
                rowwise() %>%
                mutate(
                    q2  = 
                        c(ref, qry)[c(ref, qry)!=qi]
                ) %>%
                ungroup %>%
                group_by(q2) %>%
                filter(ani_dist==max(ani_dist)) %>%
                ungroup %>%
                mutate(r2 = qi) %>%
                dplyr::select(r2, q2, ani_dist)
            
            #which(bie2$qid=="NZ_LR590082.1")
            #fcd2 %>% filter(qid==qi) %>% plot_delta()
            #bie2 %>% filter()
            
            shared_events <- 
                bie2 %>%#[which(bie2$qid=="NZ_LR590082.1"),] %>%
                mutate(
                    shared_event = 
                        sapply(
                            # bie2[which(bie2$qid=="NZ_LR590082.1"),] %>%
                            #     pull(meansize),
                            bie2$meansize, 
                            function(value){
                              any(
                                  (value >= q_events$mlow) & 
                                      (value <=  q_events$mhigh)
                                  )  
                                } 
                            )
                ) %>%
                filter(
                    qid != qi,
                    shared_event
                )

            #shared_events %>% filter(qid == qi)
            #which(shared_events$cluster_id==47)
            
            event_tib_i <- 
                lapply(
                    unique(shared_events$qid),
                    function(x){
                        # x <-
                        #  unique(shared_events$qid[2])
                        #for (x in unique(shared_events$qid)){
                        x_shared <- 
                            shared_events %>%
                            filter(qid==x) %>%
                            dplyr::select(-c("shared_event"))
                        x_events <- 
                            bie2 %>%
                            filter(qid==x)
                        
                        xdiff <- 
                            bind_rows(x_shared, x_events) %>% 
                            dplyr::add_count(meansize, name = "occ") %>%
                            filter(occ==1)
                   
                        qdiff <- 
                            q_events %>%
                            mutate(
                                qd = 
                                    lapply(
                                        q_events$meansize,
                                        function(qe){
                                            unlist(
                                                lapply(
                                                    x_events$meansize,
                                                    function(xe){
                                                        (qe >= xe-1) &
                                                            (qe <=  xe+1)
                                                    }
                                                )
                                            ) %>% any
                                        }
                                    ) %>% unlist()
                            ) %>%
                            filter(
                                !qd
                            ) 
                            
                        out_tib <- 
                            tibble(
                                qid = x,
                                se = nrow(x_shared),
                                q_events = nrow(q_events),
                                x_events = nrow(x_events)
                            ) %>%
                            mutate(
                                des = abs(se-q_events),
                                adj_event = 
                                    ifelse(
                                        length(xdiff$meansize)==1,
                                        xdiff$meansize,
                                        ifelse(
                                            length(qdiff$meansize)==1,
                                            qdiff$meansize,
                                            NA
                                        )
                                    ),
                                shared_event = 
                                    paste(x_shared$meansize, collapse = ","),
                                ref_unique_events = 
                                    paste(xdiff$meansize, collapse = ","),
                                qry_unique_events = 
                                    paste(qdiff$meansize, collapse = ",")
                            )
                        
                        return(out_tib)
                    }
                ) %>% 
                bind_rows()
            #shared_events %>%
             #   group_by(qid)
            same_cluster <- 
                event_tib_i %>%
                filter(des==0)
            adjacent_cluster <- 
                 event_tib_i %>%
                 filter(des==1) 
            
            if (nrow(event_tib_i)>0){
                out_tib <- 
                    event_tib_i %>%
                    filter(
                        des %in% c(0,1)
                    ) %>%
                    mutate(
                        rid = qi,
                        rclust = 
                            bie2$cluster_id[i]
                    ) %>%
                    left_join(
                        ., 
                        bie2[,c("qid", "cluster_id")] %>%
                            `colnames<-`(c("qid", "qclust")),
                        by = "qid"
                    ) %>%
                    left_join(
                        ., 
                        qi_id[,c(2,3)] %>%
                            `colnames<-`(c("qid", "id_dist")),
                        by = "qid"
                    )
                    
                ## shared id tble
                sid_tib <- 
                    ani_table %>%
                    filter(
                        ((ref==qi) & (qry== "NZ_CP120396.1")) |
                            ((qry==qi) & (ref== "NZ_CP120396.1")) 
                        )
                    
            } else{
                out_tib <- 
                    NULL
                    
            }
            return(out_tib)
        }
        )


bie2

clustered_genomes %>% filter(accessions=="NZ_CP117293.1")
bind_rows(bie3a) %>% filter(shared_event %in% c(21, 22))

b5a <- 
    bind_rows(bie3a) %>%
    distinct() %>%
    filter(
        !grepl(adj_event, pattern = ","),
        !is.na(adj_event),
        #str_detect(shared_event, "(^|\\D)18"),
        q_events != x_events,
        rclust !=qclust
        ) %>%
    # filter(
    #     str_detect(shared_event, "(^|\\D)11") |
    #         str_detect(shared_event, "(^|\\D)11.5"),
    #     des ==0
    # ) %>%
    filter(
        str_detect(shared_event, "(^|\\D)11") |
            str_detect(shared_event, "(^|\\D)11.5") |
            str_detect(shared_event, "(^|\\D)18"),
        rclust != 75,# qclust !=123, 
        qclust != 75, #rclust !=123, 
        # rclust != 108, qclust !=109, 
        # qclust != 108, rclust !=109, 
        des ==0
    ) %>%
    rowwise() %>%
         mutate(
            rq = paste(sort(c(rid, qid)), collapse = ",")
            ) %>%
    ungroup %>%
    distinct(rq, .keep_all = T) %>%
    add_count(adj_event) %>%
    arrange(adj_event) %>%
    mutate(
        shared_cluster = 
            ifelse(
                str_detect(shared_event, "(^|\\D)11") |
                    str_detect(shared_event, "(^|\\D)11.5"),
                "11%",
                "18%"
            ),
            comparison = "Within"
    )

b5ff <- 
    ani_table %>%
    filter(
        ((ref %in% b5a$qid) & (!(qry %in% b5a$qid)))
    )

plot(density(ani_table$ani_dist))
    
    



b5as <- 
    b5a %>%
    dplyr::select(rid, qid, adj_event, rclust, qclust) %>%
    bind_rows(b6a)
b6a <- 
    bie2 %>%
    filter(
        inversions==1,
        # (qid %in% unique(bie4$qid)) |
        #     (qid %in% unique(bie4$qid)) ,
        meansize %in% c(4,11, 18)
        ) %>%
    mutate(
        rclust = 0,
        qclust = cluster_id,
        adj_event = (meansize),
        rid = bis$accessions[1]
    ) %>%
    group_by(meansize) %>%
    filter(
        cluster_id==min(cluster_id),
        rclust != 75
        ) %>%
    ungroup %>%
    # filter(
    #     (rclust %in% c(0,1,11,12)) &
    #         (qclust %in% c(0,1,11,12))
    #     ) %>%
    dplyr::select(c(qid, rid, adj_event, rclust, qclust)) 
    

fcd2 %>% filter(cluster==13) %>% plot_delta()
fcd2 %>% filter(cluster==7) %>% plot_delta()

edge_list2 <- 
     graph_from_data_frame(
         d =
             data.frame(
                 source = b5as %>% pull(rclust),
                 target = b5as %>% pull(qclust)
                 ),
         directed = F
     )
 E(edge_list2)$label <- 
     b5as$adj_event
 label.size <- 2
 V(edge_list2)$label.cex <- 18
 plot(edge_list2, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)
 bie4 %>%
     filter(
         (qid %in% tt2) | (rid %in% tt2),
         !((rclust %in% ffs) | (qclust %in% ffs)  )
     ) %>% bind_rows(bie5)



b4a <- 
    bind_rows(bie3a) %>%
    distinct() %>%
    group_by(shared_event) %>%
    mutate(
        ses = n()
    ) %>% 
    # summarise(
    #     ses = unique(ses),
    #     mean_id = mean(id_dist)
    # ) %>% 
    arrange(desc(ses)) %>%
    filter(!grepl(shared_event, pattern = ",")) %>%
    ungroup %>%
    mutate(
        idx = row_number()
    ) %>%
    filter(shared_event %in% c(4,11, 18))

ggplot(b4a, aes(idx, y=shared_event, height = id_dist, group = shared_event)) + 
  geom_density_ridges(stat = "identity", scale = 1)

b5a %>% count(shared_cluster)

 ggplot(
    b5a, 
    aes(x=id_dist, y = shared_cluster) 
    )+  
      geom_density_ridges(jittered_points = TRUE, scale = .95, rel_min_height = .01,
        point_shape = "|", point_size = 3, size = 0.25,
        position = position_points_jitter(height = 0), 
        )  +
      scale_x_continuous(limits =c(0,5)) +
     theme(text = element_text(size = 28)) +
     xlab("% distance")

 b4a %>% filter(shared_event==11, id_dist >1)
 
 
b4a %>%
    count(shared_event, sort = T)

fcd2 %>% filter(qid=="NZ_CP117290.1") %>% plot_delta()

```



#####################



```{r}
which(bis_1 =="NZ_CP126178.1")

sd1b <- 
    sprintf("%s/subset1b2", sync_dir)

if (!dir.exists(sd1b)){dir.create(sd1b)}
id1b <- 
    sprintf("%s/subset1b", identity_dir)

if (!dir.exists(id1b)){dir.create(id1b)}

first_15_b <- 
    best_identity %>% 
    #filter(cluster_id %in% c(0:5)) %>%
    arrange(cluster_id, desc(identity)) %>%
    ungroup() %>%
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>%
    dplyr::slice(1:15) 
    


lapply(
    first_15_b$accessions %>% unique,
    function(x){
        file_in <- 
            list.files(sync_dir, pattern = paste0(x, ".txt"), full.names = T)
        file_out <- 
            sprintf(
                "%s",
                sd1b, basename(file_in)
            )
        file.copy(
            file_in, file_out
        )
    }
) %>% unlist

list.files(sync_dir)

list.files(sd1)

sd1_ssb <- 
    species_summary %>%
    filter(accession %in% first_15_b$accessions)

d1_id_b <- 
    get_pairwise_identities(
        sd1b, 
        id1b, sd1_ssb, 
        genus_name, species_name
        )
d1_id_clusterb <-
    inner_join(
        d1_id_b$mash_tib %>% 
            mutate(
                ref = as.character(ref),
                qry = as.character(qry)
            ),
        clustered_genomes %>%
            dplyr::select(cluster_id, accessions) %>%
            `colnames<-`(c("ref_clust", "ref")),
        by = "ref"
        ) %>%
        inner_join(
            ., 
            clustered_genomes %>%
                dplyr::select(cluster_id, accessions) %>%
                `colnames<-`(c("qry_clust", "qry")),
            by = "qry"
        )
d1_id_cluster$identity %>% sort %>% head()
ggplot(
    data = 
        d1_id_clusterb, 
        aes(
            x = ref,
            y = qry,
            fill = identity#_sld#identity_sld
        )
    ) + 
    geom_tile() + 
    theme_classic() +
    theme(
        axis.text.x = element_text(angle=45, hjust = 1),
        panel.background=element_rect(fill="white"),
        plot.title = element_text(size=20,face="bold", hjust = 0.5),
        axis.title = element_blank()
    ) + 
    ggtitle(
        sprintf(
            "%s \n Sequence pairwise MASH similarity", 
            sub("_", " ", species_name)
        )
    ) + 
    theme(
        legend.title=element_text( size=18),
        #axis.text=element_blank(),
            #element_text(size=14), 
        legend.text = element_text(size=18),
        strip.text = element_text(size = 20)
    ) + 
    scale_fill_viridis(
        option = "B", limits = c(95,100)
    )+
    facet_grid(
        rows = vars(qry_clust), cols = vars(ref_clust), 
        scales = "free", space = "free", switch = "y")+ 
    guides(fill=guide_legend(title="MASH similarity"))
        
    


tt <- 
    lapply(
        1:10,
        function(i){
            qi <- 
                pull(bis_1, accessions)[i]
            # the delta file filtered 
            fdi <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==qi) 
            #plot_delta(fdi)
            fd2 <- 
                fdi %>%
                filter( ## not the first or last alignment, 
                    # rs !=min(rs), re !=max(re), 
                    # qs !=1
                    segment != min(segment),
                    segment != max(segment)
                    )%>%
                filter(X_dist<1e6) %>%
                group_by(segment) %>%
                summarise(
                    strand = unique(strand),
                    rs = min(rs), 
                    re = max(re),
                    qs = ifelse(strand=="+", min(qs), max(qs)),
                    qe = ifelse(strand=="+", max(qe), min(qe)),
                    X_dist = round(mean(X_dist)),
                    rlen = unique(rlen),
                    qlen = unique(qlen),
                    meanlen = round(sum(meanlen)),
                    qid = unique(qid)
                    )
                
            fd2_props <- 
                fd2 %>%
                rowwise() %>%
                mutate(
                    rsp = 
                        ((100*rs)/rlen) %>% round(.),
                    rep = 
                        ((100*re)/rlen) %>% round(.),
                    qsp = 
                        ((100*qs)/qlen) %>% round(.),
                    qep = 
                        ((100*qe)/qlen) %>% round(.),
                    rmp = round((rs+re)/2),
                    qmp = round((qs+qe)/2),
                    rmprop = round(100*rmp/rlen),
                    qmprop = round(100*qmp/qlen),
                    meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                    midpoint = mean(rmprop, qmprop)
                ) %>%
                dplyr::select(
                    qid, rsp, rep, qsp, qep, strand, segment,
                    X_dist, meansize, midpoint
                    ) %>%
                ungroup()
            
            inner_seg <- 
                median(fd2_props$segment)
            
            if (nrow(fd2_props)==1){
                middle_seg_row <- 
                    fd2_props %>%
                    filter(
                        segment==median(segment)
                    ) %>%
                    mutate(
                        ms =
                            ifelse(
                                strand=="+",
                                mean(c(rsp, qsp)),
                                mean(c(rsp, qep))
                            ),
                        me = 
                            ifelse(
                                strand=="+",
                                mean(c(rep, qep)),
                                mean(c(rep, qsp))
                            ),
                        segment = as.character(segment)
                        ) %>%
                    dplyr::select(
                        qid, strand, segment, X_dist, meansize, midpoint, ms, me
                    )
                events <- 
                    middle_seg_row
            } else {
                
                middle_seg_row <- 
                    fd2_props %>%
                    filter(
                        segment==median(segment)
                    ) %>%
                    mutate(
                        ms =
                            ifelse(
                                strand=="+",
                                mean(c(rsp, qsp)),
                                mean(c(rsp, qep))
                            ),
                        me = 
                            ifelse(
                                strand=="+",
                                mean(c(rep, qep)),
                                mean(c(rep, qsp))
                            ),
                        segment = as.character(segment)
                        ) %>%
                    dplyr::select(
                        qid, strand, segment, X_dist, meansize, midpoint, ms, me
                    )
                outer_segs <- 
                    fd2_props %>%
                    filter(segment != median(segment)) 
                    
                possible_outers <- 
                    nrow(outer_segs) %/% 2
                outer_events <- 
                    list()
                for (j in possible_outers){
                    outer_events[[j]] <- 
                        fd2_props %>% 
                        filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                        summarise(
                            qid = unique(qid),
                            strand= unique(strand),
                            segment = paste(segment, collapse = ","),
                            X_dist = mean(X_dist),
                            ms =
                                ifelse(
                                    strand=="+",
                                    mean(c(min(rsp), min(qep))),
                                    mean(c(min(rsp), min(qsp)))
                                ),
                            me =
                                ifelse(
                                    strand=="+",
                                    mean(c(max(rep), max(qep))),
                                    mean(c(max(rep), max(qsp)))
                                ),
                            meansize = 
                                me-ms,
                            midpoint = 
                                (ms+me)/2
                        )
                }
                inv_summary <- 
                    bind_rows(
                        middle_seg_row,
                        bind_rows(outer_events)
                    )
                events <- 
                    inv_summary
            }
            return(events)
        }
        
    )

############ Capturing each event
events <- 
    list()
i=10
which(bis_1$accessions=="NZ_CP019181.1")
for (i in 1:nrow(bis_1)){
####
    qi <- 
        pull(bis_1, accessions)[i]
    # the delta file filtered
    fdi <- 
        filtered_combined_delta %>% 
        ungroup %>%
        filter(qid==qi) 
    # plot_delta(fdi)
    fd2 <- 
        fdi %>%
        filter( ## not the first or last alignment, 
            # rs !=min(rs), re !=max(re), 
            # qs !=1
            segment != min(segment),
            segment != max(segment)
            )%>%
        filter(X_dist<1e6)
    
    if (nrow(fd2)==0){
        events[[i]] <- 
            fdi %>%
            filter( ## not the first or last alignment, 
                # rs !=min(rs), re !=max(re), 
                # qs !=1
                segment != min(segment),
                segment != max(segment)
                ) %>%
            rowwise() %>%
            mutate(
                rsp = 
                    ((100*rs)/rlen) %>% round(.),
                rep = 
                    ((100*re)/rlen) %>% round(.),
                qsp = 
                    ((100*qs)/qlen) %>% round(.),
                qep = 
                    ((100*qe)/qlen) %>% round(.),
                rmp = round((rs+re)/2),
                qmp = round((qs+qe)/2),
                rmprop = round(100*rmp/rlen),
                qmprop = round(100*qmp/qlen),
                meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                midpoint = mean(rmprop, qmprop),
                segment = as.character(segment)
            ) %>%
            dplyr::select(
                qid,# rsp, rep, qsp, qep, 
                strand, segment,
                X_dist, meansize, midpoint
                ) %>%
            ungroup()
            
    } else{
        fd3 <- 
            fdi %>%
            filter( ## not the first or last alignment, 
                # rs !=min(rs), re !=max(re), 
                # qs !=1
                segment != min(segment),
                segment != max(segment)
                )%>%
            filter(X_dist<1e6) %>%
            group_by(segment) %>%
            summarise(
                strand = unique(strand),
                rs = min(rs), 
                re = max(re),
                qs = ifelse(strand=="+", min(qs), max(qs)),
                qe = ifelse(strand=="+", max(qe), min(qe)),
                X_dist = round(mean(X_dist)),
                rlen = unique(rlen),
                qlen = unique(qlen),
                meanlen = round(sum(meanlen)),
                qid = unique(qid)
                )
        
        segdiff <- 
            fd3 %>%
            mutate(segdiff = segment-lag(segment, default = 1)) %>%
            pull(segdiff)
        
        if (any(segdiff>1)) next
        
        fd2_props <- 
            fd3 %>%
            rowwise() %>%
            mutate(
                rsp = 
                    ((100*rs)/rlen) %>% round(.),
                rep = 
                    ((100*re)/rlen) %>% round(.),
                qsp = 
                    ((100*qs)/qlen) %>% round(.),
                qep = 
                    ((100*qe)/qlen) %>% round(.),
                rmp = round((rs+re)/2),
                qmp = round((qs+qe)/2),
                rmprop = round(100*rmp/rlen),
                qmprop = round(100*qmp/qlen),
                meansize = round((100*meanlen)/mean(c(rlen, qlen))),
                midpoint = mean(rmprop, qmprop)
            ) %>%
            dplyr::select(
                qid, rsp, rep, qsp, qep, strand, segment,
                X_dist, meansize, midpoint
                ) %>%
            ungroup() %>%
            filter(
                strand=="-" |
                    (
                        (lag(strand, default = "+")=="-") & 
                            (lead(strand, default = "+")=="-")
                    )
            )
        
        inner_seg <- 
            median(unique(fd2_props$segment))
        
        if (nrow(fd2_props)==1){
            middle_seg_row <- 
                fd2_props %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        ),
                    segment = as.character(segment)
                    ) %>%
                dplyr::select(
                    qid, strand, segment, X_dist, meansize, midpoint, ms, me
                )
            events[[i]] <- 
                middle_seg_row
        } else {
            
            middle_seg_row <- 
                fd2_props %>%
                filter(
                    segment==median(segment)
                ) %>%
                mutate(
                    ms =
                        ifelse(
                            strand=="+",
                            mean(c(rsp, qsp)),
                            mean(c(rsp, qep))
                        ),
                    me = 
                        ifelse(
                            strand=="+",
                            mean(c(rep, qep)),
                            mean(c(rep, qsp))
                        ),
                    segment = as.character(segment)
                    ) %>%
                dplyr::select(
                    qid, strand, segment, X_dist, meansize, midpoint, ms, me
                )
            outer_segs <- 
                fd2_props %>%
                filter(segment != median(segment)) 
                
            possible_outers <- 
                nrow(outer_segs) %/% 2
            outer_events <- 
                list()
            for (j in 1:possible_outers){
                outer_events[[j]] <- 
                    fd2_props %>% 
                    filter(segment %in% c(inner_seg-j, inner_seg+j)) %>%
                    summarise(
                        qid = unique(qid),
                        strand= unique(strand),
                        segment = paste(segment, collapse = ","),
                        X_dist = mean(X_dist),
                        ms =
                            ifelse(
                                strand=="+",
                                mean(c(min(rsp), min(qsp))),
                                mean(c(min(rsp), min(qep)))
                            ),
                        me =
                            ifelse(
                                strand=="+",
                                mean(c(max(rep), max(qep))),
                                mean(c(max(rep), max(qsp)))
                            ),
                        meansize = 
                            me-ms,
                        midpoint = 
                            (ms+me)/2
                    )
            }
            inv_summary <- 
                bind_rows(
                    middle_seg_row,
                    bind_rows(outer_events)
                )
            events[[i]] <- 
                inv_summary
        }
        
    }
    #plot_delta(fdi)
    
    print(i)
    
}
events %>% head
bind_rows(events) %>% filter(!is.na(rsp))
evd <- 
    bind_rows(events) %>% arrange(meansize) 

### all inversions included
evd2 <- 
    left_join(
        evd,
        (bis_1 %>% dplyr::rename(qid = accessions)),
        by = "qid"
    ) %>% 
    arrange(meansize) %>%
    add_count(qid, name = "i2") %>%
    filter(inversions==i2)

 bind_rows(events) %>% 
     arrange(meansize) %>%
     left_join(
        .,
        (bis_1 %>% dplyr::rename(qid = accessions)),
        by = "qid"
    )
    


plot(sort(evd$midpoint))

event_tib_1 <-
    bind_rows(events)

bis_1

e2 <- 
    evd %>%
    filter(midpoint>40 & midpoint<60)


e3 <- 
    e2 %>% 
    group_by(meansize) %>%
    mutate(qqs = n()) %>%
    filter(
        qqs>1
    ) %>%
    group_by(qid) %>%
    summarise(events = paste(meansize, collapse = ","))


e2 %>% 
    group_by(meansize) %>%
    mutate(qqs = n()) %>%
    filter(
        qqs>1
    )

bie2 <- 
    left_join(
        e2,
        (bis_1 %>% dplyr::rename(qid = accessions)),
        by = "qid"
    ) %>% 
    arrange(meansize) %>%
    add_count(qid, name = "i2") %>%
    filter(inversions==i2)

bie2 %>% count(meansize)



clustered_genomes %>%
    filter(cluster_id==41)
list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP117386.1") %>%
    plot_delta()

main_events_2 <-
    bie2 %>%
    filter(midpoint>45, midpoint <55) %>%
    dplyr::select(
        qid, meansize, midpoint, cluster_id, 
        grouped_sequences
    ) %>%
    add_count(meansize, name = "meansize_n") %>%
    group_by(
        meansize
        ) %>%
    summarise(
        clust_list = paste(cluster_id, collapse = ","),
        seqs = sum(grouped_sequences),
        #midpoints = paste(midpoint, collapse = ","),
        clusts = length(unique(cluster_id))
    ) %>%
    filter(
        #seqs>5, 
        clusts>1
        ) %>%
  mutate(
      g2 = cumsum(c(1, diff(meansize) >= 1.5))
      ) %>%
    group_by(g2) %>%
    summarise(
        meansize = 
            round(mean(meansize)),
        clust_list =
            paste(clust_list, collapse = ","),
        seqs = sum(seqs),
        clusts = sum(clusts)
    )

main_events$seqs %>% sum
species_summary %>% nrow()
(bis %>% pull(grouped_sequences))[-1] %>% sum
(bis %>% pull(grouped_sequences))[-1] %>% sum

bie2

main_clusts <- 
    lapply(
        1:nrow(main_events),
        function(i){
            clusts <- 
                (main_events %>% pull(clust_list))[i] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric()
            # seqs <- 
            #     bis %>% 
            #     ungroup %>%
            #     filter(
            #         cluster_id %in% clusts
            #     ) %>%
            #     pull(grouped_sequences) %>% 
            #     sum()
            return(clusts)
        }
    ) %>% unlist() %>%
    unique()


clusts_4 <- 
    (main_events %>% pull(clust_list))[1] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric()
clusts_12 <- 
    (main_events %>% pull(clust_list))[2] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric()
clusts_18 <- 
    (main_events %>% pull(clust_list))[3] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric()
clusts_52 <- 
    (main_events %>% pull(clust_list))[4] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric()
clusts_83 <- 
    (main_events %>% pull(clust_list))[5] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric()


clusts_4_plots <- 
    lapply(
        clusts_4,
        function(i){
            qi <- 
                bis %>% filter(cluster_id==i) %>%
                pull(accessions)
            p1 <- 
                list.files(new_out_deltadir, full.names = T, pattern = qi) %>% 
                plot_delta() +
                theme(
                    plot.title = element_text(size=32, hjust = 0.49),
                    axis.text = element_text(size=16),
                    axis.text.y = element_blank()
                    ) +
                ggtitle(i)
            return(p1)
        }
    )

do.call(
     "grid.arrange",
     c(
         clusts_4_plots,
         ncol=2
         )
     )


clusts_12_plots <- 
    lapply(
        sort(clusts_12),
        function(i){
            #for (i in sort(clusts_12)){
            qi <- 
                bis %>% filter(cluster_id==i) %>%
                pull(accessions)
            p1 <- 
                list.files(new_out_deltadir, full.names = T, pattern = qi)[1] %>% 
                plot_delta() +
                theme(
                    plot.title = element_text(size=32, hjust = 0.49),
                    axis.text = element_text(size=16),
                    axis.text.y = element_blank()
                    ) +
                ggtitle(i)
            return(p1)
        }
    )

do.call(
     "grid.arrange",
     c(
         clusts_12_plots,
         ncol=3
         )
     )

clusts_18_plots <- 
    lapply(
        sort(clusts_18),
        function(i){
            #for (i in sort(clusts_12)){
            qi <- 
                bis %>% filter(cluster_id==i) %>%
                pull(accessions)
            p1 <- 
                list.files(new_out_deltadir, full.names = T, pattern = qi)[1] %>% 
                plot_delta() +
                theme(
                    plot.title = element_text(size=32, hjust = 0.49),
                    axis.text = element_text(size=16),
                    axis.text.y = element_blank()
                    ) +
                ggtitle(i)
            return(p1)
        }
    )

do.call(
     "grid.arrange",
     c(
         clusts_52_plots,
         ncol=3
         )
     )

clusts_52_plots <- 
    lapply(
        sort(clusts_52),
        function(i){
            #for (i in sort(clusts_12)){
            qi <- 
                bis %>% filter(cluster_id==i) %>%
                pull(accessions)
            p1 <- 
                list.files(new_out_deltadir, full.names = T, pattern = qi)[1] %>% 
                plot_delta() +
                theme(
                    plot.title = element_text(size=32, hjust = 0.49),
                    axis.text = element_text(size=16),
                    axis.text.y = element_blank()
                    ) +
                ggtitle(i)
            return(p1)
        }
    )

do.call(
     "grid.arrange",
     c(
         clusts_52_plots,
         ncol=3
         )
     )


clusts_83_plots <- 
    lapply(
        sort(clusts_83),
        function(i){
            #for (i in sort(clusts_12)){
            qi <- 
                bis %>% filter(cluster_id==i) %>%
                pull(accessions)
            p1 <- 
                list.files(new_out_deltadir, full.names = T, pattern = qi)[1] %>% 
                plot_delta() +
                theme(
                    plot.title = element_text(size=32, hjust = 0.49),
                    axis.text = element_text(size=16),
                    axis.text.y = element_blank()
                    ) +
                ggtitle(i)
            return(p1)
        }
    )

do.call(
     "grid.arrange",
     c(
         clusts_83_plots,
         ncol=3
         )
     )



mc1 <- 
    lapply(
        1:nrow(main_events),
        function(i){
            clusts <- 
                (main_events %>% pull(clust_list))[i] %>%
                strsplit(., split = ",") %>%
                unlist() %>% 
                as.numeric() %>% sort
            return(clusts)
        }
    )

simple_nucmer(
    ref = 
        sprintf(
            "%s/NZ_CP117290.1.txt",
            sync_dir
        ),
    qry = 
        sprintf(
            "%s/NZ_LT905063.1.txt",
            sync_dir
        ),
    output = 
        "../data/processing/extra/c3_v_c13"
) %>% system()

main_invs <- 
    list()

## 4% inversion
main_invs[[1]] <- 
    read_delta("../data/processing/extra/c3_v_c13_filtered.delta") %>%
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16)
        ) +
    xlab("") +
    ylab("") +
    ggtitle("4%")
## 12 %
main_invs[[2]] <- 
    list.files(new_out_deltadir, full.names = T, pattern = "NZ_LT905063.1") %>% 
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16),
        axis.text.y = element_blank()
        ) +
    xlab("") +
    ylab("") +
    ggtitle("12%")
## 18 %
main_invs[[3]] <- 
    list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP045956.1") %>% 
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16),
        axis.text.y = element_blank()
        ) +
    xlab("") +
    ylab("") +
    ggtitle("18%")
## 52 %
main_invs[[4]] <- 
    list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP103791.1") %>% 
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16),
        axis.text.y = element_blank()
        ) +
    xlab("") +
    ylab("") +
    ggtitle("53%")
## 83 %
main_invs[[5]] <- 
    list.files(new_out_deltadir, full.names = T, pattern = "NC_021820.1") %>% 
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16),
        axis.text.y = element_blank()
        ) +
    xlab("") +
    ylab("") +
    ggtitle("83%")

do.call(
     "grid.arrange",
     c(
         main_invs,
         ncol=length(main_invs)
         )
     )



main_events
bis %>% filter(cluster_id==6)
bis %>% filter()

main_clusts_count <- 
    bis %>% 
    ungroup %>%
    filter(
        cluster_id %in% main_clusts
    ) %>%
    pull(grouped_sequences) %>%
    sum()

522/(bis[-1,] %>% pull(grouped_sequences) %>% sum)
    
main_events_2 <- 
    main_events %>%  
    dplyr::select(c(-"seqs")) %>%
    mutate(
        seqs = seq_sums
    )

## feature table

main_clusts <-
    main_events %>%
    pull(clust_list) %>%
    paste(., collapse = ",") %>%
    strsplit(., split = ",") %>%
    unlist() %>%
    unique %>%
    as.numeric() %>%
    sort()

main_event_sizes <- 
    main_events$meansize

event_matrix <- 
    matrix(
        nrow = length(main_event_sizes),
        ncol = length(main_clusts)
    )
for (i in 1:length(main_event_sizes)){
    for(j in 1:length(main_clusts)){
        event_matrix[i,j] <-
            as.numeric(main_clusts[j] %in% clust_unlist[[i]])
            
    }
}


bis %>%
    filter(
        
    )


event_df <- 
    t(event_matrix) %>%
    `colnames<-`(c(main_event_sizes))

clust_unlist <- 
    lapply(
        main_events$clust_list,
        function(x){
            sort(as.numeric(unique(strsplit(x, split = ",")[[1]])))
        }
    )


cluster_combo %>% `colnames<-`(c("x", "y")) %>% 
    as.data.frame() %>% mutate(idx = row_number()) %>% filter(x==5, y==121)

ft <- 
    tibble(
        clusters = 
            main_clusts
    ) %>%
    bind_cols(
        as.data.frame(event_df)
    )

cluster_combo <- 
    t(combn(ft$clusters, 2))

cluster_sharediff <- 
    lapply(
        1:nrow(cluster_combo),
        function(i){
            c1_row <- 
                ft %>% 
                filter(
                    clusters ==
                        cluster_combo[i,1]
                    )
            
            c2_row <- 
                ft %>% 
                filter(
                    clusters ==
                        cluster_combo[i,2]
                    )
            
            c1c2 <- 
                bind_rows(
                    c1_row,
                    c2_row
                )
            different_rows <- 
                data.frame(as.list(colSums(c1c2[,-1]))) %>%
                `colnames<-`(
                    c(
                        paste("different", colnames(ft[,-1]), sep = "_")
                        )
                    ) %>%
                mutate(across(everything(), ~ ifelse(. > 1, 0, .)))
            shared_rows <- 
                data.frame(as.list(as.numeric(colSums(c1c2[,-1])==2))) %>%
                `colnames<-`(
                    c(
                        paste("shared", colnames(ft[,-1]), sep = "_")
                        )
                    )
            colcount <- 
                (ncol(different_rows)*2) +2
            final_tib <-
                bind_cols(
                    shared_rows, different_rows
                ) %>%
                mutate(
                    ref_clust = cluster_combo[i,1],
                    qry_clust = cluster_combo[i,2]
                ) %>%
                dplyr::select(
                    c(colcount-1, colcount, 1:(colcount-2))
                )
            
            return(final_tib)
        }
    ) %>%
    bind_rows()

cluster_sharediff %>% as_tibble

clusters_mto_dist <- 
    clustered_mash %>% 
    mutate(mash_dist = 100-identity) %>%
    dplyr::select(ref, qry, ref_clust, qry_clust, mash_dist) %>%
    filter(ref_clust !=0, qry_clust  !=0) %>%
    inner_join(
        ., 
        cluster_sharediff, 
        by = c("ref_clust", "qry_clust")
    )


csd_1 <- 
    clusters_mto_dist %>%
    filter(
        ref_clust %in% c(1,14,15,26,28, 29, 34, 71, 97, 98, 99, 106, 108, 109, 111),
        qry_clust %in% c(1,14,15,26,28, 29, 34, 71, 97, 98, 99, 106, 108, 109, 111)
    )

igg2 <- 
    csd_1 %>%
    filter(qry_clust %in%c(14, 26, 28, 97, 98, 99, 108, 109)) %>%
    group_by(ref_clust, qry_clust) %>%
    summarise(
        md = mean(mash_dist)
    )
csd_1 %>% filter(qry_clust %in%c(14,15,29,111, 26, 28, 97, 98, 99, 108, 109))%>%
    group_by(ref_clust, qry_clust) %>%
    summarise(
        md = mean(mash_dist)
    )


csd_1 %>% filter(qry_clust %in%c(14,15,29,111, 26, 28, 97, 98, 99, 108, 109))%>%
    group_by(ref_clust, qry_clust) %>%
    summarise(
        md = mean(mash_dist)
    )

csd_1 %>% 
    filter(
        ref_clust == 99,
        qry_clust ==106
        )%>%
    group_by(ref_clust, qry_clust) %>%
    summarise(
        md = mean(mash_dist)
    )

clustered_mash %>% 
    filter(
        ref_clust == 0,
        qry_clust ==1
        )%>%
    mutate(
        mash_dist = 100-identity
    ) %>%
    group_by(ref_clust, qry_clust) %>%
    summarise(
        md = mean(mash_dist)
    )

bistib3 <- 
    tibble(
        c1 = 
            c(
                "0", 
                "1", "1", "1", "1",
                "14", "14",
                "26", "26", 
                "28", "28",
                "13", "13", "13",
                "2",
                "4",
                "7",
                "8", "8", "8",
                "9"
                ),
        c2 = 
            c(
                "1",
                "26", "28", "97", "99"
                
                "15", "29",
                "15", "29",
                "15", "29",
                "2", "4", "7",
                "8",
                "8",
                "9",
                "30", "31", "32",
                "32"
            )
    )
    
    filter(
    ref_clust ==1, qry_clust==14
)



clusters_mto_dist %>% 
    dplyr::select(ref, qry, shared_18, different_18, mash_dist) %>%
    filter(
        (shared_18==1) |
            (different_18==1)
        )


 list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP103791.1") %>% 
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16),
        axis.text.y = element_blank()
        ) 
  list.files(new_out_deltadir, full.names = T, pattern = "NC_004631.1") %>% 
    plot_delta() +
    theme(
        plot.title = element_text(size=40, hjust = 0.49),
        axis.text = element_text(size=16),
        axis.text.y = element_blank()
        ) 


clusters_mto_dist %>%
    filter(shared_4==1) %>%
    filter()



install.packages("randomForest")
install.packages("pdp")
library(randomForest)
library(pdp)
rf_model_shared <- 
    randomForest(
        mash_dist ~ shared_4 + shared_12 + shared_18 + shared_52 + shared_83, 
        data = clusters_mto_dist, importance = TRUE
        )

lm_model <- 
    lm(
        mash_dist ~ 
            shared_4 + shared_12 + shared_18 + shared_52 + shared_83 +
            different_4 + different_12 + 
            different_18 + different_52 + different_83, 
        data = clusters_mto_dist
        )

lmcoef <- 
    summary(lm_model)$coefficients

coef_df <- as.data.frame(lmcoef)
coef_df$Characteristic <- rownames(coef_df)


rownames(coef_df) <- NULL

cdf <- 
    coef_df %>%
    dplyr::slice(-1) %>%
    separate(Characteristic, sep = "_", into = c("Presence", "Inversion")) %>%
    mutate(
        Inversion = factor(Inversion, levels = c("4", "12", "18", "52", "83")),
        Presence = factor(Presence, levels = c("shared", "different"))
    )

# View the coefficients
print(coef_df)
ggplot(
    cdf, 
    aes(x = Inversion, y = Estimate, fill = Presence)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "", x = "Inversion % length", y = "Coefficient") +
    theme_classic() +
    theme(
        text = element_text(size = 38)
    )


clusters_mto_dist %>% 
    filter(
        different_83>0,
        shared_4==0,
        shared_12==0, shared_18==0, shared_52==0, shared_83==0
        )


rf_model_diff <- 
    randomForest(
        mash_dist ~ different_4 + different_12 + 
            different_18 + different_52 + different_83, 
        data = clusters_mto_dist, importance = TRUE
        )
coefficients <- summary(lm_model)$coefficients
importance_shared <- 
    importance(rf_model_shared)
importance_diff <- 
    importance(rf_model_diff)


shared_df <-
    data.frame(
        Inversion = rownames(importance_shared), 
        Importance = importance_shared[, 1]
        ) %>% as_tibble %>%
    separate(Inversion, sep = "_", into = c("Presence", "Inversion"))

diff_df <- 
    data.frame(
        Inversion = rownames(importance_diff), 
        Importance = importance_diff[, 1]
        ) %>% as_tibble %>%
    separate(Inversion, sep = "_", into = c("Presence", "Inversion"))

Importance_df <- 
    bind_rows(
        shared_df, diff_df
    ) %>%
    mutate(
        Inversion = factor(Inversion, levels = c("4", "12", "18", "52", "83")),
        PresenCe = factor(Presence)
    )
    
    

partialPlot(x=rf_model_shared, pred.data=clusters_mto_dist, x.var=shared_4, which.class="TRUE")


pdp_shared_1 <- 
    partial(
        rf_model_shared, 
        pred.var = "shared_4", plot = TRUE, plot.engine = "ggplot2"
        )


ggplot(
    Importance_df, 
    aes(x = Inversion, y = Importance, fill = Presence)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Feature Importance", x = "Characteristics", y = "Importance")




ft_shared <- 
    matrix(
        nrow(
            
        )
    )
    

bie4
    




me2 <- 
    tibble(
        clusts = 
            main_events %>%
            pull(clust_list) %>%
            paste(., collapse = ",") %>%
            strsplit(., split = ",") %>%
            unlist() 
    ) %>%
    count(clusts, sort = T, name = "appearances")
    
    
me2 %>% filter(appearances ==1)


bis_1 %>%filter(cluster_id==5)
bis_1 %>%filter(cluster_id==121)
mash_tib %>% filter(ref=="CP092294.1", qry=="NC_012125.1")

list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP038847.1") %>% plot_delta
list.files(new_out_deltadir, full.names = T, pattern = "1.5")[1] %>% plot_delta

bis_1 %>%filter(cluster_id==32)
bis_1 %>%filter(cluster_id==6)
mash_tib %>% filter(ref=="CP092294.1", qry=="NC_012125.1")

list.files(new_out_deltadir, full.names = T, pattern = "CP082726.1") %>% plot_delta
list.files(new_out_deltadir, full.names = T, pattern = "NZ_CP103791.1") %>% plot_delta

 left_join(
        e2,
        (bis_1 %>% dplyr::rename(qid = accessions)),
        by = "qid"
    ) %>% 
    arrange(meansize) %>%
    add_count(qid, name = "i2") %>%
    filter(inversions!=i2)

bie2
bie2
## event_selection
sel_1 <- 
    bie2 %>% ungroup %>% filter(inversions==1) %>%
            arrange(desc(grouped_sequences)) %>%
            dplyr::slice(1:10) %>%# arrange(meansize) %>% 
    pull(qid)
sel_2 <- 
    bie2 %>% ungroup %>% filter(inversions==2) %>%
            arrange(desc(grouped_sequences)) %>%
            dplyr::slice(1:8) %>% pull(qid) %>% unique
sel_3 <- 
    (bie2 %>% ungroup %>% filter(inversions==3) %>%
            arrange(desc(grouped_sequences)) %>%
            pull(qid) %>% unique)[1:4]
## shared events vs close events





##
single_events <- 
    lapply(
       1:length(sel_1),
        function(i){
            if (i <4){
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sel_1[i]) %>%plot_delta()+
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 19)
                    ) +
                    xlab("")
            } else{
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sel_1[i]) %>%plot_delta()+
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 19)
                    )
            }
            
        }
    )

do.call(
     "grid.arrange",
     c(
         single_events,
         ncol=2
         )
     )
double_events <- 
    lapply(
       1:length(sel_2),
        function(i){
            if (i <4){
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sel_2[i]) %>%plot_delta()+
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 19)
                    ) +
                    xlab("")
            } else{
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sel_2[i]) %>%plot_delta()+
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 19)
                    )
            }
        }
    )

triple_events <- 
    lapply(
       1:length(sel_3),
        function(i){
            if (i <4){
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sel_3[i]) %>%plot_delta()+
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 19)
                    ) +
                    xlab("")
            } else{
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sel_3[i]) %>%plot_delta()+
                theme(
                    axis.text = element_text(size = 18),
                    axis.title = element_text(size = 19)
                    )
            }
        }
    )

combined_list <- 
    list()
for (i in 1:length(single_events)) {
  combined_list[[3*i - 2]] <- single_events[[i]]
  combined_list[[3*i - 1]] <- double_events[[i]]
  combined_list[[3*i]] <- triple_events[[i]]
}

bie2 %>% filter(qid=="NZ_CP060676.1")

best_identity %>% filter(cluster_id==22)

# list.files(
#sprintf("%s/all", delta_dir), pattern = "CP101355.1", full.names = T) %>% 
#     plot_delta

do.call(
     "grid.arrange",
     c(
         combined_list,
         ncol=3
         )
     )

do.call(
     "grid.arrange",
     c(
         triple_events,
         ncol=2
         )
     )

triple_events
bie2 %>% pull(meansize) %>% unique %>% round()

bie2 %>% 
    filter(inversions ==1)

bie2$qid %>% 
    unique
sd2 <- 
    sprintf("%s/subset2", sync_dir)

if (!dir.exists(sd2)){dir.create(sd2)}
id2 <- 
    sprintf("%s/subset2", identity_dir)

if (!dir.exists(id2)){dir.create(id2)}

lapply(
    bie2$qid %>% unique,
    function(x){
        file_in <- 
            list.files(sync_dir, pattern = paste0(x, ".txt"), full.names = T)
        file_out <- 
            sprintf(
                "%s",
                sd2, basename(file_in)
            )
        file.copy(
            file_in, file_out
        )
    }
) %>% unlist
list.files(sync_dir)
list.files(sd2)

sd2_ss <- 
    species_summary %>%
    filter(accession %in% bie2$qid)

d2_id <- 
    get_pairwise_identities(
        sd2, 
        id2, sd2_ss, 
        genus_name, species_name
        )

900
d2_id$mash_tib

dbs3 <- 
    fpc::dbscan(bie2$meansize, eps = 1, MinPts = 1, scale = F, method = "raw")

bie2 %>% 
    mutate(pclus = dbs3$cluster) 


dbs3 <- 
    fpc::dbscan(bie2$meansize, eps = 1, MinPts = 1, scale = F, method = "raw")

bie2 %>% filter(qid=="CP044007.1")

b3 %>% count(meansize)

b3 <- 
    bie2 %>%
    mutate(
        pclus = dbs3$cluster
        ) %>%
    add_count(pclus, name = "pn") %>%
    filter(pn>1) %>%
    mutate(pclus2 = rleid(pclus))

b3 %>%filter(inversions==1)
### meansize 13 is closer to 18 and 19!
b3 %>% 
    filter(
        qid %in% c("NZ_CP117290.1", "NZ_CP117293.1", "NZ_CP120396.1")
        #meansize==11.5
        )

b3m <- 
    b3 %>%
    filter(
        meansize %in% c(11, 11.5, 12, 18, 19)
    ) %>%
    mutate(
        mc = 
            ifelse(
                meansize %in% c(11, 11.5, 12,13),
                "1",
                ifelse(
                    meansize==18,
                    "2", "3"
                )
            )
    ) #%>% filter(mc==1) %>%pull(qid) %>% unique

l1 <- 
    b3m %>% filter(mc==1) %>% filter(meansize==11.5) %>% pull(qid)

dps1[[1]]


filtered_combined_delta %>% 
ungroup %>%
filter(qid=="NZ_CP117290.1") %>%plot_delta()

b3 %>% filter(qid=="NZ_CP117290.1")

dps1 <- 
    lapply(
        l1,
        function(x){
            filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==x) %>%plot_delta()
        }
    )

do.call(
     "grid.arrange",
     c(
         dps1,
         ncol=3
         )
     )

sync_dir
file.copy(
    "../data/processing/sync/Pseudomonadota/Salmonella/Salmonella_enterica/NZ_CP039558.1.txt",
    "../../key_inversion/Reference_NZ_CP039558.1.txt"
)
lapply(
    l1,
    function(x){
        file_in <- 
            list.files(sync_dir, pattern = paste0(x, ".txt"), full.names = T)
        file_out <- 
            sprintf(
                "../../key_inversion/",
                sd2, basename(file_in)
            )
        file.copy(
            file_in, file_out
        )
    }
) %>% unlist

lapply(
    l1,
    function(x){
        file_in <- 
            list.files(
                sync_dir, pattern = paste0(x, ".txt"), full.names = T
                )
        file_out <- 
            sprintf(
                "../../key_inversion/three_more/%s",
                 basename(file_in)
            )
        file.copy(
            file_in, file_out
        )
    }
) %>% unlist


species_summary %>% filter(accession=="1.5") %>% View


l2 <- 
    b3m %>% filter(mc==2) %>% pull(qid)

dps2 <- 
    lapply(
        l2,
        function(x){
            filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==x) %>%plot_delta()
        }
    )

do.call(
     "grid.arrange",
     c(
         dps2,
         ncol=4
         )
     )

l3 <- 
    b3m %>% filter(mc==3) %>% pull(qid)

dps3 <- 
    lapply(
        l3,
        function(x){
            filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==x) %>%plot_delta()
        }
    )

do.call(
     "grid.arrange",
     c(
         dps3,
         ncol=3
         )
     )


b4m <- 
    d2_id$mash_tib %>%
    right_join(
        ., 
        b3m[,c("qid", "mc")] %>% 
            `colnames<-`(c("ref", "ref_clust")),
        by = "ref"
    ) %>%
    right_join(
        ., 
        b3m[,c("qid", "mc")] %>% 
            `colnames<-`(c("qry", "qry_clust")),
        by = "qry"
    ) %>%
    rowwise() %>%
    mutate(
        ref = paste0(ref_clust, "_", ref),
        qry = paste0(qry_clust, "_", qry)
    ) %>%
    group_by(ref_clust, qry_clust)
  
ggplot(
    data = 
        b4m, #%>% 
        # filter(
        #     ref %in% 
        #         (b3 %>% 
        #              #filter(pclus==3) %>% 
        #              pull(qid)),
        #     qry %in% 
        #         (b3 %>% 
        #              #filter(pclus==3) %>% 
        #              pull(qid))
        # ),
        aes(
            x = ref,
            y = qry,
            #x = factor(ref, levels = unique(b3m$qid)),
            #y = factor(qry, levels = unique(b3m$qid)),
            # x= factor(ref, levels = (b3 %>% filter(pclus==3) %>% pull(qid))),
            # y= factor(qry, levels = (b3 %>% filter(pclus==3) %>% pull(qid))),
            fill = identity#identity_sld
        )
    ) + 
    geom_tile() + 
    theme_classic() +
    theme(
        axis.text.x = element_text(angle=45, hjust = 1),
        panel.background=element_rect(fill="white"),
        plot.title = element_text(size=20,face="bold", hjust = 0.5),
        axis.title = element_blank()
    ) + 
    ggtitle(
        sprintf(
            "%s \n Sequence pairwise MASH similarity", 
            sub("_", " ", species_name)
        )
    ) + 
    theme(
        legend.title=element_text( size=18),
        #axis.text=element_blank(),
            #element_text(size=14), 
        legend.text = element_text(size=18),
        strip.text = element_text(size = 20)
    ) + 
    scale_fill_viridis(
        option = "B", limits = c(95,100)
    )+
    facet_grid(
        rows = vars(qry_clust), cols = vars(ref_clust), 
        scales = "free", space = "free", switch = "y")
    #facet_grid(rows = vars(ref_clust),  scales = "free", space = "free")




b3
####
b4 <- 
    #d2_id$mash_tib %>%
    mash_mto2_dist %>%
    left_join(
        ., 
        b3[,c("qid", "pclus2", "meansize")] %>% 
            `colnames<-`(c("ref", "ref_clust", "ref_mlen")),
        by = "ref"
    ) %>%
    left_join(
        ., 
        b3[,c("qid", "pclus2", "meansize")] %>% 
            `colnames<-`(c("qry", "qry_clust", "qry_mlen")),
        by = "qry"
    ) %>%
    arrange(ref_clust, qry_clust) %>% filter(ref != qry) %>%
    mutate(
        r2 = paste(ref, ref_clust, sep = "_"),
        q2 = paste(qry, qry_clust, sep = "_")
    )

b3 %>% filter(qid %in% "CP074294.1")
b4 %>% filter(qry %in% "CP074294.1")

bcc %>% ungroup %>% filter(is.na(within_clust))
bcc <- 
    b4 %>%
    mutate(
        r2 = qry,
        q2 = ref
    )%>%
    rowwise() %>%
    mutate(sorted_row = toString(sort(c(ref, qry)))) %>%
    distinct(sorted_row, .keep_all = T) %>%
    mutate(
        within_clust = 
            ref_clust==qry_clust,
        between_clust =
            factor(paste(ref_clust, qry_clust, sep = ","))
    ) %>%
    group_by(
        within_clust,
        between_clust
    ) %>%
    mutate(
        mean_identity = mean(identity),
        mean_identity_sld = mean(identity_sld),
        sizediff = abs(ref_mlen-qry_mlen),
        nc = n()
    ) %>%
    #filter(!is.na(meansize)) %>%
    ungroup %>%
    rowwise %>%
    mutate(
        meansize = factor((round(mean(c(ref_mlen, qry_mlen)))))
    )
bcc %>% filter(is.na(meansize))
bcc$meansize
ggplot(
    bcc %>% 
        filter(within_clust==T), 
    aes(x = sizediff, y = identity, colour = within_clust)
    )+
    geom_point()+
    facet_grid(
        cols = vars(between_clust), 
        scales = "free", space = "free", switch = "y"
        ) +
    geom_hline(yintercept = 99.5)
    
ggplot(
    bcc %>% filter(within_clust==T), 
    aes(x = sizediff, y = identity, colour = between_clust)
    )+
    geom_point()
    
abc <- 
    bcc %>% 
    filter(between_clust=="2,2", identity >95)
 
c(
    (bcc %>% 
    filter(sizediff==0) %>%
     pull(ref)),
    (bcc %>% 
    filter(sizediff==0) %>%
     pull(qry))
) %>% unique

 bcc %>% 
    filter(sizediff==0) %>%
     pull(identity) %>% summary
 
 bcc %>%
     filter(identity>99) %>%
     dplyr::select(
         ref, qry, identity, ref_clust, qry_clust, ref_mlen, qry_mlen
         ) %>%
     rowwise %>%
     mutate(
         rq = paste(sort(c(ref_mlen, qry_mlen)), collapse = ",")
     ) %>%
     #count(rq)
     ungroup %>%
     filter(ref_mlen==11,qry_mlen==70.5)
     # count(rq)
     # 
 
 ### merging clusters
 bcc %>%
     filter(
         identity>99,
         sizediff<2
         ) %>%
     dplyr::select(
         ref, qry, identity, ref_clust, qry_clust, ref_mlen, qry_mlen, sizediff
         ) %>%
     rowwise %>%
     mutate(
         rq = paste(sort(c(ref_mlen, qry_mlen)), collapse = ",")
     ) %>%
     ungroup() %>%
     filter(ref_mlen !=qry_mlen) %>%
     distinct(rq, .keep_all = T)
     filter(ref_mlen == 13)
     count(rq)
     
     bie2 %>% as.data.frame() %>% head(20)

 
 
abc$ref %>% unique
abc$qry %>% unique

filtered_combined_delta %>%
    filter(qid=="CP075028.1") %>% ungroup %>%
    plot_delta()

filtered_combined_delta %>%
    filter(qid=="NZ_CP022015.1") %>% ungroup %>%
    plot_delta()

(bcc %>% filter(sizediff>9))[c(1,2),c(1,2)]

bie2 %>%
    filter(qid=="NZ_LT905143.1")

# bie2 %>%
#     left_join(
#         .,
#         (d2_id$id_dists %>% dplyr::rename(qid = Sequence)),
#         by = "qid"
#     )

b5 <- 
    left_join(
        b3,
        (d2_id$id_dists %>% dplyr::rename(qid = Sequence)),
        by = "qid"
    )

b3

plot(x = 1:nrow(b3),y = b3$meansize, col=b3$pclus, pch = 19)


b3p <- 
    b3 %>% 
    mutate(
        dbscan_cluster = as.factor(rleid(pclus))
    )



col_vector<-
    c(
        '#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', 
        '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', 
        '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', 
        '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', 
        '#ffffff', '#000000'
        )


ggplot(b3p, aes(x=meansize, fill = dbscan_cluster)) + 
  geom_histogram(binwidth=1, alpha=0.5, colour = "black") +
    theme_classic() +
    theme(
        text = element_text(size = 22)
        ) +
    xlab("Inversion length (%)") +
    ylab("Count") + scale_fill_manual(values = col_vector[1:18]) + 
    scale_x_continuous(expand = c(0, 0), limits = c(0,100)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 15))
        #scale_x_continuous(expand = c(0, 0), limits = c(0, NA))
    # scale_fill_viridis_d(
    #             option = "B"
    #         ) +

plot(x = 1:nrow(b5),y = b5$Difference_from_Mean, col=b5$pclus, pch = 19)
, 
plot(dbs3, bie2$meansize, main = "DBSCAN", frame = FALSE)

d2_id$mash_tib

ggplot(
    data = 
        b4, #%>% 
        # filter(
        #     ref %in% 
        #         (b3 %>% 
        #              #filter(pclus==3) %>% 
        #              pull(qid)),
        #     qry %in% 
        #         (b3 %>% 
        #              #filter(pclus==3) %>% 
        #              pull(qid))
        # ),
        aes(
            #r2, q2,
             x = factor(ref, levels = unique(b3$qid)),
             y = factor(qry, levels = unique(b3$qid)),
            # x= factor(ref, levels = (b3 %>% filter(pclus==3) %>% pull(qid))),
            # y= factor(qry, levels = (b3 %>% filter(pclus==3) %>% pull(qid))),
            fill = identity#identity_sld
        )
    ) + 
    geom_tile() + 
    theme_classic() +
    theme(
        axis.text.x = element_text(angle=45, hjust = 1),
        panel.background=element_rect(fill="white"),
        plot.title = element_text(size=20,face="bold", hjust = 0.5),
        axis.title = element_blank()
    ) + 
    ggtitle(
        sprintf(
            "%s \n Sequence pairwise MASH similarity", 
            sub("_", " ", species_name)
        )
    ) + 
    theme(
        legend.title=element_text( size=18),
        #axis.text=element_blank(),
            #element_text(size=14), 
        legend.text = element_text(size=18)
    ) + 
    scale_fill_viridis(
        option = "B", limits = c(95,100)
    ) +
    facet_grid(
        rows = vars(qry_clust), cols = vars(ref_clust), 
        scales = "free", space = "free", switch = "y")
    


dps3 <- 
    lapply(
        l3,
        function(x){
            filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==x) %>%plot_delta()
        }
    )

do.call(
     "grid.arrange",
     c(
         dps3,
         ncol=3
         )
     )



e2 %>%
    mutate(
        cdiff = 
            min(
                c(
                    meansize-lag(meansize, default = -1),
                    lead(meansize, default = 101)-meansize
                    )
            )
    )
    

plot(b3$meansize, b3$Sequence_mean_distance)
plot(e2$meansize)
###
das <- 
    bie2$qid %>% unique()

comp_matrix <- 
    t(combn(das, 2)) %>%
    `colnames<-`(c("ref", "qry")) %>%
    as_tibble()


###
bie2 %>% View

bie2 %>% filter(meansize==13)
filtered_combined_delta %>% 
        ungroup %>%
        filter(qid=="CP082490.1") %>%plot_delta()

e2

    
    inv_event_summary <- 
        fd2 %>%
        dplyr::select(rid, qid, rs, re, qs, qe, strand, meanlen, qlen, rlen) %>%
        mutate(midpoint = round((rs+re)/2), meanlen = round(meanlen)) %>%
        ungroup() %>% 
        arrange(rs) %>%
        left_join(
            .,
            clustered_genomes[,c(1,2)] %>%
                `colnames<-`(c("cluster", "qid")),
            by = "qid"
            )
    
    ## midpoints of events
    fdmp <- 
        fd2 %>%
        mutate(
            rmp = round((rs+re)/2),
            qmp = round((qs+qe)/2),
            rmprop = round(100*rmp/rlen),
            qmprop = round(100*qmp/qlen),
            meanlen = round(meanlen)
            #r_Xdist = 
        ) %>%
        dplyr::select(strand, rmp, qmp, X_dist, rmprop, qmprop, meanlen, rs, re)
    
    events <- 
        fdmp %>%
        mutate(
            xdd =
                round(X_dist-lag(X_dist, default = X_dist[1])),
            RASR = 
                X_dist <1e6
            )
    if (all(events$RASR) & nrow(events)>1){
        inner_rasr <- 
            events %>% 
            mutate(
                mid_diff = 
                    abs(rmp-qmp)
            )
    }
    
    RASRs <- 
        fd2 %>% 
        filter(X_dist>1e6)
    
    ## if the
    
    
    
    
    ## the delta file unfiltered
    new_out_deltadir
    list.files( delta_dir)
    
}

main_events

which(!(bis$accessions %in% all_bps$qid))
bis$accessions[40]
filtered_combined_delta %>% ungroup %>%
    filter(qid=="NZ_CP126178.1") %>% plot_delta

all_bps %>%
    filter(meanlen<1e5)

all_bps %>%
    filter(meanlen>1e5 & meanlen <3e5, rs >2e6)

all_bps %>% filter(cluster %in% 10:12)
all_bps %>% filter(qid=="NZ_CP117290.1")

filtered_combined_delta %>% filter(qid=="CP076727.1")  %>% ungroup %>%
    plot_delta

filtered_combined_delta %>% filter(qid=="NZ_CP117290.1")  %>% ungroup %>%
    plot_delta
filtered_combined_delta %>% filter(qid=="CP090229.1")  %>% ungroup %>%
    plot_delta


```



```{r}


best_identity %>% filter(cluster_id==0) %>% arrange(desc(identity)) %>%
    dplyr::slice(1:100) %>%
    View

sst1 <-     
    b3p %>% filter(dbscan_cluster==2) %>%
    #filter(inversions==1) %>%
    group_by(meansize) %>% 
    filter(grouped_sequences==max(grouped_sequences)) %>%
    filter(grouped_sequences>=5)
    #mutate(cc = n())
    #arrange(desc(grouped_sequences)) %>%
    #dplyr::slice(1)

## checking shared events adjacency


# lapply(
#     1:nrow(bie2)
#     function(k){
#         events <- 
#             bie2 %>%
#             filter(qid == bie2$qid[k]) %>%
#             dplyr::select(meansize, qid, cluster_id, inversions)
#             
#     }
# )

eo <- list()
which(bie2$cluster_id==2    )
which(bie2$qid=="NC_016832.1")
which(bie2$qid=="1.5")
bie2 %>%
    filter(
        cluster_id ==2
    )

bie2 %>%
    filter(qid %in% c("1.5", "NZ_LR590082.1"))
bie2 %>% View
## merging clusters with more shared events

bie2$qid[which(bie2$cluster_id==5)]
    filter(cluster_id==5)
which(unique(bie2$qid) ==bie2$qid[which(bie2$cluster_id==5)])
    i=67
    
which(unique(bie2$qid)=="NZ_CP034230.1")



evd2
    


bie3tt <- 
    pbmclapply(
        1:length(unique(bie2$qid)),
        function(i){
        #    for(i in 1:length(unique(bie2$qid))){
            qi <- 
                unique(bie2$qid)[i]
            q_events <- 
                bie2 %>%
                filter(qid==qi) %>% 
                dplyr::select(qid, meansize) %>%
                mutate(
                    mlow = 
                        meansize-1,
                    mhigh = 
                        meansize+1
                )
            #which(bie2$qid=="NZ_LR590082.1")
            
            #bie2 %>% filter()
            
            shared_events <- 
                bie2 %>%#[which(bie2$qid=="NZ_LR590082.1"),] %>%
                mutate(
                    shared_event = 
                        sapply(
                            # bie2[which(bie2$qid=="NZ_LR590082.1"),] %>%
                            #     pull(meansize),
                            bie2$meansize, 
                            function(value){
                              any(
                                  (value >= q_events$mlow) & 
                                      (value <=  q_events$mhigh)
                                  )  
                                } 
                            )
                ) %>%
                filter(
                    qid != qi,
                    shared_event
                )

            #shared_events %>% filter(qid == qi)
            #which(shared_events$cluster_id==47)
            
            event_tib <- 
                lapply(
                    unique(shared_events$qid),
                    function(x){
                        #x <-
                        #  unique(shared_events$qid[38])
                        #for (x in unique(shared_events$qid)){
                        x_shared <- 
                            shared_events %>%
                            filter(qid==x) %>%
                            dplyr::select(-c("shared_event"))
                        x_events <- 
                            bie2 %>%
                            filter(qid==x)
                        
                        xdiff <- 
                            bind_rows(x_shared, x_events) %>% 
                            dplyr::add_count(meansize, name = "occ") %>%
                            filter(occ==1)
                   
                        qdiff <- 
                            q_events %>%
                            mutate(
                                qd = 
                                    lapply(
                                        q_events$meansize,
                                        function(qe){
                                            unlist(
                                                lapply(
                                                    x_events$meansize,
                                                    function(xe){
                                                        (qe >= xe-1) &
                                                            (qe <=  xe+1)
                                                    }
                                                )
                                            ) %>% any
                                        }
                                    ) %>% unlist()
                            ) %>%
                            filter(
                                !qd
                            ) #%<%
                            # x_events %>% 
                            # dplyr::select(qid, meansize) %>%
                            # mutate(
                            #     mlow = 
                            #         meansize-1,
                            #     mhigh = 
                            #         meansize+1
                            # ) %>%
                            # bind_rows(
                            #     ., 
                            #     q_events
                            # )
                            
                        out_tib <- 
                            tibble(
                                qid = x,
                                se = nrow(x_shared),
                                q_events = nrow(q_events),
                                x_events = nrow(x_events)
                            ) %>%
                            mutate(
                                des = abs(se-q_events),
                                event = 
                                    ifelse(
                                        length(xdiff$meansize)==1,
                                        xdiff$meansize,
                                        ifelse(
                                            length(qdiff$meansize)==1,
                                            qdiff$meansize,
                                            NA
                                        )
                                    )
                            )
                        
                        return(out_tib)
                    }
                ) %>% 
                bind_rows()
            #shared_events %>%
             #   group_by(qid)
            same_cluster <- 
                event_tib %>%
                filter(des==0)
            adjacent_cluster <- 
                 event_tib %>%
                 filter(des==1) 
            # if (nrow(adjacent_cluster)>0){
            #     lapply(
            #         1:nrow(nrow(adjacent_cluster)),
            #         function(j){
            #             j_events <- 
            #                 bie2 %>%
            #                 filter(
            #                     qid ==adjacent_cluster$qid[j]
            #                 ) %>%
            #                  mutate(
            #                     jlow = 
            #                         meansize-1,
            #                     jhigh = 
            #                         meansize+1
            #                 )
            #                     
            #             
            #             j_shared_events <- 
            #                 shared_events %>%
            #                 filter(
            #                     qid ==adjacent_cluster$qid[j]
            #                 )
            #             # if all j events are in shared events
            #             if (all(j_events$meansize==j_shared_events$meansize)){
            #                 q_events 
            #             }
            #             
            #                 
            #                 
            #         }
            #     )
            # }
            
            ##
            # ac <- 
            #     adjacent_cluster %>%
            #     mutate(
            #         ref = qi
            #     ) %>%
            #     dplyr::select(ref, qry, se, q_events)
            
            
            if (nrow(event_tib)>0){
                out_tib <- 
                    event_tib %>%
                    filter(
                        des %in% c(0,1)
                    ) %>%
                    mutate(
                        rid = qi,
                        rclust = 
                            bie2$cluster_id[i]
                    ) %>%
                    left_join(
                        ., 
                        bie2[,c("qid", "cluster_id")] %>%
                            `colnames<-`(c("qid", "qclust")),
                        by = "qid"
                    )
            } else{
                out_tib <- NULL
            }
            
            
            return(out_tib)
                
        }
        )


ani_dir<- 
    sprintf(
        "%s/ani/ac1",
        identity_dir
        )

if (!dir.exists(ani_dir)){dir.create(ani_dir)}

all_connected <- 
    c(
        bie4$qid,
        bie4$rid
    ) %>% unique

ani_inlines <- 
    sprintf(
        "%s/%s.txt", 
        sync_dir, all_connected
    )

ani_infile <- 
    sprintf(
        "%s/ac1.txt",
        ani_dir
    )
writeLines(
    ani_inlines,
    ani_infile
)

lapply(ani_inlines, file.exists) %>% unlist %>% all

in_file <- 
    ani_inlines
out_dir <- 
    ani_dir

ani_out <- 
    sprintf(
        "%s/ac1_out.txt",
        out_dir
    )

ani_commands <- 
    function(
        in_file,
        out_dir
    ){
        ani_com <- 
            sprintf(
                "%s --ql %s --rl %s -o %s",
                "fastANI -t 7",
                ani_infile, ani_infile,
                ani_out
            )
        system(ani_com)
    }
ani_table <-
    fread(ani_out) %>%
    mutate(V1 = sub(".txt", "", basename(V1))) %>%
    mutate(V2 = sub(".txt", "", basename(V2))) %>%
    `colnames<-`(c("ref", "qry", "ani", "matched", "total")) %>%
    mutate(ani = as.numeric(ani)) %>%
    as_tibble %>%
    mutate(
        ani_dist = 100-ani
    ) 
    




fcd2 %>%
    filter(qid=="CP093081.1") %>% plot_delta()
fcd2 %>%
    filter(qid=="NZ_LT905063.1") %>% plot_delta()


list.files(identity_dir)


    head(bie3)
redun <- 
     bind_rows(bie3) %>%
    dplyr::select(-des) %>%
    filter(!((se==q_events )& (se==x_events ))) %>%
    rowwise() %>%
    mutate(fclust = max(c(rclust, qclust))) %>%
    pull(fclust) %>% unique

bind_rows(bie3) %>% 
    filter(
        (se==q_events) & (se==x_events) 
    )

list.files(new_out_deltadir, pattern = "CP076727.1", full.names = T) %>% 
                plot_delta() 
list.files(new_out_deltadir)

  which((bind_rows(bie3) %>%
    dplyr::select(-des) %>%
    filter(!((se==q_events )& (se==x_events ))) %>%
    rowwise() %>%
    mutate(fclust = max(c(rclust, qclust))) %>%
    pull(fclust) %>% unique)==2)


bie2

bind_rows(bie3)

bie4 %>% filter(qclust==3)
bie4 <- 
    bind_rows(bie3) %>%
    dplyr::select(-des) %>%
    # filter(
    #     !(rclust %in% redun),
    #     !(qclust %in% redun)
    #     ) %>%
    mutate(
        qs = q_events-se,
        xs = x_events-se,
        qx = qs+xs
    ) %>%
    #filter(qx==1) %>%
    rowwise() %>%
         mutate(
            rq = paste(sort(c(rid, qid)), collapse = ",")
            ) %>%
    ungroup %>%
    distinct(rq, .keep_all = T) %>%
    add_count(event) %>%
    arrange(event) 



#%>% 
# group_by(se, q_events, x_events, event) %>%
    # arrange(desc(n))
    # summarise(
    #     cc = n()
    # )
    # group_by(
    #     
    # )
b4_simple <- 
    bie4 %>% dplyr::select(rclust, qclust, event)

b4_simple %>% filter(rclust == 89 | rclust == 89)


b4_simple %>% add_count(rclust, name = "rcc") %>% add_count(qclust,name =  "qcc")

bie4 %>% 
    filter(rclust %in% c(24, 90), qclust %in% c(24, 90))

b4r <- 
    bie4 %>% group_by(rclust) %>%
    summarise(
        events = 
            paste(event, collapse = ",")
        ) %>%
    `colnames<-`(c("clust", "events")) %>%
    mutate(ds = "ref")

b4q <- 
    bie4 %>% group_by(qclust) %>%
    summarise(
        events = 
            paste(event, collapse = ",")
        ) %>% 
    `colnames<-`(c("clust", "events"))%>%
    mutate(ds = "qry")

b4c <- 
    bind_rows(b4r, b4q) %>%
    arrange(clust) %>%
    add_count(clust) %>%
    arrange(desc(n))

b4c


dbs3 <- 
    fpc::dbscan(
        bie2$meansize,
        #bie4 %>% filter(!is.na(event)) %>% pull(event), 
        eps = 1, MinPts = 1, scale = F, method = "raw"
        )

bie2 %>% 
    mutate(pclus = dbs3$cluster) 

bie2

bie5 <- 
    bie2 %>%
    filter(
        inversions==1,
        # (qid %in% unique(bie4$qid)) |
        #     (qid %in% unique(bie4$qid)) ,
        meansize %in% c(4,11, 13, 18, 20, 32, 52, 71, 75, 83)
        ) %>%
    mutate(
        rclust = 0,
        qclust = cluster_id,
        event = meansize,
        rid = bis$accessions[1]
    ) %>%
    group_by(meansize) %>%
    filter(cluster_id==min(cluster_id)) %>%
    # filter(
    #     (rclust %in% c(0,1,11,12)) &
    #         (qclust %in% c(0,1,11,12))
    #     ) %>%
    dplyr::select(c(qid, rid, event, rclust, qclust)) 



bie6 <- 
    bind_rows(bie5, bie4) %>% arrange(event) 

tt2 <- 
    c(
        bie4 %>% dplyr::slice(1:13) %>% pull(qid),
        bie4 %>% dplyr::slice(1:13) %>% pull(rid)
    ) %>% unique()




ffs <- 
    c(13,5,42,96,21,6, 39, 97, 99, 109, 7, 106)

bie7 <- 
    bie4 %>%
    filter(
        (qid %in% tt2) | (rid %in% tt2),
        !((rclust %in% ffs) | (qclust %in% ffs)  )
    ) %>% 
    bind_rows(bie5)

bie4 %>% filter(rclust==7)
bie4 %>% filter(qclust==70)
    
        

plot(x = 1:nrow(bie6),y = bie6$event,pch = 19)
bie6$event

bie4 %>%
    filter(event >10 & event <12)
    
 
edge_list2 <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = bie7 %>% pull(rclust),
                target = bie7 %>% pull(qclust)
                ),
        directed = F
    )   

E(edge_list2)$label <- 
    bie7$event

label.size <- 2
V(edge_list2)$label.cex <- 18
    
plot(edge_list2, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)


#############
bis_1

adj_tib <- 
    list()
for (i in 1:12){#nrow(bie2)){
    
    acc_i <- 
        bie2[i,] %>%
        pull(qid)
    clust_i <- 
        bie2[i,] %>%
        pull(cluster_id)
        
    q_events <- 
        bie2 %>%
        filter(qid==bie2$qid[i])
    
    potential_adj <- 
        bie2 %>%
        filter(
            (inversions == 1+q_events$inversions[1]) |
                (inversions == q_events$inversions[1]-1)
        )
    
    adj_qid <- 
        list()
    
    if (
        nrow(potential_adj)==0 |
        nrow(q_events) != pull(q_events, inversions)[1]
        ){
        eo[[i]] <- 
            q_events %>%
            mutate(idx = i)
    } else{
        for (j in 1:nrow(q_events)){
            qe_len <- 
                q_events[j,] %>% 
                pull(meansize)
            q_clust <- 
                potential_adj %>%
                filter(
                    (meansize <= 1+qe_len) &
                        (meansize >= qe_len-1)
                ) %>%
                pull(qid)
            asym_invs <- 
                bie2 %>% 
                filter(qid %in% q_clust) %>%
                add_count(qid, name = "qc") %>%
                filter(qc!=inversions)
            
            adj_qid[[j]] <- 
                bie2 %>% 
                filter(qid %in% q_clust) %>%
                add_count(qid, name = "qc") %>%
                filter(qc==inversions) %>%
                dplyr::select(
                    c(
                        qid, meansize, 
                        cluster_id, grouped_sequences, 
                        inversions
                        )
                    )
        }
        ## shared events table
        se_tib <- 
            tibble(
                qid = 
                    lapply(
                        adj_qid, 
                        function(x){
                            pull(x, qid) %>% unique
                        }
                    ) %>% unlist()
            ) %>%
            dplyr::count(qid, name = "shared_events") %>%
            left_join(
                ., 
                bie2,
                by = "qid"
            ) %>% 
            mutate(
                q_invs = 
                    nrow(q_events)
            ) %>%
            filter(
                abs(shared_events-q_invs)<=1,
                abs(shared_events-inversions)<=1
            )
        # shared events
        shared_ms <- 
            lapply(
                q_events$meansize, 
                function(x){
                    qe <- 
                        c(x-1, x+1)
                    des <- 
                        se_tib %>% 
                        filter(
                            meansize>qe[1] &
                                meansize <qe[2]
                        )
                    return(des)
                    }
                ) %>%
            bind_rows() %>% 
            pull(meansize)
        
        ## different events
        de_tib <-
            se_tib %>%
            mutate(
                xi =
                unlist(
                    lapply(
                        meansize, 
                        function(x){
                            any(
                                (x>= shared_ms-1) & 
                                    (x<= shared_ms+1) 
                                )
                        }
                    )
                )
            ) %>%
            filter(!xi) %>%
            dplyr::select(qid, meansize, cluster_id, grouped_sequences)
        
        
        adj_tib[[i]] <- 
            tibble(
                rid = 
                    acc_i,
                qid = 
                    unique(se_tib$qid)
            ) %>%
            inner_join(
                ., 
                de_tib, by = "qid"
            ) %>%
            mutate(r_clust = clust_i)
            
    }
    ## if shared events = inversions - 1
    
}

bie3

at1 <- 
    adj_tib %>% 
    bind_rows() %>% 
    rowwise() %>%
         mutate(
            rq = paste(sort(c(rid, qid)), collapse = ",")
            ) %>%
    ungroup %>%
    distinct(rq, .keep_all = T)

at1 %>% add_count(meansize)
   
 
edge_list2 <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = at1 %>% pull(r_clust),
                target = at1 %>% pull(cluster_id)
                ),
        directed = F
    )   

E(edge_list2)$label <- 
    at1$meansize
    
plot(edge_list2, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)

#    
adj_tib %>% bind_rows()
Filter(nrow, adj_tib)

#NZ_CP117293.1
#NC_016832.1
#CP090240.1
#NZ_LT905143.1

bie2
fcd2 <- 
    filtered_combined_delta %>% 
    ungroup %>% 
    #mutate(mls = sum(meanlen)) %>%
    left_join(
        .,
        best_identity[,(c(1,2,3))] %>%
            `colnames<-`(c("qid", "identity", "cluster")),
        by ="qid"
    ) %>%
    group_by(cluster) %>%
    arrange(desc(identity)) %>%
    filter(qid==qid[1]) %>%
    ungroup
    
fcd2 %>%
    filter(
        cluster==47,
    ) %>%
    plot_delta


fcd2 %>%
    filter(
        cluster==89,
    ) %>%
    plot_delta
    
clustered_genomes

adj_tib
filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="NZ_LN890522.1") %>% plot_delta() +
    xlab("Cluster 0") +
    #ylab("Cluster 2") +
    theme(axis.title = element_text(size = 24))

bie2 %>%
    group_by(qid) %>%
    count(meansize)

inv_plots_sst1 <- 
    lapply(
        1:nrow(sst1),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==sst1$qid[i])
            plot_delta(fcd_i)
        }
    )

filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="CP090227.1") %>% plot_delta() +
    xlab("Cluster 0") +
    #ylab("Cluster 28") +
    theme(axis.title = element_text(size = 24))

filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="1.5") %>% plot_delta()

filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="NZ_LR590082.1") %>% plot_delta()

do.call(
     "grid.arrange",
     c(
         inv_plots_sst1,
         ncol=2
         )
     )
bie2 %>% 
    filter(qid=="NZ_CP051384.1")

bie2 %>% 
    filter(qid=="NC_021820.1")

bie2 %>% filter(cluster_id==2)

bie2 %>% 
    filter(
        qid %in%
            (bie2 %>% filter(meansize>81, meansize <85) %>% pull(qid)),
        qid %in%
            (bie2 %>% filter(meansize>17, meansize <19) %>% pull(qid))
    ) 

bie2 %>% 
    filter(
        qid %in%
            (bie2 %>% filter(meansize>50, meansize <52) %>% pull(qid)),
        qid %in%
            (bie2 %>% filter(meansize>17, meansize <19) %>% pull(qid))
    ) 

bie2 %>% filter(meansize>81, meansize <85)

bie2 %>% filter(meansize==11.5)
bcc %>% 
    filter(
        ref_mlen ==11.5 |
            qry_mlen ==11.5 
    ) %>%
     dplyr::select(
         ref, qry, identity, ref_clust, qry_clust, ref_mlen, qry_mlen, sizediff
         ) 

bcc %>% filter(ref_clust==7) %>%
     dplyr::select(
         ref, qry, identity, ref_clust, qry_clust, ref_mlen, qry_mlen, sizediff
         )  %>% 
    count(ref_mlen)

### merging clusters
mc <- 
    bcc %>%
     filter(
         identity>99,
         sizediff<2
         ) %>%
     dplyr::select(
         ref, qry, identity, ref_clust, qry_clust, ref_mlen, qry_mlen, sizediff
         ) %>%
     rowwise %>%
     mutate(
         rq = paste(sort(c(ref_mlen, qry_mlen)), collapse = ",")
     ) %>%
     ungroup() %>%
     filter(ref_mlen !=qry_mlen) %>%
     distinct(rq, .keep_all = T)

reo_events <- 
    bie2 %>% 
    group_by(meansize) %>%
    summarise(
        cids = 
            paste(cluster_id, collapse = ","),
        seqs = sum(grouped_sequences)
            ) %>% arrange(desc(seqs))

## main inversion
bis
mi1 <- 
    bie2 %>% 
    filter(meansize<13, meansize>10)

mi2 <- 
    bie2 %>% 
    filter(meansize>17 & meansize<19)
    

```


## Indels using the original ffs







```{r}
list.files(sprintf("%s", sync_dir))%>% gsub(".txt", "", .) %>% head

al1 <- 
    list.files(sprintf("%s", sync_dir))%>% gsub(".txt", "", .) 

pbmclapply(al1, function(x){
    # x <- 
    #     "NZ_CP039567.1"
    rep_com <-
        sprintf(
            "repeat-match %s/%s.txt > %s/%s_reps.txt",
            sync_dir, x, repeats_dir, x
        )
    system(rep_com)
    },
    mc.cores = 6
)

rep_tibs <- 
    list.files(repeats_dir, full.names = T)
"NZ_CP039567.1"

acc_len <- 
        species_summary %>% dplyr::select(accession, len) %>%
        filter(accession==gsub("_reps.txt", "", basename(rep_tibs[i])))

rep_tib <- 
        fread(
            sprintf(
                rep_tibs[i]
            ), skip = 1
        ) %>%
         mutate(
            strand = 
                ifelse(grepl("r", Start2), "-", "+"),
            Start2 = 
                as.numeric(gsub("r", "", Start2))
        ) %>%
        as_tibble %>%
        arrange(Start1)%>%
        `colnames<-`(
            c(
                "r1s",
                "r2s", 
                "len",
                "strand"
            )
        ) %>% 
        mutate(
            r1e = r1s+len,
            r2s_a = 
                ifelse(
                    strand=="+",
                    r2s,
                    r2s-len
                ),
            r2e = 
                ifelse(
                    strand=="+",
                    r2s +len,
                    r2s
                ),
            r2s = r2s_a,
            r1_mid = round((r1e+r1s)/2),
            r2_mid = round((r2e+r2s)/2),
        ) %>% dplyr::select(-r2s_a) %>% 
        mutate(
            accession = acc_len$accession,
            ref_len = acc_len$len,
            mirror = 
                ((r1_mid+r2_mid) >= (acc_len$len-5e3)) &
                ((r1_mid+r2_mid) <= (acc_len$len+5e3))
        )
max(rep_tib$len)


rep_tib %>% filter(len>400)

rep_ranges <- 
    IRanges(
        start = 
            c(
                rep_tib$r1s, 
                rep_tib$r2s
                ),
        end = 
            c(
                rep_tib$r1e, rep_tib$r2e
            )
    ) %>%
    reduce()# %>%
    #as.data.frame() %>%
    #count(width) %>% arrange(n)
rep_ranges %>% width %>% table


rt2 <- 
    bind_rows(
        rep_tib %>% 
            mutate(
                midpoint = r1_mid,
                type = "1"
            ),
        rep_tib %>% 
            mutate(
                midpoint = r2_mid,
                type = "2"
            )
    ) %>%
            mutate(
                ivls = 
                    cut(
                        len, 
                        breaks = 
                            c(
                                min(
                                    rep_tib$len
                                    ), 
                                50, 
                                200, 
                                500, 
                                max(rep_tib$len)
                                ), 
                        include.lowest = T
                        )
            )# %>%
    #filter(len>200)
rt2

ggplot(
    rt2 %>% 
        filter(len>100), 
    aes(x=midpoint)) +
  geom_density(alpha=0.4, bw=0.01)
    
rt2 %>% filter(is.na(ivls))

ggplot(rt2, 
    aes(
      x = midpoint, y = ivls, 
      group = ivls, fill = type)
  ) + 
  theme_classic(base_size = 24) +
  geom_density_ridges( 
    rel_min_height = 0.00001, scale =1, alpha = 0.6, bandwidth=0.5, 
  ) +
  # geom_label(
  #     data = cgi2, 
  #     aes(y = org, x = dd_mean, label = "|")
  #     ) +
  #scale_fill_gradient2(low = "cyan", mid = "darkgreen", high = "burlywood4") +
  scale_fill_viridis_c(name = "Mean absolute dnaA distance", option = "C") +
  theme(
    legend.position="none",
    #panel.grid.major = element_line()
    panel.grid.major.y = element_line(color = "grey80")
  ) +
  #scale_x_continuous(breaks = seq(-0.5, 0.5, 0.1),limits = c(-.5, 0.5)) +
  ylab("Interval") + 
  xlab("genome") +
  theme(axis.text.y = element_text(face = "italic"))

 ggplot(rt2, 
    aes(
      x = midpoint, y = ivls, 
      group = ivls, fill = type)
  ) + 
  theme_classic(base_size = 24) +
  geom_density_ridges( 
    rel_min_height = 0.00001, scale =1, alpha = 0.6, bandwidth=0.5, 
  ) +
  # geom_label(
  #     data = cgi2, 
  #     aes(y = org, x = dd_mean, label = "|")
  #     ) +
  #scale_fill_gradient2(low = "cyan", mid = "darkgreen", high = "burlywood4") +
  scale_fill_viridis_c(name = "Mean absolute dnaA distance", option = "C") +
  theme(
    legend.position="none",
    #panel.grid.major = element_line()
    panel.grid.major.y = element_line(color = "grey80")
  ) +
  #scale_x_continuous(breaks = seq(-0.5, 0.5, 0.1),limits = c(-.5, 0.5)) +
  ylab("Interval") + 
  xlab("genome") +
  theme(axis.text.y = element_text(face = "italic"))

ggplot(rt2, aes(x = ivls, y = len, fill = ivls)) +
  geom_boxplot() +
  labs(title = "Boxplot by Intervals", x = "Intervals", y = "Values") +
  scale_fill_manual(values = c("skyblue", "salmon", "lightgreen", "chartreuse")) +
  theme_minimal() +
  facet_wrap(~ ivls, scales = "free_y")

all_bps <-
    filtered_combined_delta %>%
    filter(
        rs !=min(rs), re !=max(re), qs !=1
        ) %>%
    filter(
        (lag(strand, default = "+") != strand) & 
            (lead(strand, default = "+") != strand) 
    ) %>%
    dplyr::select(rid, qid, rs, re, strand) %>%
    mutate(midpoint = round((rs+re)/2)) %>%
    ungroup() %>% 
    arrange(rs) %>%
    left_join(
        .,
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("cluster", "qid"))
    )

ab_c <- 
    all_bps %>%
    group_by(cluster) %>%
    dplyr::slice(1)

plot(ab_c$rs, ab_c$re)


all_bps

clustered_genomes[,c(1,2)]

filtered_combined_delta %>% filter(qid=="CP073326.1")  %>% ungroup %>%
    plot_delta

all_bps_m <-
    all_bps %>% 
    reshape2::melt(., id.vars = c("rid", "qid", "strand")) %>% 
    as_tibble() %>%
    left_join(
        .,
        clustered_genomes[,c(1,2)] %>%
            `colnames<-`(c("cluster", "qid")),
        by = "qid"
    ) %>%
    group_by(cluster) %>%
    filter(qid==qid[1]) %>%
    arrange(value) %>%
    filter(!(variable %in%c("cluster", "midpoint"))) %>%
    ungroup() %>%
    mutate(idx = row_number())

ggplot(
    all_bps_m, 
    aes(x = idx, y = value, colour = variable)
) +
    geom_point()

dbs3 <- 
    fpc::dbscan(all_bps_m$value, eps = 3e4, MinPts = 1, scale = F, method = "raw")

plot(dbs3, all_bps_m$value, main = "DBSCAN", frame = FALSE)



aa <- 
    all_bps_m %>% 
    mutate(pclus = dbs3$cluster) %>%
    add_count(pclus) %>%
    filter(n>4)

aa2 <- 
    aa %>% 
    group_by(qid) %>% 
    summarise(cs = paste(unique(pclus), collapse = ", "))


count_shared_numbers <- function(s1, s2) {
  nums1 <- strsplit(s1, ",")[[1]]
  nums2 <- strsplit(s2, ",")[[1]]
  length(intersect(nums1, nums2))
}

count_shared_values <- function(s1, s2) {
  nums1 <- unlist(strsplit(s1, ","))
  nums2 <- unlist(strsplit(s2, ","))
  sum(nums1 %in% nums2)
}

dm <- matrix(0, nrow = nrow(aa2), ncol = nrow(aa2))
for (i in 1:nrow(aa2)) {
  for (j in 1:nrow(aa2)) {
    dm[i, j] <- 
        count_shared_values(aa2$cs[i], aa2$cs[j])
  }
}
dm[1:10, 1:10]


dm2 <- 
    dm

dm2[lower.tri(dm2, diag = T)] <- 0

dm4 <- 
    which(dm2>3, arr.ind = T)
head(dm4)
dm4[,1]
dm2
aa2

dm3 <- 
    tibble(
        from = aa2 %>% dplyr::slice(dm4[,1]) %>% pull(qid),
        to = aa2 %>% dplyr::slice(dm4[,2]) %>% pull(qid)
    )

edge_list <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = dm3$from,
                target = dm3$to
                ),
        directed = F
    )

plot(edge_list, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)



# Compare each row with all other rows
shared_counts <- sapply(aa2$cs, function(x) {
  sapply(aa2$cs, function(y) {
    count_shared_numbers(x, y)
  })
})

# Filter rows with 2 or more shared numbers with another row
rows_to_keep <- apply(shared_counts, 1, function(x) {
  any(x >= 2)
})

aa2[rows_to_keep,]
intersect(aa$qid, aa$pclus)



plot(all_bps_m$value)
    #mutate(value = as.numeric(value)) %>%
    #filter(is.na(value))


all_bps

ggplot(
    all_bps_m %>% filter(variable %in% c("rs", "re")) , 
    aes(x=value, fill = variable)
    ) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")

ggplot(all_bps, aes(x = rs, y = re)) +
  geom_point(color = "blue") +
  labs(title = "Midpoint vs. Length", x = "Midpoint", y = "Length") +
  theme_minimal()



r3 <- 
    rt2 %>%
    filter(len>50) %>%
    dplyr::select(midpoint, qid) %>%
    mutate(event = "repeat") %>%
    bind_rows(
        .,
        all_bps_m %>% 
            mutate(
                midpoint = value,
                event = "Inversion",
            ) %>%
            dplyr::select(midpoint, event)
    ) %>% 
    bind_rows(
        ., 
        (it4 %>% 
            dplyr::select(pos, mutation, acc) %>%
            `colnames<-`(c("midpoint", "event", qid)))
    )

r4 <- 
        all_bps_m %>% 
            mutate(
                midpoint = value,
                event = "Inversion",
            ) %>%
            dplyr::select(midpoint, event, qid, cluster) %>% 
    left_join(
        ., 
        (it4 %>% 
            dplyr::select(pos, mutation, acc) %>%
            `colnames<-`(c("indel", "event", "qid"))), 
        by = "qid"
    )

ggplot(
    r4,
    aes(x = cluster, y = indel)
    ) +
    geom_point()

plot(r4$indel, r4$midpoint)
ggplot(
    r3, 
    aes(
      x = midpoint, y= event,
      group = event, fill = event
      )
  ) + 
  theme_classic(base_size = 24) +
  geom_density_ridges( 
    rel_min_height = 0.0000001, scale =1, alpha = 0.6, bandwidth=10000, 
  ) +
  # geom_label(
  #     data = cgi2, 
  #     aes(y = org, x = dd_mean, label = "|")
  #     ) +
  #scale_fill_gradient2(low = "cyan", mid = "darkgreen", high = "burlywood4") +
 # scale_fill_viridis(name = "Mean absolute dnaA distance", option = "C")# +
  theme(
    legend.position="none",
    #panel.grid.major = element_line()
    panel.grid.major.y = element_line(color = "grey80")
  ) +
  #scale_x_continuous(breaks = seq(-0.5, 0.5, 0.1),limits = c(-.5, 0.5)) +
  ylab("Mutation") + 
  xlab("Genome") 







rt2

#ft2


boxplot(
    rep_tib %>%
        filter(len>500) %>%
        pull(len)
    )

boxplot(
    rep_tib %>%
        filter(len<=50) %>%
        pull(len)
    )


```


```{r}



inv_plots <- 
    lapply(
        1:25,
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==bis_1$accessions[i])
            plot_delta(fcd_i)
        }
    )

do.call(
     "grid.arrange",
     c(
         inv_plots,
         ncol=5
         )
     )

inv_coords <- 
    lapply(
        1:nrow(bis_1),
        function(i){
            
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==bis_1$accessions[i])
            #plot_delta(fcd_i)
            ## single inv
            invs <- 
                fcd_i$invs[1]
            
            if (invs==1){
                inv_s <- 
                    fcd_i %>%
                    filter(
                            abs((rlen-re) - rs ) < 2e5,
                            X_dist < 2e5,
                            strand=="-"
                        ) %>%
                        mutate(
                            cluster = bis_1$cluster_id[i],
                            seqs = bis_1$grouped_sequences[i]
                            ) %>%
                    dplyr::select(
                        c(
                            "rid", qid, "strand", rs, re, qs, qe, 
                            meanlen, rlen, invs, cluster, seqs
                            )
                        )
            } else{
                ### outer alignments
                fcd_i$segments[1]
                middle_seg <- 
                    round((1+fcd_i$segments[1])/2)
                
                osts <- 
                    t(combn(unique(fcd_i$segment), 2)) %>%
                    `colnames<-`(c("s1", "s2")) %>%
                    as_tibble %>%
                    mutate(
                        sdiff = s2-s1,
                        sd2 = 
                            ((s1+s2)/2)
                    ) %>%
                    filter(
                        #sdiff == 2,
                        s1 != min(s1),
                        s2 != max(s2)
                        ) %>%
                    filter(sdiff>1, sd2 ==median(sd2)) %>%#==max(sdiff))
                    mutate(
                        outer_seg = rleid(s1)
                    ) %>%
                    dplyr::select(-c(sdiff, sd2)) %>%  # Add a row number for each group
                    reshape2::melt(id.vars = "outer_seg") %>%
                    dplyr::select(-c(variable)) %>% 
                    `colnames<-`(c("outer_seg", "segment")) %>% 
                    arrange(outer_seg)
                #
                outer_als <- 
                    fcd_i %>% 
                    inner_join(., osts, by = "segment") %>%
                    group_by(outer_seg, rid, qid, rlen, qlen, invs) %>%
                    summarise(
                        rs = min(rs),
                        re = max(re),
                        qs = max(qs),
                        qe = min(qe),
                        strand =unique(strand)
                    ) %>%
                    mutate(
                        cluster = bis_1$cluster_id[i],
                        seqs = bis_1$grouped_sequences[i],
                        meanlen =
                            ifelse(
                                strand=="-",
                                mean(
                                    c(
                                        (re-rs),
                                        (qs-qe)
                                    )
                                ),
                                mean(
                                    c(
                                        (re-rs),
                                        (qe-qs)
                                    )
                                )
                                )
                            
                        ) %>%
                    ungroup() %>%
                    dplyr::select(
                        c(
                            "rid", qid, "strand", rs, re, qs, qe, 
                            meanlen, rlen, invs, cluster, seqs
                            )
                        )
                
                ## double invs 
                neg_al <- 
                    fcd_i %>% filter(strand=="-") %>%
                    dplyr::select(rs, re, qs, qe) %>%
                    mutate(
                        rmid = round((rs+re)/2),
                        qmid = round((qs+qe)/2)
                    )
                neg_slope <-     
                    abs(round((lm(neg_al$rmid ~ neg_al$qmid) %>% coef())[2], 1))
                ### positive alignmnets flanked by inversions
                # pos_al <- 
                #     fcd_i %>%
                #     filter(
                #         (lead(strand, default = "+") =="-") &
                #             (lag(strand, default = "+") =="-") 
                #     ) %>%
                #         mutate(
                #             cluster = bis_1$cluster_id[i],
                #             seqs = bis_1$grouped_sequences[i]
                #             ) %>%
                #     dplyr::select(
                #         c(
                #             "rid", qid, "strand", rs, re, qs, qe, 
                #             meanlen, rlen, invs, cluster, seqs
                #             )
                #         )
                inner_al <- 
                    fcd_i %>%
                    filter(segment==middle_seg) %>%
                    mutate(
                        cluster = bis_1$cluster_id[i],
                        seqs = bis_1$grouped_sequences[i],
                        meanlen =
                            ifelse(
                                strand=="-",
                                mean(
                                    c(
                                        (re-rs),
                                        (qs-qe)
                                    )
                                ),
                                mean(
                                    c(
                                        (re-rs),
                                        (qe-qs)
                                    )
                                )
                                )
                            
                        ) %>%
                    ungroup() %>%
                    dplyr::select(
                        c(
                            "rid", qid, "strand", rs, re, qs, qe, 
                            meanlen, rlen, invs, cluster, seqs
                            )
                        ) %>%
                    dplyr::select(
                        c(
                            "rid", qid, "strand", rs, re, qs, qe, 
                            meanlen, rlen, invs, cluster, seqs
                            )
                        )
                
                inv_s <- 
                    bind_rows(outer_als, inner_al)
            }
            
            if (any(fcd_i$X_dist>1e6)){
                inv_s <- NULL
            }
            
            return(inv_s)
            
        }
    )
    
st1 <- 
    bind_rows(inv_coords) %>%arrange(meanlen)

st2 <- 
    st1$meanlen
    #dplyr::select(rs,re,qs,qe,meanlen)

meanlens

dbs2 <- 
    fpc::dbscan(st2, eps = 7e4, MinPts = 1, scale = F, method = "raw")

plot(dbs2, st2, main = "DBSCAN", frame = FALSE)

za_tib <- 
    st1 %>% #[,c(4:8)] %>%
    mutate(
         ab_cluster = 
             dbs2$cluster
     ) %>%
    group_by(ab_cluster) %>%
    mutate(
        cis = 
            n()
    ) %>%
    filter(cis>1, seqs >1)

cc_list <- 
    list()
(za_tib %>% pull(cis) %>% unique)[i]

lincs <- 
    lapply(
        (za_tib %>% pull(cis) %>% unique), 
        function(j){
            za_tib %>% filter(cis==j) %>% pull(qid) %>% unique()
        }
    )



for (i in 1:length(unique(za_tib))){
    
    cn <- 
        (za_tib %>% pull(cis) %>% unique)[i]
    cct <- 
        za_tib %>% filter(cis==cn)
    
    cc_list[[i]] <- 
        unique(za_tib$qid)
}         


## 2 major ineages
lca <- 
    lapply(
        lincs[[2]], 
        function(x){
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==x) %>%plot_delta()
        }
    )

lcb <- 
     lapply(
        lincs[[3]], 
        function(x){
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==x) %>%plot_delta()
        }
     )
length(lca)
length(lcb)

do.call(
     "grid.arrange",
     c(
         lca[1:6],
         ncol=3
         )
     )

lcbc <- 
    list()

do.call(
     "grid.arrange",
     c(
         append(lcb[1:5], lca[9]),
         ncol=3
         )
     )


C02 <- 
    best_identity %>%filter(cluster_id==0) %>% 
    arrange(desc(identity)) %>%
    dplyr::slice(1:3) %>%
    mutate(rowix = LETTERS[row_number()]) %>%
    mutate(cid = paste("C0", rowix,  sep =".")) %>% 
    dplyr::select(accessions, cid)
    

tfe1 <- 
    bis %>% filter(accessions %in% lincs[[2]][1:6]) %>%
    ungroup %>%
    mutate(rowix = LETTERS[row_number()]) %>%
    rowwise() %>%
    mutate(cid = paste("lower", rowix, inversions, collapse = "_", sep = ".")) %>% 
    dplyr::select(accessions, cid)

tfe2 <- 
    bis %>% filter(accessions %in% c(lincs[[3]][1:5], lincs[[2]][9])) %>%
    ungroup %>%
    mutate(rowix = LETTERS[row_number()]) %>%
    rowwise() %>%
    mutate(cid = paste("higher", rowix, inversions, collapse = "_", sep = ".")) %>% 
    dplyr::select(accessions, cid)

tt1 <- 
    bind_rows(
        C02, tfe1, tfe2
    )

for (i in 1:nrow(tt1)){
    file.copy(
        from = 
            sprintf("%s/%s.txt", sync_dir, pull(tt1, accessions)[i]),
        to = 
            sprintf("../data/plots/c012/%s.txt", pull(tt1, cid)[i])
    )
}


                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="NZ_CP045956.1") %>%plot_delta()
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="NZ_CP126323.1") %>%plot_delta()
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid=="NZ_LR590082.1") %>%plot_delta()
                
            ## single inv

"../data/plots/c012/C0.A.txt"
```



```{bash}
parsnp  -p 6 \
    -r ../data/plots/c012/C0.A.txt \
    -c \
    -d ../data/plots/c012/ \
    -o ../data/plots/c012_par
```

```{r}

s_tree5 <- 
    read.tree("../data/plots/c012_par/parsnp.tree")

s_tree5$tip.label <- 
    sub(".ref", "", s_tree5$tip.label)
s_tree5$tip.label <- 
    sub(".txt", "", s_tree5$tip.label)

s_tree5 <- midpoint(s_tree5)
ggtree(s_tree5) + geom_tiplab(size =7) + xlim_tree(1)

```


```{r}
mlst_list <- 
    listPubmlst_orgs() 

se_mlst_c012 <- 
    doMLST(
        infiles = 
            list.files("../data/plots/c012", full.names = T),  # The fasta files
        org = mlst_list[99], # The organism
        scheme = 2, # Scheme id number
        write = "none", # Don't write fasta files for alleles found
        ddir = 
             "../data/plots/c012_mlst2"# Write downloaded sequences and profiles here.
  ) 

se_mlst_c012

ig_se <- 
    plot(se_mlst_c012, what = 'result', type = "phylo", label = T,  pt.size = 4)


igt <- 
    plot(se_mlst_c012, what = 'result', type = "phylo", plot=F)

tree1 <- igt
tree1 <- midpoint(tree1)
ggtree(tree1) + geom_tiplab(size =7) + xlim_tree(4)

```







```{r}
## clustering
za <- 
    bis %>%
    filter(inversions==1)

za1 <- 
    lapply(
        1:nrow(za),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==za$accessions[i])
            fcd_i2 <- 
                fcd_i %>%
                filter(
                        abs((rlen-re) - rs ) < 2e5,
                        X_dist < 2e5,
                        strand=="-"
                    ) %>%
                    mutate(
                        cluster = za$cluster_id[i],
                        seqs = za$grouped_sequences[i]
                        ) %>%
                dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster, seqs))
            
            #fcd_i %>% plot_delta()
            return(fcd_i2)
        }
    ) %>% bind_rows() %>%
    arrange((meanlen))

za1_all <- 
    lapply(
        1:nrow(za_all),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==za_all$accessions[i])
            fcd_i2 <- 
                fcd_i %>%
                filter(
                        abs((rlen-re) - rs ) < 2e5,
                        X_dist < 2e5,
                        strand=="-"
                    ) %>%
                    mutate(
                        cluster = za_all$cluster_id[i],
                        seqs = za_all$grouped_sequences[i]
                        ) %>%
                dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster, seqs))
            
            #fcd_i %>% plot_delta()
            return(fcd_i2)
        }
    ) %>% bind_rows() %>%
    arrange((meanlen))

zam <- 
     za1[,c(8)] %>% as.matrix()
    za1[,c(4:8)] %>% as.matrix()

zam_all <- 
    za1_all[,c(4:7)] %>% as.matrix()
#dbscan::kNNdistplot(zam, k =  7)

kdist_plot <- 
    kdist(zam)
dbs <- 
    fpc::dbscan(zam, eps = 4e5, MinPts = 1, scale = F, method = "raw")
dbs_all <- 
    fpc::dbscan(zam_all, eps = 2e5, MinPts = 1, scale = F, method = "raw")

plot(dbs, zam, main = "DBSCAN", frame = FALSE)
plot(dbs_all, zam_all, main = "DBSCAN", frame = FALSE, type = "p")

fviz_cluster(dbs, zam, geom = "point")
fviz_cluster(dbs_all, zam_all, geom = "point")

za_tib <- 
    za1 %>% #[,c(4:8)] %>%
    mutate(
         cluster = 
             dbs$cluster
     ) 
c2 <- 
    za_tib %>% filter(cluster==2) %>% arrange(meanlen)
### clust 2 
c2_p <- 
    lapply(
        1:nrow(c2),
        function(i){
            list.files(
                sprintf("%s/all", delta_dir), 
                pattern = c2$qid[i], 
                full.names = T
                )  %>% 
                plot_delta
            # filtered_combined_delta %>% 
            #     ungroup %>%
            #     filter(qid==c2$qid[i]) %>% plot_delta
        }
    )
do.call(
     "grid.arrange",
     c(
         c2_p,
         ncol=4
         )
     )



za_plot_tib <- 
    za_tib %>%
    group_by(cluster) %>% 
    mutate(total_seqs = sum(seqs)) %>%
    filter(seqs==max(seqs)) %>%
    dplyr::slice(1)

char_inv1 <- 
    lapply(
        1:nrow(za_plot_tib),
        function(i){
            filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==za_plot_tib$qid[i]) %>% plot_delta
        }
    )



za_all_tib <- 
    za1_all %>% #[,c(4:8)] %>%
    mutate(
         cluster = 
             dbs_all$cluster
     ) 
    

za_summary <- 
    za_tib %>%
    group_by(cluster) %>%
    summarise(
        mean_sd = 
            mean(
                c(
                    round(sd(rs)),
                    round(sd(re)),
                    round(sd(qs)),
                    round(sd(qe))
                )
                ),
        rs = round(mean(rs)),
        re = round(mean(re)),
        qs = round(mean(qs)),
        qe = round(mean(qe)),
        meanlen = round(mean(meanlen)),
        seqs = sum(seqs),
        reprs = n()
    ) 


za_all_summary <- 
    za_all_tib %>%
    group_by(cluster) %>%
    summarise(
        mean_sd = 
            mean(
                c(
                    round(sd(rs)),
                    round(sd(re)),
                    round(sd(qs)),
                    round(sd(qe))
                )
                ),
        rs = round(mean(rs)),
        re = round(mean(re)),
        qs = round(mean(qs)),
        qe = round(mean(qe)),
        meanlen = round(mean(meanlen)),
        seqs = n()
    ) 

chatac_invs <- 
    za_all_summary %>%
    mutate(
        strand = "-",
        qid = cluster,
        rid = 0
    )
do.call(
     "grid.arrange",
     c(
         char_inv1,
         ncol=7
         )
     )

library(ggraph)
za_l1 <- 
    tibble(
        from = "0",
        to = 
            as.character(za_all_summary$cluster)
    )
mygraph <- graph_from_data_frame( za_l1 )

ggraph(mygraph, layout = 'dendrogram', circular = FALSE) + 
  geom_edge_diagonal() +
  geom_node_point() +
  theme_void()



dbs
dbs$cluster

c1_13 <- 
    best_identity %>% 
    filter(cluster_id==13 | cluster_id==1) %>%
    group_by(cluster_id) %>%
    #arrange(identity) %>%
    dplyr::slice(1:4)

a1 <- c2_13 %>% filter(cluster_id==1) %>%dplyr::slice(1) %>% pull(accessions)
a2 <- c2_13 %>% filter(cluster_id==13) %>%dplyr::slice(1) %>% pull(accessions)
    

p1 <- list.files(sprintf("%s/all", delta_dir), pattern = a1, full.names = T)  %>% plot_delta
p2 <- list.files(sprintf("%s/all", delta_dir), pattern = a2, full.names = T)  %>% plot_delta

do.call(
     "grid.arrange",
     c(
         list(p1, p2),
         ncol=2
         )
     )

    



grid.arrange(c(p1, p2))

#filter(inversions==1, grouped_sequences==4)

```


Internal inversions
```{r}
##more than 1
mto <- 
    bis %>%
    filter(inversions>1)
###
mto %>% filter(inversions>2)

x1 <- 
    mto%>% filter(inversions>2)
x1_plots <- 
    list()

for (i in 1:12){#length(x1)){
    fcd_i <- 
        filtered_combined_delta %>% 
        ungroup %>%
        filter(qid==x1$accessions[i])
    x1_plots[[i]] <- 
        plot_delta(fcd_i)
}

do.call(
     "grid.arrange",
     c(
         x1_plots,
         ncol=4
         )
     )


### check if events are nested
mto1 <- 
    lapply(
        1:nrow(mto),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==mto$accessions[i])
            #plot_delta(fcd_i)
            ### inversion "event"
            # is the slope ~1
            
            ## neg_al
            neg_al <- 
                fcd_i %>% filter(strand=="-") %>%
                dplyr::select(rs, re, qs, qe) %>%
                mutate(
                    rmid = round((rs+re)/2),
                    qmid = round((qs+qe)/2)
                )
            neg_slope <-     
                abs(round((lm(neg_al$rmid ~ neg_al$qmid) %>% coef())[2], 1))
            ### positive alignmnets flanked by inversions
            pos_al <- 
                fcd_i %>%
                filter(
                    (lead(strand, default = "+") =="-") &
                        (lag(strand, default = "+") =="-") 
                )
            
            
            plot(neg_al[,c(5,6)])
            
            fcd_i2 <- 
                fcd_i %>%
                filter(
                        abs((rlen-re) - rs ) < 2e5,
                        X_dist < 2e5,
                        strand=="-"
                    ) %>%
                    mutate(
                        cluster = mto$cluster_id[i],
                        seqs = mto$grouped_sequences[i]
                        ) %>%
                dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster, seqs))
            
            #fcd_i %>% plot_delta()
            return(fcd_i2)
        }
    ) %>% bind_rows() %>%
    arrange((meanlen))

##

```



















#### Previous


```{r}
library(phangorn)
delta_dirs <- 
    c(
        sprintf("%s/200/ref_v_all", delta_dir),
        sprintf("%s/the_rest", delta_dir),
        sprintf("%s/the_rest_collinear", delta_dir)
    )

list.files(delta_dirs[1]) %>% length
list.files(delta_dirs[2]) %>% length
list.files(delta_dirs[3]) %>% length


## copy to one dir

new_out_deltadir <- 
      sprintf("%s/all", delta_dir)
    
if (!dir.exists(new_out_deltadir)){dir.create(new_out_deltadir)}

mv_list <- 
    lapply(
        delta_dirs,
        function(x){
            dfs <- 
                list.files(x, full.names = T)
            lapply(
                dfs, 
                function(y){
                    file.copy(y, sprintf("%s/%s", new_out_deltadir, basename(y)))
                }
            )
        }
    )

```



```{r}
### single inversion from ref

all(filtered_combined_delta$qid %in% species_summary$accession)
which(!(species_summary$accession %in% filtered_combined_delta$qid ))

####


### clustering delta files

dp_list <- 
    list()
gv <- 
    c(4,5,6,21, 33:41)
for (i in 1:length(gv)){
    dfi <- 
        list.files(sprintf("%s/all", delta_dir), full.names = T)[gv[i]]
    dft <- 
        read_delta(dfi)
    dp_list[[i]] <- 
        plot_delta(dft)
    
}



do.call(
     "grid.arrange",
     c(
         dp_list,
         ncol=5
         )
     )
#

### variant blocks 

inv_list <- 
    list()

ins_list <- 
    list()

gap_list <- 
    list()

tloc_list <- 
    list()

event_list <- 
    list()

bis %>%
    filter(inversions==1)

tt <- 
    best_identity %>%
    filter(cluster_id==1) %>%
    dplyr::slice(1) %>%
    pull(accessions)

    best_identity %>%
    filter(cluster_id==2) %>%
    pull(accessions)

 
        filtered_combined_delta %>%
        filter(qid=="NZ_CP092012.1") %>%
        ungroup %>% plot_delta()


all_deltas_files <- 
    list.files(new_out_deltadir, full.names = T)
pb = txtProgressBar(min = 0, max = length(qrys), initial = 0) 

list.files(sprintf("%s/all", delta_dir), pattern = "_1.4", full.names = T) %>% read_delta %>% filter_delta %>% plot_delta
list.files(sprintf("%s/all", delta_dir), pattern = "_1.3", full.names = T) %>% read_delta%>% filter_delta %>% plot_delta
list.files(sprintf("%s/all", delta_dir), pattern = "_1.5", full.names = T) %>% read_delta %>% plot_delta

list.files(sprintf("%s/all", delta_dir), pattern = "NZ_CP022069.2", full.names = T) %>% read_delta %>% filter_delta %>% plot_delta


filtered_combined_delta
###
#i = 6

for (i in 1:length(qrys)){#qrys))
    ft <- 
        filtered_combined_delta %>%
        filter(qid==qrys[i]) %>%
        ungroup 
    plot_delta(ft)
    
    ## -ves
    neg_invs <- 
         ft  %>%
        filter(
            strand=="-",
            meanlen >5e4
        ) %>%
        mutate(
            RASR = 
                abs((rlen-re) - rs ) < 2e5 & 
            (X_dist < 2e5) & (strand=="-"),
            cluster = clust_i
        ) %>%
        dplyr::select(
            c(
                "rid", qid, "strand", rs, re, qs, qe,
                meanlen, rlen, qlen, invs, cluster, RASR
                )
            ) %>%
        dplyr::rename(
           "acc" =  "qid",
           "clust" = "cluster"
        ) %>%
        mutate(
            rmid = 
                round((rs+re)/2),
            qmid = 
                round((qs+qe)/2),
            rmid_ori = 
                rlen-rmid,
            qmid_ori = 
                qlen-qmid
            )
        
            
        )
        

    
    
    }




####

clustered_genomes

all_svs <- 
    list()

for (i in 1:length(qrys)){#qrys))
    ft <- 
        filtered_combined_delta %>%
        filter(qid==qrys[i]) %>%
        ungroup 
    
    # ft2 <- 
    #     filtered_combined_delta %>%
    #     filter(qid==qrys[5]) %>%
    #     ungroup 
    # 
    # ## feature extraction
    # fe1 <- 
    #     ft %>%
    #     rowwise() %>%
    #     summarise(
    #         rp = round((rs+re)/2),
    #         qp = round((qs+qe)/2),
    #         strand = strand,
    #         meanlen = meanlen
    #     )
    # fe2 <- 
    #     ft2 %>%
    #     rowwise() %>%
    #     summarise(
    #         rp = round((rs+re)/2),
    #         qp = round((qs+qe)/2),
    #         strand = strand,
    #         meanlen = meanlen
    #     )
    
    # plot_delta(ft)
    # ft <- 
    #     filtered_combined_delta %>%
    #     filter(qid==tt[1]) %>%
    #     ungroup 
    
    clust_i <- 
        best_identity %>%
        filter(accessions==qrys[i]) %>%
        pull(cluster_id)
        
    
    d2 <- 
      ft %>%
      rowwise() %>%
      mutate(
        qs2 = min(c(qs, qe)),
        qe2 = max(c(qs, qe))
      ) %>% 
      ungroup
    
    ## gap insertion positions
    # insertions
    ref_ranges <- 
        IRanges(
            start = 
            ft$rs,
            end = 
            ft$re
            )#
    qry_ranges <- 
      IRanges(
        start = 
          d2$qs2,
        end = 
          d2$qe2
      ) 
    
    ins <- 
        ref_ranges %>%
        reduce() %>%
        gaps(start = 1, end = ft$rlen[1]) 
    ins_df <- 
        ins %>%
        as.data.frame() %>% 
        mutate(
            acc = ft$qid[1],
            clust = clust_i
            ) %>%
        mutate(
            variant = "Insertion"
        )
    
     gaps <- 
         qry_ranges %>%
         reduce() %>%
         gaps(start = 1, end = ft$qlen[1])
     
     gaps_df <- 
         gaps %>%
         as.data.frame() %>% 
         mutate(
             acc = ft$qid[1],
             clust = clust_i
             ) %>%
        mutate(
            variant = "Gap"
        )
     
    ins_list[[qrys[i]]] <- 
        ins_df
    
    gap_list[[qrys[i]]] <- 
        gaps_df 
    
    invs1 <- 
        ft  %>%
        filter(
            strand=="-",
            meanlen >5e4
        ) %>%
        mutate(
            RASR = 
                abs((rlen-re) - rs ) < 2e5 & 
            (X_dist < 2e5) & (strand=="-"),
            cluster = clust_i
        ) %>%
        dplyr::select(
            c(
                "rid", qid, "strand", rs, re, qs, qe,
                meanlen, rlen, invs, cluster, RASR
                )
            ) %>%
        dplyr::rename(
           "acc" =  "qid",
           "clust" = "cluster"
        ) %>%
        rowwise() %>%
        mutate(
            start = round(mean(c(rs, qe))),
            end = round(mean(c(re, qs)))
            )  %>%
        ungroup()
            
        # mutate(
        #     abs((rlen-re) - rs ) < 2e5,
        #     X_dist < 2e5,
        #     strand=="-"
        #     ) %>%
        # mutate(cluster = za$cluster_id[i]) %>%
        # dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
    inv_summary <- 
        invs1
    
    ### inner_inversions
    
    
    if (nrow(invs1)>1){
        inner_invs1 <- 
            ft %>%
            # filtered_combined_delta %>% 
            #             ungroup %>%
            #             filter(qid==zt1$qid[i]) %>%
            filter(
                strand=="+",
                meanlen >5e4,
                lag(strand=="-") & lead(strand=="-")  
            ) %>%
            mutate(
                RASR = 
                    abs((rlen-re) - rs ) < 2e5 & 
                (X_dist < 2e5) & (strand=="-"),
                cluster = clust_i
            )%>%
            dplyr::select(
                c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster, RASR)
                ) %>%
            dplyr::rename(
               "acc" =  "qid",
               "clust" = "cluster"
            ) %>%
            rowwise() %>%
            mutate(
                start = round(mean(c(rs, qs))),
                end = round(mean(c(re, qe)))
                )  %>%
            ungroup()
        inv_summary <- 
            bind_rows(invs1, inner_invs1) %>% arrange(rs)
        
        # plot(
        #     tibble(
        #         starts = 
        #             c(invs1$rs, invs1$qe),
        #         ends = 
        #             c(invs1$re, invs1$qs)
        #         )
        #     )
        
        outer_invs <- 
            invs1
        
        
    }
    
    inv_list[[qrys[i]]] <- 
        inv_summary
    
    gaps_df
    
    invs2 <- 
        inv_summary %>%
        mutate(
            width = abs(start-end+1),
            variant = "Inversion"
            ) %>%
        dplyr::select(c(start, end, width, acc, clust, variant))
        
    tlocs <- 
        ft %>% 
        filter(
            X_dist >1e5#,
            #strand=="+"
            ) %>%
        mutate(
            clust = clust_i,
            acc = qid
        )%>%
        rowwise() %>%
        mutate(
            start = round(mean(c(rs, qe))),
            end = round(mean(c(re, qs))),
            width =  abs(start-end+1),
            variant = "Translocation"
            )  %>%
        ungroup() %>%
        dplyr::select(c(start, end, width, acc, clust, variant))

    
    tloc_list[[qrys[i]]] <- 
        tlocs
    
    
    all_events <- 
        bind_rows(
            invs2, 
            gaps_df,
            ins_df,
            tlocs
        ) %>%
        arrange(start)
        
    #     filter(
    #         abs((rlen-re) - rs ) < 2e5,
    #         X_dist < 2e5,
    #         strand=="-"
    #         ) %>%
    #     mutate(cluster = za$cluster_id[i]) %>%
    #     dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
    # 
    # ### ## flanking inversions
    # 
    # filtered_combined_delta %>% 
    #                 ungroup %>%
    #                 filter(qid==zt1$qid[i]) %>%
    #     mutate(
    #         inv_flanked = 
    #         lag(strand=="-") & lead(strand=="-")  
    #         )
    
    event_list[[qrys[i]]] <- 
        all_events
        
    
    setTxtProgressBar(pb,i)
        close(pb)
    #print(i)

}

lapply(
    all_svs,
    function(x){
        
    }
)



el1 <-
    bind_rows(event_list)

el1_v <- 
    el1 %>%
    filter(variant=="Inversion") %>%
    mutate(
        midpoint = round((start+end)/2)
    )


ggplot(el1_v, aes(x = midpoint, y = width)) +
  geom_point(color = "blue") +
  labs(title = "Midpoint vs. Length", x = "Midpoint", y = "Length") +
  theme_minimal()



all(qrys %in% (el1 %>% pull(acc) %>% unique))
which(!(qrys %in% (el1 %>% pull(acc) %>% unique)))
it2 <- 
     bind_rows(inv_list)

it_lists <- 
    lapply(
        unique(it2$cluster),
        function(i){
            
        }
    )


###inversion clusters

it2[,c(4,5, 11)] %>% 
    mutate(rs = rs/1e4, re = re/1e4) %>%
    filter(cluster==2) %>%
    group_by(cluster) %>%
    summarise(
        rs2 = mean(rs) %>% round(), 
        re2 = mean(re) %>% round(),
        sdrs = sd(rs) %>% round(),
        sdre = sd(re) %>% round(),
        cc = n()
    ) %>%
    arrange(rs2) %>%
    as.data.frame()



it2 %>% 
    count(invs)

tloc_tib <- 
    bind_rows(tloc_list) %>% filter(rs >1)

tloc_tib %>% arrange(meanlen)

filtered_combined_delta %>%
        filter(qid=="NZ_CP065062.1") %>%
        ungroup %>% plot_delta()


it3 
it3 <- 
    it2[,c(4,5)]%>%
    melt() %>%
    as_tibble() %>%
    mutate(
        position = 
            factor(ifelse(variable =="rs", "start", "end"), levels = c("start", "end"))
    ) 



ggplot(it3, aes(x=position,y=value))+
  geom_violin()+
  geom_dotplot(binaxis="y",stackdir='center',binwidth = 0.1)



ggplot(it3, aes(x=value, fill = position)) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")

ggplot(all_bps_m, aes(x=value, fill = variable)) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")

ins_tib <- 
    bind_rows(ins_list) %>% 
    as_tibble()

it4 <-
    bind_rows(
        (ins_tib %>% mutate(mutation = "Insertion")),
        (gap_tib %>% mutate(mutation = "Gap")) 
    ) %>%
    mutate(pos = round((end+start)/2))

it5 <- 
    tloc_tib %>%
    mutate(pos = round((re+rs)/2))
    
ggplot(it5, aes(x=pos)) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")


ggplot(it4, aes(x=pos, fill = mutation)) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")



ins_tib2 <- 
    ins_tib[,c(1,2)]%>%
    melt() %>%
    as_tibble() %>%
    mutate(
        position = 
            factor(variable, levels = c("start", "end"))
    ) 

ggplot(ins_tib2, aes(x=value, fill = position)) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")

ggplot(it3, aes(x=value, fill = position)) + 
  geom_density(alpha=0.4)+
 # geom_histogram(aes(y=..density..), bins=100, alpha=0.5, 
 #                position="identity")+
    scale_fill_brewer(palette="Set1") +
    theme_classic() +
    theme(
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16)
    ) +
    xlab("Reference genome position") +
    ylab("Density")

    
gap_tib <- 
    bind_rows(gap_list) %>% 
    as_tibble()





ggplot(ins_tib, aes(x=start)) + 
  geom_density()

ggplot(gap_tib, aes(x=start)) + 
  geom_density()

it2


filtered_combined_delta %>%
    filter(qid==tt) %>%
    ungroup %>%
    plot_delta()




read_delta(
    list.files(
        sprintf(
            "%s/all",
            delta_dir
            ),
        pattern = tt,
        full.names = T
        )
    )


###


best_identity %>% filter(cluster_id==22)
best_identity
za <- 
    bis %>%
    filter(inversions==1)

za_all <- 
    best_identity %>%
    filter(inversions==1)

za3 <- 
    best_identity %>%
    filter(inversions==1) %>%
    filter(
        grouped_sequences>3
    ) %>%
    group_by(cluster_id) %>%
    sample_n(4)


za[!(za$cluster_id %in% za1$cluster),]

## zero adjacent
library(fpc)
install.packages("dbscan")
library(dbscan)
za1 <- 
    lapply(
        1:nrow(za),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==za$accessions[i])
            fcd_i2 <- 
                fcd_i %>%
                filter(
                        abs((rlen-re) - rs ) < 2e5,
                        X_dist < 2e5,
                        strand=="-"
                    ) %>%
                    mutate(cluster = za$cluster_id[i]) %>%
                dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
            
            #fcd_i %>% plot_delta()
            return(fcd_i2)
        }
    ) %>% bind_rows() %>%
    arrange((meanlen))


filtered_combined_delta %>% ungroup %>% filter(qid=="NZ_CP023345.1") %>% plot_delta

library(factoextra)
####
zam <- 
    za1[,c(4:7)] %>% as.matrix()

zam_all <- 
    za1_all[,c(4:7)] %>% as.matrix()


dbscan::kNNdistplot(zam, k =  7)

kdist_plot <- 
    kdist(zam)
dbs <- 
    fpc::dbscan(zam, eps = 4e5, MinPts = 1, scale = F, method = "raw")

dbs_all <- 
    fpc::dbscan(zam_all, eps = 4e5, MinPts = 1, scale = F, method = "raw", showplot = 1)

fpc::dbscan(zam, eps = 2.5e5, MinPts = 1)

plot(dbs, zam, main = "DBSCAN", frame = FALSE)

plot(dbs_all, zam_all, main = "DBSCAN", frame = FALSE)


res.fpc <- fpc::dbscan(zam_all, eps = 4e5, MinPts = 1, method = "raw")
fviz_cluster(res.fpc, zam_all, geom = "point")


plot(zam)
set.seed(123)
db <- 
    fpc::dbscan(zam, eps = 0.15, MinPts = 5)

###    
onesy_plots <- 
    lapply(
        1:nrow(za1),
        function(i){
            filtered_combined_delta %>% 
                    ungroup %>%
                    filter(qid==za1$qid[i]) %>%
                plot_delta()
        }
    )

 do.call(
     "grid.arrange",
     c(
         onesy_plots,
         ncol=5
         )
     )

## zero two
zt <- 
    bis %>%
    filter(inversions==2)
zt1 <- 
    lapply(
        1:nrow(zt),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==zt$accessions[i])
            fcd_i2 <- 
                fcd_i %>%
                filter(
                        abs((rlen-re) - rs ) < 1e5,
                        X_dist < 1e5,
                        strand=="+"
                    ) %>%
                    mutate(cluster = zt$cluster_id[i]) %>%
                dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
            
            #fcd_i %>% plot_delta()
            return(fcd_i2)
        }
    ) %>% bind_rows() %>%
    arrange((meanlen))
 

twosy_plots <- 
    lapply(
        1:nrow(zt1),
        function(i){
            filtered_combined_delta %>% 
                    ungroup %>%
                    filter(qid==zt1$qid[i]) %>%
                plot_delta()
        }
    )

 do.call(
     "grid.arrange",
     c(
         twosy_plots,
         ncol=3
         )
     )
 
 
 cc1 <- 
     bis_1$accessions[51:75]
     #bis_1$accessions[100:123]
 
 1,2,4,7,
 8,9,13,14, 
 15, 26, 28, 29, 
 30,32
 cc2 <- 
    bis_1 %>%
     ungroup %>%
     dplyr::slice(
         c(
             13,2,4,7,
             8,9,30,32,
             1,14,26, 28,
             15, 29
             # 1,2,4,7,
             # 8,9,14, 15, 
             # 26, 28, 29, 30,32
             )
         ) %>% pull(accessions)
  cc2 <- 
    bis_1 %>%
     ungroup %>%
     dplyr::slice(
         c(
             76:100
             # 1,2,4,7,
             # 8,9,14, 15, 
             # 26, 28, 29, 30,32
             )
         ) %>% pull(accessions)
 
 many_plots <- 
    lapply(
        1:length(cc1),
        function(i){
            filtered_combined_delta %>% 
                    ungroup %>%
                    filter(qid==cc1[i]) %>%
                plot_delta()
        }
    )

 do.call(
     "grid.arrange",
     c(
         many_plots,
         ncol=5
         )
     )
 


```



```{r}
### what are the shared structural variants 
```






```{r}


ref_adj <- 
    bis %>%
    filter(inversions==1)

ref_2inv <- 
    bis %>%
    filter(inversions==2)

### centered RASR
# conditions
# the alignment block that passes through the halfway point
# within 1e5...... abs((rlen-re) - rs ) < 1e5
# X dist < 1e5
###

cluster_centered_RASRs <- 
    lapply(
        1:(length(bis$accessions)-1),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>%
                filter(qid==bis$accessions[i+1])# %>%
                # filter(
                #     rs != min(rs), 
                #     re != max(re)
                #     )
            #VABs <- 

            RASRs <- 
                filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[i+1]) %>%
                filter(
                    abs((rlen-re) - rs ) < 1e5,
                    X_dist < 1e5
                ) %>%
                mutate(cluster = bis$cluster_id[i+1]) %>%
                dplyr::select(c("rid", qid, "strand", rs, re, qs, qe, meanlen, rlen, invs, cluster))
            ## variant alignment blocks
            
            filtered_combined_delta %>%
                ungroup %>%
                filter(qid==bis$accessions[i+1])

            #plot_delta(fcd_i)
            # bps <- 
            #     tibble(
            #         acc = bis$accessions[i+1],
            #         clust = bis$cluster_id[i+1],
            #         start = 
            #             fcd_i$rs,
            #         end = 
            #             fcd_i$re
            #     )
            
            return(RASRs)
        }
    ) %>% bind_rows()


### iterative for loop checking shared RASRs
cluster_centered_RASRs$cluster
shared_RASR <- 
    matrix(
        F, 
        nrow = nrow(cluster_centered_RASRs), 
        ncol = nrow(cluster_centered_RASRs)
        )

for (i in 1:nrow(cluster_centered_RASRs)){
    for (j in 1:nrow(cluster_centered_RASRs)){
        if (i<=j) {
            shared_RASR[i,j] <- F
        } else{
            ref_range <- 
                IRanges(
                    start = 
                        cluster_centered_RASRs[i,] %>% pull(rs),
                    end = 
                        cluster_centered_RASRs[i,] %>% pull(re)
                )
            
            qry_range <- 
                IRanges(
                    start = 
                        cluster_centered_RASRs[j,] %>% pull(rs),
                    end = 
                        cluster_centered_RASRs[j,] %>% pull(re)
                )
            
            ref_in_qry <- 
                subjectHits(
                    IRanges::findOverlaps(
                        ref_range,
                        qry_range, 
                        minoverlap = 0.95*width(ref_range)
                        )
                    )
            qry_in_ref <- 
                subjectHits(
                    IRanges::findOverlaps(
                        qry_range,
                        ref_range, 
                        minoverlap = 0.95*width(qry_range)
                        )
                    )
            
            shared_RASR[i,j] <- 
                ((length(ref_in_qry)>0) & (length(qry_in_ref)>0))
                
        }
        # if (pre_adj[i,j]==1){
        #     new_rows <- 
        #         rbind(new_rows, adjacent_pair)
        # }
    }
}


anchored_rasr <- 
    which(shared_RASR, arr.ind = T) %>%
    as_tibble() %>%
    mutate(invs = 1)




e_l <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = anchored_rasr$row,
                target = anchored_rasr$col
                ),
        directed = F
    )
plot(e_l)


#2-8, 8-31, 8-32, 8-30, 

    

# Create the graph from named edges
g <- graph(edges, directed = FALSE)

# Customize the layout to position the "5.9 MB" node between 0 and 13
layout <- layout_with_fr(g)  # Use a force-directed layout as a base
layout[V(g)$name == "5.9 MB", ] <- colMeans(layout[c(which(V(g)$name == "0"), which(V(g)$name == "13")), ])

vertex_sizes <- ifelse(V(g)$name == "0", 40,
                ifelse(V(g)$name == "1", 35,
                ifelse(V(g)$name == "2", 30, 25)))

# Plot the graph with custom styling
plot(g,
     layout = layout,
     vertex.label = V(g)$name,         # Label vertices with their names
     vertex.color = ifelse(V(g)$name == "5.9 MB", "white", "skyblue"),  # Different color for label vertex
     vertex.size = ifelse(V(g)$name == "5.9 MB", 0, 30),  # Hide the label vertex circle
     edge.width = 2,
     vertex.label.cex = 1.5
     edge.color = "gray",
     main = "Salmonella enterica")

# Create the graph from named edges
g2 <- graph(edges2, directed = FALSE)

# Customize the layout to position the "5.9 MB" node between 0 and 2
layout <- layout_with_fr(g2)  # Use a force-directed layout as a base
layout[V(g2)$name == "5.9 MB", ] <- colMeans(layout[c(which(V(g2)$name == "0"), which(V(g2)$name == "2")), ])

# Plot the graph with custom styling
plot(g2,
     layout = layout,
     vertex.label = V(g2)$name,         # Label vertices with their names
     vertex.color = ifelse(V(g2)$name == "5.9 MB", "white", "skyblue"),  # Different color for label vertex
     vertex.size = ifelse(V(g2)$name == "5.9 MB", 0, 30),  # Hide the label vertex circle
     edge.width = 2,
     vertex.label.cex = 1.5,
     edge.color = "gray",
     main = "Graph with Edge Label Between (0--2)")




edges <- c("13", "2", "2", "8", "13", "4", "4", "8", 
           "0", "13", "0", "1", "1", "14", "14", "15", "14", "28", "15", "28", 
           "0", "5.9 MB", "5.9 MB", "13")

# Create the graph from named edges
g <- graph(edges, directed = FALSE)

# Customize the layout to position the "5.9 MB" node between 0 and 13
layout <- layout_with_fr(g)  # Use a force-directed layout as a base
layout[V(g)$name == "5.9 MB", ] <- colMeans(layout[c(which(V(g)$name == "0"), which(V(g)$name == "13")), ])

# Set custom vertex sizes: 40 for "0", 35 for "1", 30 for "2", and 25 for all others
vertex_sizes <- ifelse(V(g)$name == "0", 40,
                ifelse(V(g)$name == "1", 35,
                ifelse(V(g)$name == "2", 30, 25)))

# Plot the graph with custom styling
plot(g,
     layout = layout,
     vertex.label = V(g)$name,          # Label vertices with their names
     vertex.label.cex = 1.5,            # Increase font size
     vertex.color = ifelse(V(g)$name == "5.9 MB", "white", "skyblue"),  # Different color for label vertex
     vertex.size = vertex_sizes,        # Apply custom vertex sizes
     edge.width = 2,
     edge.color = "gray",
     main = "Graph with Edge Label Between (0--13)")



edges <- c("13", "2", "2", "8", "13", "4", "4", "8", 
           "0", "13", "0", "1", "1", "14", "14", "15", "14", "28", "15", "28", 
           "0", "5.9 MB", "5.9 MB", "13")

# Create the graph from named edges
edges <- c("13", "2", "2", "8", "13", "4", "4", "8", 
           "0", "13", "0", "1", "1", "14", "14", "15", "14", "28", "15", "28", 
           "0", "5.9 MB", "5.9 MB", "13")

# Create the graph from named edges
edges <- c("13", "2", "2", "8", "13", "4", "4", "8", 
           "0", "13", "0", "1", "1", "14", "14", "15", "14", "28", "15", "28", 
           "0", "5.9 MB", "5.9 MB", "13", "0", "86.4 MB", "86.4 MB", "1")

# Create the graph from named edges
edges <- c("13", "2", "2", "8", "13", "4", "4", "8", 
           "0", "13", "0", "1", "1", "14", "14", "15", "14", "28", "15", "28", 
           "0", "5.9 MB", "5.9 MB", "13", "0", "86.4 MB", "86.4 MB", "1")

# Create the graph from named edges
g <- graph(edges, directed = FALSE)

# Customize the layout to position the "5.9 MB" node between 0 and 13, and "86.4 MB" between 0 and 1
layout <- layout_with_fr(g)  # Use a force-directed layout as a base
layout[V(g)$name == "5.9 MB", ] <- colMeans(layout[c(which(V(g)$name == "0"), which(V(g)$name == "13")), ])
layout[V(g)$name == "86.4 MB", ] <- colMeans(layout[c(which(V(g)$name == "0"), which(V(g)$name == "1")), ])

# Set custom vertex sizes: 40 for "0", 35 for "1", 30 for "2", and 25 for all others
vertex_sizes <- ifelse(V(g)$name == "0", 37,
                ifelse(V(g)$name == "1", 33,
                ifelse(V(g)$name == "2", 30, 25)))

# Remove the "5.9 MB" and "86.4 MB" nodes' circles by setting their size to 0, but keep the labels
vertex_sizes[V(g)$name == "5.9 MB"] <- 0
vertex_sizes[V(g)$name == "86.4 MB"] <- 0

# Manually define the communities
community_0 <- c("0")
community_1_14_15_28 <- c("1", "14", "15", "28")
community_13_2_8_4 <- c("13", "2", "8", "4")

# Assign colors based on predefined communities for node filling and borders
vertex_colors <- ifelse(V(g)$name %in% community_0, "lightblue", 
                        ifelse(V(g)$name %in% community_1_14_15_28, "lightcoral", 
                               ifelse(V(g)$name %in% community_13_2_8_4, "lightgreen", "skyblue")))

# Set border colors for each community (still use borders for main nodes)
border_colors <- ifelse(V(g)$name %in% community_0, "blue", 
                        ifelse(V(g)$name %in% community_1_14_15_28, "darkred", 
                               ifelse(V(g)$name %in% community_13_2_8_4, "green", "gray")))
species_summary %>%
    dplyr::count(technology) %>% arrange(desc(n))


# Plot the graph with custom styling, including the new label between "0" and "1"
plot(g,
     layout = layout,
     vertex.label = V(g)$name,          # Label vertices with their names
     vertex.label.cex = 1.5,             # Increase font size
     vertex.color = vertex_colors,      # Color nodes based on their community
     vertex.frame.color = border_colors,  # Set the border color based on community
     vertex.size = vertex_sizes,        # Apply custom vertex sizes
     edge.width = 3,
     edge.color = "gray",
     main = "Salmonella enterica")



edges <- c("13", "2", "2", "8", "13", "4", "4", "8", 
           "0", "13", "0", "1", "1", "14", "14", "15", "14", "28", "15", "28", 
           "0", "5.9 MB", "5.9 MB", "13", "0", "86.4 MB", "86.4 MB", "1")

# Create the graph from named edges
g <- graph(edges, directed = FALSE)

# Customize the layout to position the "5.9 MB" and "86.4 MB" nodes
layout <- layout_with_fr(g)  # Use a force-directed layout as a base

# Stretch the layout horizontally by adjusting the x-coordinates of the vertices
layout[, 1] <- layout[, 1] * 2  # Stretch x-coordinates by a factor of 2
layout[, 2] <- layout[, 2] * 1  # Keep y-coordinates as they are

# Manually adjust the position of node "0" and "1" to increase the distance between them
# Increase the distance by adjusting their x-coordinate values
layout[V(g)$name == "0", 1] <- layout[V(g)$name == "0", 1] - 1  # Move node "0" left
layout[V(g)$name == "1", 1] <- layout[V(g)$name == "1", 1] + 1  # Move node "1" right

# Set custom vertex sizes: 40 for "0", 35 for "1", 30 for "2", and 25 for all others
vertex_sizes <- ifelse(V(g)$name == "0", 40,
                ifelse(V(g)$name == "1", 35,
                ifelse(V(g)$name == "2", 30, 25)))

# Remove the "5.9 MB" and "86.4 MB" nodes' circles by setting their size to 0, but keep the labels
vertex_sizes[V(g)$name == "5.9 MB"] <- 0
vertex_sizes[V(g)$name == "86.4 MB"] <- 0

# Manually define the communities
community_0 <- c("0")
community_1_14_15_28 <- c("1", "14", "15", "28")
community_13_2_8_4 <- c("13", "2", "8", "4")

# Assign colors based on predefined communities for node filling and borders
vertex_colors <- ifelse(V(g)$name %in% community_0, "lightblue", 
                        ifelse(V(g)$name %in% community_1_14_15_28, "lightcoral", 
                               ifelse(V(g)$name %in% community_13_2_8_4, "lightgreen", "skyblue")))

# Set border colors for each community
border_colors <- ifelse(V(g)$name %in% community_0, "blue", 
                        ifelse(V(g)$name %in% community_1_14_15_28, "darkred", 
                               ifelse(V(g)$name %in% community_13_2_8_4, "green", "gray")))

# Plot the graph with custom styling
plot(g,
     layout = layout,
     vertex.label = V(g)$name,          # Label vertices with their names
     vertex.label.cex = 1.5,             # Increase font size
     vertex.color = vertex_colors,      # Color nodes based on their community
     vertex.frame.color = border_colors,  # Set the border color based on community
     vertex.size = vertex_sizes,        # Apply custom vertex sizes
     edge.width = 2,
     edge.color = "gray",
     main = "Salmonella enterica")


shared_ranges <- 
    lapply(
        1:length(ref_ranges),
        function(k){
            ov_check1 <- 
                subjectHits(
                    IRanges::findOverlaps(
                        ref_ranges[k],
                        qry_ranges, 
                        minoverlap = 0.95*width(ref_ranges[k])
                        )
                    )
            if (length(ov_check1)>0){
                ov_check2 <- 
                    subjectHits(
                        IRanges::findOverlaps(
                            qry_ranges[ov_check1], 
                            ref_ranges,
                            minoverlap = 0.95*width(qry_ranges[ov_check1])
                            )
                    )
                if (length(ov_check2)>0){
                    outstr <- 
                        c(ov_check2, ov_check1)
                } else {
                    outstr <- NULL
                }
                
            } else {
                outstr <- NULL
            }
            return(outstr)
            
        }
    )




pa3 <- 
    pbmclapply(
        1:nrow(ref_adj),
        function(i){
            ref_clust <-
                ref_adj[i,] %>% pull(cluster_id)
            ref_acc <- 
                ref_adj[i,] %>% pull(accessions)
            bp_check <- 
                lapply(
                    1:nrow(ref_2inv),
                    function(j){
                        qry_clust <-
                            ref_2inv[j,] %>% pull(cluster_id)
                        qry_acc <-
                            ref_2inv[j,] %>% pull(accessions)
                        
                        ref_bps <- 
                            cluster_major_breakpoints[[ref_clust]]
                        qry_bps <- 
                            cluster_major_breakpoints[[qry_clust]]
                        any_starts <- 
                            any(abs(outer(ref_bps$start , qry_bps$start , "-")) <= 5000)
                        any_ends <- 
                            any(abs(outer(ref_bps$end , qry_bps$end , "-")) <= 5000)
                        
                        potentially_adjacent <- 
                            all(c(any_starts, any_ends))
                        
                        if (potentially_adjacent) {
                            outtib <- 
                                tibble(
                                    ref = 
                                        ref_acc,
                                    qry = 
                                        qry_acc,
                                    ref_clust = 
                                        ref_clust,
                                    qry_clust = 
                                        qry_clust
                                )
                        } else {
                            outtib <- NULL
                        }
                        return(outtib)
                        
                    }
                ) 
            bp_tib <- 
                bp_check[!sapply(bp_check, is.null)] %>%
                bind_rows
            
            return(bp_check)
        }
    ) %>% head()




filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[12]) %>% 
                plot_delta(
                    ., 
                    xlb = "Cluster 0",
                    ylb = 
                        sprintf(
                            "Cluster %i",
                            pull(bis, cluster_id)[i] %>%
                                as.character() %>% 
                                as.numeric()
                                )
                    )






which(pa3)
any(pa3)

adjacent_check <- 
    lapply(
        1:nrow(potential_adj),
        function(i){

            ref_i <- 
                pull(potential_adj, ref)[i]
            qry_i <- 
                pull(potential_adj, qry)[i]
            
            ref_bps <- 
                cluster_major_breakpoints[[ref_i]]
            qry_bps <- 
                cluster_major_breakpoints[[qry_i]]
            
             ## do a start and an end match in each set?
            any_starts <- 
                any(abs(outer(ref_bps$start , qry_bps$start , "-")) <= 5000)
            any_ends <- 
                any(abs(outer(ref_bps$end , qry_bps$end , "-")) <= 5000)
            
            potentially_adjacent <- 
                all(c(any_starts, any_ends))
            
            return(potentially_adjacent)
        }
    ) %>% unlist








```






```{r}
## basic agreement
td <-
    "../data/processing/phylo_ss/sub3_12/fa"
td2 <-
    "../data/processing/phylo_ss/sub3_1_10/fa"

td3 <-
    "../data/processing/phylo_ss/path1/fa"

top3_path1 <- 
    best_identity %>% 
    filter(cluster_id %in% c(0,13, 5, 2, 4, 8)) %>%
    group_by(cluster_id) %>%
    sample_n(3) %>% as.data.frame()

dir.create(td2 , recursive = T)
top3_12 <- 
    best_identity %>%
    filter(cluster_id %in% c(1,2)) %>%
    group_by(cluster_id) %>%
    sample_n(3) %>% as.data.frame()


set.seed(123)
top3 <- 
    best_identity %>%
    filter(cluster_id %in% c(1:10)) %>%
    group_by(cluster_id) %>%
    sample_n(3) %>% as.data.frame()

## move the synced files to a new directory
lapply(
    1:nrow(top3_12),
    function(i){
        xf <- 
            list.files(sync_dir,full.names = T, pattern = (top3_12 %>% pull(accessions))[i])
        clust_i <- 
            (top3_12 %>% pull(cluster_id))[i]
        
        file.copy(xf, to = sprintf("%s/%s_%s",td, clust_i, basename(xf)))
    }
)

lapply(
    1:nrow(top3),
    function(i){
        xf <- 
            list.files(sync_dir,full.names = T, pattern = (top3 %>% pull(accessions))[i])
        clust_i <- 
            (top3 %>% pull(cluster_id))[i]
        
        file.copy(xf, to = sprintf("%s/%s_%s",td2, clust_i, basename(xf)))
    }
)

lapply(
    1:nrow(top3_path1),
    function(i){
        xf <- 
            list.files(sync_dir,full.names = T, pattern = (top3_path1 %>% pull(accessions))[i])
        clust_i <- 
            (top3_path1 %>% pull(cluster_id))[i]
        file.copy(xf, to = sprintf("%s/%s_%s",td3, clust_i, basename(xf)))
    }
)


```



```{bash}
parsnp  -p 6 \
    -r ../data/plots/c012/A0a \
    -c \
    -d ../data/processing/phylo_ss/sub3_1_10/fa \
    -o ../data/processing/phylo_ss/sub3_1_10/parsnp
```


```{r}
s_tree4 <- 
    read.tree("../data/processing/phylo_ss/sub3_1_10/parsnp/parsnp.tree")

s_tree4$tip.label <- 
    sub(".ref", "", s_tree4$tip.label)
s_tree4$tip.label <- 
    sub(".txt", "", s_tree4$tip.label)

ss_phylo <- 
    as.phylo(hc_row) 


ss_phylo$edge.length%>% sort %>% head(20)
ss_phylo$edge.length <- 
    pmax(ss_phylo$edge.length, 4e-3)

s_tree4 <- midpoint(s_tree4)
ggtree(s_tree4) + geom_tiplab() + xlim_tree(1)

```




```{bash}
parsnp  -p 6 \
    -r ! \
    -c \
    -d ../data/processing/phylo_ss/sub3_1_10/fa \
    -o ../data/processing/phylo_ss/sub3_1_10/parsnp
```

```{bash}
parsnp  -p 6 \
    -r ! \
    -c \
    -d ../data/processing/phylo_ss/sub3_12/fa \
    -o ../data/processing/phylo_ss/sub3_12/parsnp
```

```{bash}
parsnp  -p 6 \
    -r ! \
    -c \
    -d ../data/processing/phylo_ss/path1/fa \
    -o ../data/processing/phylo_ss/path1/parsnp
```


```{r}
s_tree4 <- 
    read.tree("../data/processing/phylo_ss/path1/parsnp/parsnp.tree")

s_tree4$tip.label <- 
    sub(".ref", "", s_tree4$tip.label)
s_tree4$tip.label <- 
    sub(".txt", "", s_tree4$tip.label)

s_tree4 <- midpoint(s_tree4)
ggtree(s_tree4) + geom_tiplab() + xlim_tree(1)
```


```{r}
s_tree3 <- read.tree("../data/processing/phylo_ss/sub3_1_10/parsnp/parsnp.tree")

s_tree3$tip.label <- 
    sub(".ref", "", s_tree3$tip.label)
s_tree3$tip.label <- 
    sub(".txt", "", s_tree3$tip.label)

s_tree3 <- midpoint(s_tree3)
ggtree(s_tree3) + geom_tiplab() + xlim_tree(6)
```




```{r}
## ParSNP
s_tree2 <- read.tree("../data/processing/phylo_ss/sub3_12/parsnp/parsnp.tree")

s_tree2$tip.label <- 
    sub(".ref", "", s_tree2$tip.label)
s_tree2$tip.label <- 
    sub(".txt", "", s_tree2$tip.label)

s_tree2 <- midpoint(s_tree2)
ggtree(s_tree2) + geom_tiplab() + xlim_tree(6)

ggtree(s_tree2)

c3_12 <- 
    list.files("../data/processing/phylo_ss/sub3_12/fa", full.names = T)
### MLSTar

se_sub3_12 <- 
    doMLST(
        infiles = c3_12, # The fasta files
        org = mlst_list[99], # The organism
        scheme = 2, # Scheme id number
        write = "none", # Don't write fasta files for alleles found
        ddir = 
            "../data/processing/phylo_ss/sub3_12/mlst",
                 # Write downloaded sequences and profiles here.
  )  


```






########


Agreement between the top 5 clusters
```{r}

## get the top 5 accessions from each one
set.seed(123)
top5 <- 
    best_identity %>%
    filter(cluster_id %in% c(0:5)) %>%
    group_by(cluster_id) %>%
    sample_n(15) %>% as.data.frame()

## move the synced files to a new directory
lapply(
    1:nrow(clusters_top3),
    function(i){
        
        
        xf <- 
            list.files(sync_dir,full.names = T, pattern = (clusters_top3 %>% pull(accessions))[i])
        
        file.copy(xf, to = sprintf("../data/processing/path/%s", basename(xf)))
    }
)



### mlstar


##snp


```







```{r}
qrys <-
    lapply(
        list.files(new_out_deltadir, full.names = F),
        function(x){
            # x <- 
            #     list.files(new_out_deltadir, full.names = F)[1]
            outf <- 
                sub(
                    "NZ_CP039558.1_v_",
                    "",
                    x
                    ) %>%
                    sub(
                        "_filtered.delta",
                        "",
                        .
                    )
            return(outf)
            }
        ) %>% unlist

all_deltas <- 
    list.files(new_out_deltadir, full.names = F)

##

bis %>%
    filter(cluster_id %in% c(0,2,4,8,5,13))

bi2 <- 
    tibble(
        ref = 
            c(
                "NZ_CP039567.1", #0
                "NZ_CP039567.1" ,#0
                "NZ_LT905063.1", #13
                "NZ_LT905063.1", #13
                "NC_021820.1", #5
                "NZ_LT905062.1" #4
            ),
        qry = 
            c(
                "NZ_LT905063.1", #13
                "NC_021820.1" ,#5
                "NZ_LT905062.1", #4
                "NZ_CP029919.1", #2
                "NZ_CP029919.1", #2
                "NC_016832.1" #4
            ),
        ref_clust = 
            c(0,0,13,13,5,4),
        qry_clust = 
            c(13,5,4,2,2,8)
    )
delta_dir_ava <- 
    sprintf("%s/path1", delta_dir)

dir.create(delta_dir_ava)

nuc_commands <- 
    generate_nucmer_commands(
        genome_matrix = bi2, 
        delta_dir =  delta_dir_ava, 
        sync_dir = sync_dir 
    )

pbmclapply(nuc_commands, system, mc.cores = 6)

pl <- 
    lapply(
        1:length(list.files(delta_dir_ava)),
        function(i){
            dfa <- 
                read_delta(
                    list.files(delta_dir_ava, full.names = T)[i]
                )
            ref <- 
                dfa$rid[1]
            qry <- 
                dfa$qid[1]
            ref_clust <- 
                clustered_genomes %>% 
                filter(accessions==ref) %>% pull(cluster_id)
            qry_clust <- 
                clustered_genomes %>% 
                filter(accessions==qry) %>% pull(cluster_id)
            
            outp <- 
                plot_delta(dfa) +
                xlab(sprintf("Cluster %s", ref_clust))+
                ylab(sprintf("Cluster %s", qry_clust))
                
        }
    
    )

list.files(delta_dir_ava, full.names = T)[1] %>% plot_delta()

 do.call(
     "grid.arrange",
     c(
         pl,
         ncol=2
         )
     )
 
 
 

nucmer_commands <- 
    nucmer_ref_v_all(
        most_related, sampled_accession_list,
        delta_dir =  delta_dir_200, 
        sync_dir = subset_sync_dir
        )




###

minimum_alignment_percentage <- 
    90
minimum_inversion_length <- 
    5e4
filtered_combined_delta <- 
    filter_combined_delta(
        sprintf("%s/all", delta_dir), 
        minimum_alignment_percentage, 
        minimum_inversion_length
    )

#delta_file <- 
#    list.files(sprintf("%s/ref_v_all", delta_dir_200), full.names = T)[1]
clustered_genomes_list <- 
    cluster_genomes(filtered_combined_delta)

clustered_genomes <- 
   clustered_genomes_list$cluster_summary %>%
    left_join(
        ., 
        species_summary[,c("accession", "len")] %>% 
            `colnames<-`(c("accessions", "len")),
        by = "accessions"
        )
saveRDS(
    clustered_genomes_list, 
    sprintf(
        "%s/clustered_genomes_list.rds",
        r_vars_dir
    )
    )

clustered_genomes
clustered_genomes_list$cluster_coords %>%
    filter(strand=="+")
lapply(
    1:nrow(clustered_genomes_list$cluster_coords),
    function(i){
        cc_row <- 
            clustered_genomes_list$cluster_coords[i,] 
        inv_tib <- 
            tibble(
                accessions = 
                    base::strsplit(
                        (cc_row %>% pull(accessions))[i], 
                        split = ", "
                        )[[1]],
                acc_count = 
                    pull(cc_row, cluster_count)
                )
        it_clust <- 
            left_join(
                inv_tib, 
                clustered_genomes, 
                by = "accessions"
            ) %>%
            group_by(cluster_id) %>%
            dplyr::count()
        
        
    }
)


clustered_genomes_list$cluster_coords %>%
    separate()

# clustered_genomes <- 
#     readRDS(
#         sprintf(
#         "%s/clustered_genomes.rds",
#         r_vars_dir
#     )
#     )
species_summary %>% nrow
fss <- 
    lapply(
        (list.files(new_out_deltadir, full.names = T)),
        file.size
    ) %>% unlist


delta_identities <- 
    pbmclapply(
        1:length((list.files(new_out_deltadir))),
        function(i){
            delta_tib <- 
                read_delta(list.files(new_out_deltadir, full.names = T)[i]) 
            
            nucmer_id <- 
                sum(
                    delta_tib$meanlen
                    )/
                (mean(
                    c(
                        delta_tib$rlen[1], 
                        delta_tib$qlen[1])
                     ))
            c
            out_tib <- 
                tibble(
                    accession = 
                        delta_tib$qid[1],
                    identity = 
                        nucmer_id
                )
            
            return(out_tib)
            
        }, mc.cores = 4
    ) %>% bind_rows()

best_identity <- 
    left_join(
        delta_identities %>% `colnames<-`(c("accessions", "identity")),
        clustered_genomes,
        by = "accessions"
    )

best_identity

bis <- 
    best_identity %>% 
    group_by(cluster_id) %>%
    arrange(desc(identity)) %>% 
    filter(!is.na(cluster_id)) %>%
    dplyr::slice(1)

clust_identity_diff_adjacent <- 
            bis %>% 
            mutate(adjacency = abs(ref_invs_to_0-qry_invs_to_0)) %>%
            filter(
                adjacency == 1
            )

cida_out <- 
    clust_identity_diff_adjacent %>%
    filter(
        ref_clust !=0,
        qry_clust !=0
    )
## potential adjacents
pre_adj <- 
    matrix(data = NA, nrow = nrow(bis), ncol = nrow(bis))
pot_adj_list <- 
    matrix()
new_rows <- NULL 
    
for (i in 1:nrow(bis)){
    for (j in 1:nrow(bis)){
        if (i>=j) {
            pre_adj[i,j] <- 0
        } else{
            pre_adj[i,j] <- 
                abs(pull(bis, inversions)[i] - pull(bis, inversions)[j])
        }
        # if (pre_adj[i,j]==1){
        #     new_rows <- 
        #         rbind(new_rows, adjacent_pair)
        # }
    }
}


pa1 <- 
    which(pre_adj==1, arr.ind = T) %>%
    as_tibble() %>%
    mutate(
        ref = row-1,
        qry = col-1
    ) %>%
    dplyr::select(c("ref", "qry"))

pre_adj[1:5,1:5]
adjacent_to_ref <- 
    pa1 %>%
    filter(
        ref==0 | qry ==0
    )


potential_adj <- 
    pa1 %>%
    filter(
        ref != 0, 
        qry != 0
    )

adjacent_check


## check if there are shared breakpoints within 50kb

adjacent_check <- 
    lapply(
        1:nrow(potential_adj),
        function(i){

            ref_i <- 
                pull(potential_adj, ref)[i]
            qry_i <- 
                pull(potential_adj, qry)[i]
            
            ref_bps <- 
                cluster_major_breakpoints[[ref_i]]
            qry_bps <- 
                cluster_major_breakpoints[[qry_i]]
            
             ## do a start and an end match in each set?
            any_starts <- 
                any(abs(outer(ref_bps$start , qry_bps$start , "-")) <= 5000)
            any_ends <- 
                any(abs(outer(ref_bps$end , qry_bps$end , "-")) <= 5000)
            
            potentially_adjacent <- 
                all(c(any_starts, any_ends))
            
            return(potentially_adjacent)
        }
    ) %>% unlist


###
pa2 <- 
    potential_adj %>%
    mutate(
        potentially_adj = 
            adjacent_check
    ) %>%
    filter(potentially_adj) %>%
    left_join(
        ., 
        (bis[,c(1,3)] %>% `colnames<-`(c("ref_acc", "ref"))),
        by = "ref"
    ) %>%
    left_join(
        ., 
        (bis[,c(1,3)] %>% `colnames<-`(c("qry_acc", "qry"))),
        by = "qry"
    )
    

pa_mat <- 
    pa2 %>%
    dplyr::select(c(ref_acc, qry_acc)) %>%
    `colnames<-`(c("ref", "qry"))

%>% as.matrix()
pa_tib <- 
    tibble(
        ref = 
            bis$accessions[pull(pa2, ref)],
        qry =
            bis$accessions[pull(pa2, qry)],
        ref_clust = 
            pull(pa2, ref),
        qry_clust = 
            pull(pa2, qry)
    )

zero_adjacent <- 
    adjacent_to_ref %>% 
    `colnames<-`(c("cluster_id", "qry")) %>%
    left_join(
        ., bis[,c(1,3,4)], by = "cluster_id"
    ) %>%
    `colnames<-`(
        c(
            "ref_clust", "cluster_id", "ref", "seqs"
        )
    ) %>%
    left_join(
        ., bis[,c(1,3)], by = "cluster_id"
    ) %>%
    mutate(inversions=1) %>%
    `colnames<-`(
        c(
            "ref_clust", "qry_clust", "rid", "seqs", "qid", "inversions"
        )
    ) %>%
    dplyr::select(c("rid", "qid", "inversions", "ref_clust", "qry_clust"))

# za2 <- 
#     zero_adjacent %>%
#     mutate(inversions=1) %>%
#     dplyr::select(c("rid", "qid", "inversions", "ref_clust", "qry_clust"))
    
mummer_commands <- 
    mummer_all_v_all(
        pa_mat,
        sprintf("%s/all_adj", delta_dir),
        sync_dir 
        )
sprintf("%s/all_adj/all_v_all", delta_dir)
### subset files
list.files(sprintf("%s/subset", sync_dir)) %in% list.files(sync_dir) %>% sum
lapply(
    1:length(list.files(sprintf("%s/subset", sync_dir))),
    function(i){
        xf <- 
            list.files(sprintf("%s/subset", sync_dir), full.names = T)[i]
        if (!(xf %in% list.files(sync_dir, full.names = T))){
            file.copy(xf, to = sprintf("%s/%s", sync_dir, basename(xf)))
        }
    }
)
length(list.files(sync_dir))


list.files(sprintf("%s/all_adj/all_v_all", delta_dir), full.names = F)[10]

list.files(sprintf("%s/subset", sync_dir), pattern = "NZ_OW704627.1")
# lapply(
#     list.files(sprintf("%s/all_adj/all_v_all", delta_dir), full.names = T),
#     function(x){
#         if(file.size(x)==0){
#             file.remove(x)
#         }
#     }
# ) %>% unlist()


filtered_combined_delta_adj <- 
    filter_combined_delta(
        sprintf("%s/all_adj/all_v_all", delta_dir), 
        minimum_alignment_percentage, 
        minimum_inversion_length
    )

inversion_counts_adj <- 
    filtered_combined_delta_adj %>%
    summarise(
        inversions = unique(invs)
    ) %>%
    ungroup %>% 
    distinct 


adjacent_clusters <-
        bind_rows(
            zero_adjacent,
            inner_join(
                (inversion_counts_adj %>% filter(inversions==1)),
                clustered_genomes %>%
                    dplyr::select(cluster_id, accessions) %>%
                    `colnames<-`(c("ref_clust", "rid")),
                by = "rid"
            ) %>%
                inner_join(
                    ., 
                    clustered_genomes %>%
                        dplyr::select(cluster_id, accessions) %>%
                        `colnames<-`(c("qry_clust", "qid")),
                    by = "qid"
                )
        )

adjacent_clusters2 <-
        #bind_rows(
            #zero_adjacent,
            inner_join(
                (inversion_counts_adj %>% filter(inversions==1)),
                clustered_genomes %>%
                    dplyr::select(cluster_id, accessions) %>%
                    `colnames<-`(c("ref_clust", "rid")),
                by = "rid"
            ) %>%
                inner_join(
                    ., 
                    clustered_genomes %>%
                        dplyr::select(cluster_id, accessions) %>%
                        `colnames<-`(c("qry_clust", "qid")),
                    by = "qid"
                )
        #)

  
bi2


edge_list <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = adjacent_clusters$ref_clust,
                target = adjacent_clusters$qry_clust
                ),
        directed = F
    )

edge_list2 <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = adjacent_clusters2$ref_clust,
                target = adjacent_clusters2$qry_clust
                ),
        directed = F
    )

adjacent_clusters2



node_size_table2 <-
    clustered_genomes %>% 
    group_by(cluster_id) %>% 
    summarise(seqs = unique(grouped_sequences)) %>%
    ungroup %>%
    mutate(size_v = rescale(seqs, to=c(15, 40))) %>%
    filter(cluster_id %in% names(V(edge_list2))) %>%
    arrange(match(cluster_id, names(V(edge_list2))))


V(edge_list2)$size <- 
    node_size_table2$size_v

V(edge_list2)$label.cex <- 
    node_size_table2$size_v

plot(edge_list2, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)




## node size based on amount
node_size_table <-
    clustered_genomes %>% 
    group_by(cluster_id) %>% 
    summarise(seqs = unique(grouped_sequences)) %>%
    ungroup %>%
    mutate(size_v = rescale(seqs, to=c(15, 40))) %>%
    filter(cluster_id %in% names(V(edge_list))) %>%
    arrange(match(cluster_id, names(V(edge_list))))

V(edge_list)$size <- 
    node_size_table$size_v

V(edge_list)$label.cex <- 
    node_size_table$size_v

plot(edge_list, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)


edge_list <- 
    graph_from_data_frame(
        d =
            data.frame(
                source = bi2$ref_clust,
                target = bi2$qry_clust
                ),
        directed = F
    )
node_size_table <-
    clustered_genomes %>% 
    filter(cluster_id %in%c(0,2,5,13,8,4)) %>%
    group_by(cluster_id) %>% 
    summarise(seqs = unique(grouped_sequences)) %>%
    ungroup %>%
    mutate(size_v = rescale(seqs, to=c(15, 40))) %>%
    filter(cluster_id %in% names(V(edge_list))) %>%
    arrange(match(cluster_id, names(V(edge_list))))
V(edge_list)$size <- 
    node_size_table$size_v

V(edge_list)$label.cex <- 
    node_size_table$size_v

plot(edge_list, vertex.label.cex=1.5, vertex.color="lightblue", edge.width= 3)


ac2 <- 
    adjacent_clusters %>%
    filter(
        ref_clust %in% c(0,2,5,4,8),
        qry_clust %in% c(0,2,5,4,8)
    )
    


        
new_rows       


bina <- 
    best_identity %>%
    filter(is.na(cluster_id)) 

sum(biprop$grouped_sequences)

biprop <- 
    bis %>%
    filter(!is.na(cluster_id)) %>%
    ungroup %>%
    #group_by(cluster_id) %>%
    mutate(
        cluster_prop = 
            round((100*grouped_sequences)/sum(grouped_sequences), 1)
    ) %>%
    mutate(
        c2 =
             ifelse(
                 grouped_sequences==1, 
                 "Singletons", 
                 ifelse(
                     grouped_sequences>1 & grouped_sequences<5,
                     "With <5 \n sequences",
                     ifelse(
                         grouped_sequences>4 & grouped_sequences<10,
                         "With <10 \n sequences",
                         as.character(cluster_id)
                         )
                     )
                 )
        ) %>% group_by(c2) %>%
    summarise(
        sequences = sum(grouped_sequences)
    ) %>% 
    as.data.frame() %>%
    mutate(
        c3 = 
            factor(c2, levels = c(0:15, "With <10 \n sequences", "With <5 \n sequences", "Singletons"))
        ) %>% 
    arrange(c3) %>%
    mutate(
        cluster_prop = 
            round((100*sequences)/sum(sequences), 1)
    ) 
            
biprop2 <- 
    bis %>%
    filter(!is.na(cluster_id)) %>%
    ungroup %>%
    #group_by(cluster_id) %>%
    mutate(
        cluster_prop = 
            round((100*grouped_sequences)/sum(grouped_sequences), 1)
    ) %>%
    mutate(
        c2 =
             ifelse(
                 grouped_sequences==1, 
                 "Singletons", 
                 ifelse(
                     grouped_sequences>1 & grouped_sequences<15,
                     "With <15 \n sequences",
                     as.character(cluster_id)
                     )
                 )
        ) %>% group_by(c2) %>%
    summarise(
        sequences = sum(grouped_sequences)
    ) %>% 
    as.data.frame() %>%
    mutate(
        c3 = 
            factor(
                c2, 
                levels = c(0:15, "Clusters with \n <15 sequences", "Singletons")
                )
        ) %>% 
    arrange(c2) %>%
    mutate(
        cluster_prop = 
            round((100*sequences)/sum(sequences), 1)
    ) 

biprop2[,-3] %>% 
    `colnames<-`(c("Cluster", "Sequences", "Percent of total")) %>%
    dplyr::slice(c(1:6,8,7)) %>%
    kbl(align = "c") %>%
  kable_classic(full_width = F, html_font = "Cambria")%>%
  kable_styling(bootstrap_options = "striped")


ggplot(
       biprop2, 
        aes(x = c3, y = cluster_prop)) +
      geom_bar(stat = "identity",  colour = "black") +
      scale_y_continuous( 
          breaks = c(1,20, 40, 60, 80, 100),
          labels = c( "",  "20", "40", "60", "80", "100"), limits = (c(0,80))
          ) +
        scale_fill_manual(values =  carto_pal(10, "Safe")) +
        guides(fill="none") +
      labs(
          x ="Cluster",
           y = "Percent of overall sequences") +
        theme_classic()+
      theme(
          axis.text.x = element_text(angle = 45, hjust = 1),
          axis.text = element_text(size=24),
          axis.title = element_text(size=28),
          title = element_text(face = "bold")
      ) 



group_data <- 
    bis %>% filter(!is.na(cluster_id))


total_members <- 
    sum(group_data$grouped_sequences)

pairs_same_suit <- 
    sum(choose(group_data$grouped_sequences, 2))

pairs_any_suit <- choose(total_members, 2)

probability_same_suit <- 
    pairs_same_suit / pairs_any_suit


total_pairs <- choose(total_members, 2)

# Calculate total number of pairs of members from different groups
different_groups_pairs <- sum(sapply(group_data$grouped_sequences, function(x) x * (total_members - x)))

# Calculate probability
probability_different_groups <- different_groups_pairs / total_pairs

# Calculate odds
odds_different_groups <- probability_different_groups / (1 - probability_different_groups)

# Print the odds
print(odds_different_groups)

bis

clusters_top3 <- 
    best_identity %>% 
    filter(cluster_id %in%c(1:15)) %>% 
    group_by(cluster_id) %>% 
    arrange(desc(identity))%>% dplyr::slice(1:4) %>%
    mutate(
    )
    probability_different_groups
    
colors <- 
    rainbow(length(unique(clusters_top3$cluster_id)))

# Create a mapping between unique numbers and colors
color_map <- setNames(colors, unique(clusters_top3$cluster_id))

# Create a new column with colors based on the mapping
clusters_top3$colors <- 
    color_map[as.character(clusters_top3$cluster_id)]
list.files(new_out_deltadir)
clusters_top3 %>% pull(accessions)
lapply(
    1:nrow(clusters_top3),
    function(i){
        xf <- 
            list.files(sync_dir,full.names = T, pattern = (clusters_top3 %>% pull(accessions))[i])
        
        file.copy(xf, to = sprintf("../data/processing/path/%s", basename(xf)))
    }
)

list.files("../data/processing/path/")

clusters_top3$accessions

stree2 <-
    plyr::rename(s_tree$tip.label, setNames(clusters_top3$cluster_id, clusters_top3$accessions))

# Plot the tree with replaced tip labels
p <- 
    ggtree(s_tree) +
  geom_tiplab(aes(label = label_mapping$new_label))


###
s_tree <- read.tree("../data/processing/pathparsnp/parsnp.tree")

show_col(viridis_pal(option = "A", begin = 0.1, end = 0.8)(15))
show_col(viridis_pal(option = "D", begin = 0, end = 0.6)(15))
s_tree$tip.label <- 
    sub(".ref", "", s_tree$tip.label)
s_tree$tip.label <- 
    sub(".txt", "", s_tree$tip.label)

unique_values <- 
    unique(clusters_top3$cluster_id)

library(randomcoloR)
n <- 15
palette <- distinctColorPalette(n)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
col_v <- 
    unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))[1:15]
pie(rep(1,n), col=col_v)  

col_p <- 
    setNames(col_v, unique_values) %>% as.data.frame() %>%
    rownames_to_column("cluster_id") %>% as_tibble %>%
    mutate(cluster_id = as.numeric(cluster_id)) %>%
    `colnames<-`(c("cluster_id", "col"))
md3 <- 
    clusters_top3 %>%
    left_join(
        ., 
        col_p, by = "cluster_id"
    )

plot(s_tree, type = "phylogram", cex = 0.8, tip.color = as.character(md3$col))

ggtree(s_tree) + geom_tiplab()


snp_tree <- 
    read.tree("../data/processing/pathparsnp//parsnp.tree")

snp_tree$tip.label <- 
    sub(".fna", "", snp_tree$tip.label)
snp_tree$tip.label <- 
    sub(".ref", "", snp_tree$tip.label)
snp_tree$tip.label <- 
    sub(".txt", "", snp_tree$tip.label)
cols2 <- 
    with(clusters_top3, as.character(colors[match(snp_tree$tip.label, accessions)]))
plot(snp_tree, tip.color = cols2)


s_tree$tip.label <- 
    sub(".txt", "", s_tree$tip.label)

plot(s_tree, type = "cladogram", cex = 0.6)


ggtree(s_tree) + geom_tiplab()

mlst_list <- 
    listPubmlst_orgs() 

se_mlst2 <- 
    doMLST(
        infiles = "../data/processing/path", # The fasta files
        org = mlst_list[99], # The organism
        scheme = 2, # Scheme id number
        write = "none", # Don't write fasta files for alleles found
        ddir = 
            "../data/processing/pubmlst/v1"# Write downloaded sequences and profiles here.
  ) 

plot(se_mlst2, what = 'result', type = "phylo", label = T,  pt.size = 2)

snp_tree$tip.label <- 
    sub(".ref", "", snp_tree$tip.label)
###
(list.files(new_out_deltadir)) %>% head(200)

list.files



bis %>% filter(cluster_id %in%c(1,5,14, 15))
delta_plots2[[3]]

### intervals for each inversion


find_adjacency <- 
    function(
        
    ){
        
    }

for (i in 1:nrow(bis)){
    for (j in 1:nrow(bis)){
        if (i>=j) {
            pre_adj[i,j] <- 0
        } else{
            pre_adj[i,j] <- 
                abs(pull(bis, inversions)[i] - pull(bis, inversions)[j])
        }
        # if (pre_adj[i,j]==1){
        #     new_rows <- 
        #         rbind(new_rows, adjacent_pair)
        # }
    }
}


cluster_inv_ranges[c(3,5)]

#i=3,j=5
bis[i,]
bis[j,]
cluster_inv_ranges[[i]]
cluster_inv_ranges[[j]]
cluster_inv_strands[[i]]
cluster_inv_strands[[j]]
## Check if any ranges match
## ref qry

ref_ranges <- 
    cluster_inv_ranges[[i]]
qry_ranges <- 
    cluster_inv_ranges[[j]]

shared_ranges <- 
    lapply(
        1:length(ref_ranges),
        function(k){
            ov_check1 <- 
                subjectHits(
                    IRanges::findOverlaps(
                        ref_ranges[k],
                        qry_ranges, 
                        minoverlap = 0.95*width(ref_ranges[k])
                        )
                    )
            if (length(ov_check1)>0){
                ov_check2 <- 
                    subjectHits(
                        IRanges::findOverlaps(
                            qry_ranges[ov_check1], 
                            ref_ranges,
                            minoverlap = 0.95*width(qry_ranges[ov_check1])
                            )
                    )
                if (length(ov_check2)>0){
                    outstr <- 
                        c(ov_check2, ov_check1)
                } else {
                    outstr <- NULL
                }
                
            } else {
                outstr <- NULL
            }
            return(outstr)
            
        }
    )
any_non_null <- 
    any(sapply(shared_ranges, function(x) !is.null(x)))

which_non_null <- 
    shared_ranges[(sapply(shared_ranges, function(x) !is.null(x)))]



    





subjectHits(
    IRanges::findOverlaps(
        cluster_inv_ranges[[i]],
        cluster_inv_ranges[[j]], 
        minoverlap = 5e4
    )
)
pre_adj[1:5,1:5]

which(pre_adj==1, arr.ind = T)
        
        cluster_inv_strands
## major breakpoints
cluster_major_breakpoints <- 
    lapply(
        1:(length(bis$accessions)-1),
        function(i){
            fcd_i <- 
                filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[i+1]) %>%
                filter(
                    rs != min(rs), 
                    re != max(re)
                    )
            plot_delta(fcd_i)
            
            bps <- 
                tibble(
                    acc = bis$accessions[i+1],
                    clust = bis$cluster_id[i+1],
                    start = 
                        fcd_i$rs,
                    end = 
                        fcd_i$re
                )
            
            return(bps)
        }
    )




        cluster_major_breakpoints %>% head
        
                
        
cluster_inv_ranges <- 
    lapply(
        1:length(bis$accessions),
        function(i){
            if (i==1){
                inner_ranges <- 
                    NA
            } else{
                fcd_i <- 
                    filtered_combined_delta %>% 
                    ungroup %>% 
                    filter(qid==bis$accessions[i]) %>%
                    filter(
                        rs != min(rs), 
                        re != max(re)
                        )
                inner_ranges <- 
                    IRanges(
                        start = 
                            fcd_i$rs,
                        end = 
                            fcd_i$re
                    )
            }
            return(inner_ranges)
        }
    )




cluster_inv_strands <- 
    lapply(
        1:length(bis$accessions),
        function(i){
            if (i==1){
                inner_strands <- 
                    NA
            } else{
                fcd_i <- 
                    filtered_combined_delta %>% 
                    ungroup %>% 
                    filter(qid==bis$accessions[i]) %>%
                    filter(
                        rs != min(rs), 
                        re != max(re)
                        )
                inner_strands <- 
                    fcd_i %>% pull(strand)
            }
            return(inner_strands)
        }
    )

cluster_inv_dists <- 
    lapply(
        1:length(bis$accessions),
        function(i){
            if (i==1){
                inner_dists <- 
                    NA
            } else{
                fcd_i <- 
                    filtered_combined_delta %>% 
                    ungroup %>% 
                    filter(qid==bis$accessions[i]) %>%
                    filter(
                        rs != min(rs), 
                        re != max(re)
                        )
                inner_dists <- 
                    fcd_i %>% pull(X_dist)
            }
            return(inner_dists)
        }
    )


bis$accessions


###
delta_plots2 <- 
    lapply(
        1:20,#nrow(bis),
        function(i){
            filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[i]) %>% 
                plot_delta(
                    ., 
                    xlb = "Cluster 0",
                    ylb = 
                        sprintf(
                            "Cluster %i",
                            pull(bis, cluster_id)[i] %>%
                                as.character() %>% 
                                as.numeric()
                                )
                    )
            }
    )

delta_plots3 <- 
    lapply(
        21:40,#nrow(bis),
        function(i){
            filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[i]) %>% 
                plot_delta(
                    ., 
                    xlb = "Cluster 0",
                    ylb = 
                        sprintf(
                            "Cluster %i",
                            pull(bis, cluster_id)[i] %>%
                                as.character() %>% 
                                as.numeric()
                                )
                    )
            }
    )

delta_plots4 <- 
    lapply(
        c(3,5,9, 6, 14),#nrow(bis),
        function(i){
            filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[i]) %>% 
                plot_delta(
                    ., 
                    xlb = "Cluster 0",
                    ylb = 
                        sprintf(
                            "Cluster %i",
                            pull(bis, cluster_id)[i] %>%
                                as.character() %>% 
                                as.numeric()
                                )
                    )
            }
    )




 do.call(
     "grid.arrange",
     c(
         delta_plots4,
         ncol=2
         )
     )
 do.call(
     "grid.arrange",
     c(
         delta_plots3,
         ncol=5
         )
     )
 
  do.call(
     "grid.arrange",
     c(
         list(
             delta_plots2[[2]],
             delta_plots2[[6]],
             delta_plots2[[15]],
             delta_plots2[[16]]
         ),
         ncol=2
         )
     )
  
  
    do.call(
     "grid.arrange",
     c(
         list(
             delta_plots2[[14]],
             delta_plots2[[6]],
             delta_plots2[[3]],
             delta_plots2[[5]],
             delta_plots2[[9]]
         ),
         ncol=2
         )
     )
  
 
    filtered_combined_delta %>% 
                ungroup %>% 
                filter(qid==bis$accessions[34]) %>% 
                plot_delta(
                    ., 
                    xlb = "Cluster 0",
                    ylb = 
                        sprintf(
                            "Cluster %i",
                            pull(bis, cluster_id)[i] %>%
                                as.character() %>% 
                                as.numeric()
                                )
                    )
            
    
    
    
 rm(delta_plots)
 
 ggplot(bis %>% filter(grouped_sequences>0), aes(cluster_id, grouped_sequences)) + geom_col(width = .8)
 ggplot(
     bis,
 )



summary(fss)
which(fss<10000)

plot_delta(
    (list.files(new_out_deltadir, full.names = T))[9]
)

all(qrys %in% species_summary$accession)
qrys %>% length

qrys %>% head

list.files(new_out_deltadir) %>% unique %>% length
```







